{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a5c11d84-b7a0-43a8-97a2-014f1bf390d5",
   "metadata": {},
   "source": [
    "# Estimate Multi-ancestry PRS profiles\n",
    "- **Project:** Multi-ancestry PRS\n",
    "- **Version:** Python/3.9\n",
    "- **Status:** COMPLETE\n",
    "- **Last Updated:** 25-FEB-2024\n",
    "\n",
    "## Notebook Overview\n",
    "- Extract 90 SNPs from individual level data across ancestries in unrelated individual level data\n",
    "- Merge files\n",
    "- Estimate profiles"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f0f9499-411b-4012-85c1-b2fc3b50d7ec",
   "metadata": {},
   "source": [
    "## Extract 90 SNPs from individual level data across ancestries in unrelated individual level data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d5fcf749-d198-49de-a3fd-2792e5b13a53",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Input data\n",
    "cd ${WORK_DIR}/imputed_data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "03119fe2-ecfb-4b06-997b-7111e6461b17",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[+] Loading plink  2.0-alpha \n"
     ]
    }
   ],
   "source": [
    "## Load packages\n",
    "module load plink/2.0-alpha"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09970d51-4d48-432f-bec3-6eb3faec8632",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Extract 90 risk loci per chr\n",
    "\n",
    "# List of ancestries\n",
    "ANCESTRIES=(\"AAC\" \"AFR\" \"AJ\" \"AMR\" \"EAS\" \"EUR\" \"CAS\")\n",
    "\n",
    "# Iterate over ancestries\n",
    "for ANCESTRY in \"${ANCESTRIES[@]}\"; do\n",
    "    for chnum in {1..22}; do\n",
    "        plink2 \\\n",
    "        --pfile \"$ANCESTRY\"/chr\"$chnum\"_\"$ANCESTRY\"_release6 \\\n",
    "        --extract range /data/GP2/projects/PRS_multi_ancestry_Nov2023/CHR_POS_HG38.txt  \\\n",
    "        --make-bed \\\n",
    "        --out \"$ANCESTRY\"/PRS_chr\"$chnum\"_release6\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6c6d23d7-51a6-4388-8f38-09cf683b0f58",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Change kernel to python\n",
    "import pandas as pd\n",
    "\n",
    "# Read the file into a DataFrame\n",
    "file_path = f\"{WORK_DIR}/GP2_master_key_release6.txt\"\n",
    "df = pd.read_csv(file_path, delimiter='\\t')  # Assuming columns are tab-separated, adjust if needed\n",
    "\n",
    "# Add a new column with \"0\"\n",
    "df.insert(0, 'New_Column', 0)\n",
    "\n",
    "# Extract columns GP2sampleID and sex_for_qc\n",
    "result_df = df[['New_Column', 'GP2sampleID', 'sex_for_qc']]\n",
    "\n",
    "# Save the result to a CSV file without headers\n",
    "result_df.to_csv(f'{WORK_DIR}/sex.txt', index=False, header=False, sep = ' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6e8e569a-fff9-41be-aff2-774076bcef22",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[+] Loading plink  1.9.0-beta4.4  on cn2458 \n"
     ]
    }
   ],
   "source": [
    "module load plink/1.9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06493892-cf24-46c8-ab2f-98d5c98f5f0a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Change kernel back to bash\n",
    "# List of ancestries\n",
    "ANCESTRIES=(\"AAC\" \"AFR\" \"AJ\" \"AMR\" \"EAS\" \"EUR\" \"CAS\")\n",
    "\n",
    "cd ${WORK_DIR}/imputed_data/\n",
    "\n",
    "# Iterate over ancestries\n",
    "for ANCESTRY in \"${ANCESTRIES[@]}\"; do\n",
    "    for chnum in {1..22};\n",
    "        do\n",
    "            plink \\\n",
    "            --bfile \"$ANCESTRY\"/PRS_chr\"$chnum\"_release6 \\\n",
    "            --update-sex ${WORK_DIR}/sex.txt \\\n",
    "            --remove ${WORK_DIR}/raw_genotypes/\"$ANCESTRY\"/related_\"$ANCESTRY\" \\\n",
    "            --make-bed \\\n",
    "            --out \"$ANCESTRY\"/PRS_chr\"$chnum\"_release6_gender\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "af9ecd82-e976-4e5d-9f70-5baa1606bd5d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Files for Ancestry AAC:\n",
      "PRS_ALL_release6_gender.bim\n",
      "PRS_chr10_release6_gender.bim\n",
      "PRS_chr11_release6_gender.bim\n",
      "PRS_chr12_release6_gender.bim\n",
      "PRS_chr13_release6_gender.bim\n",
      "PRS_chr14_release6_gender.bim\n",
      "PRS_chr15_release6_gender.bim\n",
      "PRS_chr16_release6_gender.bim\n",
      "PRS_chr17_release6_gender.bim\n",
      "PRS_chr18_release6_gender.bim\n",
      "PRS_chr19_release6_gender.bim\n",
      "PRS_chr1_release6_gender.bim\n",
      "PRS_chr20_release6_gender.bim\n",
      "PRS_chr21_release6_gender.bim\n",
      "PRS_chr2_release6_gender.bim\n",
      "PRS_chr3_release6_gender.bim\n",
      "PRS_chr4_release6_gender.bim\n",
      "PRS_chr5_release6_gender.bim\n",
      "PRS_chr6_release6_gender.bim\n",
      "PRS_chr7_release6_gender.bim\n",
      "PRS_chr8_release6_gender.bim\n",
      "PRS_chr9_release6_gender.bim\n",
      "Files for Ancestry AFR:\n",
      "PRS_ALL_release6_gender.bim\n",
      "PRS_chr10_release6_gender.bim\n",
      "PRS_chr11_release6_gender.bim\n",
      "PRS_chr12_release6_gender.bim\n",
      "PRS_chr13_release6_gender.bim\n",
      "PRS_chr14_release6_gender.bim\n",
      "PRS_chr15_release6_gender.bim\n",
      "PRS_chr16_release6_gender.bim\n",
      "PRS_chr17_release6_gender.bim\n",
      "PRS_chr18_release6_gender.bim\n",
      "PRS_chr19_release6_gender.bim\n",
      "PRS_chr1_release6_gender.bim\n",
      "PRS_chr20_release6_gender.bim\n",
      "PRS_chr21_release6_gender.bim\n",
      "PRS_chr2_release6_gender.bim\n",
      "PRS_chr3_release6_gender.bim\n",
      "PRS_chr4_release6_gender.bim\n",
      "PRS_chr5_release6_gender.bim\n",
      "PRS_chr6_release6_gender.bim\n",
      "PRS_chr7_release6_gender.bim\n",
      "PRS_chr8_release6_gender.bim\n",
      "PRS_chr9_release6_gender.bim\n",
      "Files for Ancestry AJ:\n",
      "PRS_ALL_release6_gender.bim\n",
      "PRS_chr10_release6_gender.bim\n",
      "PRS_chr11_release6_gender.bim\n",
      "PRS_chr12_release6_gender.bim\n",
      "PRS_chr13_release6_gender.bim\n",
      "PRS_chr14_release6_gender.bim\n",
      "PRS_chr15_release6_gender.bim\n",
      "PRS_chr16_release6_gender.bim\n",
      "PRS_chr17_release6_gender.bim\n",
      "PRS_chr18_release6_gender.bim\n",
      "PRS_chr19_release6_gender.bim\n",
      "PRS_chr1_release6_gender.bim\n",
      "PRS_chr20_release6_gender.bim\n",
      "PRS_chr21_release6_gender.bim\n",
      "PRS_chr2_release6_gender.bim\n",
      "PRS_chr3_release6_gender.bim\n",
      "PRS_chr4_release6_gender.bim\n",
      "PRS_chr5_release6_gender.bim\n",
      "PRS_chr6_release6_gender.bim\n",
      "PRS_chr7_release6_gender.bim\n",
      "PRS_chr8_release6_gender.bim\n",
      "PRS_chr9_release6_gender.bim\n",
      "Files for Ancestry AMR:\n",
      "PRS_ALL_release6_gender.bim\n",
      "PRS_chr10_release6_gender.bim\n",
      "PRS_chr11_release6_gender.bim\n",
      "PRS_chr12_release6_gender.bim\n",
      "PRS_chr13_release6_gender.bim\n",
      "PRS_chr14_release6_gender.bim\n",
      "PRS_chr15_release6_gender.bim\n",
      "PRS_chr16_release6_gender.bim\n",
      "PRS_chr17_release6_gender.bim\n",
      "PRS_chr18_release6_gender.bim\n",
      "PRS_chr19_release6_gender.bim\n",
      "PRS_chr1_release6_gender.bim\n",
      "PRS_chr20_release6_gender.bim\n",
      "PRS_chr21_release6_gender.bim\n",
      "PRS_chr2_release6_gender.bim\n",
      "PRS_chr3_release6_gender.bim\n",
      "PRS_chr4_release6_gender.bim\n",
      "PRS_chr5_release6_gender.bim\n",
      "PRS_chr6_release6_gender.bim\n",
      "PRS_chr7_release6_gender.bim\n",
      "PRS_chr8_release6_gender.bim\n",
      "PRS_chr9_release6_gender.bim\n",
      "Files for Ancestry EAS:\n",
      "PRS_ALL_release6_gender.bim\n",
      "PRS_chr10_release6_gender.bim\n",
      "PRS_chr11_release6_gender.bim\n",
      "PRS_chr12_release6_gender.bim\n",
      "PRS_chr13_release6_gender.bim\n",
      "PRS_chr14_release6_gender.bim\n",
      "PRS_chr15_release6_gender.bim\n",
      "PRS_chr16_release6_gender.bim\n",
      "PRS_chr17_release6_gender.bim\n",
      "PRS_chr18_release6_gender.bim\n",
      "PRS_chr19_release6_gender.bim\n",
      "PRS_chr1_release6_gender.bim\n",
      "PRS_chr20_release6_gender.bim\n",
      "PRS_chr21_release6_gender.bim\n",
      "PRS_chr2_release6_gender.bim\n",
      "PRS_chr3_release6_gender.bim\n",
      "PRS_chr4_release6_gender.bim\n",
      "PRS_chr5_release6_gender.bim\n",
      "PRS_chr6_release6_gender.bim\n",
      "PRS_chr7_release6_gender.bim\n",
      "PRS_chr8_release6_gender.bim\n",
      "PRS_chr9_release6_gender.bim\n",
      "Files for Ancestry EUR:\n",
      "PRS_ALL_release6_gender.bim\n",
      "PRS_chr10_release6_gender.bim\n",
      "PRS_chr11_release6_gender.bim\n",
      "PRS_chr12_release6_gender.bim\n",
      "PRS_chr13_release6_gender.bim\n",
      "PRS_chr14_release6_gender.bim\n",
      "PRS_chr15_release6_gender.bim\n",
      "PRS_chr16_release6_gender.bim\n",
      "PRS_chr17_release6_gender.bim\n",
      "PRS_chr18_release6_gender.bim\n",
      "PRS_chr19_release6_gender.bim\n",
      "PRS_chr1_release6_gender.bim\n",
      "PRS_chr20_release6_gender.bim\n",
      "PRS_chr21_release6_gender.bim\n",
      "PRS_chr2_release6_gender.bim\n",
      "PRS_chr3_release6_gender.bim\n",
      "PRS_chr4_release6_gender.bim\n",
      "PRS_chr5_release6_gender.bim\n",
      "PRS_chr6_release6_gender.bim\n",
      "PRS_chr7_release6_gender.bim\n",
      "PRS_chr8_release6_gender.bim\n",
      "PRS_chr9_release6_gender.bim\n",
      "Files for Ancestry CAS:\n",
      "PRS_chr10_release6_gender.bim\n",
      "PRS_chr11_release6_gender.bim\n",
      "PRS_chr12_release6_gender.bim\n",
      "PRS_chr13_release6_gender.bim\n",
      "PRS_chr14_release6_gender.bim\n",
      "PRS_chr15_release6_gender.bim\n",
      "PRS_chr16_release6_gender.bim\n",
      "PRS_chr17_release6_gender.bim\n",
      "PRS_chr18_release6_gender.bim\n",
      "PRS_chr19_release6_gender.bim\n",
      "PRS_chr1_release6_gender.bim\n",
      "PRS_chr20_release6_gender.bim\n",
      "PRS_chr21_release6_gender.bim\n",
      "PRS_chr2_release6_gender.bim\n",
      "PRS_chr3_release6_gender.bim\n",
      "PRS_chr4_release6_gender.bim\n",
      "PRS_chr5_release6_gender.bim\n",
      "PRS_chr6_release6_gender.bim\n",
      "PRS_chr7_release6_gender.bim\n",
      "PRS_chr8_release6_gender.bim\n",
      "PRS_chr9_release6_gender.bim\n"
     ]
    }
   ],
   "source": [
    "# List of ancestries\n",
    "ANCESTRIES=(\"AAC\" \"AFR\" \"AJ\" \"AMR\" \"EAS\" \"EUR\" \"CAS\")\n",
    "\n",
    "# Iterate over ancestries\n",
    "for ANCESTRY in \"${ANCESTRIES[@]}\"; do\n",
    "    # Change directory to the specified path\n",
    "    cd ${WORK_DIR}/imputed_data/\"$ANCESTRY\"\n",
    "\n",
    "    # List files matching the pattern and save to temp file\n",
    "    ls *PRS*release6_gender.bim > temp\n",
    "\n",
    "    # Print files for the current ancestry\n",
    "    echo \"Files for Ancestry $ANCESTRY:\"\n",
    "    cat temp\n",
    "\n",
    "    # Go back to the original directory\n",
    "    cd ${WORK_DIR}/imputed_data/\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3854abcb-1df1-4809-8fee-3973ab804e88",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final list of files:\n",
      "PRS_ALL_release6_gender\n",
      "PRS_chr10_release6_gender\n",
      "PRS_chr11_release6_gender\n",
      "PRS_chr12_release6_gender\n",
      "PRS_chr13_release6_gender\n",
      "PRS_chr14_release6_gender\n",
      "PRS_chr15_release6_gender\n",
      "PRS_chr16_release6_gender\n",
      "PRS_chr17_release6_gender\n",
      "PRS_chr18_release6_gender\n",
      "PRS_chr19_release6_gender\n",
      "PRS_chr20_release6_gender\n",
      "PRS_chr21_release6_gender\n",
      "PRS_chr2_release6_gender\n",
      "PRS_chr3_release6_gender\n",
      "PRS_chr4_release6_gender\n",
      "PRS_chr5_release6_gender\n",
      "PRS_chr6_release6_gender\n",
      "PRS_chr7_release6_gender\n",
      "PRS_chr8_release6_gender\n",
      "PRS_chr9_release6_gender\n",
      "Final list of files:\n",
      "PRS_ALL_release6_gender\n",
      "PRS_chr10_release6_gender\n",
      "PRS_chr11_release6_gender\n",
      "PRS_chr12_release6_gender\n",
      "PRS_chr13_release6_gender\n",
      "PRS_chr14_release6_gender\n",
      "PRS_chr15_release6_gender\n",
      "PRS_chr16_release6_gender\n",
      "PRS_chr17_release6_gender\n",
      "PRS_chr18_release6_gender\n",
      "PRS_chr19_release6_gender\n",
      "PRS_chr20_release6_gender\n",
      "PRS_chr21_release6_gender\n",
      "PRS_chr2_release6_gender\n",
      "PRS_chr3_release6_gender\n",
      "PRS_chr4_release6_gender\n",
      "PRS_chr5_release6_gender\n",
      "PRS_chr6_release6_gender\n",
      "PRS_chr7_release6_gender\n",
      "PRS_chr8_release6_gender\n",
      "PRS_chr9_release6_gender\n",
      "Final list of files:\n",
      "PRS_ALL_release6_gender\n",
      "PRS_chr10_release6_gender\n",
      "PRS_chr11_release6_gender\n",
      "PRS_chr12_release6_gender\n",
      "PRS_chr13_release6_gender\n",
      "PRS_chr14_release6_gender\n",
      "PRS_chr15_release6_gender\n",
      "PRS_chr16_release6_gender\n",
      "PRS_chr17_release6_gender\n",
      "PRS_chr18_release6_gender\n",
      "PRS_chr19_release6_gender\n",
      "PRS_chr20_release6_gender\n",
      "PRS_chr21_release6_gender\n",
      "PRS_chr2_release6_gender\n",
      "PRS_chr3_release6_gender\n",
      "PRS_chr4_release6_gender\n",
      "PRS_chr5_release6_gender\n",
      "PRS_chr6_release6_gender\n",
      "PRS_chr7_release6_gender\n",
      "PRS_chr8_release6_gender\n",
      "PRS_chr9_release6_gender\n",
      "Final list of files:\n",
      "PRS_ALL_release6_gender\n",
      "PRS_chr10_release6_gender\n",
      "PRS_chr11_release6_gender\n",
      "PRS_chr12_release6_gender\n",
      "PRS_chr13_release6_gender\n",
      "PRS_chr14_release6_gender\n",
      "PRS_chr15_release6_gender\n",
      "PRS_chr16_release6_gender\n",
      "PRS_chr17_release6_gender\n",
      "PRS_chr18_release6_gender\n",
      "PRS_chr19_release6_gender\n",
      "PRS_chr20_release6_gender\n",
      "PRS_chr21_release6_gender\n",
      "PRS_chr2_release6_gender\n",
      "PRS_chr3_release6_gender\n",
      "PRS_chr4_release6_gender\n",
      "PRS_chr5_release6_gender\n",
      "PRS_chr6_release6_gender\n",
      "PRS_chr7_release6_gender\n",
      "PRS_chr8_release6_gender\n",
      "PRS_chr9_release6_gender\n",
      "Final list of files:\n",
      "PRS_ALL_release6_gender\n",
      "PRS_chr10_release6_gender\n",
      "PRS_chr11_release6_gender\n",
      "PRS_chr12_release6_gender\n",
      "PRS_chr13_release6_gender\n",
      "PRS_chr14_release6_gender\n",
      "PRS_chr15_release6_gender\n",
      "PRS_chr16_release6_gender\n",
      "PRS_chr17_release6_gender\n",
      "PRS_chr18_release6_gender\n",
      "PRS_chr19_release6_gender\n",
      "PRS_chr20_release6_gender\n",
      "PRS_chr21_release6_gender\n",
      "PRS_chr2_release6_gender\n",
      "PRS_chr3_release6_gender\n",
      "PRS_chr4_release6_gender\n",
      "PRS_chr5_release6_gender\n",
      "PRS_chr6_release6_gender\n",
      "PRS_chr7_release6_gender\n",
      "PRS_chr8_release6_gender\n",
      "PRS_chr9_release6_gender\n",
      "Final list of files:\n",
      "PRS_ALL_release6_gender\n",
      "PRS_chr10_release6_gender\n",
      "PRS_chr11_release6_gender\n",
      "PRS_chr12_release6_gender\n",
      "PRS_chr13_release6_gender\n",
      "PRS_chr14_release6_gender\n",
      "PRS_chr15_release6_gender\n",
      "PRS_chr16_release6_gender\n",
      "PRS_chr17_release6_gender\n",
      "PRS_chr18_release6_gender\n",
      "PRS_chr19_release6_gender\n",
      "PRS_chr20_release6_gender\n",
      "PRS_chr21_release6_gender\n",
      "PRS_chr2_release6_gender\n",
      "PRS_chr3_release6_gender\n",
      "PRS_chr4_release6_gender\n",
      "PRS_chr5_release6_gender\n",
      "PRS_chr6_release6_gender\n",
      "PRS_chr7_release6_gender\n",
      "PRS_chr8_release6_gender\n",
      "PRS_chr9_release6_gender\n",
      "Final list of files:\n",
      "PRS_chr10_release6_gender\n",
      "PRS_chr11_release6_gender\n",
      "PRS_chr12_release6_gender\n",
      "PRS_chr13_release6_gender\n",
      "PRS_chr14_release6_gender\n",
      "PRS_chr15_release6_gender\n",
      "PRS_chr16_release6_gender\n",
      "PRS_chr17_release6_gender\n",
      "PRS_chr18_release6_gender\n",
      "PRS_chr19_release6_gender\n",
      "PRS_chr20_release6_gender\n",
      "PRS_chr21_release6_gender\n",
      "PRS_chr2_release6_gender\n",
      "PRS_chr3_release6_gender\n",
      "PRS_chr4_release6_gender\n",
      "PRS_chr5_release6_gender\n",
      "PRS_chr6_release6_gender\n",
      "PRS_chr7_release6_gender\n",
      "PRS_chr8_release6_gender\n",
      "PRS_chr9_release6_gender\n"
     ]
    }
   ],
   "source": [
    "# List of ancestries\n",
    "ANCESTRIES=(\"AAC\" \"AFR\" \"AJ\" \"AMR\" \"EAS\" \"EUR\" \"CAS\")\n",
    "\n",
    "# Iterate over ancestries\n",
    "for ANCESTRY in \"${ANCESTRIES[@]}\"; do\n",
    "\n",
    "    # Remove the \".bim\" extension from the file names and save to temp2\n",
    "    cd ${WORK_DIR}/imputed_data/\"$ANCESTRY\"\n",
    "    sed 's/.bim//g' temp > temp2 || { echo \"Error: Sed command failed for $ANCESTRY\"; exit 1; }\n",
    "    \n",
    "    # Exclude specific file(s) from temp2 and save to files.txt\n",
    "    grep -v \"PRS_chr1_release6_gender\" temp2 > files.txt || { echo \"Error: Grep command failed for $ANCESTRY\"; exit 1; }\n",
    "    echo \"Final list of files:\"\n",
    "    cat files.txt\n",
    "\n",
    "    # Return to the original directory\n",
    "    cd ..\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec1d1675-1561-4f9b-96c4-597a4831aacb",
   "metadata": {},
   "source": [
    "## Merge files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7b1e9336-9d9c-4281-8161-0baad4ad7ed7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[-] Unloading plink  1.9.0-beta4.4  on cn2458 \n",
      "[+] Loading plink  1.9.0-beta4.4  on cn2458 \n"
     ]
    }
   ],
   "source": [
    "module load plink/1.9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e4a972e-55e3-487e-8256-bde101d3b030",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Merge list using chr 1 as your core input\n",
    "\n",
    "# List of ancestries\n",
    "ANCESTRIES=(\"AAC\" \"AFR\" \"AJ\" \"AMR\" \"EAS\" \"EUR\" \"CAS\")\n",
    "\n",
    "# Iterate over ancestries\n",
    "for ANCESTRY in \"${ANCESTRIES[@]}\"; do\n",
    "    # Execute plink command for each ancestry\n",
    "    cd ${WORK_DIR}/imputed_data/\"${ANCESTRY}\"/\n",
    "    plink \\\n",
    "    --bfile PRS_chr1_release6_gender \\\n",
    "    --merge-list \"files.txt\" \\\n",
    "    --make-bed \\\n",
    "    --out \"PRS_ALL_release6_gender\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d2ad6712-56d4-47a5-9272-034bea325895",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reformatted file for Ancestry: AAC\n",
      "1 1:154925709 0 154925709 C G\n",
      "1 1:155162560 0 155162560 A G\n",
      "1 1:155235843 0 155235843 C T\n",
      "1 1:161499264 0 161499264 G C\n",
      "1 1:171750629 0 171750629 T C\n",
      "1 1:205754444 0 205754444 T C\n",
      "1 1:205768611 0 205768611 A G\n",
      "1 1:226728377 0 226728377 C T\n",
      "1 1:232528865 0 232528865 T C\n",
      "2 2:17966582 0 17966582 T A\n",
      "/data/CARD_training/imputed_data\n",
      "Reformatted file for Ancestry: AFR\n",
      "1 1:155162560 0 155162560 A G\n",
      "1 1:155235843 0 155235843 C T\n",
      "1 1:161499264 0 161499264 G C\n",
      "1 1:171750629 0 171750629 T C\n",
      "1 1:205754444 0 205754444 T C\n",
      "1 1:205768611 0 205768611 A G\n",
      "1 1:226728377 0 226728377 C T\n",
      "1 1:232528865 0 232528865 T C\n",
      "2 2:17966582 0 17966582 T A\n",
      "2 2:95335195 0 95335195 A T\n",
      "/data/CARD_training/imputed_data\n",
      "Reformatted file for Ancestry: AJ\n",
      "1 1:154925709 0 154925709 C G\n",
      "1 1:155162560 0 155162560 A G\n",
      "1 1:155235843 0 155235843 C T\n",
      "1 1:161499264 0 161499264 C G\n",
      "1 1:171750629 0 171750629 T C\n",
      "1 1:205754444 0 205754444 C T\n",
      "1 1:205754444 0 205754444 C T\n",
      "1 1:205768611 0 205768611 A G\n",
      "1 1:226728377 0 226728377 C T\n",
      "1 1:232528865 0 232528865 T C\n",
      "/data/CARD_training/imputed_data\n",
      "Reformatted file for Ancestry: AMR\n",
      "1 1:154925709 0 154925709 C G\n",
      "1 1:154925709 0 154925709 C G\n",
      "1 1:155162560 0 155162560 A G\n",
      "1 1:155162560 0 155162560 A G\n",
      "1 1:155235843 0 155235843 C T\n",
      "1 1:155235843 0 155235843 C T\n",
      "1 1:161499264 0 161499264 C G\n",
      "1 1:161499264 0 161499264 C G\n",
      "1 1:171750629 0 171750629 T C\n",
      "1 1:171750629 0 171750629 T C\n",
      "/data/CARD_training/imputed_data\n",
      "Reformatted file for Ancestry: EAS\n",
      "1 1:154925709 0 154925709 C G\n",
      "1 1:161499264 0 161499264 C G\n",
      "1 1:171750629 0 171750629 T C\n",
      "1 1:205754444 0 205754444 C T\n",
      "1 1:205754444 0 205754444 C T\n",
      "1 1:205768611 0 205768611 G A\n",
      "1 1:226728377 0 226728377 C T\n",
      "1 1:232528865 0 232528865 T C\n",
      "2 2:17966582 0 17966582 T A\n",
      "2 2:95335195 0 95335195 A T\n",
      "/data/CARD_training/imputed_data\n",
      "Reformatted file for Ancestry: EUR\n",
      "1 1:154925709 0 154925709 C G\n",
      "1 1:155162560 0 155162560 A G\n",
      "1 1:155235843 0 155235843 C T\n",
      "1 1:161499264 0 161499264 C G\n",
      "1 1:171750629 0 171750629 T C\n",
      "1 1:205754444 0 205754444 T C\n",
      "1 1:205768611 0 205768611 A G\n",
      "1 1:226728377 0 226728377 C T\n",
      "1 1:232528865 0 232528865 T C\n",
      "2 2:17966582 0 17966582 T A\n",
      "/data/CARD_training/imputed_data\n",
      "Reformatted file for Ancestry: CAS\n",
      "1 1:154925709 0 154925709 C G\n",
      "1 1:155162560 0 155162560 A G\n",
      "1 1:155235843 0 155235843 C T\n",
      "1 1:161499264 0 161499264 C G\n",
      "1 1:171750629 0 171750629 T C\n",
      "1 1:205754444 0 205754444 C T\n",
      "1 1:205754444 0 205754444 C T\n",
      "1 1:205768611 0 205768611 A G\n",
      "1 1:226728377 0 226728377 C T\n",
      "1 1:232528865 0 232528865 T C\n",
      "/data/CARD_training/imputed_data\n"
     ]
    }
   ],
   "source": [
    "cd ${WORK_DIR}/imputed_data/\n",
    "\n",
    "# List of ancestries\n",
    "ANCESTRIES=(\"AAC\" \"AFR\" \"AJ\" \"AMR\" \"EAS\" \"EUR\" \"CAS\")\n",
    "\n",
    "# Iterate over ancestries\n",
    "for ANCESTRY in \"${ANCESTRIES[@]}\"; do\n",
    "    # Change directory to the specified ancestry\n",
    "    cd \"$ANCESTRY/\"\n",
    "\n",
    "    # Reformat the bim file and save to temp3\n",
    "    awk '{print $1, $1\":\"$4, $3, $4, $5, $6}' PRS_ALL_release6_gender.bim > temp3\n",
    "\n",
    "    # Display the first few lines of temp3\n",
    "    echo \"Reformatted file for Ancestry: $ANCESTRY\"\n",
    "    head temp3\n",
    "    \n",
    "    # Rename file\n",
    "    mv temp3 PRS_ALL_release6_gender.bim\n",
    "\n",
    "    # Return to the original directory\n",
    "    cd -\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2e590d15-c3fd-4cc5-8636-7761305aa449",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[-] Unloading plink  1.9.0-beta4.4  on cn2458 \n",
      "[+] Loading plink  2.0-alpha \n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) plink/1.9.0-beta4.4 => plink/2.0-alpha\n",
      "\n"
     ]
    }
   ],
   "source": [
    "module load plink/2.0-alpha"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "17eaef07-f1e0-4469-8f72-39954d31839f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.00a3.1LM AVX2 Intel (19 May 2022)     www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2022 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to AAC/PRS_ALL_release6_gender_no_duplicates.log.\n",
      "Options in effect:\n",
      "  --bfile AAC/PRS_ALL_release6_gender\n",
      "  --make-bed\n",
      "  --out AAC/PRS_ALL_release6_gender_no_duplicates\n",
      "  --rm-dup force-first\n",
      "\n",
      "Start time: Mon Mar  4 16:12:51 2024\n",
      "386344 MiB RAM detected; reserving 193172 MiB for main workspace.\n",
      "Using up to 72 threads (change this with --threads).\n",
      "1073 samples (629 females, 444 males; 1073 founders) loaded from\n",
      "AAC/PRS_ALL_release6_gender.fam.\n",
      "104 variants loaded from AAC/PRS_ALL_release6_gender.bim.\n",
      "1 binary phenotype loaded (257 cases, 808 controls).\n",
      "--rm-dup: 14 duplicated IDs, 14 variants removed.\n",
      "Writing AAC/PRS_ALL_release6_gender_no_duplicates.fam ... done.\n",
      "Writing AAC/PRS_ALL_release6_gender_no_duplicates.bim ... done.\n",
      "Writing AAC/PRS_ALL_release6_gender_no_duplicates.bed ... done.\n",
      "End time: Mon Mar  4 16:12:51 2024\n",
      "PLINK v2.00a3.1LM AVX2 Intel (19 May 2022)     www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2022 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to AFR/PRS_ALL_release6_gender_no_duplicates.log.\n",
      "Options in effect:\n",
      "  --bfile AFR/PRS_ALL_release6_gender\n",
      "  --make-bed\n",
      "  --out AFR/PRS_ALL_release6_gender_no_duplicates\n",
      "  --rm-dup force-first\n",
      "\n",
      "Start time: Mon Mar  4 16:12:51 2024\n",
      "386344 MiB RAM detected; reserving 193172 MiB for main workspace.\n",
      "Using up to 72 threads (change this with --threads).\n",
      "2617 samples (1156 females, 1461 males; 2617 founders) loaded from\n",
      "AFR/PRS_ALL_release6_gender.fam.\n",
      "98 variants loaded from AFR/PRS_ALL_release6_gender.bim.\n",
      "1 binary phenotype loaded (920 cases, 1683 controls).\n",
      "--rm-dup: 10 duplicated IDs, 10 variants removed.\n",
      "Writing AFR/PRS_ALL_release6_gender_no_duplicates.fam ... done.\n",
      "Writing AFR/PRS_ALL_release6_gender_no_duplicates.bim ... done.\n",
      "Writing AFR/PRS_ALL_release6_gender_no_duplicates.bed ... done.\n",
      "End time: Mon Mar  4 16:12:51 2024\n",
      "PLINK v2.00a3.1LM AVX2 Intel (19 May 2022)     www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2022 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to AJ/PRS_ALL_release6_gender_no_duplicates.log.\n",
      "Options in effect:\n",
      "  --bfile AJ/PRS_ALL_release6_gender\n",
      "  --make-bed\n",
      "  --out AJ/PRS_ALL_release6_gender_no_duplicates\n",
      "  --rm-dup force-first\n",
      "\n",
      "Start time: Mon Mar  4 16:12:51 2024\n",
      "386344 MiB RAM detected; reserving 193172 MiB for main workspace.\n",
      "Using up to 72 threads (change this with --threads).\n",
      "2242 samples (870 females, 1372 males; 2242 founders) loaded from\n",
      "AJ/PRS_ALL_release6_gender.fam.\n",
      "108 variants loaded from AJ/PRS_ALL_release6_gender.bim.\n",
      "1 binary phenotype loaded (1011 cases, 459 controls).\n",
      "--rm-dup: 18 duplicated IDs, 18 variants removed.\n",
      "Writing AJ/PRS_ALL_release6_gender_no_duplicates.fam ... done.\n",
      "Writing AJ/PRS_ALL_release6_gender_no_duplicates.bim ... done.\n",
      "Writing AJ/PRS_ALL_release6_gender_no_duplicates.bed ... done.\n",
      "End time: Mon Mar  4 16:12:51 2024\n",
      "PLINK v2.00a3.1LM AVX2 Intel (19 May 2022)     www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2022 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to AMR/PRS_ALL_release6_gender_no_duplicates.log.\n",
      "Options in effect:\n",
      "  --bfile AMR/PRS_ALL_release6_gender\n",
      "  --make-bed\n",
      "  --out AMR/PRS_ALL_release6_gender_no_duplicates\n",
      "  --rm-dup force-first\n",
      "\n",
      "Start time: Mon Mar  4 16:12:51 2024\n",
      "386344 MiB RAM detected; reserving 193172 MiB for main workspace.\n",
      "Using up to 72 threads (change this with --threads).\n",
      "529 samples (241 females, 288 males; 529 founders) loaded from\n",
      "AMR/PRS_ALL_release6_gender.fam.\n",
      "193 variants loaded from AMR/PRS_ALL_release6_gender.bim.\n",
      "1 binary phenotype loaded (367 cases, 139 controls).\n",
      "--rm-dup: 90 duplicated IDs, 103 variants removed.\n",
      "Writing AMR/PRS_ALL_release6_gender_no_duplicates.fam ... done.\n",
      "Writing AMR/PRS_ALL_release6_gender_no_duplicates.bim ... done.\n",
      "Writing AMR/PRS_ALL_release6_gender_no_duplicates.bed ... done.\n",
      "End time: Mon Mar  4 16:12:51 2024\n",
      "PLINK v2.00a3.1LM AVX2 Intel (19 May 2022)     www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2022 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to EAS/PRS_ALL_release6_gender_no_duplicates.log.\n",
      "Options in effect:\n",
      "  --bfile EAS/PRS_ALL_release6_gender\n",
      "  --make-bed\n",
      "  --out EAS/PRS_ALL_release6_gender_no_duplicates\n",
      "  --rm-dup force-first\n",
      "\n",
      "Start time: Mon Mar  4 16:12:51 2024\n",
      "386344 MiB RAM detected; reserving 193172 MiB for main workspace.\n",
      "Using up to 72 threads (change this with --threads).\n",
      "3992 samples (1380 females, 2612 males; 3992 founders) loaded from\n",
      "EAS/PRS_ALL_release6_gender.fam.\n",
      "93 variants loaded from EAS/PRS_ALL_release6_gender.bim.\n",
      "1 binary phenotype loaded (1577 cases, 2378 controls).\n",
      "--rm-dup: 8 duplicated IDs, 8 variants removed.\n",
      "Writing EAS/PRS_ALL_release6_gender_no_duplicates.fam ... done.\n",
      "Writing EAS/PRS_ALL_release6_gender_no_duplicates.bim ... done.\n",
      "Writing EAS/PRS_ALL_release6_gender_no_duplicates.bed ... done.\n",
      "End time: Mon Mar  4 16:12:51 2024\n",
      "PLINK v2.00a3.1LM AVX2 Intel (19 May 2022)     www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2022 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to EUR/PRS_ALL_release6_gender_no_duplicates.log.\n",
      "Options in effect:\n",
      "  --bfile EUR/PRS_ALL_release6_gender\n",
      "  --make-bed\n",
      "  --out EUR/PRS_ALL_release6_gender_no_duplicates\n",
      "  --rm-dup force-first\n",
      "\n",
      "Start time: Mon Mar  4 16:12:51 2024\n",
      "386344 MiB RAM detected; reserving 193172 MiB for main workspace.\n",
      "Using up to 72 threads (change this with --threads).\n",
      "20965 samples (8226 females, 12739 males; 20965 founders) loaded from\n",
      "EUR/PRS_ALL_release6_gender.fam.\n",
      "96 variants loaded from EUR/PRS_ALL_release6_gender.bim.\n",
      "1 binary phenotype loaded (11841 cases, 7596 controls).\n",
      "--rm-dup: 6 duplicated IDs, 6 variants removed.\n",
      "Writing EUR/PRS_ALL_release6_gender_no_duplicates.fam ... done.\n",
      "Writing EUR/PRS_ALL_release6_gender_no_duplicates.bim ... done.\n",
      "Writing EUR/PRS_ALL_release6_gender_no_duplicates.bed ... done.\n",
      "End time: Mon Mar  4 16:12:51 2024\n",
      "PLINK v2.00a3.1LM AVX2 Intel (19 May 2022)     www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2022 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to CAS/PRS_ALL_release6_gender_no_duplicates.log.\n",
      "Options in effect:\n",
      "  --bfile CAS/PRS_ALL_release6_gender\n",
      "  --make-bed\n",
      "  --out CAS/PRS_ALL_release6_gender_no_duplicates\n",
      "  --rm-dup force-first\n",
      "\n",
      "Start time: Mon Mar  4 16:12:51 2024\n",
      "386344 MiB RAM detected; reserving 193172 MiB for main workspace.\n",
      "Using up to 72 threads (change this with --threads).\n",
      "585 samples (286 females, 299 males; 585 founders) loaded from\n",
      "CAS/PRS_ALL_release6_gender.fam.\n",
      "104 variants loaded from CAS/PRS_ALL_release6_gender.bim.\n",
      "1 binary phenotype loaded (286 cases, 294 controls).\n",
      "--rm-dup: 14 duplicated IDs, 14 variants removed.\n",
      "Writing CAS/PRS_ALL_release6_gender_no_duplicates.fam ... done.\n",
      "Writing CAS/PRS_ALL_release6_gender_no_duplicates.bim ... done.\n",
      "Writing CAS/PRS_ALL_release6_gender_no_duplicates.bed ... done.\n",
      "End time: Mon Mar  4 16:12:51 2024\n"
     ]
    }
   ],
   "source": [
    "cd ${WORK_DIR}/imputed_data/\n",
    "\n",
    "# List of ancestries\n",
    "ANCESTRIES=(\"AAC\" \"AFR\" \"AJ\" \"AMR\" \"EAS\" \"EUR\" \"CAS\")\n",
    "\n",
    "# Iterate over ancestries\n",
    "for ANCESTRY in \"${ANCESTRIES[@]}\"; do\n",
    "    plink2 \\\n",
    "    --bfile \"$ANCESTRY\"/PRS_ALL_release6_gender \\\n",
    "    --rm-dup force-first \\\n",
    "    --make-bed \\\n",
    "    --out \"$ANCESTRY\"/PRS_ALL_release6_gender_no_duplicates \n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45497114-8498-43d3-9bf3-3f61ffc94cfc",
   "metadata": {},
   "source": [
    "## Estimate profiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c471d63c-457a-43db-af0b-cfff51dc7c77",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[-] Unloading plink  2.0-alpha \n",
      "[+] Loading plink  1.9.0-beta4.4  on cn2458 \n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) plink/2.0-alpha => plink/1.9.0-beta4.4\n",
      "\n"
     ]
    }
   ],
   "source": [
    "module load plink/1.9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "78e32220-86be-4998-8ab9-2c4e5eeecc11",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to AAC/PRS_score_release_AFRICANS.log.\n",
      "Options in effect:\n",
      "  --bfile AAC/PRS_ALL_release6_gender_no_duplicates\n",
      "  --out AAC/PRS_score_release_AFRICANS\n",
      "  --score /data/GP2/projects/PRS_multi_ancestry_Nov2023/90LOCI_AFRICAN_CHR_POS.txt\n",
      "\n",
      "386344 MB RAM detected; reserving 193172 MB for main workspace.\n",
      "90 variants loaded from .bim file.\n",
      "1073 people (444 males, 629 females) loaded from .fam.\n",
      "1065 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1073 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.9895.\n",
      "90 variants and 1073 people pass filters and QC.\n",
      "Among remaining phenotypes, 257 are cases and 808 are controls.  (8 phenotypes\n",
      "are missing.)\n",
      "--score: 84 valid predictors loaded.\n",
      "--score: Results written to AAC/PRS_score_release_AFRICANS.profile .\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to AFR/PRS_score_release_AFRICANS.log.\n",
      "Options in effect:\n",
      "  --bfile AFR/PRS_ALL_release6_gender_no_duplicates\n",
      "  --out AFR/PRS_score_release_AFRICANS\n",
      "  --score /data/GP2/projects/PRS_multi_ancestry_Nov2023/90LOCI_AFRICAN_CHR_POS.txt\n",
      "\n",
      "386344 MB RAM detected; reserving 193172 MB for main workspace.\n",
      "88 variants loaded from .bim file.\n",
      "2617 people (1461 males, 1156 females) loaded from .fam.\n",
      "2603 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 2617 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.979921.\n",
      "88 variants and 2617 people pass filters and QC.\n",
      "Among remaining phenotypes, 920 are cases and 1683 are controls.  (14\n",
      "phenotypes are missing.)\n",
      "Warning: 1 line skipped in --score file (1 due to variant ID mismatch, 0 due to\n",
      "allele code mismatch); see AFR/PRS_score_release_AFRICANS.nopred for details.\n",
      "--score: 83 valid predictors loaded.\n",
      "--score: Results written to AFR/PRS_score_release_AFRICANS.profile .\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to AJ/PRS_score_release_AFRICANS.log.\n",
      "Options in effect:\n",
      "  --bfile AJ/PRS_ALL_release6_gender_no_duplicates\n",
      "  --out AJ/PRS_score_release_AFRICANS\n",
      "  --score /data/GP2/projects/PRS_multi_ancestry_Nov2023/90LOCI_AFRICAN_CHR_POS.txt\n",
      "\n",
      "386344 MB RAM detected; reserving 193172 MB for main workspace.\n",
      "90 variants loaded from .bim file.\n",
      "2242 people (1372 males, 870 females) loaded from .fam.\n",
      "1470 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 2242 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.992085.\n",
      "90 variants and 2242 people pass filters and QC.\n",
      "Among remaining phenotypes, 1011 are cases and 459 are controls.  (772\n",
      "phenotypes are missing.)\n",
      "Warning: 1 line skipped in --score file (0 due to variant ID mismatch, 1 due to\n",
      "allele code mismatch); see AJ/PRS_score_release_AFRICANS.nopred for details.\n",
      "--score: 83 valid predictors loaded.\n",
      "--score: Results written to AJ/PRS_score_release_AFRICANS.profile .\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to AMR/PRS_score_release_AFRICANS.log.\n",
      "Options in effect:\n",
      "  --bfile AMR/PRS_ALL_release6_gender_no_duplicates\n",
      "  --out AMR/PRS_score_release_AFRICANS\n",
      "  --score /data/GP2/projects/PRS_multi_ancestry_Nov2023/90LOCI_AFRICAN_CHR_POS.txt\n",
      "\n",
      "386344 MB RAM detected; reserving 193172 MB for main workspace.\n",
      "90 variants loaded from .bim file.\n",
      "529 people (288 males, 241 females) loaded from .fam.\n",
      "506 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 529 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.99017.\n",
      "90 variants and 529 people pass filters and QC.\n",
      "Among remaining phenotypes, 367 are cases and 139 are controls.  (23 phenotypes\n",
      "are missing.)\n",
      "--score: 84 valid predictors loaded.\n",
      "--score: Results written to AMR/PRS_score_release_AFRICANS.profile .\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to EAS/PRS_score_release_AFRICANS.log.\n",
      "Options in effect:\n",
      "  --bfile EAS/PRS_ALL_release6_gender_no_duplicates\n",
      "  --out EAS/PRS_score_release_AFRICANS\n",
      "  --score /data/GP2/projects/PRS_multi_ancestry_Nov2023/90LOCI_AFRICAN_CHR_POS.txt\n",
      "\n",
      "386344 MB RAM detected; reserving 193172 MB for main workspace.\n",
      "85 variants loaded from .bim file.\n",
      "3992 people (2612 males, 1380 females) loaded from .fam.\n",
      "3955 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 3992 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.931831.\n",
      "85 variants and 3992 people pass filters and QC.\n",
      "Among remaining phenotypes, 1577 are cases and 2378 are controls.  (37\n",
      "phenotypes are missing.)\n",
      "Warning: 4 lines skipped in --score file (3 due to variant ID mismatch, 1 due\n",
      "to allele code mismatch); see EAS/PRS_score_release_AFRICANS.nopred for\n",
      "details.\n",
      "--score: 80 valid predictors loaded.\n",
      "--score: Results written to EAS/PRS_score_release_AFRICANS.profile .\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to EUR/PRS_score_release_AFRICANS.log.\n",
      "Options in effect:\n",
      "  --bfile EUR/PRS_ALL_release6_gender_no_duplicates\n",
      "  --out EUR/PRS_score_release_AFRICANS\n",
      "  --score /data/GP2/projects/PRS_multi_ancestry_Nov2023/90LOCI_AFRICAN_CHR_POS.txt\n",
      "\n",
      "386344 MB RAM detected; reserving 193172 MB for main workspace.\n",
      "90 variants loaded from .bim file.\n",
      "20965 people (12739 males, 8226 females) loaded from .fam.\n",
      "19437 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 20965 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.953846.\n",
      "90 variants and 20965 people pass filters and QC.\n",
      "Among remaining phenotypes, 11841 are cases and 7596 are controls.  (1528\n",
      "phenotypes are missing.)\n",
      "--score: 84 valid predictors loaded.\n",
      "--score: Results written to EUR/PRS_score_release_AFRICANS.profile .\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to CAS/PRS_score_release_AFRICANS.log.\n",
      "Options in effect:\n",
      "  --bfile CAS/PRS_ALL_release6_gender_no_duplicates\n",
      "  --out CAS/PRS_score_release_AFRICANS\n",
      "  --score /data/GP2/projects/PRS_multi_ancestry_Nov2023/90LOCI_AFRICAN_CHR_POS.txt\n",
      "\n",
      "386344 MB RAM detected; reserving 193172 MB for main workspace.\n",
      "90 variants loaded from .bim file.\n",
      "585 people (299 males, 286 females) loaded from .fam.\n",
      "580 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 585 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.972175.\n",
      "90 variants and 585 people pass filters and QC.\n",
      "Among remaining phenotypes, 286 are cases and 294 are controls.  (5 phenotypes\n",
      "are missing.)\n",
      "--score: 84 valid predictors loaded.\n",
      "--score: Results written to CAS/PRS_score_release_AFRICANS.profile .\n"
     ]
    }
   ],
   "source": [
    "# List of ancestries\n",
    "ANCESTRIES=(\"AAC\" \"AFR\" \"AJ\" \"AMR\" \"EAS\" \"EUR\" \"CAS\")\n",
    "\n",
    "# Iterate over ancestries\n",
    "for ANCESTRY in \"${ANCESTRIES[@]}\"; do\n",
    "    # Change directory to the specified ancestry\n",
    " \n",
    "    # Calculate scores using plink\n",
    "    plink \\\n",
    "    --bfile \"$ANCESTRY/\"PRS_ALL_release6_gender_no_duplicates \\\n",
    "    --score ${WORK_DIR}/90LOCI_AFRICAN_CHR_POS.txt \\\n",
    "    --out \"$ANCESTRY/\"PRS_score_release_AFRICANS\n",
    "\n",
    "    # Return to the original directory\n",
    "    cd /data/CARD_training/imputed_data/\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fb0a228-bc0e-455c-ae8f-44dbc5432725",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# List of ancestries\n",
    "ANCESTRIES=(\"AAC\" \"AFR\" \"AJ\" \"AMR\" \"EAS\" \"EUR\" \"CAS\")\n",
    "\n",
    "# Iterate over ancestries\n",
    "for ANCESTRY in \"${ANCESTRIES[@]}\"; do\n",
    "    plink \\\n",
    "    --bfile \"$ANCESTRY/\"PRS_ALL_release6_gender_no_duplicates \\\n",
    "    --score ${WORK_DIR}/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt \\\n",
    "    --out \"$ANCESTRY/\"PRS_score_release_EASTASIANS\n",
    "\n",
    "    # Return to the original directory\n",
    "    cd -\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7f6f969-36cc-494f-9ff0-84e913cb88cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of ancestries\n",
    "ANCESTRIES=(\"AAC\" \"AFR\" \"AJ\" \"AMR\" \"EAS\" \"EUR\" \"CAS\")\n",
    "\n",
    "# Iterate over ancestries\n",
    "for ANCESTRY in \"${ANCESTRIES[@]}\"; do\n",
    "    plink \\\n",
    "    --bfile \"$ANCESTRY/\"PRS_ALL_release6_gender_no_duplicates \\\n",
    "    --score ${WORK_DIR}/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt \\\n",
    "    --out \"$ANCESTRY/\"PRS_score_release_LATINO\n",
    "\n",
    "    # Return to the original directory\n",
    "    cd /data/CARD_training/imputed_data\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2270512c-8f3c-44ee-be5d-ba8f6e1dd3d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of ancestries\n",
    "ANCESTRIES=(\"AAC\" \"AFR\" \"AJ\" \"AMR\" \"EAS\" \"EUR\" \"CAS\")\n",
    "\n",
    "# Iterate over ancestries\n",
    "for ANCESTRY in \"${ANCESTRIES[@]}\"; do\n",
    "    plink \\\n",
    "    --bfile \"$ANCESTRY/\"PRS_ALL_release6_gender_no_duplicates \\\n",
    "    --score ${WORK_DIR}/90LOCI_EUROPEANS_CHR_POS.txt \\\n",
    "    --out \"$ANCESTRY/\"PRS_score_release_EUROPEAN\n",
    "\n",
    "    # Return to the original directory\n",
    "    cd /data/CARD_training/imputed_data\n",
    "done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
