{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "61d37999-7b98-482e-b001-d168eabbda09",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# Multi-ancestry PRS\n",
    "- **Version:** Python/3.9\n",
    "- **Status:** COMPLETE\n",
    "- **Last Updated:** 24-FEB-2025\n",
    "\n",
    "## Notebook Overview\n",
    "- Getting Started\n",
    "    - Import python dependencies\n",
    "    - Define helper functions\n",
    "    - Print out versions of imported python dependencies\n",
    "    - Load R dependencies\n",
    "    - Define directories and ancestry lists\n",
    "    - Install bioinformatics packages\n",
    "- Process plink files\n",
    "    - For Model 1\n",
    "    - For Model 2\n",
    "    - Get number of unrelated participants per ancestry\n",
    "- Process Clinical Data\n",
    "    - Generate covariates with PCs\n",
    "    - Admixture\n",
    "- Generate PCA plot\n",
    "- Model 1\n",
    "    - Generate Upset plot\n",
    "    - Meta-analyze base data\n",
    "    - Estimate PRS profiles\n",
    "    - PRS vs. PD risk\n",
    "    - Compare PCs vs Admixture\n",
    "    - Plot ROC\n",
    "    - Estimate Specificity and Sensitivity\n",
    "- Model 2\n",
    "    - PRSice\n",
    "    - Heatmap comparing p-value thresholds\n",
    "    - Estimate Specificity and Sensitivity\n",
    "    - Plot ROC\n",
    "- Compare Model 1 and Model 2\n",
    "    - Compare AUCs between Model 1 and Model 2\n",
    "    - Density plot\n",
    "    - Forest plot\n",
    "    - Estimate SNPs contributing most to PRS\n",
    "- Foo et al. EAS Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1dd417a8-91f9-4c59-a5ba-3dba4cc4550a",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Getting Started"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1be1f696-0727-4e66-b93c-3dfaf9a4e1e7",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Import python dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "df2c7b1f-7f29-4359-b4b2-3f4a82ac3ad0",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: seaborn in /opt/conda/lib/python3.10/site-packages (0.13.2)\n",
      "Requirement already satisfied: numpy!=1.24.0,>=1.20 in /opt/conda/lib/python3.10/site-packages (from seaborn) (1.25.2)\n",
      "Requirement already satisfied: pandas>=1.2 in /home/jupyter/.local/lib/python3.10/site-packages (from seaborn) (2.2.3)\n",
      "Requirement already satisfied: matplotlib!=3.6.1,>=3.4 in /home/jupyter/.local/lib/python3.10/site-packages (from seaborn) (3.10.0)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /opt/conda/lib/python3.10/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (1.3.1)\n",
      "Requirement already satisfied: cycler>=0.10 in /opt/conda/lib/python3.10/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /opt/conda/lib/python3.10/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (4.55.0)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /opt/conda/lib/python3.10/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (1.4.7)\n",
      "Requirement already satisfied: packaging>=20.0 in /opt/conda/lib/python3.10/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (24.1)\n",
      "Requirement already satisfied: pillow>=8 in /opt/conda/lib/python3.10/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (11.0.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /opt/conda/lib/python3.10/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (3.2.0)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /home/jupyter/.local/lib/python3.10/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (2.9.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/jupyter/.local/lib/python3.10/site-packages (from pandas>=1.2->seaborn) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /opt/conda/lib/python3.10/site-packages (from pandas>=1.2->seaborn) (2024.2)\n",
      "Requirement already satisfied: six>=1.5 in /opt/conda/lib/python3.10/site-packages (from python-dateutil>=2.7->matplotlib!=3.6.1,>=3.4->seaborn) (1.16.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: kaleido in /opt/conda/lib/python3.10/site-packages (0.2.1)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "## Import the necessary python dependencies \n",
    "%pip install seaborn --upgrade\n",
    "%load_ext rpy2.ipython\n",
    "%pip install -U kaleido\n",
    "\n",
    "from datetime import date\n",
    "import importlib.metadata\n",
    "from IPython.display import display\n",
    "import math\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import rcParams\n",
    "from matplotlib.collections import PatchCollection\n",
    "from matplotlib.colors import LinearSegmentedColormap, ListedColormap, TwoSlopeNorm\n",
    "from matplotlib.ticker import FormatStrFormatter\n",
    "import matplotlib.gridspec as gridspec\n",
    "import numbers\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "import requests\n",
    "import scipy\n",
    "from scipy import stats\n",
    "from scipy.stats import norm\n",
    "from sklearn.metrics import roc_curve, auc\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import subprocess\n",
    "import sys\n",
    "import seaborn as sns\n",
    "import types\n",
    "\n",
    "today = date.today()\n",
    "date = today.strftime(\"%d-%b-%Y\").upper()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e74c10b4-0a40-419d-ab84-5a4cc01bc63e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Define helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "80012d8b-46af-4873-ba47-cf46a8a66fc4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_imports():\n",
    "    for name, val in globals().items():\n",
    "        if isinstance(val, types.ModuleType):\n",
    "            # Split ensures you get root package, not just imported function\n",
    "            name = val.__name__.split(\".\")[0]\n",
    "\n",
    "        elif isinstance(val, type):\n",
    "            name = val.__module__.split(\".\")[0]\n",
    "\n",
    "        # Some packages are weird and have different imported names vs. system/pip names\n",
    "        # Unfortunately, there is no systematic way to get pip names from a package's imported name. You'll have to add exceptions to this list manually!\n",
    "        poorly_named_packages = {\n",
    "            \"PIL\": \"Pillow\",\n",
    "            \"sklearn\": \"scikit-learn\"\n",
    "        }\n",
    "        if name in poorly_named_packages.keys():\n",
    "            name = poorly_named_packages[name]\n",
    "\n",
    "        yield name\n",
    "\n",
    "def min_max_scale(data):\n",
    "    return (data - np.min(data)) / (np.max(data) - np.min(data))\n",
    "\n",
    "def compare_rocs(input_path, output_path, auc1_variable, auc2_variable, xlabel):\n",
    "    # Read p-values matrix and rescale values\n",
    "    df_pvals = pd.read_csv(input_path, sep=\"\\t\", index_col=\"Ancestry\")\n",
    "    df_pvals_log = np.log(-np.log(np.abs(df_pvals)) + 1) * df_pvals / np.abs(df_pvals)\n",
    "    \n",
    "    # Prepare formatting for heatmap\n",
    "    max_abs = np.max(np.abs(df_pvals_log))\n",
    "    norm = TwoSlopeNorm(vmin=-max_abs, vcenter=0, vmax=max_abs)\n",
    "    annot = df_pvals.map(lambda x: \"*\" if -0.05 <= x <= 0.05 else \"\")\n",
    "    \n",
    "    # Generate heatmap\n",
    "    fig, ax = plt.subplots(figsize=(8, 6))\n",
    "    sns.heatmap(df_pvals_log, cmap=\"RdBu\", norm=norm, cbar=True, ax=ax, annot=annot, fmt=\"\")\n",
    "    \n",
    "    # Format color bar\n",
    "    cbar = ax.collections[0].colorbar\n",
    "    cbar.set_label(\"\")\n",
    "    cbar.set_ticks([-1.38522686, 1.38522686])\n",
    "    cbar.set_ticklabels([f\"{auc1_variable} Significantly Better\", f\"{auc2_variable} Significantly Better\"])\n",
    "    \n",
    "    # Label axes and save figure\n",
    "    plt.xlabel(xlabel)\n",
    "    plt.ylabel(\"Target data ancestry\")\n",
    "    plt.savefig(output_path, dpi=300, bbox_inches='tight')\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea65f7f1-fc95-4817-9a45-70a06afa8237",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Print out versions of imported python dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "64ea1ffb-49f6-4dc6-a959-c3741b57775e",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PACKAGE VERSIONS (22-FEB-2025)\n",
      "\tmatplotlib==3.10.0\n",
      "\tpandas==2.2.3\n",
      "\tstatsmodels==0.14.4\n",
      "\tmatplotlib==3.7.3\n",
      "\tscipy==1.11.4\n",
      "\tpandas==2.0.3\n",
      "\tseaborn==0.13.2\n",
      "\trequests==2.32.3\n",
      "\tplotly==5.24.1\n",
      "\tscikit-learn==1.5.2\n",
      "\tnumpy==1.25.2\n"
     ]
    }
   ],
   "source": [
    "imports = list(set(get_imports()))\n",
    "print(f\"PACKAGE VERSIONS ({date})\")\n",
    "for m in importlib.metadata.distributions():\n",
    "    if m.metadata[\"Name\"] in imports and m.metadata[\"Name\"]!=\"pip\":\n",
    "        print(f\"\\t{m.metadata['Name']}=={m.version}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2b8e809-c6e5-4fa5-b84f-95f44c75a591",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load R dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ceec2934-56c1-4a22-85e9-3dd3d712e25c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* installing *source* package ‘caret’ ...\n",
      "** package ‘caret’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** libs\n",
      "using C compiler: ‘gcc (Debian 10.2.1-6) 10.2.1 20210110’\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gcc -I\"/usr/share/R/include\" -DNDEBUG       -fpic  -g -O2 -ffile-prefix-map=/home/jranke/git/r-backports/bullseye/r-base-4.4.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c caret.c -o caret.o\n",
      "gcc -shared -L/usr/lib/R/lib -Wl,-z,relro -o caret.so caret.o -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/jupyter/.R/library/00LOCK-caret/00new/caret/libs\n",
      "** R\n",
      "** data\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (caret)\n",
      "* installing *source* package ‘optparse’ ...\n",
      "** package ‘optparse’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** R\n",
      "** exec\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "*** copying figures\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (optparse)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Installing package into ‘/home/jupyter/.R/library’\n",
       "(as ‘lib’ is unspecified)\n",
       "trying URL 'https://cloud.r-project.org/src/contrib/caret_7.0-1.tar.gz'\n",
       "Content type 'application/x-gzip' length 2273919 bytes (2.2 MB)\n",
       "==================================================\n",
       "downloaded 2.2 MB\n",
       "\n",
       "\n",
       "The downloaded source packages are in\n",
       "\t‘/var/tmp/Rtmp07bfHH/downloaded_packages’\n",
       "Installing package into ‘/home/jupyter/.R/library’\n",
       "(as ‘lib’ is unspecified)\n",
       "trying URL 'https://cloud.r-project.org/src/contrib/optparse_1.7.5.tar.gz'\n",
       "Content type 'application/x-gzip' length 41938 bytes (40 KB)\n",
       "==================================================\n",
       "downloaded 40 KB\n",
       "\n",
       "\n",
       "The downloaded source packages are in\n",
       "\t‘/var/tmp/Rtmp07bfHH/downloaded_packages’\n",
       "In addition: Warning message:\n",
       "In (function (package, help, pos = 2, lib.loc = NULL, character.only = FALSE,  :\n",
       "  library ‘/usr/lib/R/site-library’ contains no packages\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "install.packages(\"caret\")\n",
    "install.packages(\"optparse\", repos=\"https://cloud.r-project.org/\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3948821f-5f42-4f91-b143-848dc120a27f",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Loading required package: data.table\n",
       "data.table 1.16.2 using 8 threads (see ?getDTthreads).  Latest news: r-datatable.com\n",
       "Loading required package: dplyr\n",
       "\n",
       "Attaching package: ‘dplyr’\n",
       "\n",
       "The following objects are masked from ‘package:data.table’:\n",
       "\n",
       "    between, first, last\n",
       "\n",
       "The following objects are masked from ‘package:stats’:\n",
       "\n",
       "    filter, lag\n",
       "\n",
       "The following objects are masked from ‘package:base’:\n",
       "\n",
       "    intersect, setdiff, setequal, union\n",
       "\n",
       "Loading required package: ggplot2\n",
       "Loading required package: lattice\n",
       "Type 'citation(\"pROC\")' for a citation.\n",
       "\n",
       "Attaching package: ‘pROC’\n",
       "\n",
       "The following objects are masked from ‘package:stats’:\n",
       "\n",
       "    cov, smooth, var\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "require(data.table)\n",
    "require(dplyr)\n",
    "require(ggplot2)\n",
    "\n",
    "library(optparse)\n",
    "library(data.table)\n",
    "library(\"ggplot2\")\n",
    "library(RColorBrewer)\n",
    "library(\"caret\")\n",
    "library(\"pROC\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec8e754b-fc0e-48fb-95fc-3a1d8d13387c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Define directories and ancestry lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "feb8fd14-25d8-48d0-a9f6-7382b4c0fc4d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "WORK_DIR = \"/home/jupyter\"\n",
    "RESULTS_DIR = \"/home/jupyter/workspace/ws_files/results\"\n",
    "REL9_DIR = \"/home/jupyter/workspace/gp2/release/path\"\n",
    "DATA_DIR = \"/home/jupyter/workspace/ws_files/data\"\n",
    "\n",
    "ancestries = [\"AAC\", \"AFR\", \"AJ\", \"AMR\", \"EAS\", \"EUR\", \"CAS\"]\n",
    "summstat_ancestries = [\"AAC\", \"EAS\", \"EUR\", \"AMR\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3eb03bf-52db-4cc4-91be-94295bd11616",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "WORK_DIR <- \"/home/jupyter\"\n",
    "RESULTS_DIR <- \"/home/jupyter/workspace/ws_files/results\"\n",
    "REL9_DIR <- \"/home/jupyter/workspace/gp2/release/path\"\n",
    "DATA_DIR <- \"/home/jupyter/workspace/ws_files/data\"\n",
    "\n",
    "ancestries <- list(\"AAC\", \"AFR\", \"AJ\", \"AMR\", \"EAS\", \"EUR\", \"CAS\")\n",
    "summstat_ancestries <- list(\"AAC\",\"EAS\",\"EUR\",\"AMR\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "89ea9988-f660-4e9a-b4ed-215f0e341eae",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "if not os.path.exists(RESULTS_DIR):\n",
    "    ! mkdir {RESULTS_DIR}\n",
    "    for ancestry in ancestries:\n",
    "        ! mkdir {RESULTS_DIR}/{ancestry}\n",
    "        ! mkdir {RESULTS_DIR}/{ancestry}/model1_plink\n",
    "        ! mkdir {RESULTS_DIR}/{ancestry}/model2_plink\n",
    "        \n",
    "    ! mkdir {RESULTS_DIR}/EAS_Foo\n",
    "    ! mkdir {RESULTS_DIR}/meta/\n",
    "    ! mkdir {RESULTS_DIR}/scores/\n",
    "    ! mkdir {RESULTS_DIR}/density/\n",
    "    ! mkdir {RESULTS_DIR}/roc/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "119eaa12-9437-49be-8a17-9e4342ac6905",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Install bioinformatics packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "55366d45-3b45-4635-9f17-222e161b63fd",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Plink is already installed in /home/jupyter\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "if test -e /home/jupyter/plink; then\n",
    "    echo \"Plink is already installed in /home/jupyter\"\n",
    "else\n",
    "    echo \"Plink is not installed\"\n",
    "    wget -P /home/jupyter http://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20190304.zip \n",
    "    unzip -o /home/jupyter/plink_linux_x86_64_20190304.zip -d /home/jupyter\n",
    "    rm /home/jupyter/plink_linux_x86_64_20190304.zip\n",
    "fi\n",
    "\n",
    "chmod u+x /home/jupyter/plink\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "59debc40-2947-4fc5-b901-c8e46b7bd183",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Plink2 is already installed in /home/jupyter\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "if test -e /home/jupyter/plink2; then\n",
    "    echo \"Plink2 is already installed in /home/jupyter\"\n",
    "else\n",
    "    echo \"Plink2 is not installed\"\n",
    "    wget -P /home/jupyter http://s3.amazonaws.com/plink2-assets/plink2_linux_x86_64_latest.zip\n",
    "    unzip -o /home/jupyter/plink2_linux_x86_64_latest.zip -d /home/jupyter\n",
    "    rm /home/jupyter/plink2_linux_x86_64_latest.zip\n",
    "fi\n",
    "\n",
    "chmod u+x /home/jupyter/plink2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7a8db135-41cb-44d7-867e-4f3559f1f6ec",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Metal is already installed in /home/jupyter\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "if test -e /home/jupyter/metal; then\n",
    "    echo \"Metal is already installed in /home/jupyter\"\n",
    "else\n",
    "    echo \"Metal is not installed\"\n",
    "    wget -P /home/jupyter https://csg.sph.umich.edu/abecasis/metal/download/Linux-metal.tar.gz\n",
    "    tar --strip-components=1 -xzf /home/jupyter/Linux-metal.tar.gz -C /home/jupyter\n",
    "    rm /home/jupyter/Linux-metal.tar.gz\n",
    "fi\n",
    "\n",
    "chmod u+x /home/jupyter/metal\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5d58ee83-c43c-4cc2-b615-2f2dde3f1c8f",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PRSice is already installed in /home/jupyter\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "if test -e /home/jupyter/prsice; then\n",
    "    echo \"PRSice is already installed in /home/jupyter\"\n",
    "else\n",
    "    echo \"PRSice is not installed\"\n",
    "    wget -P /home/jupyter https://github.com/choishingwan/PRSice/releases/download/2.3.5/PRSice_linux.zip\n",
    "    unzip -o /home/jupyter/PRSice_linux.zip -d /home/jupyter\n",
    "    mv /home/jupyter/PRSice_linux /home/jupyter/prsice\n",
    "fi\n",
    "\n",
    "chmod u+x /home/jupyter/prsice\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b9888f8-2595-47c5-ba4a-bff1d40ca85f",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Process Plink Files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05a45247-d6e4-4370-99c8-7b5f92d67271",
   "metadata": {
    "tags": []
   },
   "source": [
    "## For Model 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60cc1fd0-92f4-4cbf-a66d-0c633c82e84a",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Find related individuals to be removed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f79e527-7b1e-4abb-80d9-91144535c90c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ancestry in ancestries:\n",
    "    df_related = pd.read_csv(f\"{REL9_DIR}/meta_data/related_samples/{ancestry}_release9_vwb.related\")\n",
    "    related_ids = df_related[\"IID2\"].str.replace(r\"_s[12345]$\", \"\", regex=True).drop_duplicates()\n",
    "    related_ids.to_csv(f\"{RESULTS_DIR}/{ancestry}/related.txt\", index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "988daea7-23c1-48aa-a846-7029b755f3af",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Extract sex data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7935fe43-a888-4574-b75c-aec07bd2a966",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_masterkey = pd.read_csv(f\"{REL9_DIR}/clinical_data/master_key_release9_final_vwb.csv\")\n",
    "df_sex = df_masterkey[['GP2ID', 'biological_sex_for_qc']].copy()\n",
    "df_sex['biological_sex_for_qc'] = df_sex['biological_sex_for_qc'].replace({\n",
    "    'Female':2, \n",
    "    'Male':1, \n",
    "    'Other/Unknown/Not Reported':0,\n",
    "})\n",
    "df_sex.insert(0, 'New_Column', 0)\n",
    "df_sex.to_csv(f'{RESULTS_DIR}/sex.txt', index=False, header=False, sep = ' ')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5976f07c-7f47-4158-9f4c-9fcbd3ab10d1",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Create b-files with 90 risk variants extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "040295b9-f86d-49aa-9c5f-e807c496a8d8",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for ANCESTRY in {\"AAC\",\"AFR\",\"AJ\",\"AMR\",\"EAS\",\"EUR\",\"CAS\"}; do\n",
    "    for CHRNUM in {1..22}; do\n",
    "        /home/jupyter/plink2 \\\n",
    "        --pfile \"/home/jupyter/workspace/gp2/release/path/imputed_genotypes/${ANCESTRY}/chr${CHRNUM}_${ANCESTRY}_release9_vwb\" \\\n",
    "        --remove \"/home/jupyter/workspace/ws_files/results/${ANCESTRY}/related.txt\" \\\n",
    "        --update-sex \"/home/jupyter/workspace/ws_files/results/sex.txt\" \\\n",
    "        --extract range \"/home/jupyter/workspace/ws_files/data/CHR_POS_HG38.txt\"  \\\n",
    "        --make-bed \\\n",
    "        --out \"/home/jupyter/workspace/ws_files/results/${ANCESTRY}/model1_plink/chr${CHRNUM}_${ANCESTRY}\"\n",
    "    done\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45e71cc3-96b6-4ae3-ab79-ec8395cdc0d9",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Save names of files for each chromosome that have been generated for each ancestry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80f42562-eb5b-41f4-b6f3-8f8868514225",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for ANCESTRY in {\"AAC\",\"AFR\",\"AJ\",\"AMR\",\"EAS\",\"EUR\",\"CAS\"}; do\n",
    "    > /home/jupyter/workspace/ws_files/results/${ANCESTRY}/model1_plink/files.txt\n",
    "    for CHRNUM in {1..22}; do\n",
    "        BED_FILE=\"/home/jupyter/workspace/ws_files/results/${ANCESTRY}/model1_plink/chr${CHRNUM}_${ANCESTRY}.bed\"\n",
    "        if [[ -f \"${BED_FILE}\" ]]; then\n",
    "            echo \"/home/jupyter/workspace/ws_files/results/${ANCESTRY}/model1_plink/chr${CHRNUM}_${ANCESTRY}\" >> /home/jupyter/workspace/ws_files/results/${ANCESTRY}/model1_plink/files.txt\n",
    "        fi\n",
    "    done\n",
    "\n",
    "    echo \"Final list of files for ${ANCESTRY}:\"\n",
    "    cat /home/jupyter/workspace/ws_files/results/${ANCESTRY}/model1_plink/files.txt \n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a828f21-e067-42d3-853b-0d27a31d9e5d",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Merge chromosomes and remove duplicate IDs at the same position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d72d0a06-6e5e-476b-87ec-0ebd03bc4334",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for ANCESTRY in {\"AAC\",\"AFR\",\"AJ\",\"AMR\",\"EAS\",\"EUR\",\"CAS\"}; do\n",
    "    /home/jupyter/plink \\\n",
    "    --merge-list \"/home/jupyter/workspace/ws_files/results/${ANCESTRY}/model1_plink/files.txt\" \\\n",
    "    --make-bed \\\n",
    "    --out \"/home/jupyter/workspace/ws_files/results/${ANCESTRY}/model1_plink/merged_${ANCESTRY}_orig\"\n",
    "\n",
    "    /home/jupyter/plink2 \\\n",
    "    --bfile \"/home/jupyter/workspace/ws_files/results/${ANCESTRY}/model1_plink/merged_${ANCESTRY}_orig\" \\\n",
    "    --rm-dup force-first \\\n",
    "    --set-all-var-ids @:# \\\n",
    "    --make-bed \\\n",
    "    --out \"/home/jupyter/workspace/ws_files/results/${ANCESTRY}/model1_plink/merged_${ANCESTRY}\"\n",
    "    \n",
    "    rm /home/jupyter/workspace/ws_files/results/${ANCESTRY}/model1_plink/merged_${ANCESTRY}_orig*\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67c3663d-36b4-4012-9e1b-2f9bb8f04659",
   "metadata": {
    "tags": []
   },
   "source": [
    "## For Model 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cf942ae-9633-40d8-a413-599f9d3aa707",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for ancestry in {\"AAC\",\"AFR\",\"AJ\",\"AMR\",\"EAS\",\"EUR\",\"CAS\"}; do\n",
    "    for chrnum in {1..22}; do\n",
    "        /home/jupyter/plink2 \\\n",
    "        --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/${ancestry}/chr${chrnum}_${ancestry}_release9_vwb \\\n",
    "        --remove /home/jupyter/workspace/ws_files/results/${ancestry}/related.txt \\\n",
    "        --make-bed \\\n",
    "        --out /home/jupyter/workspace/ws_files/results/${ancestry}/model2_plink/chr${chrnum}\n",
    "    done\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "968873d3-4a54-4e99-b9d1-e9408584e00f",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Get number of unrelated participants per ancestry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1c0ead6b-5d05-441f-aa18-767dd09eff38",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'AAC': 1144, 'AFR': 2606, 'AJ': 1819, 'AMR': 3358, 'EAS': 4168, 'EUR': 31799, 'CAS': 905}\n"
     ]
    }
   ],
   "source": [
    "n_per_ancestry = {}\n",
    "for ancestry in ancestries:\n",
    "    df_participants = pd.read_csv(f\"{RESULTS_DIR}/{ancestry}/model1_plink/merged_{ancestry}.fam\", sep=\"\\t\", header=None)\n",
    "    df_participants = df_participants[df_participants[5].isin([1, 2])]\n",
    "    n_per_ancestry[ancestry] = len(df_participants)\n",
    "print(n_per_ancestry)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ac11c6c-462b-4542-938f-00decbdd0eba",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Process Clinical Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54b619b1-790c-4ffc-aecf-b1dd76a3412b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Generate covariates with PCs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed0e6ba7-2c76-467c-a6b4-0cc1738c56d6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load and merge master key\n",
    "df_master = pd.read_csv(f\"{REL9_DIR}/clinical_data/master_key_release9_final_vwb.csv\")\n",
    "df_master = df_master[[\"GP2ID\",\"age_at_sample_collection\",\"age_of_onset\",\"biological_sex_for_qc\",\"baseline_GP2_phenotype_for_qc\"]]\n",
    "df_master.reset_index(drop=True, inplace=True)\n",
    "display(df_master)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6b44275-126b-43b8-ab41-143b0e5a2ed2",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Reformat sex column\n",
    "df_sex = df_master[['GP2ID','biological_sex_for_qc']].copy()\n",
    "df_sex['biological_sex_for_qc'] = df_sex['biological_sex_for_qc'].replace({\n",
    "    'Female':2,\n",
    "    'Male':1,\n",
    "    'Other/Unknown/Not Reported':0,\n",
    "})\n",
    "df_sex.rename({\"biological_sex_for_qc\":\"SEX\"}, inplace=True, axis=1)\n",
    "df_sex.insert(0, 'New_Column', 0)\n",
    "df_sex.reset_index(drop=True, inplace=True)\n",
    "display(df_sex)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d3360e8-6c16-4b52-ba05-16d4fec1e55f",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "pcs = []\n",
    "for ancestry in ancestries:\n",
    "    pcs.append(pd.read_csv(f\"{REL9_DIR}/raw_genotypes/{ancestry}/{ancestry}_release9_vwb.eigenvec\", sep=\"\\t\"))\n",
    "df_pcs = pd.concat(pcs)\n",
    "df_pcs.drop(\"#FID\", axis=1, inplace=True)\n",
    "df_pcs.rename({\"IID\":\"GP2ID\"}, axis=1, inplace=True)\n",
    "display(df_pcs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bd31ac0-a2fb-405d-83d0-ee4484a43566",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Print demographic data for each ancestry\n",
    "df_master_ages = df_master[[\"GP2ID\",\"age_at_sample_collection\",\"age_of_onset\"]].copy()\n",
    "\n",
    "df_ages_merged = []\n",
    "\n",
    "for ancestry in ancestries:\n",
    "    df_participants = pd.read_csv(f\"{RESULTS_DIR}/{ancestry}/model1_plink/merged_{ancestry}.fam\", sep=\"\\t\", header=None)\n",
    "    df_participants.rename(columns={1:\"GP2ID\", 4:\"SEX\", 5:\"PHENO\"}, inplace=True)\n",
    "    df_participants = df_participants[[\"GP2ID\", \"SEX\", \"PHENO\"]]\n",
    "    df_participants = df_participants.merge(df_master_ages, on=\"GP2ID\")\n",
    "    df_participants[\"AGE\"] = df_participants.apply(\n",
    "        lambda row: row[\"age_of_onset\"] if row[\"PHENO\"] == 2 else row[\"age_at_sample_collection\"], axis=1\n",
    "    )\n",
    "    df_participants[\"ANCESTRY\"] = ancestry\n",
    "    df_ages_merged.append(df_participants[[\"GP2ID\",\"AGE\",\"PHENO\",\"ANCESTRY\"]].copy())\n",
    "    \n",
    "    df_participants = df_participants[[\"GP2ID\",\"SEX\",\"PHENO\",\"AGE\"]]\n",
    "    df_participants = df_participants[df_participants[\"PHENO\"].isin([1, 2])]\n",
    "    df_controls = df_participants[df_participants[\"PHENO\"] == 1].copy()\n",
    "    df_cases = df_participants[df_participants[\"PHENO\"] == 2].copy()\n",
    "    \n",
    "    print(ancestry)\n",
    "    print(f'Cases: {df_participants[\"PHENO\"].value_counts()[2]}')\n",
    "    print(f'Controls: {df_participants[\"PHENO\"].value_counts()[1]}')\n",
    "    print(f'Male: {df_participants[\"SEX\"].value_counts()[1]}')\n",
    "    print(f'Female: {df_participants[\"SEX\"].value_counts()[2]}')\n",
    "    print(f'Average Age Controls: {df_controls[\"AGE\"].mean():.2f} +/- {df_controls[\"AGE\"].std():.2f}')\n",
    "    print(f'Average AAO Cases:    {df_cases[\"AGE\"].mean():.2f} +/- {df_cases[\"AGE\"].std():.2f}')\n",
    "    print()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8b76745-40bf-4741-8569-8005f6bebaa7",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_ages_merged = pd.concat(df_ages_merged)\n",
    "df_ages_merged.reset_index(drop=True, inplace=True)\n",
    "\n",
    "df_ages_merged_cases = df_ages_merged[df_ages_merged[\"PHENO\"] == 2].copy()\n",
    "df_ages_merged_cases.fillna({\"AGE\":df_ages_merged_cases[\"AGE\"].mean()}, inplace=True)\n",
    "df_ages_merged_controls = df_ages_merged[df_ages_merged[\"PHENO\"] != 2].copy()\n",
    "df_ages_merged_controls.fillna({\"AGE\":df_ages_merged_controls[\"AGE\"].mean()}, inplace=True)\n",
    "df_ages_merged = pd.concat([df_ages_merged_cases, df_ages_merged_controls])\n",
    "df_ages_merged = df_ages_merged[[\"GP2ID\",\"AGE\",\"ANCESTRY\"]]\n",
    "\n",
    "display(df_ages_merged)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45afa818-9f53-437e-a2fc-3232705bd91c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_pheno = df_master[[\"GP2ID\", \"baseline_GP2_phenotype_for_qc\"]].copy()\n",
    "df_pheno.rename({\"baseline_GP2_phenotype_for_qc\":\"PHENO\"}, inplace=True, axis=1)\n",
    "df_pheno[\"PHENO\"] = df_pheno[\"PHENO\"].replace({\n",
    "    \"PD\": 2,\n",
    "    \"Control\": 1,\n",
    "    \"Other\": -9,\n",
    "})\n",
    "display(df_pheno)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2a67ced-aa70-482b-a239-0c6b4268fbc8",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_covar = df_sex.merge(df_ages_merged, on=\"GP2ID\")\n",
    "df_covar = df_covar.merge(df_pheno, on=\"GP2ID\")\n",
    "df_covar = df_covar.merge(df_pcs, on=\"GP2ID\")\n",
    "df_covar.to_csv(f'{RESULTS_DIR}/covariates_pheno.txt', index=False, sep = ' ')\n",
    "df_covar.drop(\"PHENO\", axis=1, inplace=True)\n",
    "df_covar.to_csv(f'{RESULTS_DIR}/covariates.txt', index=False, sep = ' ')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e335856-24b6-4903-9e3f-773ce10376ae",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Admixture"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8656154f-cb40-484d-b5c9-c5548624a23b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load covariate data\n",
    "df_covar = pd.read_csv(f\"{RESULTS_DIR}/covariates_pheno.txt\", sep=\" \")\n",
    "df_covar.columns = [\"FID\", \"IID\", \"SEX\", \"AGE\", \"ANCESTRY\", \"PHENO\", \"PC1\", \"PC2\", \"PC3\", \"PC4\", \"PC5\", \"PC6\", \"PC7\", \"PC8\", \"PC9\", \"PC10\"]\n",
    "\n",
    "# Load admixture data and find IDs that do not correspond with covariate data\n",
    "df_admixture = pd.read_csv(f\"{REL9_DIR}/meta_data/admixture/GP2_release9_labeled_admixture_vwb.txt\", sep=\"\\t\")\n",
    "df_admixture[\"IID\"] = df_admixture[\"IID\"].str.replace(r\"_s.*$\", \"\", regex=True)\n",
    "df_admixture = df_admixture.drop_duplicates(subset=\"IID\")\n",
    "ids_excluded = df_covar[\"IID\"][~df_covar[\"IID\"].isin(df_admixture[\"IID\"])]\n",
    "\n",
    "# Find a mapping for inconsistent IDs\n",
    "df_master = pd.read_csv(f\"{REL9_DIR}/clinical_data/master_key_release9_final_vwb.csv\")\n",
    "df_master = df_master[[\"GP2ID\", \"nba_GP2sampleID_r7\"]]\n",
    "df_master[\"nba_GP2sampleID_r7\"] = df_master[\"nba_GP2sampleID_r7\"].str.replace(r\"_s.*$\", \"\", regex=True)\n",
    "df_master = df_master[df_master[\"GP2ID\"].isin(ids_excluded)]\n",
    "mapping_dict = df_master.set_index(\"nba_GP2sampleID_r7\")[\"GP2ID\"].to_dict()\n",
    "\n",
    "# Apply the mapping to the admixture data\n",
    "df_admixture[\"IID\"] = df_admixture[\"IID\"].map(mapping_dict).fillna(df_admixture[\"IID\"])\n",
    "df_admixture.to_csv(f\"{RESULTS_DIR}/admixture.txt\", index=False, sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e57a21e2-9911-46f3-b822-73c1fb3c2e11",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Generate PCA plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85f5f8ae-38d7-4673-8868-f95568f1e86d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_covar = pd.read_csv(f\"{RESULTS_DIR}/covariates_pheno.txt\", sep=\"\\s+\")\n",
    "df_covar = df_covar[df_covar[\"PHENO\"].isin([1, 2])]\n",
    "df_covar.rename(columns={\"GP2ID\":\"IID\"}, inplace=True)\n",
    "df_covar = df_covar[[\"IID\",\"ANCESTRY\"]]\n",
    "df_covar.replace({\n",
    "    \"AFR\":\"African\",\n",
    "    \"AAC\":\"African Admixed\",\n",
    "    \"EUR\":\"European\",\n",
    "    \"AMR\":\"Latino / Admixed American\",\n",
    "    \"AJ\":\"Ashkenazi Jewish\",\n",
    "    \"CAS\":\"Central Asian\",\n",
    "    \"EAS\":\"East Asian\",\n",
    "}, inplace=True)\n",
    "\n",
    "df_ids_mapping = pd.read_csv(f\"{DATA_DIR}/logging_updatedIDchanges.txt\", sep=\"\\t\")\n",
    "df_ids_mapping.rename(columns={\"old_iid\":\"IID\", \"new_iid\":\"IID_new\"}, inplace=True)\n",
    "df_pcs = pd.read_csv(f\"{REL9_DIR}/meta_data/qc_metrics/projected_pcs_vwb.csv\")\n",
    "df_pcs = df_pcs[[\"IID\",\"PC1\",\"PC2\",\"PC3\"]]\n",
    "df_pcs = df_pcs.merge(df_ids_mapping, on=\"IID\")\n",
    "df_pcs = df_pcs[[\"IID_new\",\"PC1\",\"PC2\",\"PC3\"]]\n",
    "df_pcs.rename(columns={\"IID_new\":\"IID\"}, inplace=True)\n",
    "df_pcs = df_pcs.drop_duplicates(subset=\"IID\")\n",
    "\n",
    "df_covar = df_covar.merge(df_pcs, on=\"IID\")\n",
    "display(df_covar)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff3e8f5b-b68b-4606-af64-8ed001a85a4d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "fig = px.scatter_3d(\n",
    "    df_covar,\n",
    "    x=\"PC1\",\n",
    "    y=\"PC2\",\n",
    "    z=\"PC3\",\n",
    "    color=\"ANCESTRY\",\n",
    "    color_discrete_sequence=px.colors.qualitative.Bold,\n",
    ")\n",
    "\n",
    "fig.update_traces(marker={'size': 3})\n",
    "fig.update_layout(\n",
    "    legend=dict(\n",
    "        font=dict(size=16),\n",
    "        itemsizing=\"constant\",\n",
    "        tracegroupgap=5,\n",
    "    ),\n",
    ")\n",
    "fig.show()\n",
    "fig.write_image(f'{RESULTS_DIR}/supp_fig1.png', width=1980, height=1080)\n",
    "fig.write_html(f'{RESULTS_DIR}/supp_fig1.html')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71214fa7-7d0d-4b75-b9c6-223a88f4e30e",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Model 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6adc1d3-3500-4264-987a-d51fc5bf2916",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Generate Upset plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d9eb852-defc-4c23-8337-658850b34440",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Generate association summary stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eeaaa3d9-055c-499b-a97a-520a4bda31ef",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for ANCESTRY in {\"AAC\",\"AFR\",\"AJ\",\"AMR\",\"EAS\",\"EUR\",\"CAS\"}; do\n",
    "    /home/jupyter/plink \\\n",
    "    --bfile \"/home/jupyter/workspace/ws_files/results/${ANCESTRY}/model1_plink/merged_${ANCESTRY}\" \\\n",
    "    --assoc \\\n",
    "    --out \"/home/jupyter/workspace/ws_files/results/${ANCESTRY}/model1\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7211f021-d5fe-4e29-b575-96e541208539",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Merge summary stats across ancestries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "504b904a-4d8a-471c-954f-9143efcee4ff",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_merged = None\n",
    "for ancestry in [\"AAC\",\"AFR\",\"AMR\",\"CAS\",\"EAS\",\"EUR\",\"AJ\"]:\n",
    "    df = pd.read_csv(f\"{RESULTS_DIR}/{ancestry}/model1.assoc\", sep=\"\\s+\")\n",
    "    df[\"BETA\"] = np.log(df[\"OR\"] + 0.01)\n",
    "    df = df[[\"SNP\",\"A1\",\"BETA\",\"P\"]]\n",
    "    df.rename(columns={\"A1\":f\"A1_{ancestry}\", \"BETA\":f\"BETA_{ancestry}\", \"P\":f\"P_{ancestry}\"}, inplace=True)\n",
    "    if df_merged is None:\n",
    "        df_merged = df\n",
    "    else:\n",
    "        df_merged = df_merged.merge(df, on=\"SNP\", how=\"outer\")\n",
    "    \n",
    "df_merged[[\"CHR\", \"POS\"]] = df_merged[\"SNP\"].str.split(\":\", expand=True)\n",
    "df_merged[\"POS\"] = df_merged[\"POS\"].astype(int)\n",
    "df_merged[\"CHR\"] = df_merged[\"CHR\"].astype(int)\n",
    "df_merged.sort_values([\"CHR\",\"POS\"], inplace=True)\n",
    "df_merged.drop([\"CHR\",\"POS\"], axis=1, inplace=True)\n",
    "df_merged.reset_index(drop=True, inplace=True)\n",
    "df_merged.to_csv(f\"{RESULTS_DIR}/supp_tab3.txt\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b7d4287-94ed-456a-a0e0-725feb3da4d5",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Flip signs where A1 is inconsistent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c111d2a-dbf1-43e2-84eb-4c10e79f5f39",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_merged_adjusted = df_merged.copy()\n",
    "df_merged_adjusted[\"A1\"] = np.nan\n",
    "for ancestry in ancestries:\n",
    "    df_merged_adjusted[\"A1\"] = df_merged_adjusted[\"A1\"].fillna(df_merged_adjusted[f\"A1_{ancestry}\"])\n",
    "    \n",
    "for ancestry in ancestries:\n",
    "    mask = df_merged_adjusted[f\"A1_{ancestry}\"] != df_merged_adjusted[\"A1\"]\n",
    "    df_merged_adjusted.loc[mask, f\"BETA_{ancestry}\"] *= -1\n",
    "df_merged_adjusted.drop(\"A1\", axis=1, inplace=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c11c1560-4339-44d1-982f-e4835189659d",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Melt dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5ff48ca-5f50-4db2-af88-87ab53c74bbf",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ancestry in ancestries:\n",
    "    df_merged_adjusted.rename(columns={f\"BETA_{ancestry}\":ancestry}, inplace=True)\n",
    "    df_merged_adjusted.drop([f\"A1_{ancestry}\", f\"P_{ancestry}\"], axis=1, inplace=True)\n",
    "    \n",
    "df_upset_merged = df_merged_adjusted.melt(id_vars=[\"SNP\"], var_name=\"ANC\", value_name=\"BETA\")[[\"ANC\",\"SNP\",\"BETA\"]]\n",
    "df_upset_merged['CHR'] = df_upset_merged['SNP'].str.extract(r'^(\\d+)').astype(int)\n",
    "df_upset_merged['BP'] = df_upset_merged['SNP'].str.extract(r'(\\d+)$').astype(int)\n",
    "df_upset_merged = df_upset_merged.sort_values(by=['CHR','BP','ANC'])\n",
    "df_upset_merged.drop([\"CHR\", \"BP\"], axis=1, inplace=True)\n",
    "df_upset_merged = df_upset_merged.reset_index(drop=True)\n",
    "display(df_upset_merged)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9256121d-a27d-4001-a7f0-f0755375ce73",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Generate Upset plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcafb640-fc6a-4f7e-988d-c5724f099b7a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "pal=[\"#6B1414\",\"#C14E4E\",\"#CF9FFF\",\"#1E90FF\", \"#000000\"]\n",
    "\n",
    "# Want purple to be right at beta=0\n",
    "quantiles = df_upset_merged[\"BETA\"][df_upset_merged[\"BETA\"] < 0].describe().loc[['min','50%']].tolist()\n",
    "quantiles += [0]\n",
    "quantiles += df_upset_merged[\"BETA\"][df_upset_merged[\"BETA\"] > 0].describe().loc[['50%', 'max']].tolist()\n",
    "quantiles_norm = min_max_scale(quantiles)\n",
    "cmp = LinearSegmentedColormap.from_list(\"\", list(zip(quantiles_norm, pal)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08bee8e2-6164-4b92-b5ed-a20957d2e3b9",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(\n",
    "    figsize=(30,3),\n",
    "    sharex='col'\n",
    ")\n",
    "fig.clf()\n",
    "gs = gridspec.GridSpec(1, 2, width_ratios=(30,0.7))\n",
    "\n",
    "ax1 = fig.add_subplot(gs[0, 0])\n",
    "ax2 = fig.add_subplot(gs[0, 1])\n",
    "\n",
    "sns.set_style({'font_scale': 2, 'axes.facecolor':'white', 'grid.color': '.8'})\n",
    "sns.set_theme(style=\"white\", palette=None)\n",
    "plt.margins(x=0.01, y=0.05)\n",
    "plt.subplots_adjust(wspace=0, hspace=0)\n",
    "plot_bottom = sns.scatterplot(\n",
    "    df_upset_merged,\n",
    "    x='SNP',\n",
    "    y='ANC',\n",
    "    hue='BETA',\n",
    "    s=200,\n",
    "    edgecolor='black',\n",
    "    palette=cmp,\n",
    "    legend='brief',\n",
    "    ax=ax1,\n",
    ")\n",
    "plot_bottom.spines[\"top\"].set_visible(False) \n",
    "plot_bottom.spines[\"bottom\"].set_visible(False) \n",
    "plot_bottom.spines[\"left\"].set_visible(False) \n",
    "plot_bottom.spines[\"right\"].set_visible(False) \n",
    "plot_bottom.set_xticklabels(df_upset_merged.drop_duplicates(subset='SNP')['SNP'], rotation=45, ha='right')\n",
    "plot_bottom.get_legend().remove()\n",
    "plot_bottom.axhspan(ymin=0.5, ymax=1.5, color='steelblue', alpha=0.1, lw=0)\n",
    "plot_bottom.axhspan(ymin=2.5, ymax=3.5, color='steelblue', alpha=0.1, lw=0)\n",
    "plot_bottom.axhspan(ymin=4.5, ymax=5.5, color='steelblue', alpha=0.1, lw=0)\n",
    "plot_bottom.set_ylabel('')\n",
    "\n",
    "sm = plt.cm.ScalarMappable(cmap=cmp)\n",
    "sm.set_array([])\n",
    "\n",
    "cb_ticks=[df_upset_merged['BETA'].min().round(2), 0, df_upset_merged['BETA'].max().round(2)]\n",
    "cb = plt.colorbar(\n",
    "    sm,\n",
    "    cax=ax2,\n",
    "    ticks=min_max_scale(cb_ticks),\n",
    ")\n",
    "cb.ax.set_yticklabels(cb_ticks) \n",
    "cb.outline.set_color('black')\n",
    "cb.outline.set_linewidth(1)\n",
    "cb.ax.tick_params(labelsize=10, pad=15, width=0)\n",
    "cb.ax.set_title('Beta', size=15, pad=10)\n",
    "\n",
    "fig.savefig(f'{RESULTS_DIR}/figure2.png',facecolor='white', dpi=300, bbox_inches = \"tight\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d57f895-245b-4fd8-9480-0d931793f8c0",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Meta-analyze Base Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4140d89f-561a-4a8d-8dd2-8be669362f22",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Generate metaGWAS - summary statistics for EAS (Foo + 23andMe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71477444-94b5-46f4-872f-e4473925e130",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "## Extract 90 SNPs from Foo et al. \n",
    "sumstats <- fread(paste0(DATA_DIR, \"/6724PDcases-24851controls-5843213snps-summary-stats-metaP-SE.txt\"), header =T)\n",
    "SNPs <- fread(paste0(DATA_DIR, \"/90riskloci.txt\"), header =T)\n",
    "total <- merge(sumstats, SNPs, by=\"BP\")\n",
    "output <- total[,c(\"SNP.y\",\"BETA\", \"SE\", \"P\", \"A1\", \"A2\")]\n",
    "write.table(output, file = paste0(RESULTS_DIR, \"/scores/Foo_90riskloci.txt\"), quote = F, row.names = F, sep = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e7604de-ffa8-45ec-b5d2-9e54e5c0bc53",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "## Extract 90 SNPs from 23andMe EAS\n",
    "sumstats <- fread(paste0(DATA_DIR, \"/filtered_sumstats_23andme_EASTASIAN_PD.hg19.txt\"), header =T)\n",
    "SNPs <- fread(paste0(DATA_DIR, \"/90riskloci.txt\"), header =T)\n",
    "names(SNPs)[3] <- \"position\"\n",
    "total <- merge(sumstats, SNPs, by=\"position\")\n",
    "output <- total[,c(\"SNP\",\"effect\", \"stderr\", \"pvalue\", \"effect_allele\", \"alt_allele\")]\n",
    "write.table(output, file = paste0(RESULTS_DIR, \"/scores/23andMe_EASTASIAN_90riskloci.txt\"), quote = F, row.names = F, sep = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "461c66bb-dc3d-41af-9516-7e17e1e70f05",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "## Now meta-analyze\n",
    "/home/jupyter/metal\n",
    "SCHEME STDERR\n",
    "GENOMICCONTROL ON\n",
    "\n",
    "# === DESCRIBE AND PROCESS THE FIRST INPUT FILE ===\n",
    "MARKER SNP.y\n",
    "ALLELE A1 A2\n",
    "EFFECT BETA\n",
    "STDERR SE\n",
    "PVALUE P\n",
    "PROCESS /home/jupyter/workspace/ws_files/results/scores/Foo_90riskloci.txt\n",
    "\n",
    "# === DESCRIBE AND PROCESS THE SECOND INPUT FILE ===\n",
    "MARKER SNP\n",
    "ALLELE effect_allele alt_allele\n",
    "EFFECT effect\n",
    "STDERR stderr\n",
    "PVALUE pvalue\n",
    "PROCESS /home/jupyter/workspace/ws_files/results/scores/23andMe_EASTASIAN_90riskloci.txt\n",
    "\n",
    "OUTFILE /home/jupyter/workspace/ws_files/results/meta/ASIAN .tbl\n",
    "ANALYZE HETEROGENEITY\n",
    "QUIT\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dea3ebe0-18e6-4444-8c2b-21a4fe69ed22",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "## Remove SNPs only present in one of the datasets\n",
    "grep -v \"?\" /home/jupyter/workspace/ws_files/results/meta/ASIAN1.tbl > /home/jupyter/workspace/ws_files/results/temp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c997a20-06aa-44e0-93dd-fc6e41a14607",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "## Now convert all to hg38\n",
    "sumstats <- fread(paste0(RESULTS_DIR, \"/temp\"), header =T)\n",
    "SNPs <- fread(paste0(DATA_DIR, \"/90riskloci_38.txt\"), header =T)\n",
    "names(SNPs)[1] <- \"MarkerName\"\n",
    "total <- merge(sumstats, SNPs, by=\"MarkerName\")\n",
    "total$A1cap <- toupper(total$Allele1)\n",
    "output <- total[,c(\"markerID\",\"A1cap\", \"Effect\")]\n",
    "write.table(output, file = paste0(RESULTS_DIR, \"/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\"), quote = F, col.names = FALSE, row.names = F, sep = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82586ae3-793c-476e-b364-a54ec0046cb0",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Generate metaGWAS - summary statistics for AMR (Loesch + 23andMe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "baf929e6-ca78-4e52-9ca5-b516a4dc027b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "## Extract 90 SNPs from Loesch et al. \n",
    "sumstats <- fread(paste0(DATA_DIR, \"/Loesch_et_al_2021_Latam_no23andme_hg38.txt\"), header =T)\n",
    "SNPs <- fread(paste0(DATA_DIR, \"/90riskloci_38.txt\"), header =T)\n",
    "names(sumstats)[4] <- \"BP\"\n",
    "total <- merge(sumstats, SNPs, by=\"BP\")\n",
    "total$SE <- total$beta/total$Score\n",
    "output <- total[,c(\"markerID\", \"beta\", \"SE\", \"Score.pval\", \"ref\", \"alt\")]\n",
    "write.table(output, file = paste0(RESULTS_DIR, \"/scores/Loesch_90riskloci.txt\"), quote = F, row.names = F, sep = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d54b3b26-b2f2-41f2-bab1-9feaebea74cd",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "## Extract 90 SNPs from 23andMe LATINO\n",
    "sumstats <- fread(paste0(DATA_DIR, \"/filtered_sumstats_23andme_LATINO_PD.txt\"), header =T)\n",
    "SNPs <- fread(paste0(DATA_DIR, \"/90riskloci_38.txt\"), header =T)\n",
    "names(SNPs)[3] <- \"position\"\n",
    "total <- merge(sumstats, SNPs, by=\"position\")\n",
    "output <- total[,c(\"markerID\", \"effect\", \"stderr\", \"pvalue\", \"effect_allele\", \"alt_allele\")]\n",
    "write.table(output, file = paste0(RESULTS_DIR, \"/scores/23andMe_LATINO_90riskloci.txt\"), quote = F, row.names = F, sep = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6879957-3e6f-4f17-84da-c67d35549c60",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "## Now meta-analyze\n",
    "/home/jupyter/metal\n",
    "SCHEME STDERR\n",
    "GENOMICCONTROL ON\n",
    "\n",
    "# === DESCRIBE AND PROCESS THE FIRST INPUT FILE ===\n",
    "MARKER markerID\n",
    "ALLELE ref alt\n",
    "EFFECT beta\n",
    "STDERR SE\n",
    "PVALUE Score.pval\n",
    "PROCESS /home/jupyter/workspace/ws_files/results/scores/Loesch_90riskloci.txt\n",
    "\n",
    "# === DESCRIBE AND PROCESS THE SECOND INPUT FILE ===\n",
    "MARKER markerID\n",
    "ALLELE effect_allele alt_allele\n",
    "EFFECT effect\n",
    "STDERR stderr\n",
    "PVALUE pvalue\n",
    "PROCESS /home/jupyter/workspace/ws_files/results/scores/23andMe_LATINO_90riskloci.txt\n",
    "\n",
    "OUTFILE /home/jupyter/workspace/ws_files/results/meta/LATINO .tbl\n",
    "ANALYZE HETEROGENEITY\n",
    "\n",
    "QUIT\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee2fc8a5-8dda-439e-952a-17c5513567c5",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "## Remove SNPs only present in one of the datasets\n",
    "grep -v \"?\" /home/jupyter/workspace/ws_files/results/meta/LATINO1.tbl > /home/jupyter/workspace/ws_files/results/temp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92e74fb5-d426-4e6f-b216-9389c714fa3b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "## Now extract columns for score file\n",
    "sumstats <- fread(paste0(RESULTS_DIR, \"/temp\"), header =T)\n",
    "sumstats$A1cap <- toupper(sumstats$Allele1)\n",
    "output <- sumstats[,c(\"MarkerName\",\"A1cap\", \"Effect\")]\n",
    "write.table(output, file = paste0(RESULTS_DIR, \"/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\"), quote = F, col.names = FALSE, row.names = F, sep = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49b27e43-574f-4cf5-aebe-e7367cef1178",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "! rm {RESULTS_DIR}/temp\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ffb96cc-3095-4267-9e4d-0104e2aaaf41",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Estimate PRS profiles"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ca406da-90e5-433e-926e-d56996bd1eb8",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Estimate profiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78ba9c0c-c7f3-4c9c-8103-cd9dd6138c72",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for ancestry in {\"AAC\",\"AFR\",\"AJ\",\"AMR\",\"EAS\",\"EUR\",\"CAS\"}; do\n",
    "    /home/jupyter/plink \\\n",
    "    --bfile \"/home/jupyter/workspace/ws_files/results/${ancestry}/model1_plink/merged_${ancestry}\" \\\n",
    "    --score \"/home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\" \\\n",
    "    --out \"/home/jupyter/workspace/ws_files/results/${ancestry}/model1_AAC\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a1caab2-d499-4e42-b33f-0674033281f6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for ancestry in {\"AAC\",\"AFR\",\"AJ\",\"AMR\",\"EAS\",\"EUR\",\"CAS\"}; do\n",
    "    /home/jupyter/plink \\\n",
    "    --bfile \"/home/jupyter/workspace/ws_files/results/${ancestry}/model1_plink/merged_${ancestry}\" \\\n",
    "    --score \"/home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\" \\\n",
    "    --out \"/home/jupyter/workspace/ws_files/results/${ancestry}/model1_EAS\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c5e8bfb-13d1-4700-a2f4-82843f648d4d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for ancestry in {\"AAC\",\"AFR\",\"AJ\",\"AMR\",\"EAS\",\"EUR\",\"CAS\"}; do\n",
    "    /home/jupyter/plink \\\n",
    "    --bfile \"/home/jupyter/workspace/ws_files/results/${ancestry}/model1_plink/merged_${ancestry}\" \\\n",
    "    --score \"/home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\" \\\n",
    "    --out \"/home/jupyter/workspace/ws_files/results/${ancestry}/model1_AMR\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6ea6577-e93a-4584-a032-4021051a66a1",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for ancestry in {\"AAC\",\"AFR\",\"AJ\",\"AMR\",\"EAS\",\"EUR\",\"CAS\"}; do\n",
    "    /home/jupyter/plink \\\n",
    "    --bfile \"/home/jupyter/workspace/ws_files/results/${ancestry}/model1_plink/merged_${ancestry}\" \\\n",
    "    --score \"/home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\" \\\n",
    "    --out \"/home/jupyter/workspace/ws_files/results/${ancestry}/model1_EUR\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a1e4981-6b79-48c0-96c0-774dabbfefbf",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Valid predictors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c5a055f-1633-4490-a0b1-d21178417666",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "ANCESTRIES=(\"EUR\" \"AAC\" \"AMR\" \"EAS\" \"AFR\" \"AJ\" \"CAS\")\n",
    "SUMMSTAT_ANCESTRIES=(\"EUR\" \"AAC\" \"AMR\" \"EAS\")\n",
    "OUTPUT_FILE=\"/home/jupyter/workspace/ws_files/results/supp_tab2.txt\"\n",
    "\n",
    "echo -e \"Cohort\\tSummary Statistics\\tN\" > $OUTPUT_FILE\n",
    "\n",
    "for ANCESTRY in \"${ANCESTRIES[@]}\"; do\n",
    "    for SUMMSTAT_ANCESTRY in \"${SUMMSTAT_ANCESTRIES[@]}\"; do\n",
    "        COUNT=$(grep \"valid predictors\" /home/jupyter/workspace/ws_files/results/${ANCESTRY}/model1_${SUMMSTAT_ANCESTRY}.log | awk '{print $2}')\n",
    "        echo -e \"${ANCESTRY}\\t${SUMMSTAT_ANCESTRY}\\t${COUNT}\" >> ${OUTPUT_FILE}\n",
    "    done\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07cd134e-2913-48e0-b99a-c0aff98fb3cb",
   "metadata": {
    "tags": []
   },
   "source": [
    "## PRS vs PD Risk"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aaa0bd5-9b9e-4360-a5c6-2ab462d7182e",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Adjust by PCs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "be8aeda0-e7a7-4190-a310-1e87c2802848",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"------------------------------AAC------------------------------\"\n",
      "[1] \"AAC Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   2.083442   0.476518   4.372 1.23e-05 ***\n",
      "zSCORE        0.199900   0.071590   2.792  0.00523 ** \n",
      "SEX          -0.653243   0.146144  -4.470 7.83e-06 ***\n",
      "AGE          -0.034019   0.006715  -5.066 4.06e-07 ***\n",
      "PC1         -17.334728   2.565804  -6.756 1.42e-11 ***\n",
      "PC2         -11.951764   2.638407  -4.530 5.90e-06 ***\n",
      "PC3          -0.331229   4.123361  -0.080  0.93597    \n",
      "PC4           0.314532   3.591190   0.088  0.93021    \n",
      "PC5          -7.017291   4.057954  -1.729  0.08376 .  \n",
      "PC6          -1.950201   3.850930  -0.506  0.61256    \n",
      "PC7           6.421741   4.055667   1.583  0.11333    \n",
      "PC8          -0.928432   2.963519  -0.313  0.75406    \n",
      "PC9         -13.708800   2.713357  -5.052 4.36e-07 ***\n",
      "PC10         -1.562683   2.651578  -0.589  0.55563    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1365.4  on 1143  degrees of freedom\n",
      "Residual deviance: 1150.2  on 1130  degrees of freedom\n",
      "AIC: 1178.2\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EAS Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   2.092103   0.479744   4.361 1.30e-05 ***\n",
      "zSCORE        0.306921   0.075080   4.088 4.35e-05 ***\n",
      "SEX          -0.660102   0.146872  -4.494 6.98e-06 ***\n",
      "AGE          -0.034473   0.006739  -5.115 3.13e-07 ***\n",
      "PC1         -16.248218   2.595447  -6.260 3.84e-10 ***\n",
      "PC2         -11.263040   2.645304  -4.258 2.06e-05 ***\n",
      "PC3          -0.478621   3.917392  -0.122    0.903    \n",
      "PC4           0.075077   3.793585   0.020    0.984    \n",
      "PC5          -6.930185   3.985131  -1.739    0.082 .  \n",
      "PC6          -2.375593   3.885678  -0.611    0.541    \n",
      "PC7           7.326048   4.460388   1.642    0.100    \n",
      "PC8          -0.684292   2.972353  -0.230    0.818    \n",
      "PC9         -13.956481   2.739025  -5.095 3.48e-07 ***\n",
      "PC10         -1.240321   2.669096  -0.465    0.642    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1365.4  on 1143  degrees of freedom\n",
      "Residual deviance: 1140.9  on 1130  degrees of freedom\n",
      "AIC: 1168.9\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EUR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   2.018026   0.483339   4.175 2.98e-05 ***\n",
      "zSCORE        0.438817   0.074292   5.907 3.49e-09 ***\n",
      "SEX          -0.648203   0.148260  -4.372 1.23e-05 ***\n",
      "AGE          -0.034350   0.006771  -5.073 3.92e-07 ***\n",
      "PC1         -17.201221   2.637273  -6.522 6.92e-11 ***\n",
      "PC2          -9.897233   2.714179  -3.646 0.000266 ***\n",
      "PC3          -0.526299   4.076542  -0.129 0.897275    \n",
      "PC4           1.055718   3.697108   0.286 0.775221    \n",
      "PC5          -6.264974   4.227291  -1.482 0.138332    \n",
      "PC6          -1.821154   4.073924  -0.447 0.654856    \n",
      "PC7           7.107434   4.641454   1.531 0.125697    \n",
      "PC8          -1.488016   3.026042  -0.492 0.622905    \n",
      "PC9         -13.566973   2.784924  -4.872 1.11e-06 ***\n",
      "PC10         -1.350951   2.673172  -0.505 0.613296    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1365.4  on 1143  degrees of freedom\n",
      "Residual deviance: 1121.6  on 1130  degrees of freedom\n",
      "AIC: 1149.6\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"AMR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   2.086055   0.479475   4.351 1.36e-05 ***\n",
      "zSCORE        0.192475   0.072909   2.640  0.00829 ** \n",
      "SEX          -0.643349   0.146094  -4.404 1.06e-05 ***\n",
      "AGE          -0.034107   0.006738  -5.062 4.15e-07 ***\n",
      "PC1         -17.665136   2.567777  -6.880 6.00e-12 ***\n",
      "PC2         -12.971671   2.608615  -4.973 6.61e-07 ***\n",
      "PC3          -0.994921   3.751555  -0.265  0.79085    \n",
      "PC4           0.899192   3.587635   0.251  0.80210    \n",
      "PC5          -6.555786   3.890567  -1.685  0.09198 .  \n",
      "PC6          -2.048460   3.832310  -0.535  0.59298    \n",
      "PC7           6.912398   4.208752   1.642  0.10051    \n",
      "PC8          -0.828568   2.961638  -0.280  0.77966    \n",
      "PC9         -13.463699   2.718072  -4.953 7.29e-07 ***\n",
      "PC10         -0.809523   2.638956  -0.307  0.75903    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1365.4  on 1143  degrees of freedom\n",
      "Residual deviance: 1151.0  on 1130  degrees of freedom\n",
      "AIC: 1179\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"Merged\"\n",
      "   COHORT              BETA                 SE                    P\n",
      "   <char>            <char>             <char>               <char>\n",
      "1:    AAC 0.199899966866848 0.0715903290453165  0.00523386708127726\n",
      "2:    EAS 0.306920944623081 0.0750802347643144 4.35284388227264e-05\n",
      "3:    EUR 0.438817123992525 0.0742916789991203 3.49072300510688e-09\n",
      "4:    AMR 0.192475281865515 0.0729094996046397  0.00829254800247915\n",
      "                  L95               U95\n",
      "               <char>            <char>\n",
      "1: 0.0598538120108478 0.340751385806602\n",
      "2:  0.160714219473859 0.455305775917461\n",
      "3:  0.294533641190379  0.58606711596727\n",
      "4: 0.0500448097938932 0.336103499294836\n",
      "[1] \"------------------------------AFR------------------------------\"\n",
      "[1] \"AAC Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   3.092386   0.327282   9.449  < 2e-16 ***\n",
      "zSCORE        0.285881   0.046961   6.088 1.15e-09 ***\n",
      "SEX          -0.535575   0.093419  -5.733 9.87e-09 ***\n",
      "AGE          -0.050057   0.004814 -10.399  < 2e-16 ***\n",
      "PC1           9.219434   2.601344   3.544 0.000394 ***\n",
      "PC2          33.273914   2.770046  12.012  < 2e-16 ***\n",
      "PC3         -17.255087   6.973281  -2.474 0.013344 *  \n",
      "PC4         -10.293588   4.541207  -2.267 0.023408 *  \n",
      "PC5           9.133793   7.184072   1.271 0.203588    \n",
      "PC6          -7.121564   5.565059  -1.280 0.200653    \n",
      "PC7           3.844348   4.278097   0.899 0.368860    \n",
      "PC8           7.031066   4.215859   1.668 0.095362 .  \n",
      "PC9           9.548774   4.719740   2.023 0.043057 *  \n",
      "PC10        -10.029656   3.511437  -2.856 0.004286 ** \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 3423.4  on 2605  degrees of freedom\n",
      "Residual deviance: 2968.1  on 2592  degrees of freedom\n",
      "AIC: 2996.1\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n",
      "[1] \"EAS Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   3.125419   0.327168   9.553  < 2e-16 ***\n",
      "zSCORE        0.250155   0.044722   5.594 2.22e-08 ***\n",
      "SEX          -0.542429   0.093318  -5.813 6.15e-09 ***\n",
      "AGE          -0.050325   0.004811 -10.461  < 2e-16 ***\n",
      "PC1           9.602379   2.576030   3.728 0.000193 ***\n",
      "PC2          32.699999   2.749347  11.894  < 2e-16 ***\n",
      "PC3         -17.099875   6.802605  -2.514 0.011946 *  \n",
      "PC4          -9.668030   4.330783  -2.232 0.025589 *  \n",
      "PC5          10.215757   7.233313   1.412 0.157856    \n",
      "PC6          -6.630391   5.674391  -1.168 0.242615    \n",
      "PC7           3.801784   4.177570   0.910 0.362798    \n",
      "PC8           7.640646   4.286599   1.782 0.074676 .  \n",
      "PC9           8.699384   4.644734   1.873 0.061074 .  \n",
      "PC10         -9.470260   3.492443  -2.712 0.006695 ** \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 3423.4  on 2605  degrees of freedom\n",
      "Residual deviance: 2974.2  on 2592  degrees of freedom\n",
      "AIC: 3002.2\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n",
      "[1] \"EUR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   3.136762   0.327285   9.584  < 2e-16 ***\n",
      "zSCORE        0.289287   0.046477   6.224 4.84e-10 ***\n",
      "SEX          -0.530464   0.093431  -5.678 1.37e-08 ***\n",
      "AGE          -0.050900   0.004816 -10.569  < 2e-16 ***\n",
      "PC1           9.079353   2.582199   3.516 0.000438 ***\n",
      "PC2          33.556137   2.778280  12.078  < 2e-16 ***\n",
      "PC3         -17.112795   7.101021  -2.410 0.015957 *  \n",
      "PC4          -9.416702   4.567690  -2.062 0.039247 *  \n",
      "PC5          10.019121   7.329850   1.367 0.171659    \n",
      "PC6          -7.342755   5.825971  -1.260 0.207544    \n",
      "PC7           3.806923   3.991254   0.954 0.340177    \n",
      "PC8           7.081878   4.256491   1.664 0.096156 .  \n",
      "PC9           8.886578   4.696230   1.892 0.058454 .  \n",
      "PC10         -9.704799   3.476299  -2.792 0.005243 ** \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 3423.4  on 2605  degrees of freedom\n",
      "Residual deviance: 2966.5  on 2592  degrees of freedom\n",
      "AIC: 2994.5\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n",
      "[1] \"AMR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   3.141027   0.326032   9.634  < 2e-16 ***\n",
      "zSCORE        0.087905   0.045079   1.950 0.051174 .  \n",
      "SEX          -0.530470   0.092758  -5.719 1.07e-08 ***\n",
      "AGE          -0.050517   0.004797 -10.530  < 2e-16 ***\n",
      "PC1           9.487599   2.574192   3.686 0.000228 ***\n",
      "PC2          30.710999   2.711184  11.328  < 2e-16 ***\n",
      "PC3         -16.861632   7.144266  -2.360 0.018267 *  \n",
      "PC4          -9.637323   4.643419  -2.075 0.037942 *  \n",
      "PC5           8.951252   7.186286   1.246 0.212911    \n",
      "PC6          -6.459233   5.657612  -1.142 0.253583    \n",
      "PC7           3.870482   3.982610   0.972 0.331127    \n",
      "PC8           6.221633   4.153266   1.498 0.134131    \n",
      "PC9           8.447504   4.653304   1.815 0.069466 .  \n",
      "PC10         -9.199557   3.449908  -2.667 0.007662 ** \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 3423.4  on 2605  degrees of freedom\n",
      "Residual deviance: 3002.2  on 2592  degrees of freedom\n",
      "AIC: 3030.2\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n",
      "[1] \"Merged\"\n",
      "   COHORT               BETA                 SE                    P\n",
      "   <char>             <char>             <char>               <char>\n",
      "1:    AAC  0.285880639106231 0.0469605288216364 1.14559051754197e-09\n",
      "2:    EAS  0.250154971957044  0.044722067773824 2.22477424489155e-08\n",
      "3:    EUR   0.28928742321895 0.0464767476177843 4.83564538324625e-10\n",
      "4:    AMR 0.0879048623036826 0.0450789208712271   0.0511735633776614\n",
      "                     L95               U95\n",
      "                  <char>            <char>\n",
      "1:     0.194263508359989 0.378412583467522\n",
      "2:     0.162845784743403 0.338223915103268\n",
      "3:     0.198568659664937 0.380824403469766\n",
      "4: -0.000366042492189498 0.176406976024431\n",
      "[1] \"------------------------------AJ------------------------------\"\n",
      "[1] \"AAC Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   4.045402   0.418340   9.670  < 2e-16 ***\n",
      "zSCORE        0.171124   0.054650   3.131  0.00174 ** \n",
      "SEX          -0.249997   0.118777  -2.105  0.03531 *  \n",
      "AGE          -0.039252   0.005937  -6.611 3.82e-11 ***\n",
      "PC1           8.701959   6.613206   1.316  0.18823    \n",
      "PC2          -8.274435   6.201366  -1.334  0.18211    \n",
      "PC3          -0.949534   3.717685  -0.255  0.79841    \n",
      "PC4          -1.864522   5.015986  -0.372  0.71010    \n",
      "PC5           5.510901   4.923209   1.119  0.26298    \n",
      "PC6           3.318116   4.669472   0.711  0.47733    \n",
      "PC7         -10.776048   4.650506  -2.317  0.02049 *  \n",
      "PC8           8.205323   4.465393   1.838  0.06613 .  \n",
      "PC9          -0.216057   4.495954  -0.048  0.96167    \n",
      "PC10          4.368400   4.524589   0.965  0.33430    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1973.0  on 1818  degrees of freedom\n",
      "Residual deviance: 1893.1  on 1805  degrees of freedom\n",
      "AIC: 1921.1\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EAS Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   3.964848   0.418310   9.478  < 2e-16 ***\n",
      "zSCORE        0.155076   0.054728   2.834   0.0046 ** \n",
      "SEX          -0.239006   0.118795  -2.012   0.0442 *  \n",
      "AGE          -0.038196   0.005932  -6.439 1.21e-10 ***\n",
      "PC1           9.036971   6.597498   1.370   0.1708    \n",
      "PC2          -7.603922   6.191880  -1.228   0.2194    \n",
      "PC3          -0.789828   3.712473  -0.213   0.8315    \n",
      "PC4          -1.454178   5.001376  -0.291   0.7712    \n",
      "PC5           6.024086   4.916649   1.225   0.2205    \n",
      "PC6           3.128653   4.650884   0.673   0.5011    \n",
      "PC7         -10.839296   4.643218  -2.334   0.0196 *  \n",
      "PC8           8.702379   4.464179   1.949   0.0513 .  \n",
      "PC9           0.269192   4.499685   0.060   0.9523    \n",
      "PC10          4.394224   4.517707   0.973   0.3307    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1973.0  on 1818  degrees of freedom\n",
      "Residual deviance: 1894.9  on 1805  degrees of freedom\n",
      "AIC: 1922.9\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EUR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "             Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  3.576301   0.432549   8.268  < 2e-16 ***\n",
      "zSCORE       0.504081   0.055957   9.008  < 2e-16 ***\n",
      "SEX         -0.271145   0.121893  -2.224   0.0261 *  \n",
      "AGE         -0.033913   0.006154  -5.511 3.57e-08 ***\n",
      "PC1          8.994786   6.788283   1.325   0.1852    \n",
      "PC2         -7.190588   6.323420  -1.137   0.2555    \n",
      "PC3         -1.314104   3.816086  -0.344   0.7306    \n",
      "PC4         -3.619956   5.116517  -0.708   0.4793    \n",
      "PC5          5.776562   5.087993   1.135   0.2562    \n",
      "PC6          2.128953   4.750379   0.448   0.6540    \n",
      "PC7         -9.159610   4.717446  -1.942   0.0522 .  \n",
      "PC8          8.107038   4.582573   1.769   0.0769 .  \n",
      "PC9         -0.399519   4.601186  -0.087   0.9308    \n",
      "PC10         3.421520   4.618877   0.741   0.4588    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1973.0  on 1818  degrees of freedom\n",
      "Residual deviance: 1803.4  on 1805  degrees of freedom\n",
      "AIC: 1831.4\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n",
      "[1] \"AMR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   3.978455   0.417926   9.520  < 2e-16 ***\n",
      "zSCORE        0.127315   0.054657   2.329   0.0198 *  \n",
      "SEX          -0.246841   0.118622  -2.081   0.0374 *  \n",
      "AGE          -0.038192   0.005923  -6.448 1.13e-10 ***\n",
      "PC1           9.196877   6.595208   1.394   0.1632    \n",
      "PC2          -7.285021   6.177265  -1.179   0.2383    \n",
      "PC3          -0.481776   3.705345  -0.130   0.8965    \n",
      "PC4          -1.601129   4.995005  -0.321   0.7486    \n",
      "PC5           6.336665   4.908503   1.291   0.1967    \n",
      "PC6           2.390492   4.648761   0.514   0.6071    \n",
      "PC7         -10.344145   4.639517  -2.230   0.0258 *  \n",
      "PC8           8.799870   4.460881   1.973   0.0485 *  \n",
      "PC9           0.410726   4.492232   0.091   0.9272    \n",
      "PC10          4.289442   4.510268   0.951   0.3416    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1973.0  on 1818  degrees of freedom\n",
      "Residual deviance: 1897.5  on 1805  degrees of freedom\n",
      "AIC: 1925.5\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"Merged\"\n",
      "   COHORT              BETA                 SE                    P\n",
      "   <char>            <char>             <char>               <char>\n",
      "1:    AAC 0.171124028044705 0.0546496359797128    0.001740383697159\n",
      "2:    EAS 0.155076473291803 0.0547281941443761  0.00460304246275329\n",
      "3:    EUR 0.504080956955584 0.0559570402881565 2.09167813956654e-19\n",
      "4:    AMR 0.127315086627761 0.0546570795469684   0.0198409011560227\n",
      "                  L95               U95\n",
      "               <char>            <char>\n",
      "1: 0.0643163207295712 0.278676695999459\n",
      "2: 0.0480375513572354 0.262687198436073\n",
      "3:  0.396679441091629 0.616080085050657\n",
      "4: 0.0203371023795096 0.234715492435584\n",
      "[1] \"------------------------------AMR------------------------------\"\n",
      "[1] \"AAC Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   4.076462   0.404567  10.076  < 2e-16 ***\n",
      "zSCORE        0.191868   0.040147   4.779 1.76e-06 ***\n",
      "SEX          -0.799415   0.085159  -9.387  < 2e-16 ***\n",
      "AGE          -0.042976   0.006286  -6.837 8.08e-12 ***\n",
      "PC1          56.089679   2.673036  20.984  < 2e-16 ***\n",
      "PC2          29.889537   2.440694  12.246  < 2e-16 ***\n",
      "PC3           6.200118   2.811221   2.205 0.027420 *  \n",
      "PC4          -3.545805   4.716133  -0.752 0.452144    \n",
      "PC5         -10.122084   2.804523  -3.609 0.000307 ***\n",
      "PC6          20.809152   2.925021   7.114 1.13e-12 ***\n",
      "PC7         -20.240853   2.986940  -6.776 1.23e-11 ***\n",
      "PC8          -0.539764   2.451546  -0.220 0.825736    \n",
      "PC9           7.211668   5.238580   1.377 0.168622    \n",
      "PC10          3.515746   2.665072   1.319 0.187104    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 4581.0  on 3357  degrees of freedom\n",
      "Residual deviance: 3446.1  on 3344  degrees of freedom\n",
      "AIC: 3474.1\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EAS Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "             Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   4.04124    0.40803   9.904  < 2e-16 ***\n",
      "zSCORE        0.36973    0.04255   8.689  < 2e-16 ***\n",
      "SEX          -0.81213    0.08598  -9.445  < 2e-16 ***\n",
      "AGE          -0.04307    0.00634  -6.793 1.10e-11 ***\n",
      "PC1          52.80345    2.71168  19.473  < 2e-16 ***\n",
      "PC2          32.27079    2.49262  12.947  < 2e-16 ***\n",
      "PC3           7.49067    2.85642   2.622 0.008731 ** \n",
      "PC4          -5.28329    4.79882  -1.101 0.270916    \n",
      "PC5          -9.45452    2.81271  -3.361 0.000776 ***\n",
      "PC6          21.22894    2.96496   7.160 8.07e-13 ***\n",
      "PC7         -20.92885    3.01933  -6.932 4.16e-12 ***\n",
      "PC8           0.11867    2.49001   0.048 0.961989    \n",
      "PC9           7.71622    5.29277   1.458 0.144874    \n",
      "PC10          3.22065    2.70036   1.193 0.232997    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 4581.0  on 3357  degrees of freedom\n",
      "Residual deviance: 3390.9  on 3344  degrees of freedom\n",
      "AIC: 3418.9\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EUR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   4.030677   0.415601   9.698  < 2e-16 ***\n",
      "zSCORE        0.513111   0.041435  12.383  < 2e-16 ***\n",
      "SEX          -0.838372   0.087451  -9.587  < 2e-16 ***\n",
      "AGE          -0.042939   0.006452  -6.655 2.83e-11 ***\n",
      "PC1          57.656839   2.756380  20.918  < 2e-16 ***\n",
      "PC2          29.613247   2.497826  11.856  < 2e-16 ***\n",
      "PC3           9.967587   2.929295   3.403 0.000667 ***\n",
      "PC4          -4.835882   4.992784  -0.969 0.332758    \n",
      "PC5          -8.439596   2.854775  -2.956 0.003113 ** \n",
      "PC6          21.593829   3.053947   7.071 1.54e-12 ***\n",
      "PC7         -20.707301   3.053883  -6.781 1.20e-11 ***\n",
      "PC8           0.108703   2.558491   0.042 0.966110    \n",
      "PC9           9.025450   5.413539   1.667 0.095475 .  \n",
      "PC10          3.164376   2.760156   1.146 0.251610    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 4581.0  on 3357  degrees of freedom\n",
      "Residual deviance: 3295.8  on 3344  degrees of freedom\n",
      "AIC: 3323.8\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"AMR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   4.082253   0.406240  10.049  < 2e-16 ***\n",
      "zSCORE        0.306740   0.043393   7.069 1.56e-12 ***\n",
      "SEX          -0.808808   0.085574  -9.452  < 2e-16 ***\n",
      "AGE          -0.042080   0.006313  -6.666 2.63e-11 ***\n",
      "PC1          63.742285   2.904971  21.942  < 2e-16 ***\n",
      "PC2          29.477846   2.446080  12.051  < 2e-16 ***\n",
      "PC3           6.370482   2.838036   2.245  0.02479 *  \n",
      "PC4          -3.176975   4.784981  -0.664  0.50672    \n",
      "PC5          -9.476113   2.820580  -3.360  0.00078 ***\n",
      "PC6          21.446947   2.952176   7.265 3.74e-13 ***\n",
      "PC7         -20.860914   3.003322  -6.946 3.76e-12 ***\n",
      "PC8          -0.286552   2.481468  -0.115  0.90807    \n",
      "PC9           8.212776   5.283323   1.554  0.12007    \n",
      "PC10          3.812159   2.691596   1.416  0.15668    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 4581.0  on 3357  degrees of freedom\n",
      "Residual deviance: 3418.1  on 3344  degrees of freedom\n",
      "AIC: 3446.1\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"Merged\"\n",
      "   COHORT              BETA                 SE                    P\n",
      "   <char>            <char>             <char>               <char>\n",
      "1:    AAC 0.191868183968086 0.0401473299296971 1.76079872311536e-06\n",
      "2:    EAS 0.369734260159239 0.0425514692593226 3.65300179750182e-18\n",
      "3:    EUR  0.51311070143442 0.0414353098696187 3.21365240551557e-35\n",
      "4:    AMR 0.306739518132906 0.0433929069388578 1.56182234061847e-12\n",
      "                 L95               U95\n",
      "              <char>            <char>\n",
      "1:  0.11338833780824 0.270812727197975\n",
      "2: 0.286765293589599 0.453622260891032\n",
      "3: 0.432738594001144 0.595210600150418\n",
      "4:  0.22202280854049 0.392177685743131\n",
      "[1] \"------------------------------EAS------------------------------\"\n",
      "[1] \"AAC Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   3.247502   0.305435  10.632  < 2e-16 ***\n",
      "zSCORE        0.150946   0.040951   3.686 0.000228 ***\n",
      "SEX          -0.203145   0.084610  -2.401 0.016352 *  \n",
      "AGE          -0.052251   0.004492 -11.631  < 2e-16 ***\n",
      "PC1          84.325676   3.349396  25.176  < 2e-16 ***\n",
      "PC2         -19.132991   2.994492  -6.389 1.67e-10 ***\n",
      "PC3          79.963636   4.049564  19.746  < 2e-16 ***\n",
      "PC4          18.848380   4.094489   4.603 4.16e-06 ***\n",
      "PC5          18.162936   9.665739   1.879 0.060230 .  \n",
      "PC6          10.451838   9.291526   1.125 0.260640    \n",
      "PC7          44.323721   5.271418   8.408  < 2e-16 ***\n",
      "PC8           4.647564   4.466515   1.041 0.298092    \n",
      "PC9         -11.535799   6.463783  -1.785 0.074313 .  \n",
      "PC10         18.617329   4.860614   3.830 0.000128 ***\n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 5710.5  on 4167  degrees of freedom\n",
      "Residual deviance: 3829.5  on 4154  degrees of freedom\n",
      "AIC: 3857.5\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n",
      "[1] \"EAS Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   3.202856   0.306031  10.466  < 2e-16 ***\n",
      "zSCORE        0.239424   0.041741   5.736 9.70e-09 ***\n",
      "SEX          -0.212165   0.084886  -2.499  0.01244 *  \n",
      "AGE          -0.051281   0.004507 -11.378  < 2e-16 ***\n",
      "PC1          87.143622   3.380681  25.777  < 2e-16 ***\n",
      "PC2         -19.252730   3.000872  -6.416 1.40e-10 ***\n",
      "PC3          80.341896   4.061132  19.783  < 2e-16 ***\n",
      "PC4          19.577850   4.138356   4.731 2.24e-06 ***\n",
      "PC5          17.920903   9.681942   1.851  0.06418 .  \n",
      "PC6          10.180856   9.312527   1.093  0.27429    \n",
      "PC7          43.975674   5.300248   8.297  < 2e-16 ***\n",
      "PC8           5.261186   4.471222   1.177  0.23932    \n",
      "PC9         -12.414212   6.504411  -1.909  0.05632 .  \n",
      "PC10         18.903091   4.888232   3.867  0.00011 ***\n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 5710.5  on 4167  degrees of freedom\n",
      "Residual deviance: 3809.8  on 4154  degrees of freedom\n",
      "AIC: 3837.8\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n",
      "[1] \"EUR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   3.058888   0.307247   9.956  < 2e-16 ***\n",
      "zSCORE        0.383342   0.041613   9.212  < 2e-16 ***\n",
      "SEX          -0.203645   0.085625  -2.378 0.017391 *  \n",
      "AGE          -0.050170   0.004517 -11.107  < 2e-16 ***\n",
      "PC1          84.597408   3.380530  25.025  < 2e-16 ***\n",
      "PC2         -16.160871   3.027119  -5.339 9.36e-08 ***\n",
      "PC3          80.820599   4.094538  19.739  < 2e-16 ***\n",
      "PC4          18.965946   4.107948   4.617 3.90e-06 ***\n",
      "PC5          17.532240   9.771062   1.794 0.072765 .  \n",
      "PC6          11.040648   9.367668   1.179 0.238561    \n",
      "PC7          44.428349   5.344593   8.313  < 2e-16 ***\n",
      "PC8           6.100854   4.499565   1.356 0.175139    \n",
      "PC9         -12.944068   6.558887  -1.974 0.048437 *  \n",
      "PC10         18.999029   4.903938   3.874 0.000107 ***\n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 5710.5  on 4167  degrees of freedom\n",
      "Residual deviance: 3755.3  on 4154  degrees of freedom\n",
      "AIC: 3783.3\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n",
      "[1] \"AMR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   3.268082   0.305436  10.700  < 2e-16 ***\n",
      "zSCORE        0.131139   0.040924   3.204 0.001353 ** \n",
      "SEX          -0.215105   0.084632  -2.542 0.011033 *  \n",
      "AGE          -0.052133   0.004495 -11.597  < 2e-16 ***\n",
      "PC1          85.559522   3.343908  25.587  < 2e-16 ***\n",
      "PC2         -17.034445   3.015411  -5.649 1.61e-08 ***\n",
      "PC3          80.294555   4.048094  19.835  < 2e-16 ***\n",
      "PC4          18.956276   4.103297   4.620 3.84e-06 ***\n",
      "PC5          18.440625   9.655121   1.910 0.056142 .  \n",
      "PC6          10.751138   9.283344   1.158 0.246819    \n",
      "PC7          44.531543   5.277595   8.438  < 2e-16 ***\n",
      "PC8           5.512764   4.465878   1.234 0.217047    \n",
      "PC9         -11.882378   6.476127  -1.835 0.066536 .  \n",
      "PC10         18.644343   4.863758   3.833 0.000126 ***\n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 5710.5  on 4167  degrees of freedom\n",
      "Residual deviance: 3832.9  on 4154  degrees of freedom\n",
      "AIC: 3860.9\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n",
      "[1] \"Merged\"\n",
      "   COHORT              BETA                 SE                    P\n",
      "   <char>            <char>             <char>               <char>\n",
      "1:    AAC 0.150945877877689 0.0409505438810806 0.000227758878442733\n",
      "2:    EAS 0.239424161773319 0.0417414533655978 9.70046657694511e-09\n",
      "3:    EUR 0.383342208747534 0.0416128311636265 3.19759639775098e-20\n",
      "4:    AMR  0.13113872054358 0.0409243488392121  0.00135335829526492\n",
      "                  L95               U95\n",
      "               <char>            <char>\n",
      "1: 0.0708271946128014 0.231394738339014\n",
      "2:  0.157860427289757  0.32153035384142\n",
      "3:  0.302163854940786  0.46532888299196\n",
      "4: 0.0510709420892738 0.211537937010415\n",
      "[1] \"------------------------------EUR------------------------------\"\n",
      "[1] \"AAC Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   3.951121   0.093031  42.471  < 2e-16 ***\n",
      "zSCORE        0.190180   0.012637  15.049  < 2e-16 ***\n",
      "SEX          -0.449324   0.025270 -17.781  < 2e-16 ***\n",
      "AGE          -0.041015   0.001355 -30.273  < 2e-16 ***\n",
      "PC1          14.769366   2.853996   5.175 2.28e-07 ***\n",
      "PC2           3.320753   2.806713   1.183  0.23675    \n",
      "PC3          27.037700   2.886264   9.368  < 2e-16 ***\n",
      "PC4          12.976440   3.166856   4.098 4.17e-05 ***\n",
      "PC5          15.775939   3.570903   4.418 9.97e-06 ***\n",
      "PC6          -0.298903   4.287021  -0.070  0.94441    \n",
      "PC7         -16.690787   3.241391  -5.149 2.62e-07 ***\n",
      "PC8          -0.206075   3.833439  -0.054  0.95713    \n",
      "PC9           8.154442   3.238559   2.518  0.01180 *  \n",
      "PC10         11.814262   3.769440   3.134  0.00172 ** \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 39127  on 31798  degrees of freedom\n",
      "Residual deviance: 37346  on 31785  degrees of freedom\n",
      "AIC: 37374\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EAS Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "             Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   3.88075    0.09344  41.532  < 2e-16 ***\n",
      "zSCORE        0.29131    0.01278  22.789  < 2e-16 ***\n",
      "SEX          -0.45128    0.02540 -17.766  < 2e-16 ***\n",
      "AGE          -0.04023    0.00136 -29.580  < 2e-16 ***\n",
      "PC1          12.64367    2.85021   4.436 9.16e-06 ***\n",
      "PC2           2.65214    2.81903   0.941  0.34681    \n",
      "PC3          27.84463    2.89333   9.624  < 2e-16 ***\n",
      "PC4          12.85727    3.17131   4.054 5.03e-05 ***\n",
      "PC5          15.63189    3.56836   4.381 1.18e-05 ***\n",
      "PC6          -1.62417    4.21449  -0.385  0.69996    \n",
      "PC7         -17.12676    3.24179  -5.283 1.27e-07 ***\n",
      "PC8           0.18891    3.87098   0.049  0.96108    \n",
      "PC9           6.84782    3.25165   2.106  0.03521 *  \n",
      "PC10         11.29723    3.78824   2.982  0.00286 ** \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 39127  on 31798  degrees of freedom\n",
      "Residual deviance: 37043  on 31785  degrees of freedom\n",
      "AIC: 37071\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EUR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "             Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   3.75108    0.09421  39.818  < 2e-16 ***\n",
      "zSCORE        0.45413    0.01303  34.860  < 2e-16 ***\n",
      "SEX          -0.46149    0.02575 -17.925  < 2e-16 ***\n",
      "AGE          -0.03893    0.00137 -28.419  < 2e-16 ***\n",
      "PC1           9.01009    2.88432   3.124 0.001785 ** \n",
      "PC2           4.08638    2.86081   1.428 0.153176    \n",
      "PC3          27.35123    2.94083   9.301  < 2e-16 ***\n",
      "PC4          11.51967    3.21077   3.588 0.000333 ***\n",
      "PC5          15.59209    3.61977   4.307 1.65e-05 ***\n",
      "PC6          -0.88159    4.23604  -0.208 0.835137    \n",
      "PC7         -16.49387    3.27910  -5.030 4.90e-07 ***\n",
      "PC8           0.59555    4.11146   0.145 0.884828    \n",
      "PC9           7.21979    3.29555   2.191 0.028469 *  \n",
      "PC10         11.12666    3.84056   2.897 0.003766 ** \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 39127  on 31798  degrees of freedom\n",
      "Residual deviance: 36253  on 31785  degrees of freedom\n",
      "AIC: 36281\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"AMR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   3.939971   0.093165  42.290  < 2e-16 ***\n",
      "zSCORE        0.233717   0.012649  18.477  < 2e-16 ***\n",
      "SEX          -0.450686   0.025322 -17.798  < 2e-16 ***\n",
      "AGE          -0.040960   0.001356 -30.201  < 2e-16 ***\n",
      "PC1          10.174882   2.842613   3.579 0.000344 ***\n",
      "PC2           4.388795   2.815826   1.559 0.119087    \n",
      "PC3          26.087372   2.899363   8.998  < 2e-16 ***\n",
      "PC4          13.264291   3.169455   4.185 2.85e-05 ***\n",
      "PC5          15.031818   3.572005   4.208 2.57e-05 ***\n",
      "PC6          -0.350377   4.187156  -0.084 0.933312    \n",
      "PC7         -17.135633   3.239142  -5.290 1.22e-07 ***\n",
      "PC8           0.050919   3.824340   0.013 0.989377    \n",
      "PC9           6.975647   3.244679   2.150 0.031565 *  \n",
      "PC10         11.297829   3.777991   2.990 0.002786 ** \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 39127  on 31798  degrees of freedom\n",
      "Residual deviance: 37228  on 31785  degrees of freedom\n",
      "AIC: 37256\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"Merged\"\n",
      "   COHORT              BETA                 SE                     P\n",
      "   <char>            <char>             <char>                <char>\n",
      "1:    AAC 0.190179508565399 0.0126371388761649  3.49151011981814e-51\n",
      "2:    EAS  0.29130867701988 0.0127829247322786 5.90931980495427e-115\n",
      "3:    EUR 0.454127939419562 0.0130272313590462 3.01465754979865e-266\n",
      "4:    AMR 0.233716519978101 0.0126488934220051  3.14937555266522e-76\n",
      "                 L95               U95\n",
      "              <char>            <char>\n",
      "1: 0.165435163137198 0.214973386338874\n",
      "2: 0.266289370946431  0.31639907623231\n",
      "3: 0.428663209947996 0.479730231842746\n",
      "4:  0.20895371718342 0.258538042759763\n",
      "[1] \"------------------------------CAS------------------------------\"\n",
      "[1] \"AAC Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   1.718688   0.574292   2.993 0.002765 ** \n",
      "zSCORE        0.226077   0.077404   2.921 0.003492 ** \n",
      "SEX          -0.393605   0.161796  -2.433 0.014986 *  \n",
      "AGE          -0.013178   0.009122  -1.445 0.148550    \n",
      "PC1          29.866470   5.155514   5.793 6.91e-09 ***\n",
      "PC2          -1.644448   9.471563  -0.174 0.862164    \n",
      "PC3          -8.573866  12.739716  -0.673 0.500945    \n",
      "PC4         -41.922591  11.753951  -3.567 0.000362 ***\n",
      "PC5           8.304891   4.399410   1.888 0.059062 .  \n",
      "PC6          10.566548   4.871535   2.169 0.030080 *  \n",
      "PC7           8.394172   7.646349   1.098 0.272291    \n",
      "PC8          -7.171815   6.269788  -1.144 0.252678    \n",
      "PC9         -29.497342   7.457256  -3.956 7.64e-05 ***\n",
      "PC10         -1.083467   6.131717  -0.177 0.859745    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1179.4  on 904  degrees of freedom\n",
      "Residual deviance: 1043.2  on 891  degrees of freedom\n",
      "AIC: 1071.2\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EAS Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   1.779502   0.572986   3.106 0.001899 ** \n",
      "zSCORE        0.210761   0.076653   2.750 0.005968 ** \n",
      "SEX          -0.402290   0.161377  -2.493 0.012672 *  \n",
      "AGE          -0.014121   0.009115  -1.549 0.121345    \n",
      "PC1          29.200362   5.101397   5.724 1.04e-08 ***\n",
      "PC2          -2.970667   9.011700  -0.330 0.741668    \n",
      "PC3          -7.002578  12.761168  -0.549 0.583183    \n",
      "PC4         -42.324280  11.782951  -3.592 0.000328 ***\n",
      "PC5           8.430839   4.353533   1.937 0.052800 .  \n",
      "PC6          10.532038   4.857320   2.168 0.030137 *  \n",
      "PC7           8.532980   7.604782   1.122 0.261839    \n",
      "PC8          -7.877633   6.269396  -1.257 0.208927    \n",
      "PC9         -28.315484   7.439417  -3.806 0.000141 ***\n",
      "PC10         -1.472785   6.108124  -0.241 0.809463    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1179.4  on 904  degrees of freedom\n",
      "Residual deviance: 1044.2  on 891  degrees of freedom\n",
      "AIC: 1072.2\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EUR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   1.675449   0.578560   2.896 0.003781 ** \n",
      "zSCORE        0.364008   0.079144   4.599 4.24e-06 ***\n",
      "SEX          -0.390360   0.162491  -2.402 0.016290 *  \n",
      "AGE          -0.013448   0.009193  -1.463 0.143491    \n",
      "PC1          29.680497   5.139788   5.775 7.71e-09 ***\n",
      "PC2          -3.358642   8.922402  -0.376 0.706599    \n",
      "PC3         -12.079400  12.929788  -0.934 0.350185    \n",
      "PC4         -46.108276  11.928445  -3.865 0.000111 ***\n",
      "PC5           8.520857   4.372140   1.949 0.051308 .  \n",
      "PC6           9.905913   4.889264   2.026 0.042759 *  \n",
      "PC7           9.531647   7.630188   1.249 0.211591    \n",
      "PC8          -8.127766   6.360235  -1.278 0.201283    \n",
      "PC9         -29.116390   7.515381  -3.874 0.000107 ***\n",
      "PC10         -0.616201   6.145207  -0.100 0.920127    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1179.4  on 904  degrees of freedom\n",
      "Residual deviance: 1030.1  on 891  degrees of freedom\n",
      "AIC: 1058.1\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"AMR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   1.724402   0.575207   2.998 0.002719 ** \n",
      "zSCORE        0.295371   0.077121   3.830 0.000128 ***\n",
      "SEX          -0.405453   0.161893  -2.504 0.012264 *  \n",
      "AGE          -0.013717   0.009153  -1.499 0.133977    \n",
      "PC1          28.693029   5.187225   5.531 3.18e-08 ***\n",
      "PC2          -1.478415   9.613975  -0.154 0.877785    \n",
      "PC3          -9.857937  12.844281  -0.767 0.442787    \n",
      "PC4         -45.484090  11.799244  -3.855 0.000116 ***\n",
      "PC5           9.177121   4.338500   2.115 0.034407 *  \n",
      "PC6           9.622249   4.844815   1.986 0.047023 *  \n",
      "PC7          10.053830   7.631704   1.317 0.187712    \n",
      "PC8          -6.897760   6.303115  -1.094 0.273805    \n",
      "PC9         -28.479665   7.474654  -3.810 0.000139 ***\n",
      "PC10          0.441161   6.163514   0.072 0.942939    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1179.4  on 904  degrees of freedom\n",
      "Residual deviance: 1036.9  on 891  degrees of freedom\n",
      "AIC: 1064.9\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"Merged\"\n",
      "   COHORT              BETA                 SE                    P\n",
      "   <char>            <char>             <char>               <char>\n",
      "1:    AAC 0.226076974991325 0.0774036362012785  0.00349185500332206\n",
      "2:    EAS 0.210761474577035 0.0766533161955865  0.00596787038703974\n",
      "3:    EUR 0.364008496828593 0.0791443401919635 4.23914670793399e-06\n",
      "4:    AMR 0.295371233517872 0.0771210325345607 0.000128158932730083\n",
      "                  L95               U95\n",
      "               <char>            <char>\n",
      "1: 0.0753439693205127  0.37908325975082\n",
      "2: 0.0611961120272743 0.361994420038717\n",
      "3:  0.210087945461393 0.520668508302368\n",
      "4:  0.145241640201476 0.447889863284412\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "for (ancestry in ancestries) {\n",
    "    print(paste0(\"------------------------------\", ancestry, \"------------------------------\"))\n",
    "    forest <- list()\n",
    "    for (summstat_ancestry in summstat_ancestries) {\n",
    "        print(paste0(summstat_ancestry, \" Summary stats\"))\n",
    "        temp_data <- read.table(paste0(RESULTS_DIR, \"/\", ancestry, \"/model1_\", summstat_ancestry, \".profile\"), header = T) \n",
    "        temp_covs <- read.table(paste0(RESULTS_DIR, \"/covariates.txt\"), header = T, sep=\" \")\n",
    "        colnames(temp_covs) <- c(\"FID\", \"IID\", \"SEX\", \"AGE\", \"ANCESTRY\", \"PC1\", \"PC2\", \"PC3\", \"PC4\", \"PC5\", \"PC6\", \"PC7\", \"PC8\", \"PC9\", \"PC10\")\n",
    "        data <- merge(temp_data, temp_covs, by = \"IID\")\n",
    "        data$CASE <- data$PHENO - 1\n",
    "        \n",
    "        dat <- subset(data, CASE != -10) \n",
    "        meanPop <- mean(subset(dat, CASE == 0)$SCORE)\n",
    "        sdPop <- sd(subset(dat, CASE == 0)$SCORE)\n",
    "        dat$zSCORE <- (dat$SCORE - meanPop)/sdPop\n",
    "        \n",
    "        grsTests <- glm(CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10, data = dat, family='binomial')\n",
    "        print(summary(grsTests))\n",
    "        grs_summary <- summary(grsTests)$coefficients\n",
    "        conf_intervals <- confint(grsTests, level = 0.95)\n",
    "        zscore_summary <- grs_summary[\"zSCORE\", , drop = FALSE]\n",
    "        zscore_conf <- conf_intervals[\"zSCORE\", , drop = FALSE]\n",
    "        zscore <- as.data.table(cbind(summstat_ancestry, zscore_summary, zscore_conf))\n",
    "        colnames(zscore) <- c(\"COHORT\", \"BETA\", \"SE\", \"T\", \"P\", \"L95\", \"U95\")\n",
    "        zscore <- zscore[, !\"T\", with = FALSE]\n",
    "        forest <- rbind(forest, zscore)\n",
    "    }\n",
    "    print(\"Merged\")\n",
    "    print(forest)\n",
    "    fwrite(forest, paste0(RESULTS_DIR, \"/\", ancestry, \"/forest_risk_pcs_\", ancestry, \".txt\"), sep = \"\\t\", col.names = TRUE, quote = FALSE)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcf7d8ca-0f00-4e98-a1f9-38de7bf545fa",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Adjust by Admixture"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "b21e35aa-b0d9-4249-85bf-0da01e6fb79a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"------------------------------AAC------------------------------\"\n",
      "[1] \"AAC Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  14.866903   7.522071   1.976  0.04811 *  \n",
      "zSCORE        0.190725   0.071355   2.673  0.00752 ** \n",
      "SEX          -0.647314   0.145827  -4.439 9.04e-06 ***\n",
      "AGE          -0.034346   0.006621  -5.187 2.14e-07 ***\n",
      "AJ           33.101387  20.458608   1.618  0.10567    \n",
      "EUR         -32.399701  38.110514  -0.850  0.39524    \n",
      "EAS           8.795826  17.877709   0.492  0.62272    \n",
      "AMR         -10.309898   7.890333  -1.307  0.19133    \n",
      "AFR         -14.468987   7.518182  -1.925  0.05429 .  \n",
      "CAS         -31.871583  20.382568  -1.564  0.11790    \n",
      "AAC         -48.470555  14.613010  -3.317  0.00091 ***\n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1365.4  on 1143  degrees of freedom\n",
      "Residual deviance: 1155.7  on 1133  degrees of freedom\n",
      "AIC: 1177.7\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EAS Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  14.959249   7.443171   2.010  0.04445 *  \n",
      "zSCORE        0.299387   0.074917   3.996 6.43e-05 ***\n",
      "SEX          -0.649524   0.146492  -4.434 9.26e-06 ***\n",
      "AGE          -0.034582   0.006654  -5.197 2.02e-07 ***\n",
      "AJ           30.779350  20.342574   1.513  0.13027    \n",
      "EUR         -33.830303  37.921446  -0.892  0.37233    \n",
      "EAS           8.047144  17.262408   0.466  0.64110    \n",
      "AMR         -10.290853   7.870035  -1.308  0.19101    \n",
      "AFR         -14.508487   7.447007  -1.948  0.05139 .  \n",
      "CAS         -32.580755  20.263636  -1.608  0.10787    \n",
      "AAC         -46.955022  14.451681  -3.249  0.00116 ** \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1365.4  on 1143  degrees of freedom\n",
      "Residual deviance: 1146.6  on 1133  degrees of freedom\n",
      "AIC: 1168.6\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EUR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  11.978200   7.781494   1.539  0.12373    \n",
      "zSCORE        0.436627   0.074512   5.860 4.63e-09 ***\n",
      "SEX          -0.651156   0.147975  -4.400 1.08e-05 ***\n",
      "AGE          -0.035287   0.006682  -5.281 1.29e-07 ***\n",
      "AJ           39.069101  20.884501   1.871  0.06138 .  \n",
      "EUR         -35.168329  39.039513  -0.901  0.36767    \n",
      "EAS          15.524749  19.302429   0.804  0.42123    \n",
      "AMR          -8.053000   8.115132  -0.992  0.32103    \n",
      "AFR         -11.521204   7.768450  -1.483  0.13805    \n",
      "CAS         -28.510443  20.796728  -1.371  0.17040    \n",
      "AAC         -44.460399  15.111604  -2.942  0.00326 ** \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1365.4  on 1143  degrees of freedom\n",
      "Residual deviance: 1127.0  on 1133  degrees of freedom\n",
      "AIC: 1149\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n",
      "[1] \"AMR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  15.725081   7.588694   2.072  0.03825 *  \n",
      "zSCORE        0.173210   0.072632   2.385  0.01709 *  \n",
      "SEX          -0.641758   0.145686  -4.405 1.06e-05 ***\n",
      "AGE          -0.034620   0.006641  -5.213 1.86e-07 ***\n",
      "AJ           31.628557  20.409101   1.550  0.12121    \n",
      "EUR         -33.760593  37.893022  -0.891  0.37296    \n",
      "EAS           7.007555  18.522008   0.378  0.70518    \n",
      "AMR         -11.352199   7.943100  -1.429  0.15295    \n",
      "AFR         -15.399617   7.581697  -2.031  0.04224 *  \n",
      "CAS         -33.536609  20.466155  -1.639  0.10129    \n",
      "AAC         -47.426669  14.610150  -3.246  0.00117 ** \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1365.4  on 1143  degrees of freedom\n",
      "Residual deviance: 1157.2  on 1133  degrees of freedom\n",
      "AIC: 1179.2\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"Merged\"\n",
      "   COHORT              BETA                 SE                    P\n",
      "   <char>            <char>             <char>               <char>\n",
      "1:    AAC 0.190724800543698 0.0713552714832259  0.00752008767855082\n",
      "2:    EAS 0.299387406559358 0.0749166993162127 6.43481636711844e-05\n",
      "3:    EUR 0.436627003858021 0.0745117410414402 4.63308740545023e-09\n",
      "4:    AMR 0.173210150487076 0.0726320295088499   0.0170901440276781\n",
      "                  L95               U95\n",
      "               <char>            <char>\n",
      "1: 0.0511581092958101  0.33113568219958\n",
      "2:  0.153494807425013 0.447439481728376\n",
      "3:  0.291954426568205 0.584323748232823\n",
      "4: 0.0313073076442125 0.316280998479674\n",
      "[1] \"------------------------------AFR------------------------------\"\n",
      "[1] \"AAC Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept) -9.490e+00  1.707e+01  -0.556   0.5783    \n",
      "zSCORE       2.496e-01  4.573e-02   5.458 4.81e-08 ***\n",
      "SEX         -5.957e-01  9.141e-02  -6.517 7.17e-11 ***\n",
      "AGE         -4.872e-02  4.637e-03 -10.507  < 2e-16 ***\n",
      "AJ          -4.776e+01  4.440e+01  -1.076   0.2821    \n",
      "EUR         -1.304e+02  1.746e+02  -0.747   0.4550    \n",
      "EAS         -8.635e+01  4.417e+01  -1.955   0.0506 .  \n",
      "AMR          6.504e+01  2.563e+01   2.537   0.0112 *  \n",
      "AFR          1.456e+01  1.711e+01   0.851   0.3949    \n",
      "CAS         -8.271e+00  4.009e+01  -0.206   0.8366    \n",
      "AAC          4.604e+01  3.583e+01   1.285   0.1988    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 3423.4  on 2605  degrees of freedom\n",
      "Residual deviance: 3052.6  on 2595  degrees of freedom\n",
      "AIC: 3074.6\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EAS Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept) -1.220e+01  1.694e+01  -0.720  0.47146    \n",
      "zSCORE       2.158e-01  4.387e-02   4.919 8.70e-07 ***\n",
      "SEX         -6.012e-01  9.135e-02  -6.582 4.65e-11 ***\n",
      "AGE         -4.886e-02  4.633e-03 -10.544  < 2e-16 ***\n",
      "AJ          -3.954e+01  4.415e+01  -0.895  0.37054    \n",
      "EUR         -1.247e+02  1.746e+02  -0.714  0.47508    \n",
      "EAS         -8.388e+01  4.406e+01  -1.903  0.05698 .  \n",
      "AMR          6.766e+01  2.544e+01   2.660  0.00781 ** \n",
      "AFR          1.726e+01  1.698e+01   1.016  0.30944    \n",
      "CAS         -2.879e+00  3.996e+01  -0.072  0.94257    \n",
      "AAC          4.962e+01  3.577e+01   1.387  0.16535    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 3423.4  on 2605  degrees of freedom\n",
      "Residual deviance: 3058.5  on 2595  degrees of freedom\n",
      "AIC: 3080.5\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EUR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept) -9.837e+00  1.698e+01  -0.579   0.5623    \n",
      "zSCORE       2.569e-01  4.550e-02   5.646 1.64e-08 ***\n",
      "SEX         -5.894e-01  9.141e-02  -6.448 1.13e-10 ***\n",
      "AGE         -4.945e-02  4.635e-03 -10.671  < 2e-16 ***\n",
      "AJ          -4.892e+01  4.428e+01  -1.105   0.2692    \n",
      "EUR         -1.212e+02  1.746e+02  -0.694   0.4876    \n",
      "EAS         -8.675e+01  4.414e+01  -1.965   0.0494 *  \n",
      "AMR          6.433e+01  2.552e+01   2.520   0.0117 *  \n",
      "AFR          1.497e+01  1.702e+01   0.879   0.3792    \n",
      "CAS         -5.746e+00  3.999e+01  -0.144   0.8857    \n",
      "AAC          4.618e+01  3.583e+01   1.289   0.1975    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 3423.4  on 2605  degrees of freedom\n",
      "Residual deviance: 3050.6  on 2595  degrees of freedom\n",
      "AIC: 3072.6\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"AMR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept) -10.595169  16.921424  -0.626  0.53122    \n",
      "zSCORE        0.079476   0.044332   1.793  0.07301 .  \n",
      "SEX          -0.590949   0.090896  -6.501 7.96e-11 ***\n",
      "AGE          -0.049096   0.004628 -10.609  < 2e-16 ***\n",
      "AJ          -36.083797  44.015291  -0.820  0.41233    \n",
      "EUR         -83.661511 173.699519  -0.482  0.63006    \n",
      "EAS         -89.276770  43.949972  -2.031  0.04222 *  \n",
      "AMR          66.527071  25.391199   2.620  0.00879 ** \n",
      "AFR          15.616966  16.963474   0.921  0.35725    \n",
      "CAS          -7.998041  39.806847  -0.201  0.84076    \n",
      "AAC          44.158430  35.620019   1.240  0.21508    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 3423.4  on 2605  degrees of freedom\n",
      "Residual deviance: 3079.8  on 2595  degrees of freedom\n",
      "AIC: 3101.8\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"Merged\"\n",
      "   COHORT               BETA                 SE                    P\n",
      "   <char>             <char>             <char>               <char>\n",
      "1:    AAC  0.249584856720608 0.0457273511978731 4.81232832693503e-08\n",
      "2:    EAS  0.215792145920743 0.0438698499571589 8.70244068879012e-07\n",
      "3:    EUR  0.256924284941352 0.0455022435491686 1.63833665238527e-08\n",
      "4:    AMR 0.0794755122242312 0.0443316665755902   0.0730132072751991\n",
      "                    L95               U95\n",
      "                 <char>            <char>\n",
      "1:    0.160312990598201 0.339625736346485\n",
      "2:    0.130099442252681 0.302133581249352\n",
      "3:    0.168065234369095 0.346500597787405\n",
      "4: -0.00734732258468351 0.166495230232024\n",
      "[1] \"------------------------------AJ------------------------------\"\n",
      "[1] \"AAC Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  -6.970703   6.504994  -1.072 0.283903    \n",
      "zSCORE        0.182991   0.054739   3.343 0.000829 ***\n",
      "SEX          -0.267416   0.118651  -2.254 0.024208 *  \n",
      "AGE          -0.039686   0.005951  -6.669 2.58e-11 ***\n",
      "AJ           10.894277   6.681483   1.631 0.102992    \n",
      "EUR           7.987779   7.187317   1.111 0.266408    \n",
      "EAS          54.422667  90.074606   0.604 0.545714    \n",
      "AMR           9.085925  20.998325   0.433 0.665234    \n",
      "AFR           0.474352  30.583462   0.016 0.987625    \n",
      "CAS          54.823201  19.748693   2.776 0.005503 ** \n",
      "AAC         -13.858362  43.437383  -0.319 0.749694    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1973.0  on 1818  degrees of freedom\n",
      "Residual deviance: 1896.6  on 1808  degrees of freedom\n",
      "AIC: 1918.6\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EAS Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  -7.138046   6.518728  -1.095  0.27351    \n",
      "zSCORE        0.158086   0.054601   2.895  0.00379 ** \n",
      "SEX          -0.255420   0.118622  -2.153  0.03130 *  \n",
      "AGE          -0.038471   0.005938  -6.479 9.26e-11 ***\n",
      "AJ           10.961898   6.695870   1.637  0.10161    \n",
      "EUR           7.526934   7.206884   1.044  0.29630    \n",
      "EAS          52.335495  90.002057   0.581  0.56091    \n",
      "AMR          12.038752  21.352923   0.564  0.57289    \n",
      "AFR           0.099882  30.506846   0.003  0.99739    \n",
      "CAS          53.505307  19.740341   2.710  0.00672 ** \n",
      "AAC         -11.681755  43.738258  -0.267  0.78941    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1973.0  on 1818  degrees of freedom\n",
      "Residual deviance: 1899.4  on 1808  degrees of freedom\n",
      "AIC: 1921.4\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EUR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  -4.658412   6.674210  -0.698   0.4852    \n",
      "zSCORE        0.505984   0.055984   9.038  < 2e-16 ***\n",
      "SEX          -0.288847   0.121698  -2.373   0.0176 *  \n",
      "AGE          -0.033804   0.006171  -5.478  4.3e-08 ***\n",
      "AJ            7.979883   6.857190   1.164   0.2445    \n",
      "EUR           5.705587   7.414982   0.769   0.4416    \n",
      "EAS          43.001643  91.452891   0.470   0.6382    \n",
      "AMR           8.207745  20.872208   0.393   0.6941    \n",
      "AFR           2.869607  31.156599   0.092   0.9266    \n",
      "CAS          47.017424  20.105921   2.338   0.0194 *  \n",
      "AAC         -24.589368  44.063150  -0.558   0.5768    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1973  on 1818  degrees of freedom\n",
      "Residual deviance: 1807  on 1808  degrees of freedom\n",
      "AIC: 1829\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n",
      "[1] \"AMR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  -6.461584   6.517694  -0.991  0.32149    \n",
      "zSCORE        0.126016   0.054461   2.314  0.02067 *  \n",
      "SEX          -0.262993   0.118442  -2.220  0.02639 *  \n",
      "AGE          -0.038524   0.005924  -6.503 7.86e-11 ***\n",
      "AJ           10.303457   6.694025   1.539  0.12375    \n",
      "EUR           7.120128   7.200919   0.989  0.32277    \n",
      "EAS          48.598202  89.882391   0.541  0.58872    \n",
      "AMR          11.392560  21.186692   0.538  0.59077    \n",
      "AFR           0.578554  30.525363   0.019  0.98488    \n",
      "CAS          51.198329  19.711775   2.597  0.00939 ** \n",
      "AAC         -13.005386  43.683231  -0.298  0.76592    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1973.0  on 1818  degrees of freedom\n",
      "Residual deviance: 1902.5  on 1808  degrees of freedom\n",
      "AIC: 1924.5\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"Merged\"\n",
      "   COHORT              BETA                 SE                    P\n",
      "   <char>            <char>             <char>               <char>\n",
      "1:    AAC 0.182991330192705 0.0547392111342536 0.000828880085161063\n",
      "2:    EAS 0.158086236972167 0.0546007283102743  0.00378779786976327\n",
      "3:    EUR 0.505983808533616 0.0559844735284334 1.59665109910648e-19\n",
      "4:    AMR 0.126016193223976 0.0544606874660787   0.0206736025464218\n",
      "                  L95               U95\n",
      "               <char>            <char>\n",
      "1: 0.0760425898669438 0.290750656375314\n",
      "2: 0.0513171639358488 0.265465052253201\n",
      "3:  0.398574187044917 0.618077636504566\n",
      "4: 0.0194227831140852 0.233028166305915\n",
      "[1] \"------------------------------AMR------------------------------\"\n",
      "[1] \"AAC Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  10.743633   5.297939   2.028  0.04257 *  \n",
      "zSCORE        0.224607   0.037798   5.942 2.81e-09 ***\n",
      "SEX          -0.789353   0.078425 -10.065  < 2e-16 ***\n",
      "AGE          -0.047714   0.005956  -8.012 1.13e-15 ***\n",
      "AJ           -4.071369   9.275297  -0.439  0.66070    \n",
      "EUR          -5.176310   6.406041  -0.808  0.41907    \n",
      "EAS          27.663859  10.707704   2.584  0.00978 ** \n",
      "AMR          -7.306233   5.274316  -1.385  0.16598    \n",
      "AFR           2.366741   7.008646   0.338  0.73560    \n",
      "CAS         -27.851526   9.933172  -2.804  0.00505 ** \n",
      "AAC         -33.859704  14.311744  -2.366  0.01799 *  \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 4581  on 3357  degrees of freedom\n",
      "Residual deviance: 3878  on 3347  degrees of freedom\n",
      "AIC: 3900\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EAS Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  10.043710   5.367357   1.871  0.06131 .  \n",
      "zSCORE        0.340443   0.039498   8.619  < 2e-16 ***\n",
      "SEX          -0.810917   0.079062 -10.257  < 2e-16 ***\n",
      "AGE          -0.048292   0.006002  -8.046 8.57e-16 ***\n",
      "AJ           -3.992740   9.377691  -0.426  0.67028    \n",
      "EUR          -4.789593   6.492841  -0.738  0.46071    \n",
      "EAS          28.143612  10.872301   2.589  0.00964 ** \n",
      "AMR          -6.532551   5.344358  -1.222  0.22158    \n",
      "AFR           3.987078   7.054106   0.565  0.57193    \n",
      "CAS         -27.156917  10.084732  -2.693  0.00708 ** \n",
      "AAC         -31.830296  14.414966  -2.208  0.02723 *  \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 4581.0  on 3357  degrees of freedom\n",
      "Residual deviance: 3837.2  on 3347  degrees of freedom\n",
      "AIC: 3859.2\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EUR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "             Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  11.01599    5.51578   1.997  0.04581 *  \n",
      "zSCORE        0.51093    0.03891  13.130  < 2e-16 ***\n",
      "SEX          -0.83325    0.08063 -10.334  < 2e-16 ***\n",
      "AGE          -0.04830    0.00612  -7.892 2.98e-15 ***\n",
      "AJ           -5.28908    9.60706  -0.551  0.58195    \n",
      "EUR          -4.66430    6.73649  -0.692  0.48869    \n",
      "EAS          26.97213   11.16742   2.415  0.01572 *  \n",
      "AMR          -7.59751    5.49132  -1.384  0.16650    \n",
      "AFR           9.58833    7.32357   1.309  0.19045    \n",
      "CAS         -28.54575   10.51731  -2.714  0.00664 ** \n",
      "AAC         -40.37034   14.67586  -2.751  0.00595 ** \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 4581  on 3357  degrees of freedom\n",
      "Residual deviance: 3719  on 3347  degrees of freedom\n",
      "AIC: 3741\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"AMR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  10.647883   5.325072   2.000  0.04555 *  \n",
      "zSCORE        0.222026   0.040071   5.541 3.01e-08 ***\n",
      "SEX          -0.800674   0.078394 -10.213  < 2e-16 ***\n",
      "AGE          -0.047710   0.005965  -7.998 1.26e-15 ***\n",
      "AJ           -3.382414   9.320358  -0.363  0.71668    \n",
      "EUR          -4.578201   6.486828  -0.706  0.48033    \n",
      "EAS          31.642595  10.810407   2.927  0.00342 ** \n",
      "AMR          -7.251485   5.300587  -1.368  0.17130    \n",
      "AFR           1.531114   7.016972   0.218  0.82727    \n",
      "CAS         -28.214842  10.091994  -2.796  0.00518 ** \n",
      "AAC         -34.133452  14.292676  -2.388  0.01693 *  \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 4581.0  on 3357  degrees of freedom\n",
      "Residual deviance: 3882.8  on 3347  degrees of freedom\n",
      "AIC: 3904.8\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"Merged\"\n",
      "   COHORT              BETA                 SE                    P\n",
      "   <char>            <char>             <char>               <char>\n",
      "1:    AAC 0.224606950662753 0.0377977410300762  2.8098539060261e-09\n",
      "2:    EAS 0.340442513857531 0.0394977844552088 6.73756682802461e-18\n",
      "3:    EUR 0.510930115338691 0.0389134586844898 2.21927269054227e-39\n",
      "4:    AMR 0.222026402782982 0.0400711307171998 3.01080852072074e-08\n",
      "                 L95               U95\n",
      "              <char>            <char>\n",
      "1: 0.150773845041215 0.298982243631374\n",
      "2: 0.263391229247953 0.418267060419719\n",
      "3: 0.435422385636881 0.587999426588036\n",
      "4:  0.14373198026616 0.300856749803952\n",
      "[1] \"------------------------------EAS------------------------------\"\n",
      "[1] \"AAC Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  9.229e+00  3.722e+00   2.480  0.01315 *  \n",
      "zSCORE       2.006e-01  3.531e-02   5.683 1.32e-08 ***\n",
      "SEX          5.555e-01  7.437e-02   7.470 8.04e-14 ***\n",
      "AGE         -6.427e-02  4.583e-03 -14.025  < 2e-16 ***\n",
      "AJ           5.208e+01  8.396e+01   0.620  0.53511    \n",
      "EUR          2.046e+02  2.595e+02   0.788  0.43052    \n",
      "EAS         -5.890e+00  3.565e+00  -1.652  0.09852 .  \n",
      "AMR          3.970e+01  8.693e+00   4.567 4.96e-06 ***\n",
      "AFR         -1.378e+02  3.290e+01  -4.189 2.80e-05 ***\n",
      "CAS          4.196e+01  1.593e+01   2.634  0.00844 ** \n",
      "AAC         -3.442e+02  8.520e+01  -4.040 5.34e-05 ***\n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 5710.5  on 4167  degrees of freedom\n",
      "Residual deviance: 4808.5  on 4157  degrees of freedom\n",
      "AIC: 4830.5\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EAS Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  9.227e+00  3.721e+00   2.479  0.01316 *  \n",
      "zSCORE       2.247e-01  3.611e-02   6.222 4.91e-10 ***\n",
      "SEX          5.543e-01  7.448e-02   7.442 9.90e-14 ***\n",
      "AGE         -6.391e-02  4.592e-03 -13.919  < 2e-16 ***\n",
      "AJ           3.828e+01  8.409e+01   0.455  0.64898    \n",
      "EUR          2.030e+02  2.600e+02   0.781  0.43477    \n",
      "EAS         -5.923e+00  3.565e+00  -1.662  0.09659 .  \n",
      "AMR          4.400e+01  8.723e+00   5.043 4.58e-07 ***\n",
      "AFR         -1.371e+02  3.301e+01  -4.154 3.27e-05 ***\n",
      "CAS          4.607e+01  1.607e+01   2.868  0.00414 ** \n",
      "AAC         -3.610e+02  8.544e+01  -4.225 2.39e-05 ***\n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 5710.5  on 4167  degrees of freedom\n",
      "Residual deviance: 4801.9  on 4157  degrees of freedom\n",
      "AIC: 4823.9\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EUR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  8.525e+00  3.730e+00   2.285   0.0223 *  \n",
      "zSCORE       3.754e-01  3.628e-02  10.346  < 2e-16 ***\n",
      "SEX          5.614e-01  7.523e-02   7.462 8.50e-14 ***\n",
      "AGE         -6.266e-02  4.598e-03 -13.626  < 2e-16 ***\n",
      "AJ           1.932e+01  8.477e+01   0.228   0.8197    \n",
      "EUR          2.115e+02  2.611e+02   0.810   0.4179    \n",
      "EAS         -5.486e+00  3.573e+00  -1.535   0.1247    \n",
      "AMR          4.663e+01  8.784e+00   5.308 1.11e-07 ***\n",
      "AFR         -1.306e+02  3.327e+01  -3.926 8.62e-05 ***\n",
      "CAS          4.539e+01  1.621e+01   2.800   0.0051 ** \n",
      "AAC         -3.375e+02  8.539e+01  -3.952 7.74e-05 ***\n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 5710.5  on 4167  degrees of freedom\n",
      "Residual deviance: 4729.8  on 4157  degrees of freedom\n",
      "AIC: 4751.8\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"AMR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  8.958e+00  3.718e+00   2.409 0.015978 *  \n",
      "zSCORE       1.304e-01  3.526e-02   3.700 0.000216 ***\n",
      "SEX          5.531e-01  7.422e-02   7.452 9.21e-14 ***\n",
      "AGE         -6.454e-02  4.581e-03 -14.087  < 2e-16 ***\n",
      "AJ           4.941e+01  8.393e+01   0.589 0.556025    \n",
      "EUR          1.770e+02  2.602e+02   0.680 0.496326    \n",
      "EAS         -5.692e+00  3.561e+00  -1.598 0.109942    \n",
      "AMR          4.454e+01  8.719e+00   5.108 3.25e-07 ***\n",
      "AFR         -1.357e+02  3.289e+01  -4.128 3.67e-05 ***\n",
      "CAS          4.711e+01  1.609e+01   2.927 0.003421 ** \n",
      "AAC         -3.510e+02  8.488e+01  -4.136 3.54e-05 ***\n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 5710.5  on 4167  degrees of freedom\n",
      "Residual deviance: 4827.5  on 4157  degrees of freedom\n",
      "AIC: 4849.5\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"Merged\"\n",
      "   COHORT              BETA                 SE                    P\n",
      "   <char>            <char>             <char>               <char>\n",
      "1:    AAC 0.200649458121759 0.0353078556102422 1.32463819316207e-08\n",
      "2:    EAS 0.224689299913973  0.036112600821037 4.91143290655222e-10\n",
      "3:    EUR 0.375408260875595 0.0362852613944162   4.362180671024e-25\n",
      "4:    AMR 0.130430689762866 0.0352546078712203 0.000215873366447634\n",
      "                  L95               U95\n",
      "               <char>            <char>\n",
      "1:  0.131633752747551 0.270074131077386\n",
      "2:  0.154121122194319 0.295716660738073\n",
      "3:  0.304631932330386 0.446904488865066\n",
      "4: 0.0614491401412723 0.199679492424282\n",
      "[1] \"------------------------------EUR------------------------------\"\n",
      "[1] \"AAC Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  -0.262968   0.670226  -0.392   0.6948    \n",
      "zSCORE        0.188220   0.012615  14.920  < 2e-16 ***\n",
      "SEX          -0.449514   0.025231 -17.816  < 2e-16 ***\n",
      "AGE          -0.040723   0.001355 -30.059  < 2e-16 ***\n",
      "AJ            7.926051   1.091058   7.265 3.74e-13 ***\n",
      "EUR           4.210430   0.723764   5.817 5.98e-09 ***\n",
      "EAS          77.220530   9.285597   8.316  < 2e-16 ***\n",
      "AMR           5.362209   2.499000   2.146   0.0319 *  \n",
      "AFR         -40.791476   6.435127  -6.339 2.31e-10 ***\n",
      "CAS           3.189239   2.543818   1.254   0.2099    \n",
      "AAC           8.694124   6.508136   1.336   0.1816    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 39127  on 31798  degrees of freedom\n",
      "Residual deviance: 37391  on 31788  degrees of freedom\n",
      "AIC: 37413\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n",
      "[1] \"EAS Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "             Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  -0.19578    0.67252  -0.291   0.7710    \n",
      "zSCORE        0.29093    0.01277  22.787  < 2e-16 ***\n",
      "SEX          -0.45191    0.02536 -17.817  < 2e-16 ***\n",
      "AGE          -0.03990    0.00136 -29.340  < 2e-16 ***\n",
      "AJ            7.65811    1.09394   7.000 2.55e-12 ***\n",
      "EUR           4.06644    0.72618   5.600 2.15e-08 ***\n",
      "EAS          78.06861    9.35649   8.344  < 2e-16 ***\n",
      "AMR           5.13673    2.50250   2.053   0.0401 *  \n",
      "AFR         -40.25694    6.48351  -6.209 5.33e-10 ***\n",
      "CAS           2.45172    2.54592   0.963   0.3355    \n",
      "AAC           7.98407    6.53789   1.221   0.2220    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 39127  on 31798  degrees of freedom\n",
      "Residual deviance: 37084  on 31788  degrees of freedom\n",
      "AIC: 37106\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n",
      "[1] \"EUR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  -0.281653   0.681510  -0.413   0.6794    \n",
      "zSCORE        0.453222   0.013017  34.818  < 2e-16 ***\n",
      "SEX          -0.462060   0.025709 -17.973  < 2e-16 ***\n",
      "AGE          -0.038509   0.001369 -28.125  < 2e-16 ***\n",
      "AJ            7.217118   1.107502   6.517 7.19e-11 ***\n",
      "EUR           4.063576   0.736006   5.521 3.37e-08 ***\n",
      "EAS          74.991717   9.474046   7.915 2.46e-15 ***\n",
      "AMR           4.416884   2.511414   1.759   0.0786 .  \n",
      "AFR         -37.988992   6.529425  -5.818 5.95e-09 ***\n",
      "CAS           2.811469   2.590203   1.085   0.2777    \n",
      "AAC           8.384224   6.602195   1.270   0.2041    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 39127  on 31798  degrees of freedom\n",
      "Residual deviance: 36298  on 31788  degrees of freedom\n",
      "AIC: 36320\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n",
      "[1] \"AMR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)  -0.172114   0.672924  -0.256   0.7981    \n",
      "zSCORE        0.233004   0.012633  18.444  < 2e-16 ***\n",
      "SEX          -0.451118   0.025284 -17.842  < 2e-16 ***\n",
      "AGE          -0.040614   0.001356 -29.945  < 2e-16 ***\n",
      "AJ            7.601126   1.094394   6.946 3.77e-12 ***\n",
      "EUR           4.123630   0.726726   5.674 1.39e-08 ***\n",
      "EAS          76.932308   9.364058   8.216  < 2e-16 ***\n",
      "AMR           5.593540   2.499830   2.238   0.0252 *  \n",
      "AFR         -41.194580   6.471666  -6.365 1.95e-10 ***\n",
      "CAS           2.281276   2.531386   0.901   0.3675    \n",
      "AAC           8.562955   6.516335   1.314   0.1888    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 39127  on 31798  degrees of freedom\n",
      "Residual deviance: 37271  on 31788  degrees of freedom\n",
      "AIC: 37293\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n",
      "[1] \"Merged\"\n",
      "   COHORT              BETA                 SE                     P\n",
      "   <char>            <char>             <char>                <char>\n",
      "1:    AAC 0.188219765331436 0.0126152701406985  2.44292294837477e-50\n",
      "2:    EAS 0.290925966566423 0.0127674236214408 6.22895169548478e-115\n",
      "3:    EUR 0.453221668206387 0.0130169652974334  1.3097662216067e-265\n",
      "4:    AMR 0.233004480420535 0.0126333220090325  5.86530732038722e-76\n",
      "                 L95               U95\n",
      "              <char>            <char>\n",
      "1: 0.163518144017601 0.212970657488376\n",
      "2: 0.265936994034142 0.315985964234221\n",
      "3: 0.427776666386605 0.478803484022938\n",
      "4: 0.208272171551065  0.25779548511549\n",
      "[1] \"------------------------------CAS------------------------------\"\n",
      "[1] \"AAC Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "             Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   1.88146    1.51968   1.238  0.21569    \n",
      "zSCORE        0.21659    0.07637   2.836  0.00457 ** \n",
      "SEX          -0.29832    0.15800  -1.888  0.05902 .  \n",
      "AGE          -0.01621    0.00903  -1.795  0.07262 .  \n",
      "AJ           13.36690    6.01553   2.222  0.02628 *  \n",
      "EUR          11.11234   31.69166   0.351  0.72586    \n",
      "EAS          -0.83907    1.46650  -0.572  0.56722    \n",
      "AMR           9.91390    3.77353   2.627  0.00861 ** \n",
      "AFR          -5.85648   13.30779  -0.440  0.65988    \n",
      "CAS         -11.16264    2.83536  -3.937 8.25e-05 ***\n",
      "AAC          21.19844   20.26247   1.046  0.29547    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1179.4  on 904  degrees of freedom\n",
      "Residual deviance: 1063.3  on 894  degrees of freedom\n",
      "AIC: 1085.3\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EAS Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   1.867883   1.510686   1.236 0.216292    \n",
      "zSCORE        0.221514   0.074888   2.958 0.003097 ** \n",
      "SEX          -0.310961   0.158083  -1.967 0.049174 *  \n",
      "AGE          -0.016920   0.009017  -1.876 0.060594 .  \n",
      "AJ           13.386937   6.003386   2.230 0.025754 *  \n",
      "EUR          14.532682  31.176894   0.466 0.641118    \n",
      "EAS          -0.715125   1.461304  -0.489 0.624576    \n",
      "AMR           9.719195   3.763001   2.583 0.009799 ** \n",
      "AFR          -7.696491  13.250599  -0.581 0.561348    \n",
      "CAS         -10.747527   2.829762  -3.798 0.000146 ***\n",
      "AAC          20.641802  20.125746   1.026 0.305061    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1179.4  on 904  degrees of freedom\n",
      "Residual deviance: 1062.6  on 894  degrees of freedom\n",
      "AIC: 1084.6\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"EUR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "             Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   1.73141    1.52469   1.136 0.256132    \n",
      "zSCORE        0.33852    0.07755   4.365 1.27e-05 ***\n",
      "SEX          -0.30171    0.15906  -1.897 0.057853 .  \n",
      "AGE          -0.01624    0.00908  -1.788 0.073726 .  \n",
      "AJ           13.18676    6.04443   2.182 0.029136 *  \n",
      "EUR          18.36963   31.74699   0.579 0.562842    \n",
      "EAS          -0.88334    1.46945  -0.601 0.547748    \n",
      "AMR          10.04484    3.79248   2.649 0.008082 ** \n",
      "AFR          -4.86153   13.39448  -0.363 0.716642    \n",
      "CAS         -10.63558    2.84016  -3.745 0.000181 ***\n",
      "AAC          17.20076   20.33942   0.846 0.397728    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1179.4  on 904  degrees of freedom\n",
      "Residual deviance: 1051.9  on 894  degrees of freedom\n",
      "AIC: 1073.9\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"AMR Summary stats\"\n",
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + \n",
      "    AFR + CAS + AAC, family = \"binomial\", data = cases)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   1.926841   1.523536   1.265 0.205973    \n",
      "zSCORE        0.278911   0.076485   3.647 0.000266 ***\n",
      "SEX          -0.308708   0.158309  -1.950 0.051172 .  \n",
      "AGE          -0.016719   0.009042  -1.849 0.064445 .  \n",
      "AJ           12.682331   6.038045   2.100 0.035693 *  \n",
      "EUR          22.024657  31.846632   0.692 0.489198    \n",
      "EAS          -1.087354   1.469699  -0.740 0.459392    \n",
      "AMR           9.883959   3.762518   2.627 0.008615 ** \n",
      "AFR          -5.803455  13.374618  -0.434 0.664350    \n",
      "CAS         -10.668429   2.833143  -3.766 0.000166 ***\n",
      "AAC          15.652453  20.333420   0.770 0.441425    \n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 1179.4  on 904  degrees of freedom\n",
      "Residual deviance: 1057.9  on 894  degrees of freedom\n",
      "AIC: 1079.9\n",
      "\n",
      "Number of Fisher Scoring iterations: 4\n",
      "\n",
      "[1] \"Merged\"\n",
      "   COHORT              BETA                 SE                    P\n",
      "   <char>            <char>             <char>               <char>\n",
      "1:    AAC 0.216593550093067 0.0763678992795242  0.00456558794568548\n",
      "2:    EAS  0.22151406450344 0.0748881412682289  0.00309709967406541\n",
      "3:    EUR 0.338520005896278 0.0775447668744626 1.26844828736911e-05\n",
      "4:    AMR 0.278911377527037 0.0764850040521075 0.000265717391562227\n",
      "                  L95               U95\n",
      "               <char>            <char>\n",
      "1: 0.0678520653841698 0.367536717079796\n",
      "2: 0.0754776014723491 0.369342376020493\n",
      "3:  0.187625374540705   0.4919331562424\n",
      "4:   0.12993242120127 0.430081539775116\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n",
       "Waiting for profiling to be done...\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "for (ancestry in ancestries) {\n",
    "    print(paste0(\"------------------------------\", ancestry, \"------------------------------\"))\n",
    "    forest <- list()\n",
    "    i <- 1\n",
    "    for (summstat_ancestry in summstat_ancestries) {\n",
    "        print(paste0(summstat_ancestry, \" Summary stats\"))\n",
    "        \n",
    "        temp_data <- read.table(paste0(RESULTS_DIR, \"/\", ancestry, \"/model1_\", summstat_ancestry, \".profile\"), header = T) \n",
    "        temp_covs <- read.table(paste0(RESULTS_DIR, \"/covariates.txt\"), header = T, sep=\" \")\n",
    "        admixture <- read.table(paste0(REL9_DIR, \"/meta_data/admixture/GP2_release9_labeled_admixture_vwb.txt\"), header = T, sep=\"\\t\")\n",
    "        colnames(temp_covs) <- c(\"FID\", \"IID\", \"SEX\", \"AGE\", \"ANCESTRY\", \"PC1\", \"PC2\", \"PC3\", \"PC4\", \"PC5\", \"PC6\", \"PC7\", \"PC8\", \"PC9\", \"PC10\")\n",
    "        covs <- merge(temp_covs, admixture, by = \"IID\")\n",
    "        data <- merge(temp_data, covs, by = \"IID\")\n",
    "        \n",
    "        data$CASE <- data$PHENO - 1\n",
    "        cases <- subset(data, CASE != -10)\n",
    "        \n",
    "        meanPop <- mean(subset(data, CASE == 0)$SCORE)\n",
    "        sdPop <- sd(subset(data, CASE == 0)$SCORE)\n",
    "        cases$zSCORE <- (cases$SCORE - meanPop)/sdPop\n",
    "        \n",
    "        grsTests <- glm(CASE ~ zSCORE + SEX + AGE + AJ + EUR + EAS + AMR + AFR + CAS + AAC, data = cases, family='binomial')\n",
    "        print(summary(grsTests))\n",
    "        grs_summary <- summary(grsTests)$coefficients\n",
    "        conf_intervals <- confint(grsTests, level = 0.95)\n",
    "        zscore_summary <- grs_summary[\"zSCORE\", , drop = FALSE]\n",
    "        zscore_conf <- conf_intervals[\"zSCORE\", , drop = FALSE]\n",
    "        zscore <- as.data.table(cbind(summstat_ancestry, zscore_summary, zscore_conf))\n",
    "        colnames(zscore) <- c(\"COHORT\", \"BETA\", \"SE\", \"T\", \"P\", \"L95\", \"U95\")\n",
    "        zscore <- zscore[, !\"T\", with = FALSE]\n",
    "        forest <- rbind(forest, zscore)\n",
    "        i <- i + 1\n",
    "    }\n",
    "    print(\"Merged\")\n",
    "    print(forest)\n",
    "    fwrite(forest, paste0(RESULTS_DIR, \"/\", ancestry, \"/forest_risk_admixture_\", ancestry, \".txt\"), sep = \"\\t\", col.names = TRUE, quote = FALSE)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "698c4df9-8e40-4e46-88d9-5b033c60966b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Compare PCs vs Admixture"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6bb7511-bc0d-4b42-8c74-09f6606b72c1",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "2cb10b38-476b-4520-b433-9fe37417ebee",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_heatmap = []\n",
    "for ancestry in [\"EUR\",\"AJ\",\"AFR\",\"AAC\",\"EAS\",\"AMR\",\"CAS\"]:\n",
    "    df_pcs = pd.read_csv(f\"{RESULTS_DIR}/{ancestry}/forest_risk_pcs_{ancestry}.txt\", sep=\"\\t\")\n",
    "    df_pcs[ancestry] = df_pcs[\"BETA\"] / df_pcs[\"SE\"]\n",
    "    df_pcs = df_pcs[[\"COHORT\", ancestry]]\n",
    "    df_pcs = df_pcs.set_index(\"COHORT\").T\n",
    "    df_pcs.rename(columns={\"AAC\":\"AFR\"}, inplace=True)\n",
    "    df_pcs = df_pcs.add_prefix(\"PCs adjusted - \")\n",
    "\n",
    "    df_admixture = pd.read_csv(f\"{RESULTS_DIR}/{ancestry}/forest_risk_admixture_{ancestry}.txt\", sep=\"\\t\")\n",
    "    df_admixture[ancestry] = df_admixture[\"BETA\"] / df_admixture[\"SE\"]\n",
    "    df_admixture = df_admixture[[\"COHORT\", ancestry]]\n",
    "    df_admixture = df_admixture.set_index(\"COHORT\").T\n",
    "    df_admixture.rename(columns={\"AAC\":\"AFR\"}, inplace=True)\n",
    "    df_admixture = df_admixture.add_prefix(\"% admixture adjusted - \")\n",
    "    \n",
    "    df_merged = pd.concat([df_pcs, df_admixture], axis=1)\n",
    "    df_merged = df_merged[[\n",
    "        \"PCs adjusted - EUR\",\n",
    "        \"% admixture adjusted - EUR\",\n",
    "        \"PCs adjusted - AMR\",\n",
    "        \"% admixture adjusted - AMR\",\n",
    "        \"PCs adjusted - EAS\",\n",
    "        \"% admixture adjusted - EAS\",\n",
    "        \"PCs adjusted - AFR\",\n",
    "        \"% admixture adjusted - AFR\",\n",
    "    ]]\n",
    "    df_heatmap.append(df_merged)\n",
    "    \n",
    "df_heatmap = pd.concat(df_heatmap)\n",
    "df_heatmap = df_heatmap.rename_axis(\"Ancestry\")\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a06089bc-f98f-4863-b1d8-5574f9087ee2",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "source": [
    "# Create heatmap with custom color palette\n",
    "max_val = np.max(df_heatmap.values)\n",
    "cmap = ListedColormap(sns.color_palette(\"Spectral_r\", n_colors=math.ceil(max_val)*64))\n",
    "norm = matplotlib.colors.Normalize(vmin=0, vmax=math.ceil(max_val))\n",
    "ax = sns.heatmap(df_heatmap, cmap=cmap, yticklabels=df_heatmap.index.values, norm=norm, mask=df_heatmap < 0)\n",
    "ax.set_ylabel('Ancestry')\n",
    "plt.savefig(f\"{RESULTS_DIR}/supp_fig4.png\", dpi=300, bbox_inches='tight')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8426d860-6075-4704-8600-7aab0dbe0bf2",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "51769beb-52f5-4ec7-926e-77559bbb40c5",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_comparison = []\n",
    "for ancestry in [\"EUR\",\"AJ\",\"AFR\",\"AAC\",\"EAS\",\"AMR\",\"CAS\"]:\n",
    "    df_pcs = pd.read_csv(f\"{RESULTS_DIR}/{ancestry}/forest_risk_pcs_{ancestry}.txt\", sep=\"\\t\")\n",
    "    df_pcs_copy = df_pcs.copy()\n",
    "    df_pcs[\"OR PCs\"] = df_pcs.apply(lambda row: f\"{math.e ** row['BETA']:.2f} ({math.e ** row['L95']:.2f} - {math.e ** row['U95']:.2f})\", axis=1)\n",
    "    df_pcs.rename(columns={\"COHORT\":\"Summary Statistics\", \"P\":\"p-value PCs\"}, inplace=True)\n",
    "    df_pcs[\"Cohort\"] = ancestry\n",
    "    df_pcs = df_pcs[[\"Cohort\",\"Summary Statistics\",\"OR PCs\",\"p-value PCs\"]]\n",
    "    \n",
    "    df_admixture = pd.read_csv(f\"{RESULTS_DIR}/{ancestry}/forest_risk_admixture_{ancestry}.txt\", sep=\"\\t\")\n",
    "    df_admixture_copy = df_admixture.copy()\n",
    "    df_admixture[\"OR Admixture\"] = df_admixture.apply(lambda row: f\"{math.e ** row['BETA']:.2f} ({math.e ** row['L95']:.2f} - {math.e ** row['U95']:.2f})\", axis=1)\n",
    "    df_admixture.rename(columns={\"COHORT\":\"Summary Statistics\", \"P\":\"p-value Admixture\"}, inplace=True)\n",
    "    df_admixture[\"Cohort\"] = ancestry\n",
    "    df_admixture = df_admixture[[\"Cohort\",\"Summary Statistics\",\"OR Admixture\",\"p-value Admixture\"]]\n",
    "    \n",
    "    df_merged = df_pcs.merge(df_admixture, on=[\"Cohort\",\"Summary Statistics\"])\n",
    "    df_merged = df_merged[[\n",
    "        \"Cohort\",\n",
    "        \"Summary Statistics\",\n",
    "        \"OR PCs\",\n",
    "        \"OR Admixture\",\n",
    "        \"p-value PCs\",\n",
    "        \"p-value Admixture\",\n",
    "    ]]\n",
    "    \n",
    "    se = np.sqrt(df_pcs_copy[\"SE\"]**2 + df_admixture_copy[\"SE\"]**2)\n",
    "    zsc = (df_pcs_copy[\"BETA\"] - df_admixture_copy[\"BETA\"]) / se\n",
    "    p_val = 2 * (1 - norm.cdf(abs(zsc)))\n",
    "    df_merged[\"Z-test p-value\"] = p_val\n",
    "    \n",
    "    df_comparison.append(df_merged)\n",
    "\n",
    "df_comparison = pd.concat(df_comparison)\n",
    "df_comparison.to_csv(f\"{RESULTS_DIR}/supp_tab4.txt\", index=False, sep=\"\\t\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "208eefc4-987f-41ac-bb56-c8b23f8de342",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COHORT</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>L95</th>\n",
       "      <th>U95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAC</td>\n",
       "      <td>0.190180</td>\n",
       "      <td>0.012637</td>\n",
       "      <td>3.491510e-51</td>\n",
       "      <td>0.165435</td>\n",
       "      <td>0.214973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EAS</td>\n",
       "      <td>0.291309</td>\n",
       "      <td>0.012783</td>\n",
       "      <td>5.909320e-115</td>\n",
       "      <td>0.266289</td>\n",
       "      <td>0.316399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>0.454128</td>\n",
       "      <td>0.013027</td>\n",
       "      <td>3.014658e-266</td>\n",
       "      <td>0.428663</td>\n",
       "      <td>0.479730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>0.233717</td>\n",
       "      <td>0.012649</td>\n",
       "      <td>3.149376e-76</td>\n",
       "      <td>0.208954</td>\n",
       "      <td>0.258538</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  COHORT      BETA        SE              P       L95       U95\n",
       "0    AAC  0.190180  0.012637   3.491510e-51  0.165435  0.214973\n",
       "1    EAS  0.291309  0.012783  5.909320e-115  0.266289  0.316399\n",
       "2    EUR  0.454128  0.013027  3.014658e-266  0.428663  0.479730\n",
       "3    AMR  0.233717  0.012649   3.149376e-76  0.208954  0.258538"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COHORT</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>L95</th>\n",
       "      <th>U95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAC</td>\n",
       "      <td>0.188220</td>\n",
       "      <td>0.012615</td>\n",
       "      <td>2.442923e-50</td>\n",
       "      <td>0.163518</td>\n",
       "      <td>0.212971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EAS</td>\n",
       "      <td>0.290926</td>\n",
       "      <td>0.012767</td>\n",
       "      <td>6.228952e-115</td>\n",
       "      <td>0.265937</td>\n",
       "      <td>0.315986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>0.453222</td>\n",
       "      <td>0.013017</td>\n",
       "      <td>1.309766e-265</td>\n",
       "      <td>0.427777</td>\n",
       "      <td>0.478803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>0.233004</td>\n",
       "      <td>0.012633</td>\n",
       "      <td>5.865307e-76</td>\n",
       "      <td>0.208272</td>\n",
       "      <td>0.257795</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  COHORT      BETA        SE              P       L95       U95\n",
       "0    AAC  0.188220  0.012615   2.442923e-50  0.163518  0.212971\n",
       "1    EAS  0.290926  0.012767  6.228952e-115  0.265937  0.315986\n",
       "2    EUR  0.453222  0.013017  1.309766e-265  0.427777  0.478803\n",
       "3    AMR  0.233004  0.012633   5.865307e-76  0.208272  0.257795"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COHORT</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>L95</th>\n",
       "      <th>U95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAC</td>\n",
       "      <td>0.171124</td>\n",
       "      <td>0.054650</td>\n",
       "      <td>1.740384e-03</td>\n",
       "      <td>0.064316</td>\n",
       "      <td>0.278677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EAS</td>\n",
       "      <td>0.155076</td>\n",
       "      <td>0.054728</td>\n",
       "      <td>4.603042e-03</td>\n",
       "      <td>0.048038</td>\n",
       "      <td>0.262687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>0.504081</td>\n",
       "      <td>0.055957</td>\n",
       "      <td>2.091678e-19</td>\n",
       "      <td>0.396679</td>\n",
       "      <td>0.616080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>0.127315</td>\n",
       "      <td>0.054657</td>\n",
       "      <td>1.984090e-02</td>\n",
       "      <td>0.020337</td>\n",
       "      <td>0.234715</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  COHORT      BETA        SE             P       L95       U95\n",
       "0    AAC  0.171124  0.054650  1.740384e-03  0.064316  0.278677\n",
       "1    EAS  0.155076  0.054728  4.603042e-03  0.048038  0.262687\n",
       "2    EUR  0.504081  0.055957  2.091678e-19  0.396679  0.616080\n",
       "3    AMR  0.127315  0.054657  1.984090e-02  0.020337  0.234715"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COHORT</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>L95</th>\n",
       "      <th>U95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAC</td>\n",
       "      <td>0.182991</td>\n",
       "      <td>0.054739</td>\n",
       "      <td>8.288801e-04</td>\n",
       "      <td>0.076043</td>\n",
       "      <td>0.290751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EAS</td>\n",
       "      <td>0.158086</td>\n",
       "      <td>0.054601</td>\n",
       "      <td>3.787798e-03</td>\n",
       "      <td>0.051317</td>\n",
       "      <td>0.265465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>0.505984</td>\n",
       "      <td>0.055984</td>\n",
       "      <td>1.596651e-19</td>\n",
       "      <td>0.398574</td>\n",
       "      <td>0.618078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>0.126016</td>\n",
       "      <td>0.054461</td>\n",
       "      <td>2.067360e-02</td>\n",
       "      <td>0.019423</td>\n",
       "      <td>0.233028</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  COHORT      BETA        SE             P       L95       U95\n",
       "0    AAC  0.182991  0.054739  8.288801e-04  0.076043  0.290751\n",
       "1    EAS  0.158086  0.054601  3.787798e-03  0.051317  0.265465\n",
       "2    EUR  0.505984  0.055984  1.596651e-19  0.398574  0.618078\n",
       "3    AMR  0.126016  0.054461  2.067360e-02  0.019423  0.233028"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COHORT</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>L95</th>\n",
       "      <th>U95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAC</td>\n",
       "      <td>0.285881</td>\n",
       "      <td>0.046961</td>\n",
       "      <td>1.145591e-09</td>\n",
       "      <td>0.194264</td>\n",
       "      <td>0.378413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EAS</td>\n",
       "      <td>0.250155</td>\n",
       "      <td>0.044722</td>\n",
       "      <td>2.224774e-08</td>\n",
       "      <td>0.162846</td>\n",
       "      <td>0.338224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>0.289287</td>\n",
       "      <td>0.046477</td>\n",
       "      <td>4.835645e-10</td>\n",
       "      <td>0.198569</td>\n",
       "      <td>0.380824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>0.087905</td>\n",
       "      <td>0.045079</td>\n",
       "      <td>5.117356e-02</td>\n",
       "      <td>-0.000366</td>\n",
       "      <td>0.176407</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  COHORT      BETA        SE             P       L95       U95\n",
       "0    AAC  0.285881  0.046961  1.145591e-09  0.194264  0.378413\n",
       "1    EAS  0.250155  0.044722  2.224774e-08  0.162846  0.338224\n",
       "2    EUR  0.289287  0.046477  4.835645e-10  0.198569  0.380824\n",
       "3    AMR  0.087905  0.045079  5.117356e-02 -0.000366  0.176407"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COHORT</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>L95</th>\n",
       "      <th>U95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAC</td>\n",
       "      <td>0.249585</td>\n",
       "      <td>0.045727</td>\n",
       "      <td>4.812328e-08</td>\n",
       "      <td>0.160313</td>\n",
       "      <td>0.339626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EAS</td>\n",
       "      <td>0.215792</td>\n",
       "      <td>0.043870</td>\n",
       "      <td>8.702441e-07</td>\n",
       "      <td>0.130099</td>\n",
       "      <td>0.302134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>0.256924</td>\n",
       "      <td>0.045502</td>\n",
       "      <td>1.638337e-08</td>\n",
       "      <td>0.168065</td>\n",
       "      <td>0.346501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>0.079476</td>\n",
       "      <td>0.044332</td>\n",
       "      <td>7.301321e-02</td>\n",
       "      <td>-0.007347</td>\n",
       "      <td>0.166495</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  COHORT      BETA        SE             P       L95       U95\n",
       "0    AAC  0.249585  0.045727  4.812328e-08  0.160313  0.339626\n",
       "1    EAS  0.215792  0.043870  8.702441e-07  0.130099  0.302134\n",
       "2    EUR  0.256924  0.045502  1.638337e-08  0.168065  0.346501\n",
       "3    AMR  0.079476  0.044332  7.301321e-02 -0.007347  0.166495"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COHORT</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>L95</th>\n",
       "      <th>U95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAC</td>\n",
       "      <td>0.199900</td>\n",
       "      <td>0.071590</td>\n",
       "      <td>5.233867e-03</td>\n",
       "      <td>0.059854</td>\n",
       "      <td>0.340751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EAS</td>\n",
       "      <td>0.306921</td>\n",
       "      <td>0.075080</td>\n",
       "      <td>4.352844e-05</td>\n",
       "      <td>0.160714</td>\n",
       "      <td>0.455306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>0.438817</td>\n",
       "      <td>0.074292</td>\n",
       "      <td>3.490723e-09</td>\n",
       "      <td>0.294534</td>\n",
       "      <td>0.586067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>0.192475</td>\n",
       "      <td>0.072909</td>\n",
       "      <td>8.292548e-03</td>\n",
       "      <td>0.050045</td>\n",
       "      <td>0.336103</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  COHORT      BETA        SE             P       L95       U95\n",
       "0    AAC  0.199900  0.071590  5.233867e-03  0.059854  0.340751\n",
       "1    EAS  0.306921  0.075080  4.352844e-05  0.160714  0.455306\n",
       "2    EUR  0.438817  0.074292  3.490723e-09  0.294534  0.586067\n",
       "3    AMR  0.192475  0.072909  8.292548e-03  0.050045  0.336103"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COHORT</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>L95</th>\n",
       "      <th>U95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAC</td>\n",
       "      <td>0.190725</td>\n",
       "      <td>0.071355</td>\n",
       "      <td>7.520088e-03</td>\n",
       "      <td>0.051158</td>\n",
       "      <td>0.331136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EAS</td>\n",
       "      <td>0.299387</td>\n",
       "      <td>0.074917</td>\n",
       "      <td>6.434816e-05</td>\n",
       "      <td>0.153495</td>\n",
       "      <td>0.447439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>0.436627</td>\n",
       "      <td>0.074512</td>\n",
       "      <td>4.633087e-09</td>\n",
       "      <td>0.291954</td>\n",
       "      <td>0.584324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>0.173210</td>\n",
       "      <td>0.072632</td>\n",
       "      <td>1.709014e-02</td>\n",
       "      <td>0.031307</td>\n",
       "      <td>0.316281</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  COHORT      BETA        SE             P       L95       U95\n",
       "0    AAC  0.190725  0.071355  7.520088e-03  0.051158  0.331136\n",
       "1    EAS  0.299387  0.074917  6.434816e-05  0.153495  0.447439\n",
       "2    EUR  0.436627  0.074512  4.633087e-09  0.291954  0.584324\n",
       "3    AMR  0.173210  0.072632  1.709014e-02  0.031307  0.316281"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COHORT</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>L95</th>\n",
       "      <th>U95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAC</td>\n",
       "      <td>0.150946</td>\n",
       "      <td>0.040951</td>\n",
       "      <td>2.277589e-04</td>\n",
       "      <td>0.070827</td>\n",
       "      <td>0.231395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EAS</td>\n",
       "      <td>0.239424</td>\n",
       "      <td>0.041741</td>\n",
       "      <td>9.700467e-09</td>\n",
       "      <td>0.157860</td>\n",
       "      <td>0.321530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>0.383342</td>\n",
       "      <td>0.041613</td>\n",
       "      <td>3.197596e-20</td>\n",
       "      <td>0.302164</td>\n",
       "      <td>0.465329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>0.131139</td>\n",
       "      <td>0.040924</td>\n",
       "      <td>1.353358e-03</td>\n",
       "      <td>0.051071</td>\n",
       "      <td>0.211538</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  COHORT      BETA        SE             P       L95       U95\n",
       "0    AAC  0.150946  0.040951  2.277589e-04  0.070827  0.231395\n",
       "1    EAS  0.239424  0.041741  9.700467e-09  0.157860  0.321530\n",
       "2    EUR  0.383342  0.041613  3.197596e-20  0.302164  0.465329\n",
       "3    AMR  0.131139  0.040924  1.353358e-03  0.051071  0.211538"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COHORT</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>L95</th>\n",
       "      <th>U95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAC</td>\n",
       "      <td>0.200649</td>\n",
       "      <td>0.035308</td>\n",
       "      <td>1.324638e-08</td>\n",
       "      <td>0.131634</td>\n",
       "      <td>0.270074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EAS</td>\n",
       "      <td>0.224689</td>\n",
       "      <td>0.036113</td>\n",
       "      <td>4.911433e-10</td>\n",
       "      <td>0.154121</td>\n",
       "      <td>0.295717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>0.375408</td>\n",
       "      <td>0.036285</td>\n",
       "      <td>4.362181e-25</td>\n",
       "      <td>0.304632</td>\n",
       "      <td>0.446904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>0.130431</td>\n",
       "      <td>0.035255</td>\n",
       "      <td>2.158734e-04</td>\n",
       "      <td>0.061449</td>\n",
       "      <td>0.199679</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  COHORT      BETA        SE             P       L95       U95\n",
       "0    AAC  0.200649  0.035308  1.324638e-08  0.131634  0.270074\n",
       "1    EAS  0.224689  0.036113  4.911433e-10  0.154121  0.295717\n",
       "2    EUR  0.375408  0.036285  4.362181e-25  0.304632  0.446904\n",
       "3    AMR  0.130431  0.035255  2.158734e-04  0.061449  0.199679"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COHORT</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>L95</th>\n",
       "      <th>U95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAC</td>\n",
       "      <td>0.191868</td>\n",
       "      <td>0.040147</td>\n",
       "      <td>1.760799e-06</td>\n",
       "      <td>0.113388</td>\n",
       "      <td>0.270813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EAS</td>\n",
       "      <td>0.369734</td>\n",
       "      <td>0.042551</td>\n",
       "      <td>3.653002e-18</td>\n",
       "      <td>0.286765</td>\n",
       "      <td>0.453622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>0.513111</td>\n",
       "      <td>0.041435</td>\n",
       "      <td>3.213652e-35</td>\n",
       "      <td>0.432739</td>\n",
       "      <td>0.595211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>0.306740</td>\n",
       "      <td>0.043393</td>\n",
       "      <td>1.561822e-12</td>\n",
       "      <td>0.222023</td>\n",
       "      <td>0.392178</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  COHORT      BETA        SE             P       L95       U95\n",
       "0    AAC  0.191868  0.040147  1.760799e-06  0.113388  0.270813\n",
       "1    EAS  0.369734  0.042551  3.653002e-18  0.286765  0.453622\n",
       "2    EUR  0.513111  0.041435  3.213652e-35  0.432739  0.595211\n",
       "3    AMR  0.306740  0.043393  1.561822e-12  0.222023  0.392178"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COHORT</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>L95</th>\n",
       "      <th>U95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAC</td>\n",
       "      <td>0.224607</td>\n",
       "      <td>0.037798</td>\n",
       "      <td>2.809854e-09</td>\n",
       "      <td>0.150774</td>\n",
       "      <td>0.298982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EAS</td>\n",
       "      <td>0.340443</td>\n",
       "      <td>0.039498</td>\n",
       "      <td>6.737567e-18</td>\n",
       "      <td>0.263391</td>\n",
       "      <td>0.418267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>0.510930</td>\n",
       "      <td>0.038913</td>\n",
       "      <td>2.219273e-39</td>\n",
       "      <td>0.435422</td>\n",
       "      <td>0.587999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>0.222026</td>\n",
       "      <td>0.040071</td>\n",
       "      <td>3.010809e-08</td>\n",
       "      <td>0.143732</td>\n",
       "      <td>0.300857</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  COHORT      BETA        SE             P       L95       U95\n",
       "0    AAC  0.224607  0.037798  2.809854e-09  0.150774  0.298982\n",
       "1    EAS  0.340443  0.039498  6.737567e-18  0.263391  0.418267\n",
       "2    EUR  0.510930  0.038913  2.219273e-39  0.435422  0.587999\n",
       "3    AMR  0.222026  0.040071  3.010809e-08  0.143732  0.300857"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COHORT</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>L95</th>\n",
       "      <th>U95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAC</td>\n",
       "      <td>0.226077</td>\n",
       "      <td>0.077404</td>\n",
       "      <td>0.003492</td>\n",
       "      <td>0.075344</td>\n",
       "      <td>0.379083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EAS</td>\n",
       "      <td>0.210761</td>\n",
       "      <td>0.076653</td>\n",
       "      <td>0.005968</td>\n",
       "      <td>0.061196</td>\n",
       "      <td>0.361994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>0.364008</td>\n",
       "      <td>0.079144</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.210088</td>\n",
       "      <td>0.520669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>0.295371</td>\n",
       "      <td>0.077121</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.145242</td>\n",
       "      <td>0.447890</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  COHORT      BETA        SE         P       L95       U95\n",
       "0    AAC  0.226077  0.077404  0.003492  0.075344  0.379083\n",
       "1    EAS  0.210761  0.076653  0.005968  0.061196  0.361994\n",
       "2    EUR  0.364008  0.079144  0.000004  0.210088  0.520669\n",
       "3    AMR  0.295371  0.077121  0.000128  0.145242  0.447890"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COHORT</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>L95</th>\n",
       "      <th>U95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAC</td>\n",
       "      <td>0.216594</td>\n",
       "      <td>0.076368</td>\n",
       "      <td>0.004566</td>\n",
       "      <td>0.067852</td>\n",
       "      <td>0.367537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EAS</td>\n",
       "      <td>0.221514</td>\n",
       "      <td>0.074888</td>\n",
       "      <td>0.003097</td>\n",
       "      <td>0.075478</td>\n",
       "      <td>0.369342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>0.338520</td>\n",
       "      <td>0.077545</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.187625</td>\n",
       "      <td>0.491933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>0.278911</td>\n",
       "      <td>0.076485</td>\n",
       "      <td>0.000266</td>\n",
       "      <td>0.129932</td>\n",
       "      <td>0.430082</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  COHORT      BETA        SE         P       L95       U95\n",
       "0    AAC  0.216594  0.076368  0.004566  0.067852  0.367537\n",
       "1    EAS  0.221514  0.074888  0.003097  0.075478  0.369342\n",
       "2    EUR  0.338520  0.077545  0.000013  0.187625  0.491933\n",
       "3    AMR  0.278911  0.076485  0.000266  0.129932  0.430082"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "for ancestry in [\"EUR\",\"AJ\",\"AFR\",\"AAC\",\"EAS\",\"AMR\",\"CAS\"]:\n",
    "    df_pcs = pd.read_csv(f\"{RESULTS_DIR}/{ancestry}/forest_risk_pcs_{ancestry}.txt\", sep=\"\\t\")\n",
    "    display(df_pcs.head())\n",
    "    df_admixture = pd.read_csv(f\"{RESULTS_DIR}/{ancestry}/forest_risk_admixture_{ancestry}.txt\", sep=\"\\t\")\n",
    "    display(df_admixture.head())\n",
    "    print()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad312b4d-1bb1-4a6a-a17c-05caf97bf956",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Plot ROC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "4d7be4c1-04a5-4d00-8965-e26403f2653c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"--------------------AAC--------------------\"\n",
      "[1] \"--------------------AFR--------------------\"\n",
      "[1] \"--------------------AJ--------------------\"\n",
      "[1] \"--------------------AMR--------------------\"\n",
      "[1] \"--------------------EAS--------------------\"\n",
      "[1] \"--------------------EUR--------------------\"\n",
      "[1] \"--------------------CAS--------------------\"\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "for (ancestry in ancestries) {\n",
    "    print(paste0(\"--------------------\", ancestry, \"--------------------\"))\n",
    "    to_plot <- list()\n",
    "    for (summstat_ancestry in summstat_ancestries) {\n",
    "        data <- read.table(paste0(RESULTS_DIR, \"/\", ancestry, \"/model1_\", summstat_ancestry, \".profile\"), header = T) \n",
    "        data$CASE <- data$PHENO - 1\n",
    "        dat <- subset(data, CASE != -10)\n",
    "        Model <- glm(CASE ~ SCORE, data = dat, family = 'binomial')\n",
    "        dat$probDisease <- predict(Model, dat, type = c(\"response\"))\n",
    "        dat$predicted <- ifelse(dat$probDisease > 0.5, \"DISEASE\", \"CONTROL\")\n",
    "        dat$reported <- ifelse(dat$CASE == 1, \"DISEASE\",\"CONTROL\")\n",
    "        dat$Group <- summstat_ancestry\n",
    "        to_plot <- rbind(to_plot, dat)\n",
    "    }\n",
    "    write.table(to_plot, file = paste0(RESULTS_DIR, \"/\", ancestry, \"/model1_risk_results.txt\"), sep = \"\\t\", quote = FALSE, row.names = FALSE)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "1916b435-fb0f-4d20-8b11-d771a3d58ba0",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------- AAC --------------------\n",
      "-------------------- AFR --------------------\n",
      "-------------------- AJ --------------------\n",
      "-------------------- AMR --------------------\n",
      "-------------------- EAS --------------------\n",
      "-------------------- EUR --------------------\n",
      "-------------------- CAS --------------------\n"
     ]
    }
   ],
   "source": [
    "for ancestry in ancestries:\n",
    "    print(f\"-------------------- {ancestry} --------------------\")\n",
    "    to_plot = pd.read_csv(f\"{RESULTS_DIR}/{ancestry}/model1_risk_results.txt\", sep=\"\\t\")\n",
    "    \n",
    "    # Plot ROC curves\n",
    "    plt.figure(figsize=(8, 5))\n",
    "    for group in to_plot[\"Group\"].unique():\n",
    "        group_data = to_plot[to_plot[\"Group\"] == group]\n",
    "        fpr, tpr, _ = roc_curve(group_data[\"CASE\"], group_data[\"probDisease\"])\n",
    "        roc_auc = auc(fpr, tpr)\n",
    "        plt.plot(fpr, tpr, label=f\"{group} (AUC = {roc_auc:.2f})\")\n",
    "    \n",
    "    plt.plot([0, 1], [0, 1], \"k--\", lw=2)\n",
    "    plt.xlabel(\"False Positive Rate\")\n",
    "    plt.ylabel(\"True Positive Rate\")\n",
    "    plt.title(f\"ROC Curve for {ancestry}\")\n",
    "    plt.legend(loc=\"lower right\")\n",
    "    output_path = f\"{RESULTS_DIR}/supp_fig2_{ancestry}.png\"\n",
    "    plt.savefig(output_path, dpi=300)\n",
    "    plt.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1540e605-325f-4abe-a866-e8667b1febb0",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Estimate Specificity and Sensitivity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "ca454482-52c0-4256-a81a-15e18b8786a8",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"------------------------------AAC------------------------------\"\n",
      "   Ancestry       AUC  Accuracy  CI_Lower  CI_Upper Balanced_Accuracy\n",
      "     <list>     <num>     <num>     <num>     <num>             <num>\n",
      "1:      AAC 0.5851752 0.5646853 0.5353824 0.5936547         0.5706960\n",
      "2:      AFR 0.5434991 0.5402916 0.5209291 0.5595636         0.5373158\n",
      "3:       AJ 0.5433305 0.5035734 0.4803316 0.5268036         0.5348708\n",
      "4:      AMR 0.5791882 0.5613460 0.5443662 0.5782189         0.5628159\n",
      "5:      EAS 0.5796702 0.5575816 0.5423473 0.5727351         0.5597993\n",
      "6:      EUR 0.5542992 0.5306456 0.5251421 0.5361436         0.5393554\n",
      "7:      CAS 0.5569510 0.5624309 0.5293984 0.5950558         0.5530172\n",
      "   Sensitivity Specificity\n",
      "         <num>       <num>\n",
      "1:   0.5846154   0.5567766\n",
      "2:   0.5262055   0.5484262\n",
      "3:   0.4763610   0.5933806\n",
      "4:   0.5529046   0.5727273\n",
      "5:   0.5772402   0.5423585\n",
      "6:   0.5169963   0.5617144\n",
      "7:   0.5859107   0.5201238\n",
      "[1] \n",
      "[1] \"------------------------------EAS------------------------------\"\n",
      "   Ancestry       AUC  Accuracy  CI_Lower  CI_Upper Balanced_Accuracy\n",
      "     <list>     <num>     <num>     <num>     <num>             <num>\n",
      "1:      AAC 0.6216925 0.5777972 0.5485697 0.6066237         0.6039805\n",
      "2:      AFR 0.5360912 0.5594781 0.5401678 0.5786547         0.5353977\n",
      "3:       AJ 0.5558570 0.5568994 0.5337200 0.5798950         0.5481924\n",
      "4:      AMR 0.6251790 0.5941036 0.5772696 0.6107736         0.5918848\n",
      "5:      EAS 0.5503950 0.5386276 0.5233510 0.5538501         0.5375880\n",
      "6:      EUR 0.5843115 0.5578792 0.5523998 0.5633480         0.5611783\n",
      "7:      CAS 0.5655474 0.5359116 0.5027954 0.5687934         0.5482430\n",
      "   Sensitivity Specificity\n",
      "         <num>       <num>\n",
      "1:   0.6646154   0.5433455\n",
      "2:   0.4454927   0.6253027\n",
      "3:   0.5644699   0.5319149\n",
      "4:   0.6068465   0.5769231\n",
      "5:   0.5294118   0.5457642\n",
      "6:   0.5527090   0.5696476\n",
      "7:   0.5051546   0.5913313\n",
      "[1] \n",
      "[1] \"------------------------------EUR------------------------------\"\n",
      "   Ancestry       AUC  Accuracy  CI_Lower  CI_Upper Balanced_Accuracy\n",
      "     <list>     <num>     <num>     <num>     <num>             <num>\n",
      "1:      AAC 0.6505776 0.6171329 0.5882562 0.6454055         0.6119658\n",
      "2:      AFR 0.5356229 0.5069071 0.4875216 0.5262772         0.5220564\n",
      "3:       AJ 0.6647548 0.6146234 0.5918127 0.6370638         0.6236977\n",
      "4:      AMR 0.6357019 0.5970816 0.5802646 0.6137294         0.5943878\n",
      "5:      EAS 0.6179967 0.5856526 0.5705219 0.6006631         0.5896651\n",
      "6:      EUR 0.6318650 0.5958678 0.5904499 0.6012681         0.5952181\n",
      "7:      CAS 0.5913047 0.5911602 0.5583241 0.6234010         0.5787984\n",
      "   Sensitivity Specificity\n",
      "         <num>       <num>\n",
      "1:   0.6000000   0.6239316\n",
      "2:   0.5786164   0.4654964\n",
      "3:   0.6067335   0.6406619\n",
      "4:   0.6125519   0.5762238\n",
      "5:   0.6212205   0.5581098\n",
      "6:   0.5968859   0.5935504\n",
      "7:   0.6219931   0.5356037\n",
      "[1] \n",
      "[1] \"------------------------------AMR------------------------------\"\n",
      "   Ancestry       AUC  Accuracy  CI_Lower  CI_Upper Balanced_Accuracy\n",
      "     <list>     <num>     <num>     <num>     <num>             <num>\n",
      "1:      AAC 0.5750878 0.6118881 0.5829537 0.6402456         0.5684005\n",
      "2:      AFR 0.5108001 0.5199540 0.5005672 0.5392959         0.5062163\n",
      "3:       AJ 0.5448487 0.5063222 0.4830772 0.5295467         0.5531390\n",
      "4:      AMR 0.5050707 0.5068493 0.4897909 0.5238958         0.5109320\n",
      "5:      EAS 0.5369914 0.5383877 0.5231107 0.5536109         0.5309250\n",
      "6:      EUR 0.5691012 0.5563382 0.5508570 0.5618092         0.5532819\n",
      "7:      CAS 0.5849930 0.5558011 0.5227390 0.5884989         0.5650846\n",
      "   Sensitivity Specificity\n",
      "         <num>       <num>\n",
      "1:   0.4676923   0.6691087\n",
      "2:   0.4549266   0.5575061\n",
      "3:   0.4656160   0.6406619\n",
      "4:   0.4834025   0.5384615\n",
      "5:   0.4722375   0.5896126\n",
      "6:   0.5611280   0.5454358\n",
      "7:   0.5326460   0.5975232\n",
      "[1] \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "for (summstat_ancestry in summstat_ancestries) {\n",
    "    print(paste0(\"------------------------------\", summstat_ancestry, \"------------------------------\"))\n",
    "    \n",
    "    # Initialize results data.table\n",
    "    results_dt <- data.table(Ancestry = ancestries)\n",
    "    results_dt[, AUC := as.numeric(NA)]\n",
    "    results_dt[, Accuracy := as.numeric(NA)]\n",
    "    results_dt[, CI_Lower := as.numeric(NA)]\n",
    "    results_dt[, CI_Upper := as.numeric(NA)]\n",
    "    results_dt[, Balanced_Accuracy := as.numeric(NA)]\n",
    "    results_dt[, Sensitivity := as.numeric(NA)]\n",
    "    results_dt[, Specificity := as.numeric(NA)]\n",
    "    \n",
    "    for (ancestry in ancestries) {\n",
    "        dat <- read.table(paste0(RESULTS_DIR, \"/\", ancestry, \"/model1_\", summstat_ancestry, \".profile\"), header = T)\n",
    "        dat$CASE <- dat$PHENO - 1\n",
    "        dat <- subset(dat, CASE != -10)\n",
    "        meanControls <- mean(dat$SCORE[dat$CASE == 0])\n",
    "        sdControls <- sd(dat$SCORE[dat$CASE == 0])\n",
    "        dat$zSCORE <- (dat$SCORE - meanControls)/sdControls\n",
    "        grsTests <- glm(CASE ~ zSCORE, family=\"binomial\", data = dat)\n",
    "        \n",
    "        dat$probDisease <- predict(grsTests, dat, type = c(\"response\"))\n",
    "        dat$reported <- ifelse(dat$CASE == 1, \"DISEASE\", \"CONTROL\")\n",
    "        roc <- roc(response = dat$reported, predictor = dat$probDisease)\n",
    "        auc_value <- auc(roc)\n",
    "        \n",
    "        result.coords <- coords(\n",
    "            roc, \n",
    "            \"best\", \n",
    "            best.method = \"closest.topleft\", \n",
    "            ret = c(\"threshold\", \"accuracy\", \"specificity\", \"sensitivity\", \"youden\"),\n",
    "        )\n",
    "        \n",
    "        dat$predicted <- ifelse(dat$probDisease > result.coords$threshold, \"DISEASE\", \"CONTROL\")\n",
    "        confMat <- confusionMatrix(data = as.factor(dat$predicted), reference = as.factor(dat$reported), positive = \"DISEASE\")\n",
    "        results_dt[Ancestry == ancestry, AUC := auc_value]\n",
    "        results_dt[Ancestry == ancestry, Accuracy := confMat$overall[\"Accuracy\"]]\n",
    "        results_dt[Ancestry == ancestry, CI_Lower := confMat$overall[\"AccuracyLower\"]]\n",
    "        results_dt[Ancestry == ancestry, CI_Upper := confMat$overall[\"AccuracyUpper\"]]\n",
    "        results_dt[Ancestry == ancestry, Balanced_Accuracy := confMat$byClass[\"Balanced Accuracy\"]]\n",
    "        results_dt[Ancestry == ancestry, Sensitivity := confMat$byClass[\"Sensitivity\"]]\n",
    "        results_dt[Ancestry == ancestry, Specificity := confMat$byClass[\"Specificity\"]]\n",
    "        \n",
    "        dat$ancestry <- ancestry\n",
    "        dat$sumstat <- summstat_ancestry\n",
    "    }\n",
    "\n",
    "    fwrite(results_dt, paste0(RESULTS_DIR, \"/\", summstat_ancestry, \"_results.txt\"), sep = \"\\t\", quote = FALSE, row.names = FALSE)\n",
    "    print(results_dt)\n",
    "    print(noquote(\"\"))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e5838638-5ba6-4963-9712-cb5373821a33",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_model1 = []\n",
    "abbr = [\"AAC\",\"EAS\",\"EUR\",\"AMR\"]\n",
    "for i, summstat_ancestry in enumerate(summstat_ancestries):\n",
    "    df = pd.read_csv(f\"{RESULTS_DIR}/{summstat_ancestry}_results.txt\", sep=\"\\s+\")\n",
    "    df.insert(1, \"Summary statistics\", abbr[i])\n",
    "    df_model1.append(df)\n",
    "df_model1 = pd.concat(df_model1)\n",
    "numeric_cols = [\"AUC\", \"Accuracy\", \"CI_Lower\", \"CI_Upper\", \"Balanced_Accuracy\", \"Sensitivity\", \"Specificity\"]\n",
    "df_model1[numeric_cols] = df_model1[numeric_cols].round(3)\n",
    "df_model1[\"Accuracy\"] = df_model1.apply(lambda row: f\"{row['Accuracy']:.3f} ({row['CI_Lower']:.3f} - {row['CI_Upper']:.3f})\", axis=1)\n",
    "df_model1.drop(columns=[\"CI_Lower\", \"CI_Upper\"], inplace=True)\n",
    "df_model1.sort_values([\"Ancestry\",\"Summary statistics\"], inplace=True)\n",
    "df_model1.to_csv(f\"{RESULTS_DIR}/table2.txt\", index=False, sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4733ce0d-b297-45b9-9562-68e10e1d4ba4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"------------------------------AAC------------------------------\"\n",
      "[1] \"--------------------AAC vs. EUR --------------------\"\n",
      "[1] 0.000851787\n",
      "[1] \"--------------------AMR vs. EUR --------------------\"\n",
      "[1] 5.388729e-05\n",
      "[1] \"--------------------EAS vs. EUR --------------------\"\n",
      "[1] 0.08645032\n",
      "[1] \"------------------------------AFR------------------------------\"\n",
      "[1] \"--------------------AAC vs. EUR --------------------\"\n",
      "[1] -0.5016517\n",
      "[1] \"--------------------AMR vs. EUR --------------------\"\n",
      "[1] 0.03654276\n",
      "[1] \"--------------------EAS vs. EUR --------------------\"\n",
      "[1] -0.9648708\n",
      "[1] \"------------------------------AJ------------------------------\"\n",
      "[1] \"--------------------AAC vs. EUR --------------------\"\n",
      "[1] 1.568025e-11\n",
      "[1] \"--------------------AMR vs. EUR --------------------\"\n",
      "[1] 9.093203e-15\n",
      "[1] \"--------------------EAS vs. EUR --------------------\"\n",
      "[1] 4.60068e-12\n",
      "[1] \"------------------------------AMR------------------------------\"\n",
      "[1] \"--------------------AAC vs. EUR --------------------\"\n",
      "[1] 3.425841e-07\n",
      "[1] \"--------------------AMR vs. EUR --------------------\"\n",
      "[1] 7.391487e-15\n",
      "[1] \"--------------------EAS vs. EUR --------------------\"\n",
      "[1] 0.2601636\n",
      "[1] \"------------------------------EAS------------------------------\"\n",
      "[1] \"--------------------AAC vs. EUR --------------------\"\n",
      "[1] 0.0002983223\n",
      "[1] \"--------------------AMR vs. EUR --------------------\"\n",
      "[1] 7.097009e-20\n",
      "[1] \"--------------------EAS vs. EUR --------------------\"\n",
      "[1] 8.929058e-19\n",
      "[1] \"------------------------------EUR------------------------------\"\n",
      "[1] \"--------------------AAC vs. EUR --------------------\"\n",
      "[1] 4.202264e-88\n",
      "[1] \"--------------------AMR vs. EUR --------------------\"\n",
      "[1] 3.057792e-72\n",
      "[1] \"--------------------EAS vs. EUR --------------------\"\n",
      "[1] 1.48012e-49\n",
      "[1] \"------------------------------CAS------------------------------\"\n",
      "[1] \"--------------------AAC vs. EUR --------------------\"\n",
      "[1] 0.1604358\n",
      "[1] \"--------------------AMR vs. EUR --------------------\"\n",
      "[1] 0.7497863\n",
      "[1] \"--------------------EAS vs. EUR --------------------\"\n",
      "[1] 0.1771015\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "non_eur_sumstat_ancestries <- list(\"AAC\", \"AMR\", \"EAS\")\n",
    "\n",
    "# Initialize results data.table\n",
    "results_dt <- data.table(Ancestry = ancestries)\n",
    "for (non_eur_sumstat_ancestry in non_eur_sumstat_ancestries) {\n",
    "    results_dt[[paste0(non_eur_sumstat_ancestry, \" vs. EUR\")]] <- as.numeric(NA)\n",
    "}\n",
    "\n",
    "for (ancestry in ancestries) {\n",
    "    print(paste0(\"------------------------------\", ancestry, \"------------------------------\"))\n",
    "    model1_data_full <- read.table(paste0(RESULTS_DIR, \"/\", ancestry, \"/model1_risk_results.txt\"), header = TRUE) \n",
    "    \n",
    "    for (non_eur_sumstat_ancestry in non_eur_sumstat_ancestries) {\n",
    "        sumstat1_data <- subset(model1_data_full, Group == non_eur_sumstat_ancestry)\n",
    "        sumstat2_data <- subset(model1_data_full, Group == \"EUR\")\n",
    "        \n",
    "        gt1 <- sumstat1_data$CASE\n",
    "        gt2 <- sumstat2_data$CASE\n",
    "        pred_ss1 <- sumstat1_data$probDisease\n",
    "        pred_ss2 <- sumstat2_data$probDisease\n",
    "        roc1 <- roc(gt1, pred_ss1)\n",
    "        roc2 <- roc(gt2, pred_ss2)        \n",
    "        \n",
    "        auc1 <- auc(roc1)\n",
    "        auc2 <- auc(roc2)\n",
    "        results <- roc.test(roc1, roc2, method = \"delong\")\n",
    "        \n",
    "        p_value <- as.numeric(results$p.value)\n",
    "        if (auc1 > auc2) {\n",
    "            p_value <- p_value * -1\n",
    "        }\n",
    "        print(paste0(\"--------------------\", non_eur_sumstat_ancestry, \" vs. EUR --------------------\"))\n",
    "        print(p_value)\n",
    "        \n",
    "        # Store numeric p-value in the correct row and column\n",
    "        results_dt[Ancestry == ancestry, paste0(non_eur_sumstat_ancestry, \" vs. EUR\") := p_value]\n",
    "    }\n",
    "}\n",
    "\n",
    "fwrite(results_dt, paste0(RESULTS_DIR, \"/model1_comparison.txt\"), sep = \"\\t\", quote = FALSE, row.names = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "48a5348a-127e-4b2a-a60d-495373f219fa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAzQAAAINCAYAAADocDqmAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAb9VJREFUeJzt3XmcjfX///HnmRmzL3Yz1rGMYbIvER8hZIlQIYmRJUuRIsunJCq0ECUtlkH52LKEoqQiZDf2kDXMlGUwtsHM9fvDz/l2mhnmjDmuuZzH/Xa7bjfnWt7X6zrnmjGv83q/35fNMAxDAAAAAGBBHmYHAAAAAACZRUIDAAAAwLJIaAAAAABYFgkNAAAAAMsioQEAAABgWSQ0AAAAACyLhAYAAACAZZHQAAAAALAsEhoAAAAAluVldgCuMOG3w2aHAKRy6kKS2SEADpqWyW92CICDmsVym3Zu78pdXNb2tW1TXdY2ACo0AAAAACzsvqzQAAAAOMPm4Wl2CAAyiYQGAAC4PRIawLrocgYAAADAsqjQAAAAt0eFBrAuKjQAAAAALIsKDQAAcHtUaADrokIDAAAAwLKo0AAAALdn86RCA1gVFRoAAAAAlkWFBgAAuD0PxtAAlkVCAwAA3B6TAgDWRZczAAAAAJZFhQYAALg9KjSAdVGhAQAAAGBZVGgAAIDbs3nwHS9gVfz0AgAAALAsKjQAAMDtMYYGsC4qNAAAAAAsy9QKzYULFxQYGCiPf/VbTU5O1qVLlxQcHGxSZAAAwJ1QoQGsy7QKzcKFC1WtWjVdvXo11barV6+qevXqWrJkiQmRAQAAd2Pz8HTZAsC1TEtoPv30Uw0cOFD+/v6ptgUEBGjQoEGaMGGCCZEBAAAAsArTEppdu3apXr166W5/+OGHtXPnznsXEAAAcFs2T0+XLQBcy7SEJiEhQTdu3Eh3+/Xr15WQkHAPIwIAAABgNaZNChAeHq7NmzerTJkyaW7fvHmzihUrdo+jAgAA7oixLoB1mVaheeKJJ/Taa6/pr7/+SrUtPj5er7/+up588kkTIgMAAABgFaYlNIMHD1ZQUJAiIiLUu3dvjR8/XuPHj1evXr1UunRpBQYGavDgwWaFh/9vwahXzQ4BcLDyo9fMDgHAfYhZzgDrMq3LWVBQkNauXashQ4Zozpw59vEyOXPm1LPPPqt33nlHQUFBZoXn1k4e2K0b15JU9IEq9nXHdm9VDm9fhUVEmRgZ3NWpQ3uVfD1JoZGV7Ovi98XKK4eP8pYoa15gAADAdDbDMAyzgzAMQ6dPn5ZhGMqXL59sNpsk6ezZs8qdO7fT7U347XBWh+hWEs+c0ppZn8svOKfiD/6u0JJldCXxvP7z9PMKypPP7PAs69SFJLNDsKxLCacUu3CqfAJDdObIfuUJL62kixdUqfVzCsjFPZlZTcvkNzsEwEHNYs7/n59VQp8c77K24+e/5LK2AZjY5eyfbDab8uXLp/z588tms+mHH35Q27ZtVahQIbNDc0tBefKp6Yuvy9vPX6eO/iFv/wA1feE1khmYJiBXPtXuMkg5fP2VcPygcvgFqHaXgSQzALIMXc4A68oWCY0kHT16VMOGDVN4eLjatGkjDw8PzZgxw+yw3NLFhNNaPnGUki5fUr5ipZR06aKWTxyliwmnzQ4NburyuTNaN+19Xb9ySbkKl9T1yxe1btr7unzujNmhAQAAk5k2hkaSrl27pgULFmjy5Mlau3atGjZsqOPHj2vbtm0qX758htpISkpSUpJjV57r15KUw9vHFSG7hQun/1JU3cYq+kAVLRj1qupH99Gx3Vt14fRfCsyV1+zw4IYunf1bJR5qpNDISlr50Wuq1q6X4vfF6tLZv+WfM4/Z4QG4D1BJAazLtApNnz59VLBgQY0fP16tW7fW8ePHtWTJEtlsNnk68VTdUaNGKSQkxGFZMeNTF0Z+/ysY8YDDhACSVPSBKioY8YBJEcHd5StR1mFCAEkKjaykfEwIAACA2zNtUgAvLy8NGjTIPn3zLTly5ND27dsVFZWx2bTSqtBM2XaSCg2yHSYFQHbDpADIbsycFKBQ+y9c1vaJWc+7rG0AJlZovvzyS23cuFFhYWFq166dli5dquTkZKfb8fHxUXBwsMNCMgMAAAC4B9MSmvbt22vFihXauXOnypQpoxdeeEGhoaFKSUnRnj17zAoLAAC4IWY5A6zL9FnOihcvruHDh+vIkSP66quv9OSTT+rZZ59V4cKF1bdvX7PDAwAAAJCNmZ7Q3GKz2dS4cWPNnTtXJ0+e1KuvvqqJEyeaHRYAAHADVGgA6zJ12ua0JCYm6uuvv9bMmTNl0nwFAADAzdicmGEVQPaSbSo0q1evVnR0tMLCwvTBBx/okUce0fr1680OCwAAAEA2ZmqFJj4+XtOmTdOUKVN04cIFtW3bVklJSVq0aFGGp20GAAC4W3QNA6zLtApNixYtFBkZqR07dmjcuHE6efKkPv74Y7PCAQAAAGBBplVoli1bpr59+6pXr16KiIgwKwwAAAAqNICFmVahWbNmjRITE1W1alXVqFFDEyZM0OnTp80KBwAAAIAFmZbQ1KxZU5MmTVJcXJx69Oih2bNnq2DBgkpJSdGKFSuUmJhoVmgAAMDNMG0zYF2mz3IWEBCgLl26aM2aNdq5c6f69++v0aNHK3/+/Hr88cfNDg8AAABANmZ6QvNPkZGReu+993T8+HHNmjXL7HAAAICb8PCwuWwB4FrZ7sGakuTp6alWrVqpVatWZocCAADcgI3EA7CsbFWhAQAAAABnZMsKDQAAwL1ks1GhAayKCg0AAAAAy6JCAwAA3B6D9wHrokIDAAAAwLKo0AAAALfHLGeAdVGhAQAAAGBZVGgAAIDbo0IDWBcJDQAAcHseTNsMWBZdzgAAAABYFhUaAADg9uhyBlgXFRoAAAAAlkWFBgAAuD0qNIB1UaEBAAAAYFlUaAAAgNvzoEIDWBYVGgAAAACWRYUGAAC4PRtf8QKWdV8mND1KUjZG9vNTQm6zQwAcVMrvZ3YIQLZh48GagGXxfQQAAAAAy7ovKzQAAADOYFIAwLqo0AAAAACwLCo0AADA7fFgTcC6qNAAAAAAsCwqNAAAwO1RoQGsiwoNAAAAAMuiQgMAANyeB8+hASyLCg0AAAAAy6JCAwAA3B5jaADrIqEBAABuj4QGsC66nAEAAACwLCo0AADA7XlQoQEsiwoNAAAAAMuiQgMAANyejWmbAcuiQgMAAADAsqjQAAAAt2fjK17AsvjxBQAAAGBZVGgAAIDbY5YzwLpIaAAAgNvjwZqAddHlDAAAAIBlZduE5urVq/rggw/MDgMAALgBm83msgWAa5ma0Jw6dUpLly7VDz/8oOTkZEnS9evXNX78eIWHh2v06NFmhgcAAAAgmzNtDM2aNWvUvHlzXbhwQTabTdWqVVNMTIxatWolLy8vvfnmm4qOjjYrPAAA4EaYFACwLtMqNK+//rqaNWumHTt26JVXXtGmTZvUunVrjRw5Unv27FHPnj3l5+dnVngAAAAALMBmGIZhxonz5MmjX3/9VVFRUbpy5YoCAwO1YMECtWzZ8q7bvv73kbsPEMhiPyX4mx0C4KBu0SCzQwAc+Jr4RWbDCWtc1vaPL/7HZW0DMLFCk5CQoLx580qS/Pz85O/vr3LlypkVDgCL+Oi/fcwOAQAAZCOmPodmz549io+PlyQZhqF9+/bp0qVLDvtUqFDBjNAAZCOH9uzQ9WvXFFmpmn3dvtjNyuHjoxJly5sYGYD7hSdjaADLMjWhadCggf7Z46158+aSbk6daBiGbDabffYzAO4rV74CWjhlgrb/tkpXr1zW3E/H6OKF82rd5QWzQwNwnyChAazLtITm8OHDZp0agMXkyldAXQa/pcXTP9Pxg/tVtvKDaturv9lhAQCAbMC0hKZYsWJmnRqAxZw7c0qLpk6QX0CQCpcsrcsXEzXt/WFq1eVF5cyTz+zwANwHqNAA1mXapACdOnVSYmKi/fX27dt1/fp1p9tJSkrShQsXHJakpKSsDBWAyc7+FaeHGrVQu94D5Ovnr3a9B+ihRi109q84s0MDAAAmMy2hmTlzpq5cuWJ/XadOHf35559OtzNq1CiFhIQ4LO9+9GlWhgrAZCWiKjhMCCBJkZWqqUQUk4YAyBqeHjaXLQBcy7QuZ/9+/E1mH4czZMgQvfLKKw7rPM7zrS1wv+o78mOzQwAAANmIqbOcZQUfHx/5+Pg4rLt+9axJ0QAAACuikgJYV7Z6Ds3vv/+uixcvOuzDc2gAAAAApCdbPofmFp5DAwAA7gUvKjSAZWXr59D8cxY0AAAAV6HLGWBd2e45NImJiZo1a5amTJmizZs3U6EBAAAAkK5sMynA6tWrNWXKFM2fP18FCxbUE088oQkTJpgdFgAAcANUaADrMjWhiY+P17Rp0zRlyhRduHBBbdu2VVJSkhYtWqSoqCgzQwMAAABgAaY9WLNFixaKjIzUjh07NG7cOJ08eVIff8zzJQAAwL3n6eHhsgWAa5lWoVm2bJn69u2rXr16KSIiwqwwAAAAAFiYaV8brFmzRomJiapatapq1KihCRMm6PTp02aFAwAA3Jinh81lCwDXMi2hqVmzpiZNmqS4uDj16NFDs2fPVsGCBZWSkqIVK1YwZTMAAACAOzK9Y2dAQIC6dOmiNWvWaOfOnerfv79Gjx6t/Pnz6/HHHzc7PAAA4Aao0ADWZXpC80+RkZF67733dPz4cc2aNcvscAAAgJsgoQGsK1slNLd4enqqVatWWrx4sdmhAAAAAMjGss2DNQEAAMziaaOSAlhVtqzQAAAAAEBGUKEBAABuj7EugHVRoQEAAABgWVRoAACA26NCA1gXFRoAAAAAlkWFBgAAuD0vKjSAZZHQAAAAt0eXM8C66HIGAAAAwLKo0AAAALdHhQawLio0AAAAACyLCg0AAHB7VGgA66JCAwAAAMCyqNAAAAC3R4UGsC4qNAAAAAAsiwoNAABwe1RoAOsioQEAAG6PhAawrvsyoXn0f8fMDgFI5erl62aHADjw8bsv/wuAhf3ycl2zQwBgQfxvBgAA3B4VGsC6mBQAAAAAgGVRoQEAAG6PCg1gXVRoAAAAAFgWFRoAAOD2qNAA1kWFBgAAAIBlUaEBAABujwoNYF0kNAAAwO152khoAKuiyxkAAAAAy6JCAwAA3J4HFRrAsqjQAAAAALAsKjQAAMDteVKgASyLCg0AAAAAy6JCAwAA3J4H0zYDlkWFBgAAAIBlUaEBAABuj+fQANZFhQYAAACAZVGhAQAAbo/n0ADWRUIDAADcHtM2A9ZFlzMAAAAAlkWFBgAAuD2mbQasiwoNAAAAAMuiQgMAANwekwIA1kWFBgAAAIBlUaEBAABuj1nOAOuiQgMAAADAskhocFsfPlXB7BAAB590qGJ2CEAq456qaHYIuEseNpvLFgCu5XRCc+jQIVfEgWykXFiwqhTJ6bCuSpGceiAs2JyA4PYqFA5R9fBcDuuqh+dS+UIhJkUESOUKBqtq0ZwO66oW5XelVXl62Fy2AHAtp8fQlCpVSnXr1lXXrl311FNPydfXN1Mn3rFjx52D8/JSaGiocufOnalzIHP+SkxS74dL6OFSeeWfw1P96pdSiF8OTVxNMgtzxJ+/qpcaRqhuZH75e3tqQONI5fTPoY9+PGB2aHBjf11I0gt1S6pOqbzy8/ZUv0f+/+/KVfyuBIB7yWYYhuHMAbGxsYqJidGsWbN07do1tWvXTl27dtWDDz7o1Ik9PDxks9l0p9PbbDZVrFhRM2bMULly5TLUdv1xq52KBWnrVjtcHaoX1cyNxzR53RGzw7G8q5evmx2C5fWqV1KdaoVr+toj+mzVQbPDsTwfP+aFyQrdaxdXhweL6quNxzR57WGzw7G0X16ua9q5l+79y2VtNy9bwGVtA8hEQnPLjRs3tHjxYk2bNk3Lly9X6dKl1aVLF3Xs2FH58uW74/FHjx694z4XLlzQpUuX9P777+vvv//Wr7/+mqHYSGjuTt4Ab/V6uIQSk26obIEg7f0rUUE+Xvp09SGdvnTN7PAsi4Qm8/IF+qhPwwglXr2uqLBg7Ym7oCDfHPr4xwM6dTHJ7PAsi4Tm7uQN8FbvuiWVePWGyoQG6ff4RAX5emniqoP8rswkEhoAmZHphOaWpKQkTZw4UUOGDNG1a9fk7e2ttm3b6t1331VYWJjT7SUmJmrWrFmaMmWKNm/erOTkZP3xxx+qWLGiLl26lOb5k5Ic/6Bp8cVGeXh5Z/qa3F25sGB5e3lo65/n9OFTFfTy1ztUpUhOXbuRol1xF8wOz7JIaDKvQuEQ+Xh5aNORBH3SoYpemLlV1cNzKelGinYcP292eJZFQnN3yhUMlo+Xh7YcO6dxT1VUv6+3q2rRnEq6kaJdJ/ldmRlmJjTLfnddQtO0DAkN4EqZnuVs8+bN6t27t8LCwjR27FgNGDBABw8e1IoVK3Ty5Em1bNnSqfZWr16t6OhohYWF6YMPPlD9+vW1fv16SVLx4sW1bt26NI8bNWqUQkJCHJajP87M7GVB0q64C9r65zmHdVv/PEcyA9PsOH5em44kOKzbdCSBZAam2nXygrYcO+ewbsuxcyQzgAv98ssvstlsOnfuXJa2W69ePfXr18/++vLly3ryyScVHBxsP194eLjGjRuXpefNjDfffFOVKlUyO4xsxemEZuzYsSpfvrxq1aqlkydPasaMGTp69KjefvttFS9eXHXq1NG0adO0devWO7YVHx+v0aNHKyIiQm3atFFwcLCSkpK0aNEijR49WtWrV5ckeXp6qmLFtKfEHDJkiM6fP++wFGvYwdnLQjpe/vrOkzcA99ILM+/8uwW41/p9vd3sEHCXmLb55h/1Npst1dKzZ09J0pEjR2Sz2RQbG5vmsf9MCMLDw+3H+/v7q3z58po8efIdY9i+fbsef/xx5c+fX76+vgoPD1e7du30999/S5Jq1aqluLg4hYRk7SyXCxYs0FtvvWV/PX36dP36669at26d/XybNm3S888/n6XnvRObzaZFixZlebv//Hw8PT1VsGBBde3aVQkJCXc++F/t/DvJmzZtmnLmzJl1wWaA0wnNp59+qmeeeUZHjx7VokWL1Lx5c3l4ODaTP39+TZky5bbttGjRQpGRkdqxY4fGjRunkydP6uOPP3Y2HPn4+Cg4ONhhobsZAACA87p37664uDiH5b333stUWyNGjFBcXJx27dqlZ599Vt27d9eyZcvS3f/UqVNq0KCBcufOre+//1579+5VTEyMChYsaB924O3trdDQUNmyOFHMnTu3goKC7K8PHjyosmXLqly5cvbz5cuXT/7+/ll6XjPd+nyOHTummTNnavXq1erbt6/ZYdklJycrJSUlQ/s6ldDcuHFDHTp0UMeOHW87Psbb21vR0dG3bWvZsmXq2rWrhg8frscee0yenp7OhAIAAJBlsstzaOrVq6e+fftq4MCByp07t0JDQ/Xmm2867HPs2DG1bNlSgYGBCg4OVtu2bfXXX/83BuhWl6Qvv/xS4eHhCgkJ0dNPP63ExMQ7nt/f31+hoaEOS3Bw5p6tFBQUpNDQUJUoUUKDBg1S7ty5tWLFinT3X7t2rc6fP6/JkyercuXKKl68uOrXr68PP/xQxYsXl5R2l7NJkyapSJEi8vf3V+vWrTV27FiHCkFG3o9/Vpjq1aunMWPGaPXq1bLZbKpXr56k1NWIc+fOqUePHipQoIB8fX1Vrlw5LV26VJJ05swZtW/fXoUKFbJXqGbNmuVwvXf6rMPDwyVJrVu3ls1ms7/+p9WrVytHjhyKj493WN+vXz/VqVMn3fda+r/Pp1ChQqpfv76io6NT9bBas2aN6tSpIz8/PxUpUkR9+/a1J5f16tXT0aNH9fLLL9urPb/88ouee+45nT9/3r7u1jUlJSVpwIABKlSokAICAlSjRg398ssv9nPdquwsXrxYUVFR8vHx0bFjx257Dbc4ldB4eXlpzJgxunHjhjOHpWnNmjVKTExU1apVVaNGDU2YMEGnT5++63YBAACc5WFz3ZKUlKQLFy44LP+e0Oifpk+froCAAG3YsEHvvfeeRowYYU8EUlJS1LJlS509e1arVq3SihUrdOjQIbVr186hjYMHD2rRokVaunSpli5dqlWrVmn06NEufQ/Tk5KSovnz5yshIUHe3un3ogkNDdWNGze0cOHCOz7W45a1a9eqZ8+eeumllxQbG6tGjRrpnXfeSbWfM+/HggUL1L17dz300EOKi4vTggUL0rympk2bau3atfrqq6+0Z88ejR492v4F/dWrV1W1alV9++232rVrl55//nl17NhRGzdudGjndp/1pk2bJEkxMTGKi4uzv/6nhx9+WCVKlNCXX35pX3f9+nXNnDlTXbp0ydB7KEknTpzQkiVLVKNGDYf3rEmTJnryySe1Y8cOzZkzR2vWrNGLL75of58KFy5sr/TExcWpVq1aGjdunIKDg+3rBgwYIEl68cUX9dtvv2n27NnasWOH2rRpoyZNmujAgf97ptzly5f17rvvavLkydq9e7fy58+fofid7nL2yCOPaNWqVc4elkrNmjU1adIkxcXFqUePHpo9e7YKFiyolJQUrVixIkPfIgAAAGR3aU1gNGrUqHT3r1ChgoYNG6aIiAh16tRJ1apV08qVKyVJK1eu1M6dO/W///3P/qXwjBkztGrVKoc/eFNSUjRt2jSVK1dOderUUceOHe1t3M7EiRMVGBjosMycmbnJlgYNGqTAwED5+PjoqaeeUq5cudStW7d0969Zs6b++9//6plnnlHevHnVtGlTvf/++w7Vp3/7+OOP1bRpUw0YMEClS5dW79691bRp01T7OfN+5M6dW/7+/vbubWk94P3HH3/Uxo0btWDBAjVq1EglSpRQ8+bN7ecuVKiQBgwYoEqVKqlEiRLq06ePmjRporlz5zq0c7vP+tZjUHLmzKnQ0NB0H4vStWtXxcTE2F8vWbJEV69eVdu2bdN936T/+3z8/PxUuHBh2Ww2jR071r591KhR6tChg/r166eIiAjVqlVLH330kWbMmKGrV68qd+7c8vT0tFd6QkND5e3trZCQENlsNvu6wMBAHTt2TDExMZo3b57q1KmjkiVLasCAAfrPf/7jEPv169c1ceJE1apVS5GRkRnu4ud0QtO0aVMNHjxYAwYM0KxZs7R48WKHxVkBAQHq0qWL1qxZo507d6p///4aPXq08ufPr8cff9zp9gAAAJzlabO5bElrAqMhQ4akG0uFChUcXoeFhdkHxe/du1dFihRRkSJF7NujoqKUM2dO7d27174uPDzcYUzIP9uYOXOmQ8Lyz+f8dejQQbGxsQ5LZv8ee/XVVxUbG6uffvpJNWrU0IcffqhSpUrd9ph33nlH8fHx+uyzz/TAAw/os88+U5kyZbRz584099+3b1+qh7un9bD3270fmREbG6vChQurdOnSaW5PTk7WW2+9pfLlyyt37twKDAzU999/n6oL1e0+64zq3Lmz/vjjD/vswNOmTVPbtm0VEBBw2+NufT47duywJ1GPPfaYkpOTJd2coGHatGkO90rjxo2VkpKiw4ede4jwzp07lZycrNKlSzu0t2rVKh08+H8Pyvb29k71nmSE0w8h6N27tyQ5ZHC32Gw2+5uQGZGRkXrvvfc0atQoLVmyRFOnTs10WwAAANmBj4+PfHx8Mrx/jhw5HF7bbLYMD47OSBuPP/64Q9eiQoUK2f8dEhKSbtJxayzN+fOpp8w/d+5cqpnH8ubNq1KlSqlUqVKaN2+eypcvr2rVqikqKuq2sefJk0dt2rRRmzZtNHLkSFWuXFkffPCBpk+fftvjbicr3tN/8vPzu+32999/X+PHj9e4ceNUvnx5BQQEqF+/frp2zfGhu1kRV/78+dWiRQvFxMSoePHiWrZsmcPYlPTc+nwkKSIiQuPGjdNDDz2kn3/+WQ0bNtTFixfVo0ePNCcKKFq0qFMxXrx4UZ6entqyZUuqcfOBgYH2f/v5+WVqwgenE5q7+fAzytPTU61atVKrVq1cfi4AAACrTK9ctmxZ/fnnn/rzzz/tVZo9e/bo3Llzd0wUbgkKCnKoVmRU7ty5lTdvXm3ZskV16/7fQ1AvXLigP/74I91qhSQVKVJE7dq105AhQ/TNN99k+Jze3t4qWbJkmg9Xl25+Gf7vsSVpjTXJahUqVNDx48e1f//+NK977dq1atmypZ599llJN/9+3r9/f4Y/o1ty5MiRoWJBt27d1L59exUuXFglS5ZU7dq1nTqPJHuiceXKFUlSlSpVtGfPnttW1by9vVPFl9a6ypUrKzk5WX///fcdJyvIDKe7nM2YMSPNgWzXrl3TjBkzsiQoAAAApNawYUOVL19eHTp00NatW7Vx40Z16tRJdevWVbVq1e66/cuXLys+Pt5h+eezSV555RWNHDlSM2fO1MGDB7Vx40Z16NBB+fLl0xNPPHHbtl966SUtWbJEmzdvTnP70qVL9eyzz2rp0qXav3+/9u3bpw8++EDfffddug9s79Onj7777juNHTtWBw4c0Oeff65ly5Zl+bTO/1a3bl09/PDDevLJJ7VixQodPnxYy5Yt0/LlyyXdrHisWLFC69at0969e9WjR4/bjgVKT3h4uFauXJnqc/i3xo0bKzg4WG+//baee+65DLWdmJio+Ph4xcXFaePGjXr11VeVL18+1apVS9LNMTbr1q3Tiy++qNjYWB04cEDffPONfVKAW/GtXr1aJ06csE/uFR4erosXL2rlypU6ffq0Ll++rNKlS6tDhw7q1KmTFixYoMOHD2vjxo0aNWqUvv32W6ffl39zOqG5NRXbvyUmJmb4DQQAAMhOPD1ct2Qlm82mb775Rrly5dLDDz+shg0bqkSJEpozZ06WtD9p0iSFhYU5LO3bt7dvHzhwoIYNG6Z3331XFSpU0JNPPqmAgAD9/PPPd+yGFRUVpUcffVRvvPFGutv9/f3Vv39/VapUSTVr1tTcuXM1efJkdezYMc1jateurc8++0xjx45VxYoVtXz5cr388svy9fXN/JuQQfPnz1f16tXVvn17RUVFaeDAgfbKxOuvv64qVaqocePGqlevnkJDQzPV82jMmDFasWKFihQposqVK6e7n4eHhzp37qzk5GR16tQpQ22/8cYbCgsLU8GCBdW8eXMFBATohx9+UJ48eSTdrEKtWrVK+/fvV506dVS5cmW98cYbKliwoL2NESNG6MiRIypZsqR90oJatWqpZ8+eateunfLly2d/jlFMTIw6deqk/v37KzIyUq1atdKmTZuc7r6WFpuR0Xnx/j8PDw/99ddfqWZa2L59u+rXr6+zZ8/edVB3q/641WaHAKRy9fJ1s0MAHPj4Od3rGHCpX16ue+edXGTdkTMua7tWeB6XtY3Uunfvrt9//91hsgN30LVrV506dSpTk3RZXYb/N6tcubL9ATkNGjSQl9f/HZqcnKzDhw+rSZMmLgkSAADAlawyhgapffDBB2rUqJECAgK0bNkyTZ8+XRMnTjQ7rHvm/Pnz9qm83TGZkZxIaG6VyWJjY9W4cWOHGQm8vb0VHh6uJ598MssDBAAAcDVPEhrL2rhxo9577z0lJiaqRIkS+uijj277vJv7TcuWLbVx40b17NlTjRo1MjscU2Q4oRk2bJikmwN9nn76aaemHwQAAABc4d8Pq3Q3GZmi+X7n9FC1Rx55RKdOnbK/3rhxo/r166cvvvgiSwMDAAC4VzxsNpctAFzL6YTmmWee0c8//yxJio+PV8OGDbVx40a99tprGjFiRJYHCAAAAADpcTqh2bVrlx588EFJN0t85cuX17p16zRz5kxNmzYtq+MDAABwOatM2wwgNad/zK5fv24fP/Pjjz/q8ccflySVKVNGcXFxWRsdAAAAANyG0wnNAw88oM8++0y//vqrVqxYYZ+q+eTJk/YH8QAAAFgJY2gA63I6oXn33Xf1+eefq169emrfvr0qVqwoSVq8eLG9KxoAAAAA3AtOPya6Xr16On36tC5cuKBcuXLZ1z///PPy9/fP0uAAAADuBQopgHVlaqiaYRjasmWLPv/8cyUmJkq6+XBNEhoAAGBFHrK5bAHgWk5XaI4ePaomTZro2LFjSkpKUqNGjRQUFKR3331XSUlJ+uyzz1wRJwAAAACk4nSF5qWXXlK1atWUkJAgPz8/+/rWrVtr5cqVWRocAADAvWCzuW4B4FpOV2h+/fVXrVu3Tt7e3g7rw8PDdeLEiSwLDAAAAADuxOmEJiUlRcnJyanWHz9+XEFBQVkSFAAAwL3kQSUFsCynu5w9+uijGjdunP21zWbTxYsXNWzYMDVr1iwrYwMAAACA23K6QjNmzBg1btxYUVFRunr1qp555hkdOHBAefPm1axZs1wRIwAAgEsx1gWwLqcTmsKFC2v79u2aM2eOtm/frosXL6pr167q0KGDwyQBAAAAAOBqTic0kuTl5aUOHTqoQ4cOWR0PAADAPcfzYgDrcnoMzahRozR16tRU66dOnap33303S4ICAAC4l5i2GbAupxOazz//XGXKlEm1/oEHHuChmgAAAADuKae7nMXHxyssLCzV+nz58ikuLi5LgrpbJUKZPhrZz55DZ80OAXDweYfKZocAZBtM2wxYl9MVmiJFimjt2rWp1q9du1YFCxbMkqAAAAAAICOcrtB0795d/fr10/Xr1/XII49IklauXKmBAweqf//+WR4gAACAq1GgAazL6YTm1Vdf1ZkzZ9S7d29du3ZNkuTr66tBgwZpyJAhWR4gAAAAAKTH6YTGZrPp3Xff1dChQ7V37175+fkpIiJCPj4+rogPAADA5TyYjgywrEw9h0aSAgMDVb169ayMBQAAAACc4nRCc+nSJY0ePVorV67U33//rZSUFIfthw4dyrLgAAAA7gUKNIB1OZ3QdOvWTatWrVLHjh0VFhYmG78BAACAxTk97SuAbMPphGbZsmX69ttvVbt2bVfEAwAAAAAZ5nRCkytXLuXOndsVsQAAAJiCHieAdTldYX3rrbf0xhtv6PLly66IBwAAAAAyzOkKzZgxY3Tw4EEVKFBA4eHhypEjh8P2rVu3ZllwAAAA94IHBRrAspxOaFq1auWCMAAAAADAeU4nNMOGDXNFHAAAAKZhCA1gXcxSCAAAAMCynK7QJCcn68MPP9TcuXN17NgxXbt2zWH72bNnsyw4AACAe4FveAHrcvrnd/jw4Ro7dqzatWun8+fP65VXXtETTzwhDw8Pvfnmmy4IEQAAwLVsNpvLFgCu5XRCM3PmTE2aNEn9+/eXl5eX2rdvr8mTJ+uNN97Q+vXrXREjAAAAAKTJ6YQmPj5e5cuXlyQFBgbq/PnzkqTmzZvr22+/zdroAAAA7gEPm+sWAK7ldEJTuHBhxcXFSZJKliypH374QZK0adMm+fj4ZG10AAAAAHAbTic0rVu31sqVKyVJffr00dChQxUREaFOnTqpS5cuWRbY1atX9cEHH2RZewAAAOmxuXAB4Fo2wzCMu2lg/fr1WrdunSIiItSiRQunjj116pQ2bNggb29vNWjQQJ6enrp+/bomTpyoUaNG6caNGzp9+rTTMXWdvc3pYwBX23OIGQCRvUzrVt3sEAAHkfmDTTv32cTLLms7d5C/y9oGkIlpm/+tZs2aqlmzptPHrVmzRs2bN9eFCxdks9lUrVo1xcTEqFWrVvLy8tKbb76p6Ojouw0PAADgjhjrAliXadOuv/7662rWrJl27NihV155RZs2bVLr1q01cuRI7dmzRz179pSfn59Z4QEAAACwgLvucpZZefLk0a+//qqoqChduXJFgYGBWrBggVq2bHnXbdPlDNkRXc6Q3dDlDNmNmV3Ozl+64rK2QwL4ghZwJdMqNAkJCcqbN68kyc/PT/7+/ipXrpxZ4QAAAACwoLseQ3M39uzZo/j4eEmSYRjat2+fLl265LBPhQoVzAgNAAC4EcbQANZlakLToEED/bPHW/PmzSVJNptNhmHIZrMpOTnZrPAg6dVHSun9n/4wOwzA7pMOVfTCzK1mhwHgPkM+A1iX0wlNcnKyPvzwQ82dO1fHjh3TtWvXHLafPZuxcQKHDx929tS4R0rlDVAOTw/t/SvRvq5sgSBdu5Gig2cu3eZIwDUqFA6Rj5eHNh1JsK+rHp5LV6+naOeJ8yZGBgAAzOZ0QjN8+HBNnjxZ/fv31+uvv67XXntNR44c0aJFi/TGG29kuJ1ixYo5e2rcI2cuX9PTlQqpauEQ+Xp56tmqhRXo46U5sSfMDg1uKv78Vb3UMEJ1I/PL39tTAxpHKqd/Dn304wGzQwNwn/CwUaMBrMrphGbmzJmaNGmSHnvsMb355ptq3769SpYsqQoVKmj9+vXq27dvhto5duxYhvYrWrSosyHiLiVcvq5P1x3RExXCFJ7bX7vjL+irLcfNDgtu7O/EJL22cJd61SupMmHB2nDorD74fp/ZYQEAgGzA6YQmPj5e5cuXlyQFBgbq/Pmb3T2aN2+uoUOHZrid4sWL2/99axyN7R/fjjCGxjw5/XKoXaVCunTtho6cvSz/HJ7q8VC45sSe0Lkr180OD24oX6CP+jSMUOLV6/o97oKC/Lw0olU5ffzjAZ26mGR2eADuAxRoAOtyOqEpXLiw4uLiVLRoUZUsWVI//PCDqlSpok2bNsnHxyfD7dhsNhUuXFidO3dWixYt5OWVufkJkpKSlJTk+AdN8vVr8szhnan2IOUN8NbqQ2e0969EvfpIKX215bjKFghS3gBvEhqYIiynr5bEntCmIwn6pEMVvb98n6qH51JYTl8SGgAA3JzTz6Fp3bq1Vq5cKUnq06ePhg4dqoiICHXq1EldunTJcDvHjx9Xr169NHv2bD322GP68ssv5e3trYoVKzosdzJq1CiFhIQ4LNu/mersZeEf/jh9yWFCAEna+1ei/jjNhAAwx47j5x0mBJCkTUcStOM4EwIAyBo2w3DZAsC1bIZxdz9p69ev17p16xQREaEWLVpkqo01a9YoJiZG8+bNU1RUlLp27aquXbvKw+PO+VZaFZq+3+ylQoNsZ8+hjM0ACNwr07pVNzsEwEFk/mDTzn31yhWXte3r5+eytgFkIqFZvXq1atWqlaqL2I0bN7Ru3To9/PDDmQ7mr7/+Uvv27bVq1SqdOnVKuXPnzlQ7XWdvy3QMgKuQ0CC7IaFBdmNqQnPZdb0QfP0DXNY2gEx0Oatfv36az5o5f/686tevn6kg1q1bp27duql06dK6ePGiPvnkE+XMmTNTbQEAADjLZqS4bAHgWk6PxL81+9i/nTlzRgEBGf8GIi4uTjNmzFBMTIwSEhLUoUMHrV27VuXKlXM2JAAAAABuKsMJzRNPPCHp5uxknTt3dpjRLDk5WTt27FCtWrUyfOKiRYuqUKFCio6O1uOPP64cOXIoJSVFO3bscNivQoUKGW4TAAAgU6ikAJaV4YQmJCRE0s0KTVBQkPz+McDN29tbNWvWVPfu3TN84uTkZB07dkxvvfWW3n77bXvb/8RzaAAAAADcToYTmpiYGElSeHi4BgwY4FT3srQcPnz4jvskJibecR8AAIC7xvTKgGU5PYZm2LBhWXLiYsWKpbk+MTFRs2bN0pQpU7R582YqNAAAAADS5XRCI0lff/215s6dq2PHjunatWsO27Zu3ZqpQFavXq0pU6Zo/vz5KliwoJ544glNmDAhU20BAAA4hTE0gGU5PW3zRx99pOeee04FChTQtm3b9OCDDypPnjw6dOiQmjZt6lRb8fHxGj16tCIiItSmTRsFBwcrKSlJixYt0ujRo1W9Os9IAAAAAJA+pxOaiRMn6osvvtDHH38sb29vDRw4UCtWrFDfvn11/vz5DLfTokULRUZGaseOHRo3bpxOnjypjz/+2NlwAAAA7hrPoQGsy+mE5tixY/bpmf38/OwD9zt27KhZs2ZluJ1ly5apa9euGj58uB577DF5eno6GwoAAEDWMFJctwBwKacTmtDQUJ09e1bSzWfJrF+/XtLNWcv+Pe3y7axZs0aJiYmqWrWqatSooQkTJuj06dPOhgMAAADAjTmd0DzyyCNavHixJOm5557Tyy+/rEaNGqldu3Zq3bp1htupWbOmJk2apLi4OPXo0UOzZ89WwYIFlZKSohUrVjBlMwAAuHeo0ACWZTOcKatISklJUUpKiry8bk6QNnv2bK1bt04RERHq0aOHvL29Mx3Mvn37NGXKFH355Zc6d+6cGjVqZE+enNF19rZMxwC4yp5DZ80OAXAwrRsTryB7icwfbNq5ryXEu6xt71yhLmsbQCYSmnshOTlZS5Ys0dSpU0locN8goUF2Q0KD7MbUhObsSZe17Z27oMvaBpDB59Ds2LEjww1WqFAh08Hc4unpqVatWqlVq1Z33RYAAACA+1eGEppKlSrJZrPJMAzZbLbb7pucnJwlgQEAANwzKYx1AawqQ5MCHD58WIcOHdLhw4c1f/58FS9eXBMnTtS2bdu0bds2TZw4USVLltT8+fNdHS8AAAAA2GWoQlOsWDH7v9u0aaOPPvpIzZo1s6+rUKGCihQpoqFDh9JNDAAAWA4PwASsK0MJzT/t3LlTxYsXT7W+ePHi2rNnT5YEBQAAcE+R0ACW5fRzaMqWLatRo0bp2rVr9nXXrl3TqFGjVLZs2SwNDgAAAABux+kKzWeffaYWLVqocOHC9hnNduzYIZvNpiVLlmR5gAAAAC6X/Z5iASCDnE5oHnzwQR06dEgzZ87U77//Lklq166dnnnmGQUEBGR5gAAAAACQHqcTGkkKCAjQ888/n9WxAAAAmIMxNIBlOT2GBgAAAACyi0xVaAAAAO4nTNsMWBcVGgAAAACWRYUGAACACg1gWU5XaEqUKKEzZ86kWn/u3DmVKFEiS4ICAAC4p4wU1y0AXMrphObIkSNKTk5OtT4pKUknTpzIkqAAAAAAICMy3OVs8eLF9n9///33CgkJsb9OTk7WypUrFR4enqXBAQAA3BNUUgDLshlGxh6N6+Fxs5hjs9n070Ny5Mih8PBwjRkzRs2bN8/6KJ30/qo/zA4BSGXm9/vNDgFwYLPZzA4BcLDtnaamnfvG8d0ua9ur8AMuaxuAExWalJSb31wUL15cmzZtUt68eV0WFAAAwL3EtM2AdTk9y9nhw4ft/7569ap8fX2zNCAAAAAAyCinJwVISUnRW2+9pUKFCikwMFCHDh2SJA0dOlRTpkzJ8gABAABcLiXFdQsAl3I6oXn77bc1bdo0vffee/L29ravL1eunCZPnpylwQEAAADA7Tid0MyYMUNffPGFOnToIE9PT/v6ihUr6vfff8/S4AAAAO4Jw3DdAsClnB5Dc+LECZUqVSrV+pSUFF2/fj1LggIAALinmBQAsCynKzRRUVH69ddfU63/+uuvVbly5SwJCgAAAAAywukKzRtvvKHo6GidOHFCKSkpWrBggfbt26cZM2Zo6dKlrogRAADApZi2GbAupys0LVu21JIlS/Tjjz8qICBAb7zxhvbu3aslS5aoUaNGrogRAAAAANLkdIVGkurUqaMVK1ZkdSwAAADmoEIDWJbTFRoAAAAAyC6crtDkypVLNpst1XqbzSZfX1+VKlVKnTt31nPPPZclAQIAALgcFRrAsjI1KcA777yjpk2b6sEHH5Qkbdy4UcuXL9cLL7ygw4cPq1evXrpx44a6d++e5QEDAAAAwC1OJzRr1qzR22+/rZ49ezqs//zzz/XDDz9o/vz5qlChgj766CMSGgAAYA0pyWZHACCTnB5D8/3336thw4ap1jdo0EDff/+9JKlZs2Y6dOjQ3UcHAABwDxgpKS5bALiW0wlN7ty5tWTJklTrlyxZoty5c0uSLl26pKCgoLuPDgAAAABuw+kuZ0OHDlWvXr30888/28fQbNq0Sd99950+++wzSdKKFStUt27drI0UAADAVehyBliW0wlN9+7dFRUVpQkTJmjBggWSpMjISK1atUq1atWSJPXv3z9rowQAAACANDiV0Fy/fl09evTQ0KFDNWvWLFfFBAAAcG9RoQEsy6kxNDly5ND8+fNdFQsAAAAAOMXpSQFatWqlRYsWuSAUAAAAcxjJyS5bALiW02NoIiIiNGLECK1du1ZVq1ZVQECAw/a+ffvedVBHjx7VpUuXVKZMGXl4OJ1zAQAAwEWOHDmi4sWLa9u2bapUqVKWtdu5c2edO3fO/sW5YRjq0aOHvv76ayUkJGjbtm3q16+fKlWqpHHjxmXZeTNj2rRp6tevn86dO2dqHLjJ6WxhypQpypkzp7Zs2aIvvvhCH374oX1x9uaaOnWqxo4d67Du+eefV4kSJVS+fHmVK1dOf/75p7MhAgAAOCclxXWLEzp37iybzZZqadKkiX0fm82WZm+Zzp07q1WrVvbX9erVsx/v6+ur0qVLa9SoUTIM47YxHD58WM8884wKFiwoX19fFS5cWC1bttTvv/8uSSpSpIji4uJUrlw5p67tTsaPH69p06bZXy9fvlzTpk3T0qVL7edbsGCB3nrrrSw9752Eh4e7JIH65+djs9lUoEABtWnTRkePHnW6nX79+jms++WXX2Sz2dwm4XI6oTl8+HC6i7MP0/ziiy+UK1cu++vly5crJiZGM2bM0KZNm5QzZ04NHz7c2RABAACck5LsusVJTZo0UVxcnMOS2cmYunfvrri4OO3bt09DhgzRG2+8YX/MRlquX7+uRo0a6fz581qwYIH27dunOXPmqHz58vY/jj09PRUaGiovL6c7+txWSEiIcubMaX998OBBhYWFqVatWvbz5c6d+7561uGtz+fkyZP65ptv9Oeff+rZZ581Oyw7wzB048YNs8O4I1P7cx04cEDVqlWzv/7mm2/UsmVLdejQQVWqVNHIkSO1cuVKEyMEAAC4t3x8fBQaGuqw/PMLYGf4+/srNDRUxYoV03PPPacKFSpoxYoV6e6/e/duHTx4UBMnTlTNmjVVrFgx1a5dW2+//bZq1qwp6WaXM5vNptjYWPtxixcvVkREhHx9fVW/fn1Nnz7doUIwbdo05cyZU99//73Kli2rwMBAe+J2yz8rTJ07d1afPn107Ngx2Ww2hYeHS0pdjUhKStKgQYNUpEgR+fj4qFSpUpoyZYokKTk5WV27dlXx4sXl5+enyMhIjR8/3uF6b53zgw8+UFhYmPLkyaMXXnhB169ft5/v6NGjevnll+2VlH87cuSIPDw8tHnzZof148aNU7FixZRymyrdrc8nLCxMNWvW1IsvvqitW7c67LNr1y41bdpUgYGBKlCggDp27KjTp0/b41+1apXGjx9vj+/IkSOqX7++JClXrlyy2Wzq3LmzJCklJUWjRo2yvycVK1bU119/bT/XrcrOsmXLVLVqVfn4+GjNmjXpxp9dZCq1Pn78uBYvXqxjx47p2rVrDtv+3YXsdq5cuaLg4GD763Xr1qlr16721yVKlFB8fHxmQgQAAMgww4XTNiclJSkpKclhnY+Pj3x8fFx2zn8zDENr1qzR77//roiIiHT3y5cvnzw8PPT111+rX79+8vT0vGPbhw8f1lNPPaWXXnpJ3bp107Zt2zRgwIBU+12+fFkffPCBvvzyS3l4eOjZZ5/VgAEDNHPmzFT7jh8/XiVLltQXX3yhTZs2pRtHp06d9Ntvv+mjjz5SxYoVdfjwYfsf+ykpKSpcuLDmzZunPHnyaN26dXr++ecVFhamtm3b2tv4+eefFRYWpp9//ll//PGH2rVrp0qVKql79+5asGCBKlasqOeff17du3dPM4bw8HA1bNhQMTExDl/Ux8TEqHPnzhkeD3727FnNnTtXNWrUsK87d+6cHnnkEXXr1k0ffvihrly5okGDBqlt27b66aefNH78eO3fv1/lypXTiBEjJN38DOfPn68nn3xS+/btU3BwsPz8/CRJo0aN0ldffaXPPvtMERERWr16tZ599lnly5dPdevWtZ938ODB+uCDD1SiRIlMJ9P3ktMJzcqVK/X444+rRIkS+v3331WuXDkdOXJEhmGoSpUqTrVVrFgxbdmyRcWKFdPp06e1e/du1a5d2749Pj5eISEhzoaILLT0g8FqPmC02WEAdpO71VC3yRvMDgNwMKnrg+o+ZaPZYSCbGjVqVKou9MOGDdObb76Z5v5Lly5VYGCgw7r//ve/+u9//+v0uSdOnKjJkyfr2rVrun79unx9fW87gVOhQoX00UcfaeDAgRo+fLiqVaum+vXrq0OHDipRokSax3z++eeKjIzU+++/L+nmA9d37dqld955x2G/69ev67PPPlPJkiUlSS+++KL9j/B/CwkJUVBQkL17W1r279+vuXPnasWKFWrYsKEkOcSYI0cOh/e9ePHi+u233zR37lyHhCZXrlyaMGGCPD09VaZMGT322GNauXKlunfvrty5c8vT01NBQUHpxiFJ3bp1U8+ePTV27Fj5+Pho69at2rlzp7755pt0j5H+7/MxDEOXL19W6dKl9f3339u3T5gwQZUrV9bIkSPt66ZOnaoiRYpo//79Kl26tLy9ve2Vnlty584tScqfP7+9G19SUpJGjhypH3/8UQ899JD9/VqzZo0+//xzh4RmxIgRatSo0W1jz06c7nI2ZMgQDRgwQDt37pSvr6/mz5+vP//8U3Xr1lWbNm2cais6OlovvPCC3nrrLbVp00ZlypRR1apV7dvXrVuX5QPOcGfxf+zRiT3bHNad2LNNfx3cY1JEcHeViuZSjZJ5HNbVKJlHFYvmNCcgQFLFojm5L+8nLpwUYMiQITp//rzDMmTIkHRDqV+/vmJjYx2Wnj17ZuqyOnTooNjYWK1du1ZNmzbVa6+9plq1at32mBdeeEHx8fGaOXOmHnroIc2bN08PPPBAul3V9u3bp+rVqzuse/DBB1Pt5+/vb09mJCksLEx///13Jq7qptjYWHl6ejr8If5vn3zyiapWrap8+fIpMDBQX3zxhY4dO+awzwMPPOBQAcpMXK1atZKnp6cWLlwo6WYXu/r169u7yqXn1uezfft2rVmzRqVKldKjjz6qxMRESdL27dv1888/KzAw0L6UKVNG0s0xRs74448/dPnyZTVq1MihvRkzZqRq65+VJitwukKzd+9e+8A0Ly8vXblyRYGBgRoxYoRatmypXr16ZbitgQMH6vLly1qwYIFCQ0M1b948h+1r167V008/7WyIuEuBufNpw9xJOrJtna5fvay1Mz/R1YsXVKNNN7NDg5uKO39FA5qVVYMHQuXv46X/Pv6AcgZ4a8x3e80ODW4s/vxV9W9aRo88UED+Pl4a8niUcvp7a+yy380ODdmMs93LAgICVKpUqXS3BwUF6fz586nWnzt3LlXPlpCQEHtbc+fOValSpVSzZk17ReN252jRooVatGiht99+W40bN9bbb799V9/a58iRw+G1zWa744xrt3OrG1V6Zs+erQEDBmjMmDF66KGHFBQUpPfff18bNjhW+dOK63bjXtLi7e2tTp06KSYmRk888YT+97//pRqvk5Z/fj63xv+EhYVpzpw56tatmy5evKgWLVro3XffTXVsWFiYUzFevHhRkvTtt9+qUKFCDtv+fX/++7Es2Z3TFZqAgAD7uJmwsDCHjO5Wn8UMn9zDQyNGjNC2bdu0bNkylS1b1mH7vHnz9NxzzzkbIu5SYO58atDzv/L2C9CZYwfl7R+oBj2GKDB3PrNDg5v66/xVvTprmxKv3lBUoRAlXr2ugbO26a/zV80ODW7sr/NXNXB2rC7eui+v3NCg2bHclxZlpCS7bMlqkZGR2rJli8O65ORkbd++XaVLl073uMDAQL300ksaMGCAU4mEzWZTmTJldOnSpXTj+feA+E2bNmW4/cwqX768UlJStGrVqjS3r127VrVq1VLv3r1VuXJllSpVyumqhnQzWUnOwANSu3Xrph9//FETJ07UjRs39MQTTzh9rluVoitXrkiSqlSpot27dys8PFylSpVyWG4lHWnF5+3tLUkO66OiouTj46Njx46laqtIkSJOx5qdZDihGTFihC5duqSaNWvaZzto1qyZ+vfvr3feeUddunSxz36RFfbv369BgwapcOHCWdYmMuZSwmn99MW7SrpySXmKllTS5Yv66Yt3dSnBuYQVyCr5g300+ulKCvLz0p4T5xXkl0Ojn66k/MH3bkAt8G/5gn00ul1FBfnm0J4T5xXsl0Oj21VUPu5L3KWkpCTFx8c7LP/80viVV17R5MmTNXHiRB04cECxsbF6/vnnlZCQoG7dbt+bokePHtq/f7/mz5+f5vbY2Fi1bNlSX3/9tfbs2aM//vhDU6ZM0dSpU9WyZct02/z99981aNAg+7iWW8+TSWtWsKwSHh6u6OhodenSRYsWLdLhw4f1yy+/aO7cuZJuPgx+8+bN+v7777V//34NHTo0U4lWeHi4Vq9erRMnTtz2y/uyZcuqZs2aGjRokNq3b3/HCpJ0c6KEW5/x9u3b1atXL/n6+urRRx+VdLP739mzZ9W+fXtt2rRJBw8e1Pfff6/nnnvOnqyEh4drw4YNOnLkiE6fPq2UlBQVK1ZMNptNS5cu1alTp3Tx4kUFBQVpwIABevnllzV9+nQdPHhQW7du1ccff6zp06c7/b5kJxlOaIYPH65Lly5p7Nix9tkXhg8frgYNGmjOnDkKDw+3T5OXWZcvX1ZMTIzq1KmjqKgorVq1Sq+88sptj0lKStKFCxcclhvXkm57DG4v8czfivzPo/pPhxeUw9df/+nwgiL/86gSz2S+nytwNwrm9NfCTX9q5De7dTnphkZ+s1sLN/2pgjn9zQ4NbqxgTj8t3HxcIxf///ty8W4t3HxcBXPe+Y8YZEPZ6Dk0y5cvV1hYmMPyn//8x769ffv2mjx5sqZOnaqqVauqSZMmio+P1+rVq1WgQIHbtp07d2516tRJb775ZprdqgoXLqzw8HANHz5cNWrUUJUqVTR+/HgNHz5cr732WpptFi9eXF9//bUWLFigChUq6NNPP7Xv6+qZ3D799FM99dRT6t27t8qUKaPu3bvbK0k9evTQE088oXbt2qlGjRo6c+aMevfu7fQ5RowYoSNHjqhkyZLKl+/2vVW6du2qa9euqUuXLhlqe9KkSfbPuH79+jp9+rS+++47RUZGSpIKFiyotWvXKjk5WY8++qjKly+vfv36KWfOnPbZ0wYMGCBPT09FRUUpX758OnbsmAoVKqThw4dr8ODBKlCggF588UVJ0ltvvaWhQ4dq1KhRKlu2rJo0aaJvv/1WxYsXd/p9yU5sRgZrjh4eHoqPj1f+/PmzPIj169dr8uTJmjdvnooWLaq9e/fq559/Vp06de547Jtvvplq5pCG0X306HPpz+CBjGOWs6wz8/v9ZodwX2CWs6zjym9O3Q2znGWNbe80Ne3c136d7bK2veu433jgd955R5999pn+/PNPs0O5p9566y3NmzdPO3bsMDsUt+LUGJqs/s9vzJgxeuCBB/TUU08pV65cWr16tXbu3CmbzaY8efLcuQEpzZlDHunQI0vjdGckM8huSGaQHZHMwN1NnDhRmzZt0qFDh/Tll1/q/fffV3R0tNlh3TMXL17Url27NGHCBPXp08fscNyOU7OclS5d+o5JzdmzZzPc3qBBgzRo0CCNGDEiQw9uSktaM4d4edN/GQAAZJyRgUHfSN+BAwf09ttv6+zZsypatKj69+9/26mp7zcvvviiZs2apVatWmW4uxmyjlNdzsaNG3fHB106k42PGjVKMTExunr1qtq3b6+OHTuqXLlyypEjh7Zv366oqKgMt/VP76/6I1PHAa5ElzNkN3Q5Q3ZjZpezpF9SP60+q/jU6+CytgE4WaF5+umns3QMzZAhQzRkyBCtWrVKU6dOVY0aNVSqVCkZhqGEhIQsOw8AAMBtuWB6ZQD3RobH0Ljym7y6detq+vTpio+PV+/evVW1alXVrVtXtWrV0tixY112XgAAAADWluGE5m6e5JpRQUFB6tGjhzZs2KBt27bpwQcf1OjRDEoHAAAulo2mbQbgnAwnNCkpKVk+ZXOzZs10/vx5++vRo0fr3Llzkm4+/XXo0KEKDg7O0nMCAAAAuH84NW1zVvv++++VlPR/D8EcOXKkwyxpN27c0OHDh80IDQAAuBEjJcVlCwDXMjWh+Xc3tnvRrQ0AAADA/cOpWc4AAADuS4x1ASzL1ITGZrOlmj2N5yIAAIB7joQGsCxTExrDMNS5c2f5+PhIkq5evaqePXsqICBAkhzG1wAAAADAv5ma0ERHRzu8fvbZZ1Pt06lTp3sVDgAAcFMM3gesy9SEJiYmxszTAwAAALA4JgUAAABgDA1gWaZO2wwAAAAAd4MKDQAAABUawLKo0AAAAACwLCo0AADA7RnJVGgAqyKhAQAAYNpmwLLocgYAAADAsqjQAAAAMCkAYFlUaAAAAABYFhUaAADg9gwqNIBlUaEBAAAAYFlUaAAAgNszmOUMsCwqNAAAAAAsiwoNAABwe0YyFRrAqkhoAACA2yOhAayLLmcAAAAALIsKDQAAcHtMCgBYFxUaAAAAAJZ1X1ZoXus30uwQgFS8fAPNDgFw4J+3oNkhAP/S1LQzM4YGsC4qNAAAAAAs676s0AAAADiDCg1gXVRoAAAAAFgWFRoAAOD2UpKTzQ4BQCaR0AAAALfHtM2AddHlDAAAAIBlUaEBAABuj0kBAOuiQgMAAADAsqjQAAAAt0eFBrAuKjQAAAAALIsKDQAAcHvMcgZYFxUaAAAAAJZFhQYAALi9FMbQAJZFQgMAANwekwIA1kWXMwAAAACWRYUGAAC4PSo0gHVRoQEAAABgWVRoAACA22PaZsC6qNAAAAAAsCwqNAAAwO0xhgawLio0AAAAACwrWyc0CxYsUIUKFcwOAwAA3OeM5BSXLQBcy/SE5vPPP9dTTz2lZ555Rhs2bJAk/fTTT6pcubI6duyo2rVrmxwhAAC436WkpLhsAeBapiY0o0ePVp8+fXTkyBEtXrxYjzzyiEaOHKkOHTqoXbt2On78uD799FMzQwQAAACQjZk6KUBMTIwmTZqk6Oho/frrr6pbt67WrVunP/74QwEBAWaGBgAA3AhdwwDrMrVCc+zYMT3yyCOSpDp16ihHjhwaPnw4yQwAAACADDE1oUlKSpKvr6/9tbe3t3Lnzm1iRPi3FZMGmh0C4GD5J33MDgFI5ZuRz5gdAu6SkZzssgWAa5n+HJqhQ4fK399fknTt2jW9/fbbCgkJcdhn7NixZoTmth6qWEp+vt76acMe+7pHakTp8tUkrd9+0MTI4K5qViguP58c+nnTfvu6+tVL6/LV69qw87CJkcGdPVi2kHy9vbR6+1H7uocrFtOVpBva9PsJEyMDAPdiakLz8MMPa9++ffbXtWrV0qFDhxz2sdls9zost/dn/Bm93/9ptXqkioL8ffXRkGeVN2eQBo6dbXZocFN/xifo3Zdaq2Xdigry99W4AW2UJ2egBn+00OzQ4MaOn7qgt7o2UPNakQr089Z7vR5V7mA/vTHlJ7NDQyYYzEYGWJapCc0vv/xi5umRjuN/Jaj9wE/1dp8nVSUqXCt+26W+o74yOyy4sRN/n9Ozr8VoeK8WqlymiH7csFf9PphndlhwcydPJ6rru4v0eqe6qlgqVD9tPaSBn/5gdlgA4HZMfw4Nsp+C+XLqq9E9FBLkr617jihncIC+Gt1DBfPlNDs0uKmC+UI0fUS0cgb6advvfypnkL+mj4hWwXwhdz4YcJHQ3IGa9OrjCgnw0fY/4pUz0FeTXn1cobkDzQ4NmcCDNQHrMrVC88QTT6S5PiQkRKVLl1a3bt2UL1++27aRlJSkpKQkh3VGSrJsHp5ZFqe7KVYwr6Yu/FU/bdijFZMGqs/IL/VIjSgVK5hXJ0+dMzs8uKGiYbk1bclv+nnTfi3/pI9een+u6lcvraJhuXXy1Hmzw4ObKlogRF/+sF2rtx/VNyOf0auf/qCHKxZT0QIhij970ezw4CQSD8C6TE1o/j34/5Zz585p0qRJev/997V69WqVK1cu3TZGjRql4cOHO6zzKFBJnmGVszRWd/Lb9j9SrfvnBAHAvbZ+R+qB//+cIAAww8a9qQf+/3OCAADAvWEzDMMwO4i0pKSkqHv37vr777+1ZMmSdPdLq0KTp04fKjTIdrx86YaC7MU/b0GzQwAcnF4y2LRz73s+7V4jWSHyiwUuaxtANpi2OT0eHh7q27evmjZtetv9fHx85OPj47COZAYAAABwD9k2oZGkgIAAXb582ewwAADAfY5pmwHrytaznK1YsUKlS5c2OwwAAAAA2ZSpFZrFixenuf78+fPasmWLJk+erMmTJ9/jqAAAgLthljPAukxNaFq1apXm+qCgIEVGRmry5Ml6+umn721QAAAAACzD1IQmhf6qAAAgGzCSs+WkrwAywNQxNM2aNdP58//3ULzRo0fr3Llz9tdnzpxRVFSUCZEBAAB3kpKc4rIFgGuZmtAsX77c4RkyI0eO1NmzZ+2vb9y4oX379pkRGgAAAAALyFbTNmfTZ3wCAID7nJHC3yCAVWXraZsBAAAA4HZMrdDYbDbZbLZU6wAAAO6lFCYFACzL1ITGMAx17txZPj4+kqSrV6+qZ8+eCggIkCSH8TUAAAAA8G+mJjTR0dEOr5999tlU+3Tq1OlehQMAANwUD9YErMvUhCYmJsbM0wMAAACwuGw1yxkAAIAZeLAmYF3McgYAAADAsqjQAAAAt8csZ4B1kdAAAAC3x6QAgHXR5QwAAACAZVGhAQAAbi8lhS5ngFVRoQEAAABgWVRoAACA22PaZsC6qNAAAAAAsCwqNAAAwO2lMMsZYFlUaAAAAABYFhUaAADg9hhDA1gXCQ0AAHB7JDSAddHlDAAAAIBlUaEBAABuj0kBAOuiQgMAAADAsqjQAAAAt2ekMIYGsKr7MqHZ/O2HZocApFI4KIfZIQAOTly8bnYIAADctfsyoQEAAHBGCrOcAZbFGBoAAAAAlkWFBgAAuD2DWc4AyyKhAQAAbo8HawLWRZczAAAAAJZFhQYAALg9JgUArIsKDQAAAADLokIDAADcnpHCpACAVVGhAQAAAGBZVGgAAIDbYwwNYF1UaAAAAABYFhUaAADg9ngODWBdJDQAAMDtGclMCgBYFV3OAAAAAFgWFRoAAOD2mBQAsC4qNAAAAAAsiwoNAABwe0wKAFgXFRoAAAAAlkWFBgAAuL0UgwoNYFVUaAAAAABYVrar0Bw9elSXLl1SmTJl5OFBvgUAAFwvmQoNYFmmZQxTp07V2LFjHdY9//zzKlGihMqXL69y5crpzz//NCk6AADgTpIN1y0AXMu0hOaLL75Qrly57K+XL1+umJgYzZgxQ5s2bVLOnDk1fPhws8IDAAAAYAGmdTk7cOCAqlWrZn/9zTffqGXLlurQoYMkaeTIkXruuefMCg8AALgRupwB1mVahebKlSsKDg62v163bp0efvhh++sSJUooPj7ejNAAAAAAWIRpCU2xYsW0ZcsWSdLp06e1e/du1a5d2749Pj5eISEhZoUHAADcCGNoAOsyrctZdHS0XnjhBe3evVs//fSTypQpo6pVq9q3r1u3TuXKlTMrPAAAAAAWYFpCM3DgQF2+fFkLFixQaGio5s2b57B97dq1evrpp02KDgAAuBPG0ADWZTOM7PsTnJycLE9PT6eP23HyvAuiAe5O4aAcZocAODhx8brZIQAOyoeZ19X8y7xlXdZ2x9N7XdY2ABPH0NzO/v37NWjQIBUuXNjsUAAAgBtgDA1gXaZ1Ofu3y5cva86cOZo6dap+++03VatWTa+88orZYQEAADdAlzPAukxPaNavX6/Jkydr3rx5Klq0qPbu3auff/5ZderUydDxSUlJSkpKclh3LSlJ3j4+rggXAAAAQDZiWpezMWPG6IEHHtBTTz2lXLlyafXq1dq5c6dsNpvy5MmT4XZGjRqlkJAQh2XKhLEujBwAANxv6HIGWJdpkwJ4eXlp0KBBGjFihMPA/xw5cmj79u2KiorKUDtpVWj2n7lKhQbZDpMCILthUgBkN2ZOCvBFrjIua/v5hN9d1jYAEys0b731lubNm6fixYtr0KBB2rVrV6ba8fHxUXBwsMNCMgMAAJxBhQawLtMSmiFDhmj//v368ssvFR8frxo1aqhixYoyDEMJCQlmhQUAAADAQkyftrlu3bqaPn264uPj1bt3b1WtWlV169ZVrVq1NHYsY2EAAIDrJRuGyxYArmV6QnNLUFCQevTooQ0bNig2NlY1atTQ6NGjzQ4LAAAAQDZmWkLz008/KSoqShcuXEi1rUiRIvr+++/1v//9z4TIAACAu2EMDWBdpiU048aNU/fu3RUcHJxqW0hIiHr27KlPPvnEhMgAAIC7ocsZYF2mJTTbt29XkyZN0t3+6KOPasuWLfcwIgAAAABW42XWif/66y/lyJH+czm8vLx06tSpexgRAABwV3QNA6zLtApNoUKFbvvsmR07digsLOweRgQAAADAakxLaJo1a6ahQ4fq6tWrqbZduXJFw4YNU/PmzU2IDAAAuBvG0ADWZTMMc37S/vrrL1WpUkWenp568cUXFRkZKUn6/fff9cknnyg5OVlbt25VgQIFnG57x8nzWR0ucNcKB6XfxRIww4mL180OAXBQPizEtHOPDohwWduDLx1wWdsATBxDU6BAAa1bt069evXSkCFDdCuvstlsaty4sT755JNMJTMAAADOYgwNYF2mJTSSVKxYMX333XdKSEjQH3/8IcMwFBERoVy5cpkZFgAAAACLMDWhuSVXrlyqXr262WEAAAA3xVgXwLqyRUIDAABgphSzAwCQaabNcgYAAAAAd4sKDQAAcHt0OQOsiwoNAAAAAMuiQgMAANwe0zYD1kWFBgAAAIBlUaEBAABujzE0gHVRoQEAAABgWVRoAACA22MMDWBdVGgAAAAAWBYVGgAA4PYYQwNYFwkNAABwe3Q5A6yLLmcAAAAALIsKDQAAcHt0OQOsiwoNAAAAAMuiQgMAANweY2gA66JCAwAAAMCybIZBp1GkLSkpSaNGjdKQIUPk4+NjdjgA9ySyJe5LADAXCQ3SdeHCBYWEhOj8+fMKDg42OxyAexLZEvclAJiLLmcAAAAALIuEBgAAAIBlkdAAAAAAsCwSGqTLx8dHw4YNY5Arsg3uSWRH3JcAYC4mBQAAAABgWVRoAAAAAFgWCQ0AAAAAyyKhAQAAAGBZJDQAAAAALIuExmS//fabPD099dhjj912v1mzZsnT01MvvPBCmtsvXLig1157TWXKlJGvr69CQ0PVsGFDLViwQGbN+/DLL7/IZrOlucTHx0uSOnfurFatWqV77Llz5yRJ06ZNsx/r4eGhsLAwtWvXTseOHbuHV3T/u939eOTIEdlsNnl6eurEiRMO2+Li4uTl5SWbzaYjR4447H9ryZ07t+rWratff/31XlzKbaV3X86ePVvSzfstZ86c6R67aNEiSdn7Gu8nnTt3TvPzatKkSap9R40aJU9PT73//vuptiUnJ2v06NEqU6aM/Pz8lDt3btWoUUOTJ0++F5eRpn/fQ/9c1q9fL0l68803ValSpXSPjY2NlZT6d26+fPnUrFkz7dy58x5eEQDceyQ0JpsyZYr69Omj1atX6+TJk7fdb+DAgZo1a5auXr3qsO3cuXOqVauWZsyYoSFDhmjr1q1avXq12rVrp4EDB+r8+fOuvozb2rdvn+Li4hyW/PnzO91OcHCw4uLidOLECc2fP1/79u1TmzZtXBCx+8rI/VioUCHNmDHDYd306dNVqFChNPf/8ccfFRcXp9WrV6tgwYJq3ry5/vrrryyP3VkxMTGp7su0kuuMyK7XeD9p0qRJqs9r1qxZqfabOnWqBg4cqKlTp6baNnz4cH344Yd66623tGfPHv388896/vnn7V+cmOnWPfTPpWrVqplq69bv3O+//15JSUl67LHHdO3atSyOGACyDxIaE128eFFz5sxRr1699Nhjj2natGlp7nf48GGtW7dOgwcPVunSpbVgwQKH7f/973915MgRbdiwQdHR0YqKilLp0qXVvXt3xcbGKjAwMFWb+/fvl81m0++//+6w/sMPP1TJkiUlSQkJCerQoYPy5csnPz8/RUREKCYmxunrzJ8/v0JDQx0WDw/nbz2bzabQ0FCFhYWpVq1a6tq1qzZu3KgLFy443RZSy+j9GB0dneo+iImJUXR0dJr758mTR6GhoSpXrpz++9//6sKFC9qwYUOa+6akpKhw4cL69NNPHdZv27ZNHh4eOnr0qAzD0JtvvqmiRYvKx8dHBQsWVN++fZ2+3pw5c6a6L319fZ1ux9lrROb4+Pik+rxy5crlsM+qVat05coVjRgxQhcuXNC6descti9evFi9e/dWmzZtVLx4cVWsWFFdu3bVgAED0jznhQsX5Ofnp2XLljmsX7hwoYKCgnT58mVdu3ZNL774osLCwuTr66tixYpp1KhRTl/frXvon0uOHDmcbkf6v9+5VapUUb9+/fTnn3+m+l0PAPcTEhoTzZ07V2XKlFFkZKSeffZZTZ06Nc3uYTExMXrssccUEhKiZ599VlOmTLFvS0lJ0ezZs9WhQwcVLFgw1bGBgYHy8vJKtb506dKqVq2aZs6c6bB+5syZeuaZZyRJQ4cO1Z49e7Rs2TLt3btXn376qfLmzXu3l50l/v77by1cuFCenp7y9PQ0O5z7Qkbvx8cff1wJCQlas2aNJGnNmjVKSEhQixYtbtv+lStX7JUdb2/vNPfx8PBQ+/bt9b///c9h/cyZM1W7dm0VK1ZM8+fP14cffqjPP/9cBw4c0KJFi1S+fPnMXHKWy8g1wnWmTJmi9u3bK0eOHGrfvr3D70pJCg0N1U8//aRTp05lqL3g4GA1b948zfuxVatW8vf310cffaTFixdr7ty52rdvn2bOnKnw8PCsuqS7cv78eXs3Su5HAPc1A6apVauWMW7cOMMwDOP69etG3rx5jZ9//tlhn+TkZKNIkSLGokWLDMMwjFOnThne3t7GoUOHDMMwjL/++suQZIwdO9bp83/44YdGyZIl7a/37dtnSDL27t1rGIZhtGjRwnjuuecyc2mGYRjGzz//bEgyAgICHJaoqCj7PtHR0UbLli3TPTYhIcEwDMOIiYmxt+Xv729IMiQZffv2zXR8cHSn+/Hw4cOGJGPbtm1Gv3797PfGc889Z7z88svGtm3bDEnG4cOHHfb38/MzAgICDJvNZkgyqlataly7di3dOLZt22bYbDbj6NGjhmHc/BkoVKiQ8emnnxqGYRhjxowxSpcufds27kSS4evrm+revHXOmJgYIyQkJN1jFy5ceFfXCOdER0cbnp6eqT6vd955x77P+fPnDT8/PyM2NtYwjJv3UWBgoJGYmGjfZ/fu3UbZsmUNDw8Po3z58kaPHj2M77777rbnXrhwoREYGGhcunTJfh5fX19j2bJlhmEYRp8+fYxHHnnESElJydS1/fse+udyy7Bhw4yKFSume+y2bdsMw0j9O/fW78nHH388U7EBgFVQoTHJvn37tHHjRrVv316S5OXlpXbt2qX6RnHFihW6dOmSmjVrJknKmzevGjVqZO8fbtzFgP+nn35aR44csQ88nTlzpqpUqaIyZcpIknr16qXZs2erUqVKGjhwYKruGxn166+/KjY21r589913mWonKChIsbGx2rx5s8aMGaMqVaronXfeyVRbcJTR+/GWLl26aN68eYqPj9e8efPUpUuXdNueM2eOtm3bpvnz56tUqVKaNm3abbvSVKpUSWXLlrV/K75q1Sr9/fff9vFSbdq00ZUrV1SiRAl1795dCxcu1I0bN5y+5g8//NDhvoyNjU2zypkRzl4jnFe/fv1Un1fPnj3t22fNmqWSJUuqYsWKkm7eR8WKFdOcOXPs+0RFRWnXrl1av369unTpor///lstWrRQt27d0j1vs2bNlCNHDi1evFiSNH/+fAUHB6thw4aSbk5YEBsbq8jISPXt21c//PBDpq5vzpw5qa4vs3799Vdt2bJF06ZNU+nSpfXZZ59lui0AsASzMyp39eqrrxqSDE9PT/vi4eFh+Pn5GefOnbPv16ZNm1T72Ww2o0iRIkZycrKRnJxs5MyZ03j++eczFUejRo2MPn36GIZhGKVKlTLGjBnjsP3vv/82pk2bZnTo0MHw9fU1+vfvn+G2/11lSUufPn2MevXqpVq/cOFCw9PT0/4td1rfmPfu3dt49tlnMxwP0peR+/Hf3wZXq1bNqFevnlG9enXDMIx0KzS39jcMw5g7d65RqlQp4+rVq7eN55133jHKly9vGIZhdOvWzWjRooXD9suXLxuLFy82+vTpY4SGhhoPPfSQUxUR/aPKkpb58+cbXl5eRnJyssP6hIQEQ5Lx008/3fU1IuPSq+T+U/Xq1Q2bzZbqd2WtWrVue9yXX35pSLJXvdPSvXt3+z3YsGFD++/MW86fP2/Mnj3b6NatmxESEmI8+eSTGbswI+176N/GjBljhIeHp1p/62fuVuxp/c597733jDp16mQ4HgCwIhIaE1y/ft0oUKCAMWbMGGPnzp0OS8mSJe1da06fPm14e3sbs2fPdtgnNjbWCAwMtHd56NmzpxEQEGCcOHEi1bkSExON69evpxvLtGnTjPz58xvr1q0zPDw80mzjls8++8wICgrK8HVmJKGZMGGCkS9fvlR//A0dOtQoVaqU/XVaCc2xY8eMHDlyGFu2bMlwTEgto/fjv//wmjhxoiHJvj0jCU1KSopRpkyZO3aRPHz4sGGz2YzNmzcbOXPmNGbPnp3uvr///rshyan74E4Jza5du9Jsc+XKlYYk4/jx4/Y4M3uNyLg7JTQ7duwwbDabsWrVKof7d9WqVYbNZrN3o03Lli1bDEnGzp07093nl19+MXLkyGHs2rXL8PDwMNavX5/uvsuXLzckGWfOnMnQtWUkoVm6dKnh5eVlxMfHO6yfMmWK4evra9y4ccMwjLR/5166dMnIlSuXsWDBggzFAwBWREJjgoULFxre3t4OlZhbBg4caFSrVs0wjJtjXMLCwtLsm922bVvjqaeeMgzDMM6cOWOUKVPGKFy4sDF9+nRj9+7dxv79+40pU6YYpUqVum1CceHCBcPPz8+oWLGi0aBBA4dtQ4cONRYtWmQcOHDA2LVrl9G8eXPjwQcftG9/5JFHjI8//jjdtm/957pv3z4jLi7OYbn1bXpCQoKRP39+o23btsbmzZuNAwcOGFOmTDGCgoLsfygbRvpjGtq2bWs89thj6caAO8vo/fjvP7yuX79unDp1yp4wZyShMYybiVD+/PntYxLSU7t2baNixYpGUFCQcfnyZfv6mJgYY/LkycbOnTuNgwcPGq+//rrh5+dnnD592jAMwxg8eLDRsWPH27YtyYiJiUl1X168eNG+z6OPPmpUrFjR+PHHH41Dhw4Zy5YtMyIjI4127drZ97nba0TGREdHG02aNEn1eZ06dcowDMN46aWXjBo1aqR57IMPPmgMGDDAMAzDePLJJ42xY8ca69evN44cOWL8/PPPRs2aNY3SpUvf9ouflJQUo0iRIkbFihUdxh0axs3qyf/+9z9j7969xr59+4yuXbsaoaGh9upex44djcGDB6fb9q176Mcff0x1fVeuXDEM4+bP2gMPPGDUr1/fWLt2rXHw4EFj3rx5RlhYmDFo0CB7W+l9iTRw4ECjfPnymR7nAwDZHQmNCZo3b240a9YszW0bNmwwJBnbt283ypcvb/Tu3TvN/ebMmWN4e3vb/0M/d+6cMXjwYCMiIsLw9vY2ChQoYDRs2NBYuHDhHf8Ta9u2rSHJmDp1qsP6t956yyhbtqzh5+dn5M6d22jZsqVDt4xixYoZw4YNS7fdW/+5prX89ttv9v327dtntG7d2ihYsKAREBBgVKxY0Zg0aZJD3OklNL/99pshydiwYcNtrxHpy+j9eKdvkjOa0Nz6xvjdd9+9bVy3KkCdOnVyWL9w4UKjRo0aRnBwsBEQEGDUrFnT+PHHH+3bo6Ojjbp169627fTuy1GjRtn3SUhIMPr27WuULFnS8PPzMyIiIoyBAwc6DDK/22tExkRHR6f5eUVGRhpJSUlGnjx5jPfeey/NY999910jf/78xrVr14wvvvjCqF+/vpEvXz7D29vbKFq0qNG5c2fjyJEjd4xh4MCBhiTjjTfecFj/xRdfGJUqVTICAgKM4OBgo0GDBsbWrVvt2+vWrWtER0en2+6teyitZdasWfb9Tpw4YURHRxtFixY1/Pz8jKioKGP06NEOXS3TS2iOHTtmeHl5GXPmzLnjdQKAFdkMw6THyAMAAADAXWKWMwAAAACWRUIDAAAAwLJIaAAAAABYFgkNAAAAAMsioQEAAABgWSQ0AAAAACyLhAYAAACAZZHQAMiwN998U5UqVTI7jPtG586d1apVK7PDAADA0khoAJN17txZNpvNvuTJk0dNmjTRjh07zA4tS9hsNi1atMjsMLKl8ePHa9q0aWaHAQCApZHQANlAkyZNFBcXp7i4OK1cuVJeXl5q3ry52WHBRZKTk5WSkqKQkBDlzJnT7HAAALA0EhogG/Dx8VFoaKhCQ0NVqVIlDR48WH/++adOnTpl32fQoEEqXbq0/P39VaJECQ0dOlTXr1+3b9++fbvq16+voKAgBQcHq2rVqtq8ebN9+5o1a1SnTh35+fmpSJEi6tu3ry5dunTbuEaPHq0CBQooKChIXbt21dWrVx22b9q0SY0aNVLevHkVEhKiunXrauvWrfbt4eHhkqTWrVvLZrPZXx88eFAtW7ZUgQIFFBgYqOrVq+vHH3+84/u0ZMkSVa9eXb6+vsqbN69at25t35aQkKBOnTopV65c8vf3V9OmTXXgwAH79mnTpilnzpxaunSpIiMj5e/vr6eeekqXL1/W9OnTFR4erly5cqlv375KTk52uIa33npL7du3V0BAgAoVKqRPPvnEIa6xY8eqfPnyCggIUJEiRdS7d29dvHgx1bkXL16sqKgo+fj46NixY6m6nH399dcqX768/Pz8lCdPHjVs2ND+GaWkpGjEiBEqXLiwfHx8VKlSJS1fvtx+7JEjR2Sz2bRgwQLVr19f/v7+qlixon777bc7vq8AAFgZCQ2QzVy8eFFfffWVSpUqpTx58tjXBwUFadq0adqzZ4/Gjx+vSZMm6cMPP7Rv79ChgwoXLqxNmzZpy5YtGjx4sHLkyCHpZgLRpEkTPfnkk9qxY4fmzJmjNWvW6MUXX0w3jrlz5+rNN9/UyJEjtXnzZoWFhWnixIkO+yQmJio6Olpr1qzR+vXrFRERoWbNmikxMVHSzYRHkmJiYhQXF2d/ffHiRTVr1kwrV67Utm3b1KRJE7Vo0ULHjh1LN55vv/1WrVu3VrNmzbRt2zatXLlSDz74oH17586dtXnzZi1evFi//fabDMNQs2bNHJK+y5cv66OPPtLs2bO1fPly/fLLL2rdurW+++47fffdd/ryyy/1+eef6+uvv3Y49/vvv6+KFStq27ZtGjx4sF566SWtWLHCvt3Dw0MfffSRdu/erenTp+unn37SwIEDHdq4fPmy3n33XU2ePFm7d+9W/vz5HbbHxcWpffv26tKli/bu3atffvlFTzzxhAzDkHSze9qYMWP0wQcfaMeOHWrcuLEef/xxh6RNkl577TUNGDBAsbGxKl26tNq3b68bN26k+74CAGB5BgBTRUdHG56enkZAQIAREBBgSDLCwsKMLVu23Pa4999/36hatar9dVBQkDFt2rQ09+3atavx/PPPO6z79ddfDQ8PD+PKlStpHvPQQw8ZvXv3dlhXo0YNo2LFiunGlJycbAQFBRlLliyxr5NkLFy48LbXYhiG8cADDxgff/xxutsfeugho0OHDmlu279/vyHJWLt2rX3d6dOnDT8/P2Pu3LmGYRhGTEyMIcn4448/7Pv06NHD8Pf3NxITE+3rGjdubPTo0cP+ulixYkaTJk0czteuXTujadOm6cY6b948I0+ePPbXt84dGxvrsF90dLTRsmVLwzAMY8uWLYYk48iRI2m2WbBgQeOdd95xWFe9enX7Z3T48GFDkjF58mT79t27dxuSjL1796YbKwAAVkeFBsgG6tevr9jYWMXGxmrjxo1q3LixmjZtqqNHj9r3mTNnjmrXrq3Q0FAFBgbq9ddfd6hovPLKK+rWrZsaNmyo0aNH6+DBg/Zt27dv17Rp0xQYGGhfGjdurJSUFB0+fDjNmPbu3asaNWo4rHvooYccXv/111/q3r27IiIiFBISouDgYF28ePG2lRbpZoVmwIABKlu2rHLmzKnAwEDt3bv3tsfFxsaqQYMG6cbq5eXlEG+ePHkUGRmpvXv32tf5+/urZMmS9tcFChRQeHi4AgMDHdb9/ffft73uhx56yKHdH3/8UQ0aNFChQoUUFBSkjh076syZM7p8+bJ9H29vb1WoUCHd66tYsaIaNGig8uXLq02bNpo0aZISEhIkSRcuXNDJkydVu3Zth2Nq167tEIckh3OEhYVJUqrrAQDgfkJCA2QDAQEBKlWqlEqVKqXq1atr8uTJunTpkiZNmiRJ+u2339ShQwc1a9ZMS5cu1bZt2/Taa6/p2rVr9jbefPNN7d69W4899ph++uknRUVFaeHChZJuJhA9evSwJ02xsbHavn27Dhw44PAHvrOio6MVGxur8ePHa926dYqNjVWePHkc4krLgAEDtHDhQo0cOVK//vqrYmNjVb58+dse5+fnl+k4b7nVBe8Wm82W5rqUlJQMt3nkyBE1b95cFSpU0Pz587Vlyxb7GJt/Xo+fn59sNlu67Xh6emrFihVatmyZoqKi9PHHHysyMjLdhDM9/7yeW+dz5noAALAaEhogG7LZbPLw8NCVK1ckSevWrVOxYsX02muvqVq1aoqIiHCo3txSunRpvfzyy/rhhx/0xBNPKCYmRpJUpUoV7dmzx540/XPx9vZOM4ayZctqw4YNDuvWr1/v8Hrt2rXq27evmjVrpgceeEA+Pj46ffq0wz45cuRwGGR/67jOnTurdevWKl++vEJDQ3XkyJHbvicVKlTQypUr0431xo0bDvGeOXNG+/btU1RU1G3bzYh/X/f69etVtmxZSdKWLVuUkpKiMWPGqGbNmipdurROnjyZqfPYbDbVrl1bw4cP17Zt2+Tt7a2FCxcqODhYBQsW1Nq1ax32X7t2bZZcHwAAVuZldgAApKSkJMXHx0u6OVvXhAkTdPHiRbVo0UKSFBERoWPHjmn27NmqXr26vv32W3v1RZKuXLmiV199VU899ZSKFy+u48ePa9OmTXryyScl3ZwhrWbNmnrxxRfVrVs3BQQEaM+ePVqxYoUmTJiQZkwvvfSSOnfurGrVqql27dqaOXOmdu/erRIlStj3iYiI0Jdffqlq1arpwoULevXVV1NVUsLDw7Vy5UrVrl1bPj4+ypUrlyIiIrRgwQK1aNFCNptNQ4cOvWMVYdiwYWrQoIFKliypp59+Wjdu3NB3332nQYMGKSIiQi1btlT37t31+eefKygoSIMHD1ahQoXUsmVL5z+Qf1m7dq3ee+89tWrVSitWrNC8efP07bffSpJKlSql69ev6+OPP1aLFi20du1affbZZ06fY8OGDVq5cqUeffRR5c+fXxs2bNCpU6fsidOrr76qYcOGqWTJkqpUqZJiYmIUGxurmTNn3vX1AQBgZVRogGxg+fLlCgsLU1hYmGrUqKFNmzZp3rx5qlevniTp8ccf18svv6wXX3xRlSpV0rp16zR06FD78Z6enjpz5ow6deqk0qVLq23btmratKmGDx8u6WZ1Y9WqVdq/f7/q1KmjypUr64033lDBggXTjaldu3YaOnSoBg4cqKpVq+ro0aPq1auXwz5TpkxRQkKCqlSpoo4dO6pv376pZu8aM2aMVqxYoSJFiqhy5cqSbk5znCtXLtWqVUstWrRQ48aNVaVKldu+R/Xq1dO8efO0ePFiVapUSY888og2btxo3x4TE6OqVauqefPmeuihh2QYhr777rtUXcoyo3///tq8ebMqV66st99+W2PHjlXjxo0l3Rz7MnbsWL377rsqV66cZs6cqVGjRjl9juDgYK1evVrNmjVT6dKl9frrr2vMmDFq2rSpJKlv37565ZVX1L9/f5UvX17Lly/X4sWLFRERcdfXBwCAldkM4//PCQoASCU8PFz9+vVTv379zA4FAACkgQoNAAAAAMsioQEAAABgWXQ5AwAAAGBZVGgAAAAAWBYJDQAAAADLIqEBAAAAYFkkNAAAAAAsi4QGAAAAgGWR0AAAAACwLBIaAAAAAJZFQgMAAADAskhoAAAAAFjW/wOeKOTiF+iHwgAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 800x600 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "compare_rocs(\n",
    "    f\"{RESULTS_DIR}/model1_comparison.txt\", \n",
    "    f\"{RESULTS_DIR}/supp_fig3.png\", \n",
    "    \"EUR\", \n",
    "    \"non-EUR\",\n",
    "    \"Base data comparison\",\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5d77520-631e-41ff-abcf-b9a7f0b87fcd",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Model 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0a4f7b3-1c9a-45f6-88d6-db60710203ed",
   "metadata": {
    "tags": []
   },
   "source": [
    "## PRSice"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94096fb8-8c67-449d-aa11-7ff6f1b8182e",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Format covars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25329dad-7399-4fa5-ac99-e6c66c5da333",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_covar = pd.read_csv(f\"{RESULTS_DIR}/covariates.txt\", sep='\\s+')\n",
    "df_covar.drop([\"New_Column\", \"ANCESTRY\"], axis=1, inplace=True)\n",
    "df_covar.rename({\"GP2ID\":\"IID\"}, inplace=True, axis=1)\n",
    "df_covar.to_csv(f\"{RESULTS_DIR}/covariates_prs.txt\", sep=\"\\t\", index=False)\n",
    "df_covar\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a255d7d-88a6-4345-9f07-2a900ecd6d01",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Run PRSice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d6eebdaf-f3c4-400e-8f59-ebfea0cdb45e",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PRSice 2.3.5 (2021-09-20) \n",
      "https://github.com/choishingwan/PRSice\n",
      "(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly\n",
      "GNU General Public License v3\n",
      "If you use PRSice in any published work, please cite:\n",
      "Choi SW, O'Reilly PF.\n",
      "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.\n",
      "GigaScience 8, no. 7 (July 1, 2019)\n",
      "2025-02-11 15:21:18\n",
      "/home/jupyter/prsice \\\n",
      "    --a1 A1 \\\n",
      "    --a2 A2 \\\n",
      "    --bar-levels 5e-08,5e-07,5e-06,5e-05,0.0005,0.005,0.05 \\\n",
      "    --base /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
      "    --beta  \\\n",
      "    --binary-target T \\\n",
      "    --bp BP \\\n",
      "    --chr CHR \\\n",
      "    --clump-kb 250kb \\\n",
      "    --clump-p 1.000000 \\\n",
      "    --clump-r2 0.100000 \\\n",
      "    --cov /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
      "    --fastscore  \\\n",
      "    --ignore-fid  \\\n",
      "    --ld /home/jupyter/workspace/ws_files/results/AAC/model2_plink/chr# \\\n",
      "    --no-full  \\\n",
      "    --num-auto 22 \\\n",
      "    --out /home/jupyter/workspace/ws_files/results/AAC/model2_AAC \\\n",
      "    --perm 10000 \\\n",
      "    --prevalence 0.005 \\\n",
      "    --print-snp  \\\n",
      "    --pvalue P \\\n",
      "    --score con-std \\\n",
      "    --seed 3 \\\n",
      "    --snp SNP \\\n",
      "    --stat BETA \\\n",
      "    --target /home/jupyter/workspace/ws_files/results/AAC/model2_plink/chr# \\\n",
      "    --thread 16\n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AAC/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Start processing multi_MAMA \n",
      "================================================== \n",
      "\n",
      "Base file: \n",
      "/home/jupyter/workspace/ws_files/data/multi_MAMA.txt \n",
      "Header of file is: \n",
      "SNP CHR BP A1 A2 BETA P TEST \n",
      "\n",
      "Reading 100.00%\n",
      "17782530 variant(s) observed in base file, with: \n",
      "16969225 variant(s) excluded due to p-value threshold \n",
      "122051 ambiguous variant(s) excluded \n",
      "691254 total variant(s) included from base file \n",
      "\n",
      "Loading Genotype info from target \n",
      "================================================== \n",
      "\n",
      "1196 people (498 male(s), 698 female(s)) observed \n",
      "1196 founder(s) included \n",
      "\n",
      "57758242 variant(s) not found in previous data \n",
      "654052 variant(s) included \n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AAC/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Loading Genotype info from reference \n",
      "================================================== \n",
      "\n",
      "1196 people (498 male(s), 698 female(s)) observed \n",
      "1196 founder(s) included \n",
      "\n",
      "57758242 variant(s) not found in previous data \n",
      "654052 variant(s) included \n",
      "\n",
      "There are a total of 1 phenotype to process \n",
      "\n",
      "Start performing clumping \n",
      "\n",
      "Clumping Progress: 100.00%\n",
      "Number of variant(s) after clumping : 160832 \n",
      "\n",
      "Processing the 1 th phenotype \n",
      "\n",
      "Phenotype is a binary phenotype \n",
      "819 control(s) \n",
      "325 case(s) \n",
      "\n",
      "Processing the covariate file: \n",
      "/home/jupyter/workspace/ws_files/results/covariates_prs.txt \n",
      "============================== \n",
      "\n",
      "Include Covariates: \n",
      "Name\tMissing\tNumber of levels \n",
      "SEX\t0\t- \n",
      "AGE\t0\t- \n",
      "PC1\t0\t- \n",
      "PC2\t0\t- \n",
      "PC3\t0\t- \n",
      "PC4\t0\t- \n",
      "PC5\t0\t- \n",
      "PC6\t0\t- \n",
      "PC7\t0\t- \n",
      "PC8\t0\t- \n",
      "PC9\t0\t- \n",
      "PC10\t0\t- \n",
      "\n",
      "After reading the covariate file, 1144 sample(s) included \n",
      "in the analysis \n",
      "\n",
      "\n",
      "Start Processing\n",
      "Processing 100.00%\n",
      "There are 1 region(s) with p-value less than 1e-5. Please \n",
      "note that these results are inflated due to the overfitting \n",
      "inherent in finding the best-fit PRS (but it's still best \n",
      "to find the best-fit PRS!). \n",
      "You can use the --perm option (see manual) to calculate an \n",
      "empirical P-value. \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Begin plotting\n",
      "Current Rscript version = 2.3.3\n",
      "Plotting the quantile plot\n",
      "WARNING: There are only 0 unique PRS but asked for 4 quantiles\n",
      "Will not generate the quantile plot for  /home/jupyter/workspace/ws_files/results/AAC/model2_AAC\n",
      "Plotting Bar Plot\n",
      "Plotting Multi-Set-Plot\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In max(pheno$Pheno, na.rm = T) :\n",
      "  no non-missing arguments to max; returning -Inf\n",
      "PRSice 2.3.5 (2021-09-20) \n",
      "https://github.com/choishingwan/PRSice\n",
      "(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly\n",
      "GNU General Public License v3\n",
      "If you use PRSice in any published work, please cite:\n",
      "Choi SW, O'Reilly PF.\n",
      "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.\n",
      "GigaScience 8, no. 7 (July 1, 2019)\n",
      "2025-02-11 15:28:50\n",
      "/home/jupyter/prsice \\\n",
      "    --a1 A1 \\\n",
      "    --a2 A2 \\\n",
      "    --bar-levels 5e-08,5e-07,5e-06,5e-05,0.0005,0.005,0.05 \\\n",
      "    --base /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
      "    --beta  \\\n",
      "    --binary-target T \\\n",
      "    --bp BP \\\n",
      "    --chr CHR \\\n",
      "    --clump-kb 250kb \\\n",
      "    --clump-p 1.000000 \\\n",
      "    --clump-r2 0.100000 \\\n",
      "    --cov /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
      "    --fastscore  \\\n",
      "    --ignore-fid  \\\n",
      "    --ld /home/jupyter/workspace/ws_files/results/AFR/model2_plink/chr# \\\n",
      "    --no-full  \\\n",
      "    --num-auto 22 \\\n",
      "    --out /home/jupyter/workspace/ws_files/results/AFR/model2_AFR \\\n",
      "    --perm 10000 \\\n",
      "    --prevalence 0.005 \\\n",
      "    --print-snp  \\\n",
      "    --pvalue P \\\n",
      "    --score con-std \\\n",
      "    --seed 3 \\\n",
      "    --snp SNP \\\n",
      "    --stat BETA \\\n",
      "    --target /home/jupyter/workspace/ws_files/results/AFR/model2_plink/chr# \\\n",
      "    --thread 16\n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AFR/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Start processing multi_MAMA \n",
      "================================================== \n",
      "\n",
      "Base file: \n",
      "/home/jupyter/workspace/ws_files/data/multi_MAMA.txt \n",
      "Header of file is: \n",
      "SNP CHR BP A1 A2 BETA P TEST \n",
      "\n",
      "Reading 100.00%\n",
      "17782530 variant(s) observed in base file, with: \n",
      "16969225 variant(s) excluded due to p-value threshold \n",
      "122051 ambiguous variant(s) excluded \n",
      "691254 total variant(s) included from base file \n",
      "\n",
      "Loading Genotype info from target \n",
      "================================================== \n",
      "\n",
      "2648 people (1458 male(s), 1190 female(s)) observed \n",
      "2648 founder(s) included \n",
      "\n",
      "65317063 variant(s) not found in previous data \n",
      "632507 variant(s) included \n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AFR/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Loading Genotype info from reference \n",
      "================================================== \n",
      "\n",
      "2648 people (1458 male(s), 1190 female(s)) observed \n",
      "2648 founder(s) included \n",
      "\n",
      "65317063 variant(s) not found in previous data \n",
      "632507 variant(s) included \n",
      "\n",
      "There are a total of 1 phenotype to process \n",
      "\n",
      "Start performing clumping \n",
      "\n",
      "Clumping Progress: 100.00%\n",
      "Number of variant(s) after clumping : 157287 \n",
      "\n",
      "Processing the 1 th phenotype \n",
      "\n",
      "Phenotype is a binary phenotype \n",
      "1652 control(s) \n",
      "954 case(s) \n",
      "\n",
      "Processing the covariate file: \n",
      "/home/jupyter/workspace/ws_files/results/covariates_prs.txt \n",
      "============================== \n",
      "\n",
      "Include Covariates: \n",
      "Name\tMissing\tNumber of levels \n",
      "SEX\t0\t- \n",
      "AGE\t0\t- \n",
      "PC1\t0\t- \n",
      "PC2\t0\t- \n",
      "PC3\t0\t- \n",
      "PC4\t0\t- \n",
      "PC5\t0\t- \n",
      "PC6\t0\t- \n",
      "PC7\t0\t- \n",
      "PC8\t0\t- \n",
      "PC9\t0\t- \n",
      "PC10\t0\t- \n",
      "\n",
      "After reading the covariate file, 2606 sample(s) included \n",
      "in the analysis \n",
      "\n",
      "\n",
      "Start Processing\n",
      "Processing 100.00%\n",
      "There are 1 region(s) with p-value less than 1e-5. Please \n",
      "note that these results are inflated due to the overfitting \n",
      "inherent in finding the best-fit PRS (but it's still best \n",
      "to find the best-fit PRS!). \n",
      "You can use the --perm option (see manual) to calculate an \n",
      "empirical P-value. \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Begin plotting\n",
      "Current Rscript version = 2.3.3\n",
      "Plotting the quantile plot\n",
      "WARNING: There are only 0 unique PRS but asked for 4 quantiles\n",
      "Will not generate the quantile plot for  /home/jupyter/workspace/ws_files/results/AFR/model2_AFR\n",
      "Plotting Bar Plot\n",
      "Plotting Multi-Set-Plot\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In max(pheno$Pheno, na.rm = T) :\n",
      "  no non-missing arguments to max; returning -Inf\n",
      "PRSice 2.3.5 (2021-09-20) \n",
      "https://github.com/choishingwan/PRSice\n",
      "(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly\n",
      "GNU General Public License v3\n",
      "If you use PRSice in any published work, please cite:\n",
      "Choi SW, O'Reilly PF.\n",
      "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.\n",
      "GigaScience 8, no. 7 (July 1, 2019)\n",
      "2025-02-11 15:40:22\n",
      "/home/jupyter/prsice \\\n",
      "    --a1 A1 \\\n",
      "    --a2 A2 \\\n",
      "    --bar-levels 5e-08,5e-07,5e-06,5e-05,0.0005,0.005,0.05 \\\n",
      "    --base /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
      "    --beta  \\\n",
      "    --binary-target T \\\n",
      "    --bp BP \\\n",
      "    --chr CHR \\\n",
      "    --clump-kb 250kb \\\n",
      "    --clump-p 1.000000 \\\n",
      "    --clump-r2 0.100000 \\\n",
      "    --cov /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
      "    --fastscore  \\\n",
      "    --ignore-fid  \\\n",
      "    --ld /home/jupyter/workspace/ws_files/results/AJ/model2_plink/chr# \\\n",
      "    --no-full  \\\n",
      "    --num-auto 22 \\\n",
      "    --out /home/jupyter/workspace/ws_files/results/AJ/model2_AJ \\\n",
      "    --perm 10000 \\\n",
      "    --prevalence 0.005 \\\n",
      "    --print-snp  \\\n",
      "    --pvalue P \\\n",
      "    --score con-std \\\n",
      "    --seed 3 \\\n",
      "    --snp SNP \\\n",
      "    --stat BETA \\\n",
      "    --target /home/jupyter/workspace/ws_files/results/AJ/model2_plink/chr# \\\n",
      "    --thread 16\n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AJ/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Start processing multi_MAMA \n",
      "================================================== \n",
      "\n",
      "Base file: \n",
      "/home/jupyter/workspace/ws_files/data/multi_MAMA.txt \n",
      "Header of file is: \n",
      "SNP CHR BP A1 A2 BETA P TEST \n",
      "\n",
      "Reading 100.00%\n",
      "17782530 variant(s) observed in base file, with: \n",
      "16969225 variant(s) excluded due to p-value threshold \n",
      "122051 ambiguous variant(s) excluded \n",
      "691254 total variant(s) included from base file \n",
      "\n",
      "Loading Genotype info from target \n",
      "================================================== \n",
      "\n",
      "3126 people (1958 male(s), 1168 female(s)) observed \n",
      "3126 founder(s) included \n",
      "\n",
      "23900195 variant(s) not found in previous data \n",
      "519070 variant(s) included \n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AJ/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Loading Genotype info from reference \n",
      "================================================== \n",
      "\n",
      "3126 people (1958 male(s), 1168 female(s)) observed \n",
      "3126 founder(s) included \n",
      "\n",
      "23900195 variant(s) not found in previous data \n",
      "519070 variant(s) included \n",
      "\n",
      "There are a total of 1 phenotype to process \n",
      "\n",
      "Start performing clumping \n",
      "\n",
      "Clumping Progress: 100.00%\n",
      "Number of variant(s) after clumping : 119149 \n",
      "\n",
      "Processing the 1 th phenotype \n",
      "\n",
      "Phenotype is a binary phenotype \n",
      "423 control(s) \n",
      "1396 case(s) \n",
      "\n",
      "Processing the covariate file: \n",
      "/home/jupyter/workspace/ws_files/results/covariates_prs.txt \n",
      "============================== \n",
      "\n",
      "Include Covariates: \n",
      "Name\tMissing\tNumber of levels \n",
      "SEX\t0\t- \n",
      "AGE\t0\t- \n",
      "PC1\t0\t- \n",
      "PC2\t0\t- \n",
      "PC3\t0\t- \n",
      "PC4\t0\t- \n",
      "PC5\t0\t- \n",
      "PC6\t0\t- \n",
      "PC7\t0\t- \n",
      "PC8\t0\t- \n",
      "PC9\t0\t- \n",
      "PC10\t0\t- \n",
      "\n",
      "After reading the covariate file, 1819 sample(s) included \n",
      "in the analysis \n",
      "\n",
      "\n",
      "Start Processing\n",
      "Processing 100.00%\n",
      "There are 1 region(s) with p-value less than 1e-5. Please \n",
      "note that these results are inflated due to the overfitting \n",
      "inherent in finding the best-fit PRS (but it's still best \n",
      "to find the best-fit PRS!). \n",
      "You can use the --perm option (see manual) to calculate an \n",
      "empirical P-value. \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Begin plotting\n",
      "Current Rscript version = 2.3.3\n",
      "Plotting the quantile plot\n",
      "WARNING: There are only 0 unique PRS but asked for 4 quantiles\n",
      "Will not generate the quantile plot for  /home/jupyter/workspace/ws_files/results/AJ/model2_AJ\n",
      "Plotting Bar Plot\n",
      "Plotting Multi-Set-Plot\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In max(pheno$Pheno, na.rm = T) :\n",
      "  no non-missing arguments to max; returning -Inf\n",
      "PRSice 2.3.5 (2021-09-20) \n",
      "https://github.com/choishingwan/PRSice\n",
      "(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly\n",
      "GNU General Public License v3\n",
      "If you use PRSice in any published work, please cite:\n",
      "Choi SW, O'Reilly PF.\n",
      "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.\n",
      "GigaScience 8, no. 7 (July 1, 2019)\n",
      "2025-02-11 15:46:36\n",
      "/home/jupyter/prsice \\\n",
      "    --a1 A1 \\\n",
      "    --a2 A2 \\\n",
      "    --bar-levels 5e-08,5e-07,5e-06,5e-05,0.0005,0.005,0.05 \\\n",
      "    --base /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
      "    --beta  \\\n",
      "    --binary-target T \\\n",
      "    --bp BP \\\n",
      "    --chr CHR \\\n",
      "    --clump-kb 250kb \\\n",
      "    --clump-p 1.000000 \\\n",
      "    --clump-r2 0.100000 \\\n",
      "    --cov /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
      "    --fastscore  \\\n",
      "    --ignore-fid  \\\n",
      "    --ld /home/jupyter/workspace/ws_files/results/AMR/model2_plink/chr# \\\n",
      "    --no-full  \\\n",
      "    --num-auto 22 \\\n",
      "    --out /home/jupyter/workspace/ws_files/results/AMR/model2_AMR \\\n",
      "    --perm 10000 \\\n",
      "    --prevalence 0.005 \\\n",
      "    --print-snp  \\\n",
      "    --pvalue P \\\n",
      "    --score con-std \\\n",
      "    --seed 3 \\\n",
      "    --snp SNP \\\n",
      "    --stat BETA \\\n",
      "    --target /home/jupyter/workspace/ws_files/results/AMR/model2_plink/chr# \\\n",
      "    --thread 16\n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AMR/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Start processing multi_MAMA \n",
      "================================================== \n",
      "\n",
      "Base file: \n",
      "/home/jupyter/workspace/ws_files/data/multi_MAMA.txt \n",
      "Header of file is: \n",
      "SNP CHR BP A1 A2 BETA P TEST \n",
      "\n",
      "Reading 100.00%\n",
      "17782530 variant(s) observed in base file, with: \n",
      "16969225 variant(s) excluded due to p-value threshold \n",
      "122051 ambiguous variant(s) excluded \n",
      "691254 total variant(s) included from base file \n",
      "\n",
      "Loading Genotype info from target \n",
      "================================================== \n",
      "\n",
      "3496 people (1661 male(s), 1835 female(s)) observed \n",
      "3496 founder(s) included \n",
      "\n",
      "62323330 variant(s) not found in previous data \n",
      "658049 variant(s) included \n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AMR/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Loading Genotype info from reference \n",
      "================================================== \n",
      "\n",
      "3496 people (1661 male(s), 1835 female(s)) observed \n",
      "3496 founder(s) included \n",
      "\n",
      "62323330 variant(s) not found in previous data \n",
      "658049 variant(s) included \n",
      "\n",
      "There are a total of 1 phenotype to process \n",
      "\n",
      "Start performing clumping \n",
      "\n",
      "Clumping Progress: 100.00%\n",
      "Number of variant(s) after clumping : 150819 \n",
      "\n",
      "Processing the 1 th phenotype \n",
      "\n",
      "Phenotype is a binary phenotype \n",
      "1430 control(s) \n",
      "1928 case(s) \n",
      "\n",
      "Processing the covariate file: \n",
      "/home/jupyter/workspace/ws_files/results/covariates_prs.txt \n",
      "============================== \n",
      "\n",
      "Include Covariates: \n",
      "Name\tMissing\tNumber of levels \n",
      "SEX\t0\t- \n",
      "AGE\t0\t- \n",
      "PC1\t0\t- \n",
      "PC2\t0\t- \n",
      "PC3\t0\t- \n",
      "PC4\t0\t- \n",
      "PC5\t0\t- \n",
      "PC6\t0\t- \n",
      "PC7\t0\t- \n",
      "PC8\t0\t- \n",
      "PC9\t0\t- \n",
      "PC10\t0\t- \n",
      "\n",
      "After reading the covariate file, 3358 sample(s) included \n",
      "in the analysis \n",
      "\n",
      "\n",
      "Start Processing\n",
      "Processing 100.00%\n",
      "There are 1 region(s) with p-value less than 1e-5. Please \n",
      "note that these results are inflated due to the overfitting \n",
      "inherent in finding the best-fit PRS (but it's still best \n",
      "to find the best-fit PRS!). \n",
      "You can use the --perm option (see manual) to calculate an \n",
      "empirical P-value. \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Begin plotting\n",
      "Current Rscript version = 2.3.3\n",
      "Plotting the quantile plot\n",
      "WARNING: There are only 0 unique PRS but asked for 4 quantiles\n",
      "Will not generate the quantile plot for  /home/jupyter/workspace/ws_files/results/AMR/model2_AMR\n",
      "Plotting Bar Plot\n",
      "Plotting Multi-Set-Plot\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In max(pheno$Pheno, na.rm = T) :\n",
      "  no non-missing arguments to max; returning -Inf\n",
      "PRSice 2.3.5 (2021-09-20) \n",
      "https://github.com/choishingwan/PRSice\n",
      "(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly\n",
      "GNU General Public License v3\n",
      "If you use PRSice in any published work, please cite:\n",
      "Choi SW, O'Reilly PF.\n",
      "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.\n",
      "GigaScience 8, no. 7 (July 1, 2019)\n",
      "2025-02-11 15:59:21\n",
      "/home/jupyter/prsice \\\n",
      "    --a1 A1 \\\n",
      "    --a2 A2 \\\n",
      "    --bar-levels 5e-08,5e-07,5e-06,5e-05,0.0005,0.005,0.05 \\\n",
      "    --base /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
      "    --beta  \\\n",
      "    --binary-target T \\\n",
      "    --bp BP \\\n",
      "    --chr CHR \\\n",
      "    --clump-kb 250kb \\\n",
      "    --clump-p 1.000000 \\\n",
      "    --clump-r2 0.100000 \\\n",
      "    --cov /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
      "    --fastscore  \\\n",
      "    --ignore-fid  \\\n",
      "    --ld /home/jupyter/workspace/ws_files/results/EAS/model2_plink/chr# \\\n",
      "    --no-full  \\\n",
      "    --num-auto 22 \\\n",
      "    --out /home/jupyter/workspace/ws_files/results/EAS/model2_EAS \\\n",
      "    --perm 10000 \\\n",
      "    --prevalence 0.005 \\\n",
      "    --print-snp  \\\n",
      "    --pvalue P \\\n",
      "    --score con-std \\\n",
      "    --seed 3 \\\n",
      "    --snp SNP \\\n",
      "    --stat BETA \\\n",
      "    --target /home/jupyter/workspace/ws_files/results/EAS/model2_plink/chr# \\\n",
      "    --thread 16\n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/EAS/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Start processing multi_MAMA \n",
      "================================================== \n",
      "\n",
      "Base file: \n",
      "/home/jupyter/workspace/ws_files/data/multi_MAMA.txt \n",
      "Header of file is: \n",
      "SNP CHR BP A1 A2 BETA P TEST \n",
      "\n",
      "Reading 100.00%\n",
      "17782530 variant(s) observed in base file, with: \n",
      "16969225 variant(s) excluded due to p-value threshold \n",
      "122051 ambiguous variant(s) excluded \n",
      "691254 total variant(s) included from base file \n",
      "\n",
      "Loading Genotype info from target \n",
      "================================================== \n",
      "\n",
      "5625 people (3473 male(s), 2152 female(s)) observed \n",
      "5625 founder(s) included \n",
      "\n",
      "47760819 variant(s) not found in previous data \n",
      "460924 variant(s) included \n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/EAS/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Loading Genotype info from reference \n",
      "================================================== \n",
      "\n",
      "5625 people (3473 male(s), 2152 female(s)) observed \n",
      "5625 founder(s) included \n",
      "\n",
      "47760819 variant(s) not found in previous data \n",
      "460924 variant(s) included \n",
      "\n",
      "There are a total of 1 phenotype to process \n",
      "\n",
      "Start performing clumping \n",
      "\n",
      "Clumping Progress: 100.00%\n",
      "Number of variant(s) after clumping : 108716 \n",
      "\n",
      "Processing the 1 th phenotype \n",
      "\n",
      "Phenotype is a binary phenotype \n",
      "2349 control(s) \n",
      "1819 case(s) \n",
      "\n",
      "Processing the covariate file: \n",
      "/home/jupyter/workspace/ws_files/results/covariates_prs.txt \n",
      "============================== \n",
      "\n",
      "Include Covariates: \n",
      "Name\tMissing\tNumber of levels \n",
      "SEX\t0\t- \n",
      "AGE\t0\t- \n",
      "PC1\t0\t- \n",
      "PC2\t0\t- \n",
      "PC3\t0\t- \n",
      "PC4\t0\t- \n",
      "PC5\t0\t- \n",
      "PC6\t0\t- \n",
      "PC7\t0\t- \n",
      "PC8\t0\t- \n",
      "PC9\t0\t- \n",
      "PC10\t0\t- \n",
      "\n",
      "After reading the covariate file, 4168 sample(s) included \n",
      "in the analysis \n",
      "\n",
      "\n",
      "Start Processing\n",
      "Processing 100.00%\n",
      "There are 1 region(s) with p-value less than 1e-5. Please \n",
      "note that these results are inflated due to the overfitting \n",
      "inherent in finding the best-fit PRS (but it's still best \n",
      "to find the best-fit PRS!). \n",
      "You can use the --perm option (see manual) to calculate an \n",
      "empirical P-value. \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Begin plotting\n",
      "Current Rscript version = 2.3.3\n",
      "Plotting the quantile plot\n",
      "WARNING: There are only 0 unique PRS but asked for 4 quantiles\n",
      "Will not generate the quantile plot for  /home/jupyter/workspace/ws_files/results/EAS/model2_EAS\n",
      "Plotting Bar Plot\n",
      "Plotting Multi-Set-Plot\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In max(pheno$Pheno, na.rm = T) :\n",
      "  no non-missing arguments to max; returning -Inf\n",
      "PRSice 2.3.5 (2021-09-20) \n",
      "https://github.com/choishingwan/PRSice\n",
      "(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly\n",
      "GNU General Public License v3\n",
      "If you use PRSice in any published work, please cite:\n",
      "Choi SW, O'Reilly PF.\n",
      "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.\n",
      "GigaScience 8, no. 7 (July 1, 2019)\n",
      "2025-02-11 16:12:33\n",
      "/home/jupyter/prsice \\\n",
      "    --a1 A1 \\\n",
      "    --a2 A2 \\\n",
      "    --bar-levels 5e-08,5e-07,5e-06,5e-05,0.0005,0.005,0.05 \\\n",
      "    --base /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
      "    --beta  \\\n",
      "    --binary-target T \\\n",
      "    --bp BP \\\n",
      "    --chr CHR \\\n",
      "    --clump-kb 250kb \\\n",
      "    --clump-p 1.000000 \\\n",
      "    --clump-r2 0.100000 \\\n",
      "    --cov /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
      "    --fastscore  \\\n",
      "    --ignore-fid  \\\n",
      "    --ld /home/jupyter/workspace/ws_files/results/EUR/model2_plink/chr# \\\n",
      "    --no-full  \\\n",
      "    --num-auto 22 \\\n",
      "    --out /home/jupyter/workspace/ws_files/results/EUR/model2_EUR \\\n",
      "    --perm 10000 \\\n",
      "    --prevalence 0.005 \\\n",
      "    --print-snp  \\\n",
      "    --pvalue P \\\n",
      "    --score con-std \\\n",
      "    --seed 3 \\\n",
      "    --snp SNP \\\n",
      "    --stat BETA \\\n",
      "    --target /home/jupyter/workspace/ws_files/results/EUR/model2_plink/chr# \\\n",
      "    --thread 16\n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/EUR/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Start processing multi_MAMA \n",
      "================================================== \n",
      "\n",
      "Base file: \n",
      "/home/jupyter/workspace/ws_files/data/multi_MAMA.txt \n",
      "Header of file is: \n",
      "SNP CHR BP A1 A2 BETA P TEST \n",
      "\n",
      "Reading 100.00%\n",
      "17782530 variant(s) observed in base file, with: \n",
      "16969225 variant(s) excluded due to p-value threshold \n",
      "122051 ambiguous variant(s) excluded \n",
      "691254 total variant(s) included from base file \n",
      "\n",
      "Loading Genotype info from target \n",
      "================================================== \n",
      "\n",
      "48909 people (27340 male(s), 21569 female(s)) observed \n",
      "48909 founder(s) included \n",
      "\n",
      "\u001b[1;31mError: A total of 11 duplicated SNP ID detected out of \n",
      "       654340 input SNPs! Valid SNP ID (post --extract / \n",
      "       --exclude, non-duplicated SNPs) stored at \n",
      "       /home/jupyter/workspace/ws_files/results/EUR/model2_EUR.valid. \n",
      "       You can avoid this error by using --extract \n",
      "       /home/jupyter/workspace/ws_files/results/EUR/model2_EUR.valid \u001b[0m\n",
      "\n",
      "Error: \n",
      "Execution halted\n",
      "PRSice 2.3.5 (2021-09-20) \n",
      "https://github.com/choishingwan/PRSice\n",
      "(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly\n",
      "GNU General Public License v3\n",
      "If you use PRSice in any published work, please cite:\n",
      "Choi SW, O'Reilly PF.\n",
      "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.\n",
      "GigaScience 8, no. 7 (July 1, 2019)\n",
      "2025-02-11 16:16:56\n",
      "/home/jupyter/prsice \\\n",
      "    --a1 A1 \\\n",
      "    --a2 A2 \\\n",
      "    --bar-levels 5e-08,5e-07,5e-06,5e-05,0.0005,0.005,0.05 \\\n",
      "    --base /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
      "    --beta  \\\n",
      "    --binary-target T \\\n",
      "    --bp BP \\\n",
      "    --chr CHR \\\n",
      "    --clump-kb 250kb \\\n",
      "    --clump-p 1.000000 \\\n",
      "    --clump-r2 0.100000 \\\n",
      "    --cov /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
      "    --fastscore  \\\n",
      "    --ignore-fid  \\\n",
      "    --ld /home/jupyter/workspace/ws_files/results/CAS/model2_plink/chr# \\\n",
      "    --no-full  \\\n",
      "    --num-auto 22 \\\n",
      "    --out /home/jupyter/workspace/ws_files/results/CAS/model2_CAS \\\n",
      "    --perm 10000 \\\n",
      "    --prevalence 0.005 \\\n",
      "    --print-snp  \\\n",
      "    --pvalue P \\\n",
      "    --score con-std \\\n",
      "    --seed 3 \\\n",
      "    --snp SNP \\\n",
      "    --stat BETA \\\n",
      "    --target /home/jupyter/workspace/ws_files/results/CAS/model2_plink/chr# \\\n",
      "    --thread 16\n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/CAS/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Start processing multi_MAMA \n",
      "================================================== \n",
      "\n",
      "Base file: \n",
      "/home/jupyter/workspace/ws_files/data/multi_MAMA.txt \n",
      "Header of file is: \n",
      "SNP CHR BP A1 A2 BETA P TEST \n",
      "\n",
      "Reading 100.00%\n",
      "17782530 variant(s) observed in base file, with: \n",
      "16969225 variant(s) excluded due to p-value threshold \n",
      "122051 ambiguous variant(s) excluded \n",
      "691254 total variant(s) included from base file \n",
      "\n",
      "Loading Genotype info from target \n",
      "================================================== \n",
      "\n",
      "1003 people (458 male(s), 545 female(s)) observed \n",
      "1003 founder(s) included \n",
      "\n",
      "32995349 variant(s) not found in previous data \n",
      "514612 variant(s) included \n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/CAS/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Loading Genotype info from reference \n",
      "================================================== \n",
      "\n",
      "1003 people (458 male(s), 545 female(s)) observed \n",
      "1003 founder(s) included \n",
      "\n",
      "32995349 variant(s) not found in previous data \n",
      "514612 variant(s) included \n",
      "\n",
      "There are a total of 1 phenotype to process \n",
      "\n",
      "Start performing clumping \n",
      "\n",
      "Clumping Progress: 100.00%\n",
      "Number of variant(s) after clumping : 127197 \n",
      "\n",
      "Processing the 1 th phenotype \n",
      "\n",
      "Phenotype is a binary phenotype \n",
      "323 control(s) \n",
      "582 case(s) \n",
      "\n",
      "Processing the covariate file: \n",
      "/home/jupyter/workspace/ws_files/results/covariates_prs.txt \n",
      "============================== \n",
      "\n",
      "Include Covariates: \n",
      "Name\tMissing\tNumber of levels \n",
      "SEX\t0\t- \n",
      "AGE\t0\t- \n",
      "PC1\t0\t- \n",
      "PC2\t0\t- \n",
      "PC3\t0\t- \n",
      "PC4\t0\t- \n",
      "PC5\t0\t- \n",
      "PC6\t0\t- \n",
      "PC7\t0\t- \n",
      "PC8\t0\t- \n",
      "PC9\t0\t- \n",
      "PC10\t0\t- \n",
      "\n",
      "After reading the covariate file, 905 sample(s) included in \n",
      "the analysis \n",
      "\n",
      "\n",
      "Start Processing\n",
      "Processing 100.00%\n",
      "There are 1 region(s) with p-value less than 1e-5. Please \n",
      "note that these results are inflated due to the overfitting \n",
      "inherent in finding the best-fit PRS (but it's still best \n",
      "to find the best-fit PRS!). \n",
      "You can use the --perm option (see manual) to calculate an \n",
      "empirical P-value. \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Begin plotting\n",
      "Current Rscript version = 2.3.3\n",
      "Plotting the quantile plot\n",
      "WARNING: There are only 0 unique PRS but asked for 4 quantiles\n",
      "Will not generate the quantile plot for  /home/jupyter/workspace/ws_files/results/CAS/model2_CAS\n",
      "Plotting Bar Plot\n",
      "Plotting Multi-Set-Plot\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In max(pheno$Pheno, na.rm = T) :\n",
      "  no non-missing arguments to max; returning -Inf\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "for ancestry in {\"AAC\",\"AFR\",\"AJ\",\"AMR\",\"EAS\",\"EUR\",\"CAS\"}; do\n",
    "    Rscript /home/jupyter/PRSice.R \\\n",
    "    --out /home/jupyter/workspace/ws_files/results/${ancestry}/model2_${ancestry} \\\n",
    "    --target /home/jupyter/workspace/ws_files/results/${ancestry}/model2_plink/chr# \\\n",
    "    -b /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
    "    --beta \\\n",
    "    --snp SNP \\\n",
    "    --A1 A1 \\\n",
    "    --A2 A2 \\\n",
    "    --stat BETA \\\n",
    "    --pvalue P \\\n",
    "    --ld /home/jupyter/workspace/ws_files/results/${ancestry}/model2_plink/chr# \\\n",
    "    --print-snp \\\n",
    "    --score con-std \\\n",
    "    --perm 10000 \\\n",
    "    --prsice /home/jupyter/prsice \\\n",
    "    -n 24 \\\n",
    "    --binary-target T \\\n",
    "    --quantile 4 \\\n",
    "    --prevalence 0.005 \\\n",
    "    --bar-levels 5e-2,5e-3,5e-4,5e-5,5e-6,5e-7,5e-8 \\\n",
    "    --fastscore \\\n",
    "    --no-full \\\n",
    "    --multi-plot 10 \\\n",
    "    --thread 16 \\\n",
    "    --cov-file /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
    "    --seed 3 \\\n",
    "    --ignore-fid\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "62f371cc-c588-4681-8c44-b3e8214b3b50",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PRSice 2.3.5 (2021-09-20) \n",
      "https://github.com/choishingwan/PRSice\n",
      "(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly\n",
      "GNU General Public License v3\n",
      "If you use PRSice in any published work, please cite:\n",
      "Choi SW, O'Reilly PF.\n",
      "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.\n",
      "GigaScience 8, no. 7 (July 1, 2019)\n",
      "2025-02-11 16:27:34\n",
      "/home/jupyter/prsice \\\n",
      "    --a1 A1 \\\n",
      "    --a2 A2 \\\n",
      "    --bar-levels 5e-08,5e-07,5e-06,5e-05,0.0005,0.005,0.05 \\\n",
      "    --base /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
      "    --beta  \\\n",
      "    --binary-target T \\\n",
      "    --bp BP \\\n",
      "    --chr CHR \\\n",
      "    --clump-kb 250kb \\\n",
      "    --clump-p 1.000000 \\\n",
      "    --clump-r2 0.100000 \\\n",
      "    --cov /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
      "    --extract /home/jupyter/workspace/ws_files/results/EUR/model2_EUR.valid \\\n",
      "    --fastscore  \\\n",
      "    --ignore-fid  \\\n",
      "    --ld /home/jupyter/workspace/ws_files/results/EUR/model2_plink/chr# \\\n",
      "    --no-full  \\\n",
      "    --num-auto 22 \\\n",
      "    --out /home/jupyter/workspace/ws_files/results/EUR/model2_EUR \\\n",
      "    --perm 10000 \\\n",
      "    --prevalence 0.005 \\\n",
      "    --print-snp  \\\n",
      "    --pvalue P \\\n",
      "    --score con-std \\\n",
      "    --seed 3 \\\n",
      "    --snp SNP \\\n",
      "    --stat BETA \\\n",
      "    --target /home/jupyter/workspace/ws_files/results/EUR/model2_plink/chr# \\\n",
      "    --thread 16\n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/EUR/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Start processing multi_MAMA \n",
      "================================================== \n",
      "\n",
      "SNP extraction/exclusion list contains 5 columns, will \n",
      "assume first column contains the SNP ID \n",
      "\n",
      "Base file: \n",
      "/home/jupyter/workspace/ws_files/data/multi_MAMA.txt \n",
      "Header of file is: \n",
      "SNP CHR BP A1 A2 BETA P TEST \n",
      "\n",
      "Reading 100.00%\n",
      "17782530 variant(s) observed in base file, with: \n",
      "17128201 variant(s) excluded based on user input \n",
      "654329 total variant(s) included from base file \n",
      "\n",
      "Loading Genotype info from target \n",
      "================================================== \n",
      "\n",
      "48909 people (27340 male(s), 21569 female(s)) observed \n",
      "48909 founder(s) included \n",
      "\n",
      "122734888 variant(s) not found in previous data \n",
      "654329 variant(s) included \n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/EUR/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Loading Genotype info from reference \n",
      "================================================== \n",
      "\n",
      "48909 people (27340 male(s), 21569 female(s)) observed \n",
      "48909 founder(s) included \n",
      "\n",
      "122734888 variant(s) not found in previous data \n",
      "654329 variant(s) included \n",
      "\n",
      "There are a total of 1 phenotype to process \n",
      "\n",
      "Start performing clumping \n",
      "\n",
      "Clumping Progress: 100.00%\n",
      "Number of variant(s) after clumping : 152863 \n",
      "\n",
      "Processing the 1 th phenotype \n",
      "\n",
      "Phenotype is a binary phenotype \n",
      "9706 control(s) \n",
      "22093 case(s) \n",
      "\n",
      "Processing the covariate file: \n",
      "/home/jupyter/workspace/ws_files/results/covariates_prs.txt \n",
      "============================== \n",
      "\n",
      "Include Covariates: \n",
      "Name\tMissing\tNumber of levels \n",
      "SEX\t0\t- \n",
      "AGE\t0\t- \n",
      "PC1\t0\t- \n",
      "PC2\t0\t- \n",
      "PC3\t0\t- \n",
      "PC4\t0\t- \n",
      "PC5\t0\t- \n",
      "PC6\t0\t- \n",
      "PC7\t0\t- \n",
      "PC8\t0\t- \n",
      "PC9\t0\t- \n",
      "PC10\t0\t- \n",
      "\n",
      "After reading the covariate file, 31799 sample(s) included \n",
      "in the analysis \n",
      "\n",
      "\n",
      "Start Processing\n",
      "Processing 100.00%\n",
      "There are 1 region(s) with p-value less than 1e-5. Please \n",
      "note that these results are inflated due to the overfitting \n",
      "inherent in finding the best-fit PRS (but it's still best \n",
      "to find the best-fit PRS!). \n",
      "You can use the --perm option (see manual) to calculate an \n",
      "empirical P-value. \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Begin plotting\n",
      "Current Rscript version = 2.3.3\n",
      "Plotting the quantile plot\n",
      "WARNING: There are only 0 unique PRS but asked for 4 quantiles\n",
      "Will not generate the quantile plot for  /home/jupyter/workspace/ws_files/results/EUR/model2_EUR\n",
      "Plotting Bar Plot\n",
      "Plotting Multi-Set-Plot\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In max(pheno$Pheno, na.rm = T) :\n",
      "  no non-missing arguments to max; returning -Inf\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "Rscript /home/jupyter/PRSice.R \\\n",
    "--out /home/jupyter/workspace/ws_files/results/EUR/model2_EUR \\\n",
    "--target /home/jupyter/workspace/ws_files/results/EUR/model2_plink/chr# \\\n",
    "-b /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
    "--beta \\\n",
    "--snp SNP \\\n",
    "--A1 A1 \\\n",
    "--A2 A2 \\\n",
    "--stat BETA \\\n",
    "--pvalue P \\\n",
    "--ld /home/jupyter/workspace/ws_files/results/EUR/model2_plink/chr# \\\n",
    "--print-snp \\\n",
    "--score con-std \\\n",
    "--perm 10000 \\\n",
    "--prsice /home/jupyter/prsice \\\n",
    "-n 24 \\\n",
    "--binary-target T \\\n",
    "--quantile 4 \\\n",
    "--prevalence 0.005 \\\n",
    "--bar-levels 5e-2,5e-3,5e-4,5e-5,5e-6,5e-7,5e-8 \\\n",
    "--fastscore \\\n",
    "--no-full \\\n",
    "--multi-plot 10 \\\n",
    "--thread 16 \\\n",
    "--cov-file /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
    "--seed 3 \\\n",
    "--ignore-fid \\\n",
    "--extract /home/jupyter/workspace/ws_files/results/EUR/model2_EUR.valid\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d82f6e56-125f-4e51-be7e-fbc3bda16c2d",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Find best PRSice model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "69a5513f-f089-40db-acaf-d23e996581ef",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAC: 5e-08\n",
      "AFR: 5e-07\n",
      "AJ: 5e-06\n",
      "AMR: 5e-08\n",
      "EAS: 5e-08\n",
      "EUR: 5e-07\n",
      "CAS: 5e-06\n"
     ]
    }
   ],
   "source": [
    "for ancestry in ancestries:\n",
    "    df_prsice = pd.read_csv(f\"{RESULTS_DIR}/{ancestry}/model2_{ancestry}.prsice\", sep=\"\\t\")\n",
    "    df_prsice = df_prsice[df_prsice[\"Num_SNP\"] < n_per_ancestry[ancestry]] \n",
    "    threshold = df_prsice.sort_values(by='R2', ascending=False)[\"Threshold\"].iloc[0]\n",
    "    print(f\"{ancestry}: {threshold}\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb4ca7ec-dbfe-44d0-b0c2-24897e86b34f",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Re-run PRSice with best p-threshold for each ancestry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c3595bd2-aa7e-427d-b19a-2414c0931adc",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PRSice 2.3.5 (2021-09-20) \n",
      "https://github.com/choishingwan/PRSice\n",
      "(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly\n",
      "GNU General Public License v3\n",
      "If you use PRSice in any published work, please cite:\n",
      "Choi SW, O'Reilly PF.\n",
      "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.\n",
      "GigaScience 8, no. 7 (July 1, 2019)\n",
      "2025-02-11 17:50:34\n",
      "/home/jupyter/prsice \\\n",
      "    --a1 A1 \\\n",
      "    --a2 A2 \\\n",
      "    --bar-levels 5e-08 \\\n",
      "    --base /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
      "    --beta  \\\n",
      "    --binary-target T \\\n",
      "    --bp BP \\\n",
      "    --chr CHR \\\n",
      "    --clump-kb 250kb \\\n",
      "    --clump-p 1.000000 \\\n",
      "    --clump-r2 0.100000 \\\n",
      "    --cov /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
      "    --fastscore  \\\n",
      "    --ignore-fid  \\\n",
      "    --ld /home/jupyter/workspace/ws_files/results/AAC/model2_plink/chr# \\\n",
      "    --no-full  \\\n",
      "    --num-auto 22 \\\n",
      "    --out /home/jupyter/workspace/ws_files/results/AAC/model2_AAC_best \\\n",
      "    --perm 10000 \\\n",
      "    --prevalence 0.005 \\\n",
      "    --print-snp  \\\n",
      "    --pvalue P \\\n",
      "    --score con-std \\\n",
      "    --seed 3 \\\n",
      "    --snp SNP \\\n",
      "    --stat BETA \\\n",
      "    --target /home/jupyter/workspace/ws_files/results/AAC/model2_plink/chr# \\\n",
      "    --thread 16\n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AAC/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Start processing multi_MAMA \n",
      "================================================== \n",
      "\n",
      "Base file: \n",
      "/home/jupyter/workspace/ws_files/data/multi_MAMA.txt \n",
      "Header of file is: \n",
      "SNP CHR BP A1 A2 BETA P TEST \n",
      "\n",
      "Reading 100.00%\n",
      "17782530 variant(s) observed in base file, with: \n",
      "17776458 variant(s) excluded due to p-value threshold \n",
      "883 ambiguous variant(s) excluded \n",
      "5189 total variant(s) included from base file \n",
      "\n",
      "Loading Genotype info from target \n",
      "================================================== \n",
      "\n",
      "1196 people (498 male(s), 698 female(s)) observed \n",
      "1196 founder(s) included \n",
      "\n",
      "58407265 variant(s) not found in previous data \n",
      "5029 variant(s) included \n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AAC/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Loading Genotype info from reference \n",
      "================================================== \n",
      "\n",
      "1196 people (498 male(s), 698 female(s)) observed \n",
      "1196 founder(s) included \n",
      "\n",
      "58407265 variant(s) not found in previous data \n",
      "5029 variant(s) included \n",
      "\n",
      "There are a total of 1 phenotype to process \n",
      "\n",
      "Start performing clumping \n",
      "\n",
      "Clumping Progress: 100.00%\n",
      "Number of variant(s) after clumping : 273 \n",
      "\n",
      "Processing the 1 th phenotype \n",
      "\n",
      "Phenotype is a binary phenotype \n",
      "819 control(s) \n",
      "325 case(s) \n",
      "\n",
      "Processing the covariate file: \n",
      "/home/jupyter/workspace/ws_files/results/covariates_prs.txt \n",
      "============================== \n",
      "\n",
      "Include Covariates: \n",
      "Name\tMissing\tNumber of levels \n",
      "SEX\t0\t- \n",
      "AGE\t0\t- \n",
      "PC1\t0\t- \n",
      "PC2\t0\t- \n",
      "PC3\t0\t- \n",
      "PC4\t0\t- \n",
      "PC5\t0\t- \n",
      "PC6\t0\t- \n",
      "PC7\t0\t- \n",
      "PC8\t0\t- \n",
      "PC9\t0\t- \n",
      "PC10\t0\t- \n",
      "\n",
      "After reading the covariate file, 1144 sample(s) included \n",
      "in the analysis \n",
      "\n",
      "\n",
      "Start Processing\n",
      "Processing 100.00%\n",
      "There are 1 region(s) with p-value between 0.1 and 1e-5 \n",
      "(\u001b[1;31mmay not be significant\u001b[0m). \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Begin plotting\n",
      "Current Rscript version = 2.3.3\n",
      "Plotting the quantile plot\n",
      "WARNING: There are only 0 unique PRS but asked for 4 quantiles\n",
      "Will not generate the quantile plot for  /home/jupyter/workspace/ws_files/results/AAC/model2_AAC_best\n",
      "Plotting Bar Plot\n",
      "Plotting Multi-Set-Plot\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In max(pheno$Pheno, na.rm = T) :\n",
      "  no non-missing arguments to max; returning -Inf\n",
      "PRSice 2.3.5 (2021-09-20) \n",
      "https://github.com/choishingwan/PRSice\n",
      "(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly\n",
      "GNU General Public License v3\n",
      "If you use PRSice in any published work, please cite:\n",
      "Choi SW, O'Reilly PF.\n",
      "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.\n",
      "GigaScience 8, no. 7 (July 1, 2019)\n",
      "2025-02-11 17:53:58\n",
      "/home/jupyter/prsice \\\n",
      "    --a1 A1 \\\n",
      "    --a2 A2 \\\n",
      "    --bar-levels 5e-07 \\\n",
      "    --base /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
      "    --beta  \\\n",
      "    --binary-target T \\\n",
      "    --bp BP \\\n",
      "    --chr CHR \\\n",
      "    --clump-kb 250kb \\\n",
      "    --clump-p 1.000000 \\\n",
      "    --clump-r2 0.100000 \\\n",
      "    --cov /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
      "    --fastscore  \\\n",
      "    --ignore-fid  \\\n",
      "    --ld /home/jupyter/workspace/ws_files/results/AFR/model2_plink/chr# \\\n",
      "    --no-full  \\\n",
      "    --num-auto 22 \\\n",
      "    --out /home/jupyter/workspace/ws_files/results/AFR/model2_AFR_best \\\n",
      "    --perm 10000 \\\n",
      "    --prevalence 0.005 \\\n",
      "    --print-snp  \\\n",
      "    --pvalue P \\\n",
      "    --score con-std \\\n",
      "    --seed 3 \\\n",
      "    --snp SNP \\\n",
      "    --stat BETA \\\n",
      "    --target /home/jupyter/workspace/ws_files/results/AFR/model2_plink/chr# \\\n",
      "    --thread 16\n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AFR/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Start processing multi_MAMA \n",
      "================================================== \n",
      "\n",
      "Base file: \n",
      "/home/jupyter/workspace/ws_files/data/multi_MAMA.txt \n",
      "Header of file is: \n",
      "SNP CHR BP A1 A2 BETA P TEST \n",
      "\n",
      "Reading 100.00%\n",
      "17782530 variant(s) observed in base file, with: \n",
      "17774241 variant(s) excluded due to p-value threshold \n",
      "1192 ambiguous variant(s) excluded \n",
      "7097 total variant(s) included from base file \n",
      "\n",
      "Loading Genotype info from target \n",
      "================================================== \n",
      "\n",
      "2648 people (1458 male(s), 1190 female(s)) observed \n",
      "2648 founder(s) included \n",
      "\n",
      "65942765 variant(s) not found in previous data \n",
      "6805 variant(s) included \n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AFR/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Loading Genotype info from reference \n",
      "================================================== \n",
      "\n",
      "2648 people (1458 male(s), 1190 female(s)) observed \n",
      "2648 founder(s) included \n",
      "\n",
      "65942765 variant(s) not found in previous data \n",
      "6805 variant(s) included \n",
      "\n",
      "There are a total of 1 phenotype to process \n",
      "\n",
      "Start performing clumping \n",
      "\n",
      "Clumping Progress: 100.00%\n",
      "Number of variant(s) after clumping : 506 \n",
      "\n",
      "Processing the 1 th phenotype \n",
      "\n",
      "Phenotype is a binary phenotype \n",
      "1652 control(s) \n",
      "954 case(s) \n",
      "\n",
      "Processing the covariate file: \n",
      "/home/jupyter/workspace/ws_files/results/covariates_prs.txt \n",
      "============================== \n",
      "\n",
      "Include Covariates: \n",
      "Name\tMissing\tNumber of levels \n",
      "SEX\t0\t- \n",
      "AGE\t0\t- \n",
      "PC1\t0\t- \n",
      "PC2\t0\t- \n",
      "PC3\t0\t- \n",
      "PC4\t0\t- \n",
      "PC5\t0\t- \n",
      "PC6\t0\t- \n",
      "PC7\t0\t- \n",
      "PC8\t0\t- \n",
      "PC9\t0\t- \n",
      "PC10\t0\t- \n",
      "\n",
      "After reading the covariate file, 2606 sample(s) included \n",
      "in the analysis \n",
      "\n",
      "\n",
      "Start Processing\n",
      "Processing 100.00%\n",
      "There are 1 region(s) with p-value less than 1e-5. Please \n",
      "note that these results are inflated due to the overfitting \n",
      "inherent in finding the best-fit PRS (but it's still best \n",
      "to find the best-fit PRS!). \n",
      "You can use the --perm option (see manual) to calculate an \n",
      "empirical P-value. \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Begin plotting\n",
      "Current Rscript version = 2.3.3\n",
      "Plotting the quantile plot\n",
      "WARNING: There are only 0 unique PRS but asked for 4 quantiles\n",
      "Will not generate the quantile plot for  /home/jupyter/workspace/ws_files/results/AFR/model2_AFR_best\n",
      "Plotting Bar Plot\n",
      "Plotting Multi-Set-Plot\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In max(pheno$Pheno, na.rm = T) :\n",
      "  no non-missing arguments to max; returning -Inf\n",
      "PRSice 2.3.5 (2021-09-20) \n",
      "https://github.com/choishingwan/PRSice\n",
      "(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly\n",
      "GNU General Public License v3\n",
      "If you use PRSice in any published work, please cite:\n",
      "Choi SW, O'Reilly PF.\n",
      "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.\n",
      "GigaScience 8, no. 7 (July 1, 2019)\n",
      "2025-02-11 17:57:41\n",
      "/home/jupyter/prsice \\\n",
      "    --a1 A1 \\\n",
      "    --a2 A2 \\\n",
      "    --bar-levels 5e-06 \\\n",
      "    --base /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
      "    --beta  \\\n",
      "    --binary-target T \\\n",
      "    --bp BP \\\n",
      "    --chr CHR \\\n",
      "    --clump-kb 250kb \\\n",
      "    --clump-p 1.000000 \\\n",
      "    --clump-r2 0.100000 \\\n",
      "    --cov /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
      "    --fastscore  \\\n",
      "    --ignore-fid  \\\n",
      "    --ld /home/jupyter/workspace/ws_files/results/AJ/model2_plink/chr# \\\n",
      "    --no-full  \\\n",
      "    --num-auto 22 \\\n",
      "    --out /home/jupyter/workspace/ws_files/results/AJ/model2_AJ_best \\\n",
      "    --perm 10000 \\\n",
      "    --prevalence 0.005 \\\n",
      "    --print-snp  \\\n",
      "    --pvalue P \\\n",
      "    --score con-std \\\n",
      "    --seed 3 \\\n",
      "    --snp SNP \\\n",
      "    --stat BETA \\\n",
      "    --target /home/jupyter/workspace/ws_files/results/AJ/model2_plink/chr# \\\n",
      "    --thread 16\n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AJ/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Start processing multi_MAMA \n",
      "================================================== \n",
      "\n",
      "Base file: \n",
      "/home/jupyter/workspace/ws_files/data/multi_MAMA.txt \n",
      "Header of file is: \n",
      "SNP CHR BP A1 A2 BETA P TEST \n",
      "\n",
      "Reading 100.00%\n",
      "17782530 variant(s) observed in base file, with: \n",
      "17770710 variant(s) excluded due to p-value threshold \n",
      "1704 ambiguous variant(s) excluded \n",
      "10116 total variant(s) included from base file \n",
      "\n",
      "Loading Genotype info from target \n",
      "================================================== \n",
      "\n",
      "3126 people (1958 male(s), 1168 female(s)) observed \n",
      "3126 founder(s) included \n",
      "\n",
      "24409605 variant(s) not found in previous data \n",
      "9660 variant(s) included \n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AJ/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Loading Genotype info from reference \n",
      "================================================== \n",
      "\n",
      "3126 people (1958 male(s), 1168 female(s)) observed \n",
      "3126 founder(s) included \n",
      "\n",
      "24409605 variant(s) not found in previous data \n",
      "9660 variant(s) included \n",
      "\n",
      "There are a total of 1 phenotype to process \n",
      "\n",
      "Start performing clumping \n",
      "\n",
      "Clumping Progress: 100.00%\n",
      "Number of variant(s) after clumping : 457 \n",
      "\n",
      "Processing the 1 th phenotype \n",
      "\n",
      "Phenotype is a binary phenotype \n",
      "423 control(s) \n",
      "1396 case(s) \n",
      "\n",
      "Processing the covariate file: \n",
      "/home/jupyter/workspace/ws_files/results/covariates_prs.txt \n",
      "============================== \n",
      "\n",
      "Include Covariates: \n",
      "Name\tMissing\tNumber of levels \n",
      "SEX\t0\t- \n",
      "AGE\t0\t- \n",
      "PC1\t0\t- \n",
      "PC2\t0\t- \n",
      "PC3\t0\t- \n",
      "PC4\t0\t- \n",
      "PC5\t0\t- \n",
      "PC6\t0\t- \n",
      "PC7\t0\t- \n",
      "PC8\t0\t- \n",
      "PC9\t0\t- \n",
      "PC10\t0\t- \n",
      "\n",
      "After reading the covariate file, 1819 sample(s) included \n",
      "in the analysis \n",
      "\n",
      "\n",
      "Start Processing\n",
      "Processing 100.00%\n",
      "There are 1 region(s) with p-value less than 1e-5. Please \n",
      "note that these results are inflated due to the overfitting \n",
      "inherent in finding the best-fit PRS (but it's still best \n",
      "to find the best-fit PRS!). \n",
      "You can use the --perm option (see manual) to calculate an \n",
      "empirical P-value. \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Begin plotting\n",
      "Current Rscript version = 2.3.3\n",
      "Plotting the quantile plot\n",
      "WARNING: There are only 0 unique PRS but asked for 4 quantiles\n",
      "Will not generate the quantile plot for  /home/jupyter/workspace/ws_files/results/AJ/model2_AJ_best\n",
      "Plotting Bar Plot\n",
      "Plotting Multi-Set-Plot\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In max(pheno$Pheno, na.rm = T) :\n",
      "  no non-missing arguments to max; returning -Inf\n",
      "PRSice 2.3.5 (2021-09-20) \n",
      "https://github.com/choishingwan/PRSice\n",
      "(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly\n",
      "GNU General Public License v3\n",
      "If you use PRSice in any published work, please cite:\n",
      "Choi SW, O'Reilly PF.\n",
      "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.\n",
      "GigaScience 8, no. 7 (July 1, 2019)\n",
      "2025-02-11 17:59:52\n",
      "/home/jupyter/prsice \\\n",
      "    --a1 A1 \\\n",
      "    --a2 A2 \\\n",
      "    --bar-levels 5e-08 \\\n",
      "    --base /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
      "    --beta  \\\n",
      "    --binary-target T \\\n",
      "    --bp BP \\\n",
      "    --chr CHR \\\n",
      "    --clump-kb 250kb \\\n",
      "    --clump-p 1.000000 \\\n",
      "    --clump-r2 0.100000 \\\n",
      "    --cov /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
      "    --fastscore  \\\n",
      "    --ignore-fid  \\\n",
      "    --ld /home/jupyter/workspace/ws_files/results/AMR/model2_plink/chr# \\\n",
      "    --no-full  \\\n",
      "    --num-auto 22 \\\n",
      "    --out /home/jupyter/workspace/ws_files/results/AMR/model2_AMR_best \\\n",
      "    --perm 10000 \\\n",
      "    --prevalence 0.005 \\\n",
      "    --print-snp  \\\n",
      "    --pvalue P \\\n",
      "    --score con-std \\\n",
      "    --seed 3 \\\n",
      "    --snp SNP \\\n",
      "    --stat BETA \\\n",
      "    --target /home/jupyter/workspace/ws_files/results/AMR/model2_plink/chr# \\\n",
      "    --thread 16\n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AMR/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Start processing multi_MAMA \n",
      "================================================== \n",
      "\n",
      "Base file: \n",
      "/home/jupyter/workspace/ws_files/data/multi_MAMA.txt \n",
      "Header of file is: \n",
      "SNP CHR BP A1 A2 BETA P TEST \n",
      "\n",
      "Reading 100.00%\n",
      "17782530 variant(s) observed in base file, with: \n",
      "17776458 variant(s) excluded due to p-value threshold \n",
      "883 ambiguous variant(s) excluded \n",
      "5189 total variant(s) included from base file \n",
      "\n",
      "Loading Genotype info from target \n",
      "================================================== \n",
      "\n",
      "3496 people (1661 male(s), 1835 female(s)) observed \n",
      "3496 founder(s) included \n",
      "\n",
      "62976349 variant(s) not found in previous data \n",
      "5030 variant(s) included \n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/AMR/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Loading Genotype info from reference \n",
      "================================================== \n",
      "\n",
      "3496 people (1661 male(s), 1835 female(s)) observed \n",
      "3496 founder(s) included \n",
      "\n",
      "62976349 variant(s) not found in previous data \n",
      "5030 variant(s) included \n",
      "\n",
      "There are a total of 1 phenotype to process \n",
      "\n",
      "Start performing clumping \n",
      "\n",
      "Clumping Progress: 100.00%\n",
      "Number of variant(s) after clumping : 206 \n",
      "\n",
      "Processing the 1 th phenotype \n",
      "\n",
      "Phenotype is a binary phenotype \n",
      "1430 control(s) \n",
      "1928 case(s) \n",
      "\n",
      "Processing the covariate file: \n",
      "/home/jupyter/workspace/ws_files/results/covariates_prs.txt \n",
      "============================== \n",
      "\n",
      "Include Covariates: \n",
      "Name\tMissing\tNumber of levels \n",
      "SEX\t0\t- \n",
      "AGE\t0\t- \n",
      "PC1\t0\t- \n",
      "PC2\t0\t- \n",
      "PC3\t0\t- \n",
      "PC4\t0\t- \n",
      "PC5\t0\t- \n",
      "PC6\t0\t- \n",
      "PC7\t0\t- \n",
      "PC8\t0\t- \n",
      "PC9\t0\t- \n",
      "PC10\t0\t- \n",
      "\n",
      "After reading the covariate file, 3358 sample(s) included \n",
      "in the analysis \n",
      "\n",
      "\n",
      "Start Processing\n",
      "Processing 100.00%\n",
      "There are 1 region(s) with p-value less than 1e-5. Please \n",
      "note that these results are inflated due to the overfitting \n",
      "inherent in finding the best-fit PRS (but it's still best \n",
      "to find the best-fit PRS!). \n",
      "You can use the --perm option (see manual) to calculate an \n",
      "empirical P-value. \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Begin plotting\n",
      "Current Rscript version = 2.3.3\n",
      "Plotting the quantile plot\n",
      "WARNING: There are only 0 unique PRS but asked for 4 quantiles\n",
      "Will not generate the quantile plot for  /home/jupyter/workspace/ws_files/results/AMR/model2_AMR_best\n",
      "Plotting Bar Plot\n",
      "Plotting Multi-Set-Plot\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In max(pheno$Pheno, na.rm = T) :\n",
      "  no non-missing arguments to max; returning -Inf\n",
      "PRSice 2.3.5 (2021-09-20) \n",
      "https://github.com/choishingwan/PRSice\n",
      "(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly\n",
      "GNU General Public License v3\n",
      "If you use PRSice in any published work, please cite:\n",
      "Choi SW, O'Reilly PF.\n",
      "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.\n",
      "GigaScience 8, no. 7 (July 1, 2019)\n",
      "2025-02-11 18:03:23\n",
      "/home/jupyter/prsice \\\n",
      "    --a1 A1 \\\n",
      "    --a2 A2 \\\n",
      "    --bar-levels 5e-08 \\\n",
      "    --base /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
      "    --beta  \\\n",
      "    --binary-target T \\\n",
      "    --bp BP \\\n",
      "    --chr CHR \\\n",
      "    --clump-kb 250kb \\\n",
      "    --clump-p 1.000000 \\\n",
      "    --clump-r2 0.100000 \\\n",
      "    --cov /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
      "    --fastscore  \\\n",
      "    --ignore-fid  \\\n",
      "    --ld /home/jupyter/workspace/ws_files/results/EAS/model2_plink/chr# \\\n",
      "    --no-full  \\\n",
      "    --num-auto 22 \\\n",
      "    --out /home/jupyter/workspace/ws_files/results/EAS/model2_EAS_best \\\n",
      "    --perm 10000 \\\n",
      "    --prevalence 0.005 \\\n",
      "    --print-snp  \\\n",
      "    --pvalue P \\\n",
      "    --score con-std \\\n",
      "    --seed 3 \\\n",
      "    --snp SNP \\\n",
      "    --stat BETA \\\n",
      "    --target /home/jupyter/workspace/ws_files/results/EAS/model2_plink/chr# \\\n",
      "    --thread 16\n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/EAS/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Start processing multi_MAMA \n",
      "================================================== \n",
      "\n",
      "Base file: \n",
      "/home/jupyter/workspace/ws_files/data/multi_MAMA.txt \n",
      "Header of file is: \n",
      "SNP CHR BP A1 A2 BETA P TEST \n",
      "\n",
      "Reading 100.00%\n",
      "17782530 variant(s) observed in base file, with: \n",
      "17776458 variant(s) excluded due to p-value threshold \n",
      "883 ambiguous variant(s) excluded \n",
      "5189 total variant(s) included from base file \n",
      "\n",
      "Loading Genotype info from target \n",
      "================================================== \n",
      "\n",
      "5625 people (3473 male(s), 2152 female(s)) observed \n",
      "5625 founder(s) included \n",
      "\n",
      "48216760 variant(s) not found in previous data \n",
      "4983 variant(s) included \n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/EAS/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Loading Genotype info from reference \n",
      "================================================== \n",
      "\n",
      "5625 people (3473 male(s), 2152 female(s)) observed \n",
      "5625 founder(s) included \n",
      "\n",
      "48216760 variant(s) not found in previous data \n",
      "4983 variant(s) included \n",
      "\n",
      "There are a total of 1 phenotype to process \n",
      "\n",
      "Start performing clumping \n",
      "\n",
      "Clumping Progress: 100.00%\n",
      "Number of variant(s) after clumping : 223 \n",
      "\n",
      "Processing the 1 th phenotype \n",
      "\n",
      "Phenotype is a binary phenotype \n",
      "2349 control(s) \n",
      "1819 case(s) \n",
      "\n",
      "Processing the covariate file: \n",
      "/home/jupyter/workspace/ws_files/results/covariates_prs.txt \n",
      "============================== \n",
      "\n",
      "Include Covariates: \n",
      "Name\tMissing\tNumber of levels \n",
      "SEX\t0\t- \n",
      "AGE\t0\t- \n",
      "PC1\t0\t- \n",
      "PC2\t0\t- \n",
      "PC3\t0\t- \n",
      "PC4\t0\t- \n",
      "PC5\t0\t- \n",
      "PC6\t0\t- \n",
      "PC7\t0\t- \n",
      "PC8\t0\t- \n",
      "PC9\t0\t- \n",
      "PC10\t0\t- \n",
      "\n",
      "After reading the covariate file, 4168 sample(s) included \n",
      "in the analysis \n",
      "\n",
      "\n",
      "Start Processing\n",
      "Processing 100.00%\n",
      "There are 1 region(s) with p-value less than 1e-5. Please \n",
      "note that these results are inflated due to the overfitting \n",
      "inherent in finding the best-fit PRS (but it's still best \n",
      "to find the best-fit PRS!). \n",
      "You can use the --perm option (see manual) to calculate an \n",
      "empirical P-value. \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Begin plotting\n",
      "Current Rscript version = 2.3.3\n",
      "Plotting the quantile plot\n",
      "WARNING: There are only 0 unique PRS but asked for 4 quantiles\n",
      "Will not generate the quantile plot for  /home/jupyter/workspace/ws_files/results/EAS/model2_EAS_best\n",
      "Plotting Bar Plot\n",
      "Plotting Multi-Set-Plot\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In max(pheno$Pheno, na.rm = T) :\n",
      "  no non-missing arguments to max; returning -Inf\n",
      "PRSice 2.3.5 (2021-09-20) \n",
      "https://github.com/choishingwan/PRSice\n",
      "(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly\n",
      "GNU General Public License v3\n",
      "If you use PRSice in any published work, please cite:\n",
      "Choi SW, O'Reilly PF.\n",
      "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.\n",
      "GigaScience 8, no. 7 (July 1, 2019)\n",
      "2025-02-11 18:06:21\n",
      "/home/jupyter/prsice \\\n",
      "    --a1 A1 \\\n",
      "    --a2 A2 \\\n",
      "    --bar-levels 5e-07 \\\n",
      "    --base /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
      "    --beta  \\\n",
      "    --binary-target T \\\n",
      "    --bp BP \\\n",
      "    --chr CHR \\\n",
      "    --clump-kb 250kb \\\n",
      "    --clump-p 1.000000 \\\n",
      "    --clump-r2 0.100000 \\\n",
      "    --cov /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
      "    --fastscore  \\\n",
      "    --ignore-fid  \\\n",
      "    --ld /home/jupyter/workspace/ws_files/results/EUR/model2_plink/chr# \\\n",
      "    --no-full  \\\n",
      "    --num-auto 22 \\\n",
      "    --out /home/jupyter/workspace/ws_files/results/EUR/model2_EUR_best \\\n",
      "    --perm 10000 \\\n",
      "    --prevalence 0.005 \\\n",
      "    --print-snp  \\\n",
      "    --pvalue P \\\n",
      "    --score con-std \\\n",
      "    --seed 3 \\\n",
      "    --snp SNP \\\n",
      "    --stat BETA \\\n",
      "    --target /home/jupyter/workspace/ws_files/results/EUR/model2_plink/chr# \\\n",
      "    --thread 16\n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/EUR/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Start processing multi_MAMA \n",
      "================================================== \n",
      "\n",
      "Base file: \n",
      "/home/jupyter/workspace/ws_files/data/multi_MAMA.txt \n",
      "Header of file is: \n",
      "SNP CHR BP A1 A2 BETA P TEST \n",
      "\n",
      "Reading 100.00%\n",
      "17782530 variant(s) observed in base file, with: \n",
      "17774241 variant(s) excluded due to p-value threshold \n",
      "1192 ambiguous variant(s) excluded \n",
      "7097 total variant(s) included from base file \n",
      "\n",
      "Loading Genotype info from target \n",
      "================================================== \n",
      "\n",
      "48909 people (27340 male(s), 21569 female(s)) observed \n",
      "48909 founder(s) included \n",
      "\n",
      "123382417 variant(s) not found in previous data \n",
      "6800 variant(s) included \n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/EUR/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Loading Genotype info from reference \n",
      "================================================== \n",
      "\n",
      "48909 people (27340 male(s), 21569 female(s)) observed \n",
      "48909 founder(s) included \n",
      "\n",
      "123382417 variant(s) not found in previous data \n",
      "6800 variant(s) included \n",
      "\n",
      "There are a total of 1 phenotype to process \n",
      "\n",
      "Start performing clumping \n",
      "\n",
      "Clumping Progress: 100.00%\n",
      "Number of variant(s) after clumping : 267 \n",
      "\n",
      "Processing the 1 th phenotype \n",
      "\n",
      "Phenotype is a binary phenotype \n",
      "9706 control(s) \n",
      "22093 case(s) \n",
      "\n",
      "Processing the covariate file: \n",
      "/home/jupyter/workspace/ws_files/results/covariates_prs.txt \n",
      "============================== \n",
      "\n",
      "Include Covariates: \n",
      "Name\tMissing\tNumber of levels \n",
      "SEX\t0\t- \n",
      "AGE\t0\t- \n",
      "PC1\t0\t- \n",
      "PC2\t0\t- \n",
      "PC3\t0\t- \n",
      "PC4\t0\t- \n",
      "PC5\t0\t- \n",
      "PC6\t0\t- \n",
      "PC7\t0\t- \n",
      "PC8\t0\t- \n",
      "PC9\t0\t- \n",
      "PC10\t0\t- \n",
      "\n",
      "After reading the covariate file, 31799 sample(s) included \n",
      "in the analysis \n",
      "\n",
      "\n",
      "Start Processing\n",
      "Processing 100.00%\n",
      "There are 1 region(s) with p-value less than 1e-5. Please \n",
      "note that these results are inflated due to the overfitting \n",
      "inherent in finding the best-fit PRS (but it's still best \n",
      "to find the best-fit PRS!). \n",
      "You can use the --perm option (see manual) to calculate an \n",
      "empirical P-value. \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Begin plotting\n",
      "Current Rscript version = 2.3.3\n",
      "Plotting the quantile plot\n",
      "WARNING: There are only 0 unique PRS but asked for 4 quantiles\n",
      "Will not generate the quantile plot for  /home/jupyter/workspace/ws_files/results/EUR/model2_EUR_best\n",
      "Plotting Bar Plot\n",
      "Plotting Multi-Set-Plot\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In max(pheno$Pheno, na.rm = T) :\n",
      "  no non-missing arguments to max; returning -Inf\n",
      "PRSice 2.3.5 (2021-09-20) \n",
      "https://github.com/choishingwan/PRSice\n",
      "(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly\n",
      "GNU General Public License v3\n",
      "If you use PRSice in any published work, please cite:\n",
      "Choi SW, O'Reilly PF.\n",
      "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.\n",
      "GigaScience 8, no. 7 (July 1, 2019)\n",
      "2025-02-11 18:11:51\n",
      "/home/jupyter/prsice \\\n",
      "    --a1 A1 \\\n",
      "    --a2 A2 \\\n",
      "    --bar-levels 5e-06 \\\n",
      "    --base /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
      "    --beta  \\\n",
      "    --binary-target T \\\n",
      "    --bp BP \\\n",
      "    --chr CHR \\\n",
      "    --clump-kb 250kb \\\n",
      "    --clump-p 1.000000 \\\n",
      "    --clump-r2 0.100000 \\\n",
      "    --cov /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
      "    --fastscore  \\\n",
      "    --ignore-fid  \\\n",
      "    --ld /home/jupyter/workspace/ws_files/results/CAS/model2_plink/chr# \\\n",
      "    --no-full  \\\n",
      "    --num-auto 22 \\\n",
      "    --out /home/jupyter/workspace/ws_files/results/CAS/model2_CAS_best \\\n",
      "    --perm 10000 \\\n",
      "    --prevalence 0.005 \\\n",
      "    --print-snp  \\\n",
      "    --pvalue P \\\n",
      "    --score con-std \\\n",
      "    --seed 3 \\\n",
      "    --snp SNP \\\n",
      "    --stat BETA \\\n",
      "    --target /home/jupyter/workspace/ws_files/results/CAS/model2_plink/chr# \\\n",
      "    --thread 16\n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/CAS/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Start processing multi_MAMA \n",
      "================================================== \n",
      "\n",
      "Base file: \n",
      "/home/jupyter/workspace/ws_files/data/multi_MAMA.txt \n",
      "Header of file is: \n",
      "SNP CHR BP A1 A2 BETA P TEST \n",
      "\n",
      "Reading 100.00%\n",
      "17782530 variant(s) observed in base file, with: \n",
      "17770710 variant(s) excluded due to p-value threshold \n",
      "1704 ambiguous variant(s) excluded \n",
      "10116 total variant(s) included from base file \n",
      "\n",
      "Loading Genotype info from target \n",
      "================================================== \n",
      "\n",
      "1003 people (458 male(s), 545 female(s)) observed \n",
      "1003 founder(s) included \n",
      "\n",
      "33500568 variant(s) not found in previous data \n",
      "9393 variant(s) included \n",
      "\n",
      "Initializing Genotype file: \n",
      "/home/jupyter/workspace/ws_files/results/CAS/model2_plink/chr# \n",
      "(bed) \n",
      "\n",
      "Loading Genotype info from reference \n",
      "================================================== \n",
      "\n",
      "1003 people (458 male(s), 545 female(s)) observed \n",
      "1003 founder(s) included \n",
      "\n",
      "33500568 variant(s) not found in previous data \n",
      "9393 variant(s) included \n",
      "\n",
      "There are a total of 1 phenotype to process \n",
      "\n",
      "Start performing clumping \n",
      "\n",
      "Clumping Progress: 100.00%\n",
      "Number of variant(s) after clumping : 485 \n",
      "\n",
      "Processing the 1 th phenotype \n",
      "\n",
      "Phenotype is a binary phenotype \n",
      "323 control(s) \n",
      "582 case(s) \n",
      "\n",
      "Processing the covariate file: \n",
      "/home/jupyter/workspace/ws_files/results/covariates_prs.txt \n",
      "============================== \n",
      "\n",
      "Include Covariates: \n",
      "Name\tMissing\tNumber of levels \n",
      "SEX\t0\t- \n",
      "AGE\t0\t- \n",
      "PC1\t0\t- \n",
      "PC2\t0\t- \n",
      "PC3\t0\t- \n",
      "PC4\t0\t- \n",
      "PC5\t0\t- \n",
      "PC6\t0\t- \n",
      "PC7\t0\t- \n",
      "PC8\t0\t- \n",
      "PC9\t0\t- \n",
      "PC10\t0\t- \n",
      "\n",
      "After reading the covariate file, 905 sample(s) included in \n",
      "the analysis \n",
      "\n",
      "\n",
      "Start Processing\n",
      "Processing 100.00%\n",
      "There are 1 region(s) with p-value less than 1e-5. Please \n",
      "note that these results are inflated due to the overfitting \n",
      "inherent in finding the best-fit PRS (but it's still best \n",
      "to find the best-fit PRS!). \n",
      "You can use the --perm option (see manual) to calculate an \n",
      "empirical P-value. \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Begin plotting\n",
      "Current Rscript version = 2.3.3\n",
      "Plotting the quantile plot\n",
      "WARNING: There are only 0 unique PRS but asked for 4 quantiles\n",
      "Will not generate the quantile plot for  /home/jupyter/workspace/ws_files/results/CAS/model2_CAS_best\n",
      "Plotting Bar Plot\n",
      "Plotting Multi-Set-Plot\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In max(pheno$Pheno, na.rm = T) :\n",
      "  no non-missing arguments to max; returning -Inf\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "i=0\n",
    "thresholds=(5e-08 5e-07 5e-06 5e-08 5e-08 5e-07 5e-06)\n",
    "for ancestry in {\"AAC\",\"AFR\",\"AJ\",\"AMR\",\"EAS\",\"EUR\",\"CAS\"}; do\n",
    "    Rscript /home/jupyter/PRSice.R \\\n",
    "    --out /home/jupyter/workspace/ws_files/results/${ancestry}/model2_${ancestry}_best \\\n",
    "    --target /home/jupyter/workspace/ws_files/results/${ancestry}/model2_plink/chr# \\\n",
    "    -b /home/jupyter/workspace/ws_files/data/multi_MAMA.txt \\\n",
    "    --beta \\\n",
    "    --snp SNP \\\n",
    "    --A1 A1 \\\n",
    "    --A2 A2 \\\n",
    "    --stat BETA \\\n",
    "    --pvalue P \\\n",
    "    --ld /home/jupyter/workspace/ws_files/results/${ancestry}/model2_plink/chr# \\\n",
    "    --print-snp \\\n",
    "    --score con-std \\\n",
    "    --perm 10000 \\\n",
    "    --prsice /home/jupyter/prsice \\\n",
    "    -n 24 \\\n",
    "    --binary-target T \\\n",
    "    --quantile 4 \\\n",
    "    --prevalence 0.005 \\\n",
    "    --bar-levels ${thresholds[i]} \\\n",
    "    --fastscore \\\n",
    "    --no-full \\\n",
    "    --multi-plot 10 \\\n",
    "    --cov-file /home/jupyter/workspace/ws_files/results/covariates_prs.txt \\\n",
    "    --seed 3 \\\n",
    "    --thread 16 \\\n",
    "    --ignore-fid\n",
    "    \n",
    "    ((i++))\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cc56343-fa77-4fdc-b556-b38cccfaf4ef",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Estimate Specificity and Sensitivity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f35fa8dd-f48d-4b45-99d4-6856f6d76de1",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AAC\"\n",
      "[1] \"AFR\"\n",
      "[1] \"AJ\"\n",
      "[1] \"AMR\"\n",
      "[1] \"EAS\"\n",
      "[1] \"EUR\"\n",
      "[1] \"CAS\"\n",
      "   Ancestry       AUC  Accuracy  CI_Lower  CI_Upper Balanced_Accuracy\n",
      "     <list>     <num>     <num>     <num>     <num>             <num>\n",
      "1:      AAC 0.5913966 0.5611888 0.5318693 0.5901929         0.5793895\n",
      "2:      AFR 0.5831696 0.5763622 0.5571217 0.5954310         0.5682021\n",
      "3:       AJ 0.6695295 0.6173722 0.5945871 0.6397780         0.6345511\n",
      "4:      AMR 0.6391026 0.6045265 0.5877548 0.6211160         0.6077351\n",
      "5:      EAS 0.5775092 0.5556622 0.5404226 0.5708237         0.5558018\n",
      "6:      EUR 0.6210634 0.5976918 0.5922777 0.6030880         0.5863640\n",
      "7:      CAS 0.6123701 0.6044199 0.5717123 0.6364455         0.5980632\n",
      "   Sensitivity Specificity\n",
      "         <num>       <num>\n",
      "1:   0.6215385   0.5372405\n",
      "2:   0.5377358   0.5986683\n",
      "3:   0.6024355   0.6666667\n",
      "4:   0.5860996   0.6293706\n",
      "5:   0.5568994   0.5547041\n",
      "6:   0.6154438   0.5572842\n",
      "7:   0.6202749   0.5758514\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "# Initialize results data.table\n",
    "results_dt <- data.table(Ancestry = ancestries)\n",
    "results_dt[, AUC := as.numeric(NA)]\n",
    "results_dt[, Accuracy := as.numeric(NA)]\n",
    "results_dt[, CI_Lower := as.numeric(NA)]\n",
    "results_dt[, CI_Upper := as.numeric(NA)]\n",
    "results_dt[, Balanced_Accuracy := as.numeric(NA)]\n",
    "results_dt[, Sensitivity := as.numeric(NA)]\n",
    "results_dt[, Specificity := as.numeric(NA)]\n",
    "\n",
    "for (ancestry in ancestries) {\n",
    "    print(ancestry)\n",
    "\n",
    "    dat <- read.table(paste0(RESULTS_DIR, \"/\", ancestry, \"/model2_\", ancestry, \"_best.best\"), header = TRUE, sep = \" \")\n",
    "    cov <- read.table(paste0(RESULTS_DIR, \"/covariates_pheno.txt\"), header = TRUE, sep = \" \")\n",
    "    \n",
    "    colnames(cov) <- c(\"FID\", \"IID\", \"SEX\", \"AGE\", \"ANCESTRY\", \"PHENO\", \"PC1\", \"PC2\", \"PC3\", \"PC4\", \"PC5\", \"PC6\", \"PC7\", \"PC8\", \"PC9\", \"PC10\")\n",
    "    dat <- merge(dat, cov, by = \"IID\")\n",
    "    dat$CASE <- dat$PHENO - 1\n",
    "    dat <- subset(dat, CASE != -10)\n",
    "    meanControls <- mean(dat$PRS[dat$CASE == 0])\n",
    "    sdControls <- sd(dat$PRS[dat$CASE == 0])\n",
    "    dat$zSCORE <- (dat$PRS - meanControls) / sdControls\n",
    "    grsTests <- glm(CASE ~ zSCORE, family = \"binomial\", data = dat)\n",
    "    dat$probDisease <- predict(grsTests, dat, type = \"response\")\n",
    "    dat$reported <- ifelse(dat$CASE == 1, \"DISEASE\", \"CONTROL\")\n",
    "    roc <- roc(response = dat$reported, predictor = dat$probDisease)\n",
    "    auc_value <- auc(roc)\n",
    "    \n",
    "    result.coords <- coords(\n",
    "        roc, \n",
    "        \"best\", \n",
    "        best.method = \"closest.topleft\", \n",
    "        ret = c(\"threshold\", \"accuracy\", \"specificity\", \"sensitivity\", \"youden\"),\n",
    "    )\n",
    "    \n",
    "    dat$predicted <- ifelse(dat$probDisease > result.coords$threshold, \"DISEASE\", \"CONTROL\")\n",
    "    confMat <- confusionMatrix(data = as.factor(dat$predicted), reference = as.factor(dat$reported), positive = \"DISEASE\")\n",
    "    \n",
    "    results_dt[Ancestry == ancestry, AUC := auc_value]\n",
    "    results_dt[Ancestry == ancestry, Accuracy := confMat$overall[\"Accuracy\"]]\n",
    "    results_dt[Ancestry == ancestry, CI_Lower := confMat$overall[\"AccuracyLower\"]]\n",
    "    results_dt[Ancestry == ancestry, CI_Upper := confMat$overall[\"AccuracyUpper\"]]\n",
    "    results_dt[Ancestry == ancestry, Balanced_Accuracy := confMat$byClass[\"Balanced Accuracy\"]]\n",
    "    results_dt[Ancestry == ancestry, Sensitivity := confMat$byClass[\"Sensitivity\"]]\n",
    "    results_dt[Ancestry == ancestry, Specificity := confMat$byClass[\"Specificity\"]]\n",
    "}\n",
    "\n",
    "fwrite(results_dt, paste0(RESULTS_DIR, \"/model2_results.txt\"), sep = \"\\t\", quote = FALSE, row.names = FALSE)\n",
    "print(results_dt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6752e2d4-0c5e-41cc-ae7f-c158a49e760f",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_bestfit = pd.read_csv(f\"{RESULTS_DIR}/model2_results.txt\", sep=\"\\s+\")\n",
    "df_bestfit[[\"AUC\",\"Accuracy\",\"CI_Lower\",\"CI_Upper\",\"Balanced_Accuracy\",\"Sensitivity\",\"Specificity\"]] = df_bestfit[[\"AUC\",\"Accuracy\",\"CI_Lower\",\"CI_Upper\",\"Balanced_Accuracy\",\"Sensitivity\",\"Specificity\"]].round(3).astype(str)\n",
    "df_bestfit[\"Accuracy\"] = df_bestfit.apply(lambda row: f\"{row['Accuracy']} ({row['CI_Lower']} - {row['CI_Upper']})\", axis=1)\n",
    "df_bestfit.rename(columns={\"Accuracy\":\"Accuracy (95% CI)\", \"Balanced_Accuracy\":\"Balanced Accuracy\"}, inplace=True)\n",
    "df_bestfit.drop([\"CI_Lower\", \"CI_Upper\"], inplace=True, axis=1)\n",
    "\n",
    "df_prsice = []\n",
    "for ancestry in ancestries:\n",
    "    df = pd.read_csv(f\"{RESULTS_DIR}/{ancestry}/model2_{ancestry}_best.summary\", sep=\"\\s+\")\n",
    "    df[\"Ancestry\"] = ancestry\n",
    "    df = df[[\n",
    "        \"Ancestry\",\n",
    "        \"Threshold\",\n",
    "        \"PRS.R2.adj\",\n",
    "        \"Full.R2\",\n",
    "        \"Null.R2\",\n",
    "        \"Coefficient\",\n",
    "        \"Standard.Error\",\n",
    "        \"Num_SNP\",\n",
    "    ]]\n",
    "    df.rename(columns={\n",
    "        \"PRS.R2.adj\":\"PRS R2 Adj\",\n",
    "        \"Full.R2\":\"Full R2\",\n",
    "        \"Null.R2\":\"Null R2\",\n",
    "        \"Standard.Error\":\"SE\",\n",
    "        \"Num_SNP\":\"No. of SNP\",\n",
    "    }, inplace=True)\n",
    "    lower = f\"{math.e ** (df.loc[0,'Coefficient'] - 1.96 * df.loc[0,'SE']):.2f}\"\n",
    "    upper = f\"{math.e ** (df.loc[0,'Coefficient'] + 1.96 * df.loc[0,'SE']):.2f}\"\n",
    "    df[\"OR (95% CI)\"] = f\"{math.e ** df.loc[0,'Coefficient']:.2f} ({lower} - {upper})\"\n",
    "    df_prsice.append(df)\n",
    "df_prsice = pd.concat(df_prsice)\n",
    "df_prsice[[\"PRS R2 Adj\",\"Full R2\",\"Null R2\",\"Coefficient\",\"SE\"]] = df_prsice[[\"PRS R2 Adj\",\"Full R2\",\"Null R2\",\"Coefficient\",\"SE\"]].round(3)\n",
    "df_prsice[\"Threshold\"] = df_prsice[\"Threshold\"].apply(lambda x: f'{x:.2e}')\n",
    "\n",
    "df_merged = df_prsice.merge(df_bestfit, on=\"Ancestry\")\n",
    "df_merged.to_csv(f\"{RESULTS_DIR}/table3.txt\", index=False, sep=\"\\t\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4abb6ba9-1f98-4996-ba6b-abfb8dbe6de3",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Plot ROC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "c0411e84-dfe6-42a3-bb5f-c7cda7ceb68d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"------------------------------AAC------------------------------\"\n",
      "[1] \"------------------------------AFR------------------------------\"\n",
      "[1] \"------------------------------AJ------------------------------\"\n",
      "[1] \"------------------------------AMR------------------------------\"\n",
      "[1] \"------------------------------EAS------------------------------\"\n",
      "[1] \"------------------------------EUR------------------------------\"\n",
      "[1] \"------------------------------CAS------------------------------\"\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "data_list <- list()\n",
    "i <- 1\n",
    "for (ancestry in ancestries) {\n",
    "    print(paste0(\"------------------------------\", ancestry, \"------------------------------\"))\n",
    "    \n",
    "    prs_data <- fread(paste0(RESULTS_DIR, '/', ancestry,'/model2_', ancestry, '_best.best'), header=T)\n",
    "    colnames(prs_data)[4] <- \"PRS\"\n",
    "    covar <- fread(paste0(RESULTS_DIR, \"/covariates_pheno.txt\"), header=T)\n",
    "    colnames(covar)[1] <- \"FID\"\n",
    "    colnames(covar)[2] <- \"IID\"\n",
    "    temp <- merge(prs_data, covar, by = c(\"FID\",\"IID\"))\n",
    "    temp$CASE <- temp$PHENO - 1\n",
    "    DATA <- subset(temp, CASE != -10)\n",
    "\n",
    "    ## Probability of disease calculation\n",
    "    Model <- glm(CASE ~ PRS, data = DATA, family = \"binomial\")\n",
    "    DATA$probDisease <- predict(Model, DATA, type = c(\"response\"))\n",
    "    DATA$predicted <- ifelse(DATA$probDisease > 0.5, \"DISEASE\", \"CONTROL\")\n",
    "    DATA$reported <- ifelse(DATA$CASE == 1, \"DISEASE\",\"CONTROL\")\n",
    "\n",
    "    dat <- DATA\n",
    "    dat$Group.name <- paste0(\"Model 2 \", ancestry)\n",
    "    dat$Group <- \"Model 2\"\n",
    "    data_list[[i]] <- dat\n",
    "    i <- i + 1\n",
    "}\n",
    "\n",
    "to_plot <- rbindlist(data_list, fill = TRUE)\n",
    "write.table(to_plot, file = paste0(RESULTS_DIR, \"/model2_risk_results.txt\"), sep = \"\\t\", quote = FALSE, row.names = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "08452819-25fe-469b-b6a7-7b86c0b214e7",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "to_plot = pd.read_csv(f\"{RESULTS_DIR}/model2_risk_results.txt\", sep=\"\\t\")\n",
    "plt.figure(figsize=(8, 5))\n",
    "ancestry_groups = to_plot['Group.name'].unique()\n",
    "for ancestry_group in ancestry_groups:\n",
    "    group_data = to_plot[to_plot['Group.name'] == ancestry_group]\n",
    "    fpr, tpr, _ = roc_curve(group_data[\"CASE\"], group_data[\"probDisease\"])\n",
    "    roc_auc = auc(fpr, tpr)\n",
    "    plt.plot(fpr, tpr, label=f'{ancestry_group} (AUC = {roc_auc:.2f})')\n",
    "\n",
    "sns.set_palette(\"Dark2\")\n",
    "plt.plot([0, 1], [0, 1], color='gray', linestyle='--')\n",
    "plt.title(\"ROC Curves by Group\")\n",
    "plt.xlabel(\"False Positive Rate\")\n",
    "plt.ylabel(\"True Positive Rate\")\n",
    "plt.legend(loc=\"lower right\")\n",
    "plt.savefig(f\"{RESULTS_DIR}/figure4.png\", dpi=300)\n",
    "plt.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "752b73f3-7a32-44e2-a1e2-a1c0b44f7a8f",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Compare Model 1 and Model 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91a7f1e2-2290-443e-8891-9f47d7abb28b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Compare AUCs between Model 1 and Model 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "bded0010-f2bf-44d9-93e3-9186101eaf02",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"------------------------------AAC------------------------------\"\n",
      "[1] \"--------------------AAC--------------------\"\n",
      "[1] 0.8163538\n",
      "Area under the curve: 0.5852\n",
      "Area under the curve: 0.5914\n",
      "[1] \"--------------------EAS--------------------\"\n",
      "[1] -0.2451266\n",
      "Area under the curve: 0.6217\n",
      "Area under the curve: 0.5914\n",
      "[1] \"--------------------EUR--------------------\"\n",
      "[1] -0.02163641\n",
      "Area under the curve: 0.6506\n",
      "Area under the curve: 0.5914\n",
      "[1] \"--------------------AMR--------------------\"\n",
      "[1] 0.5369458\n",
      "Area under the curve: 0.5751\n",
      "Area under the curve: 0.5914\n",
      "[1] \"------------------------------AFR------------------------------\"\n",
      "[1] \"--------------------AAC--------------------\"\n",
      "[1] 0.01553902\n",
      "Area under the curve: 0.5435\n",
      "Area under the curve: 0.5832\n",
      "[1] \"--------------------EAS--------------------\"\n",
      "[1] 0.004326059\n",
      "Area under the curve: 0.5361\n",
      "Area under the curve: 0.5832\n",
      "[1] \"--------------------EUR--------------------\"\n",
      "[1] 0.003719084\n",
      "Area under the curve: 0.5356\n",
      "Area under the curve: 0.5832\n",
      "[1] \"--------------------AMR--------------------\"\n",
      "[1] 1.148835e-05\n",
      "Area under the curve: 0.5108\n",
      "Area under the curve: 0.5832\n",
      "[1] \"------------------------------AJ------------------------------\"\n",
      "[1] \"--------------------AAC--------------------\"\n",
      "[1] 4.185158e-09\n",
      "Area under the curve: 0.5433\n",
      "Area under the curve: 0.6695\n",
      "[1] \"--------------------EAS--------------------\"\n",
      "[1] 1.587348e-07\n",
      "Area under the curve: 0.5559\n",
      "Area under the curve: 0.6695\n",
      "[1] \"--------------------EUR--------------------\"\n",
      "[1] 0.8162049\n",
      "Area under the curve: 0.6648\n",
      "Area under the curve: 0.6695\n",
      "[1] \"--------------------AMR--------------------\"\n",
      "[1] 5.535135e-09\n",
      "Area under the curve: 0.5448\n",
      "Area under the curve: 0.6695\n",
      "[1] \"------------------------------AMR------------------------------\"\n",
      "[1] \"--------------------AAC--------------------\"\n",
      "[1] 1.207689e-05\n",
      "Area under the curve: 0.5792\n",
      "Area under the curve: 0.6391\n",
      "[1] \"--------------------EAS--------------------\"\n",
      "[1] 0.3031686\n",
      "Area under the curve: 0.6252\n",
      "Area under the curve: 0.6391\n",
      "[1] \"--------------------EUR--------------------\"\n",
      "[1] 0.8000847\n",
      "Area under the curve: 0.6357\n",
      "Area under the curve: 0.6391\n",
      "[1] \"--------------------AMR--------------------\"\n",
      "[1] 3.969299e-22\n",
      "Area under the curve: 0.5051\n",
      "Area under the curve: 0.6391\n",
      "[1] \"------------------------------EAS------------------------------\"\n",
      "[1] \"--------------------AAC--------------------\"\n",
      "[1] -0.8416216\n",
      "Area under the curve: 0.5797\n",
      "Area under the curve: 0.5775\n",
      "[1] \"--------------------EAS--------------------\"\n",
      "[1] 0.005163263\n",
      "Area under the curve: 0.5504\n",
      "Area under the curve: 0.5775\n",
      "[1] \"--------------------EUR--------------------\"\n",
      "[1] -5.220348e-07\n",
      "Area under the curve: 0.618\n",
      "Area under the curve: 0.5775\n",
      "[1] \"--------------------AMR--------------------\"\n",
      "[1] 0.0001895628\n",
      "Area under the curve: 0.537\n",
      "Area under the curve: 0.5775\n",
      "[1] \"------------------------------EUR------------------------------\"\n",
      "[1] \"--------------------AAC--------------------\"\n",
      "[1] 1.023106e-42\n",
      "Area under the curve: 0.5543\n",
      "Area under the curve: 0.6211\n",
      "[1] \"--------------------EAS--------------------\"\n",
      "[1] 3.421432e-14\n",
      "Area under the curve: 0.5843\n",
      "Area under the curve: 0.6211\n",
      "[1] \"--------------------EUR--------------------\"\n",
      "[1] -0.02379958\n",
      "Area under the curve: 0.6319\n",
      "Area under the curve: 0.6211\n",
      "[1] \"--------------------AMR--------------------\"\n",
      "[1] 1.059901e-26\n",
      "Area under the curve: 0.5691\n",
      "Area under the curve: 0.6211\n",
      "[1] \"------------------------------CAS------------------------------\"\n",
      "[1] \"--------------------AAC--------------------\"\n",
      "[1] 0.026037\n",
      "Area under the curve: 0.557\n",
      "Area under the curve: 0.6124\n",
      "[1] \"--------------------EAS--------------------\"\n",
      "[1] 0.049193\n",
      "Area under the curve: 0.5655\n",
      "Area under the curve: 0.6124\n",
      "[1] \"--------------------EUR--------------------\"\n",
      "[1] 0.2830106\n",
      "Area under the curve: 0.5913\n",
      "Area under the curve: 0.6124\n",
      "[1] \"--------------------AMR--------------------\"\n",
      "[1] 0.2321434\n",
      "Area under the curve: 0.585\n",
      "Area under the curve: 0.6124\n",
      "   Ancestry           AAC           EAS           EUR          AMR\n",
      "     <list>         <num>         <num>         <num>        <num>\n",
      "1:      AAC  8.163538e-01 -2.451266e-01 -2.163641e-02 5.369458e-01\n",
      "2:      AFR  1.553902e-02  4.326059e-03  3.719084e-03 1.148835e-05\n",
      "3:       AJ  4.185158e-09  1.587348e-07  8.162049e-01 5.535135e-09\n",
      "4:      AMR  1.207689e-05  3.031686e-01  8.000847e-01 3.969299e-22\n",
      "5:      EAS -8.416216e-01  5.163263e-03 -5.220348e-07 1.895628e-04\n",
      "6:      EUR  1.023106e-42  3.421432e-14 -2.379958e-02 1.059901e-26\n",
      "7:      CAS  2.603700e-02  4.919300e-02  2.830106e-01 2.321434e-01\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n",
       "Setting levels: control = 0, case = 1\n",
       "Setting direction: controls < cases\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "summstat_list <- c(\"AAC\", \"EAS\", \"EUR\", \"AMR\")\n",
    "\n",
    "# Initialize results data.table\n",
    "results_dt <- data.table(Ancestry = ancestries)\n",
    "for (summstat in summstat_list) {\n",
    "    results_dt[[summstat]] <- as.numeric(NA)\n",
    "}\n",
    "\n",
    "model2_data_full <- fread(paste0(RESULTS_DIR, \"/model2_risk_results.txt\"), header = TRUE)\n",
    "for (ancestry in ancestries) {\n",
    "    print(paste0(\"------------------------------\", ancestry, \"------------------------------\"))\n",
    "    model1_data_full <- read.table(paste0(RESULTS_DIR, \"/\", ancestry, \"/model1_risk_results.txt\"), header = TRUE) \n",
    "    \n",
    "    for (summstat_ancestry in summstat_list) {\n",
    "        model1_data <- subset(model1_data_full, Group == summstat_ancestry)\n",
    "        model2_data <- subset(model2_data_full, ANCESTRY == ancestry)\n",
    "        \n",
    "        gt_m1 <- model1_data$CASE\n",
    "        gt_m2 <- model2_data$CASE\n",
    "        pred_m1 <- model1_data$probDisease\n",
    "        pred_m2 <- model2_data$probDisease\n",
    "        roc1 <- roc(gt_m1, pred_m1)\n",
    "        roc2 <- roc(gt_m2, pred_m2)        \n",
    "        \n",
    "        auc1 <- auc(roc1)\n",
    "        auc2 <- auc(roc2)\n",
    "        results <- roc.test(roc1, roc2, method = \"delong\")\n",
    "        \n",
    "        p_value <- as.numeric(results$p.value)\n",
    "        if (auc1 > auc2) {\n",
    "            p_value <- p_value * -1\n",
    "        }\n",
    "        print(paste0(\"--------------------\", summstat_ancestry, \"--------------------\"))\n",
    "        print(p_value)\n",
    "        print(auc1)\n",
    "        print(auc2)\n",
    "        \n",
    "        results_dt[Ancestry == ancestry, (summstat_ancestry) := p_value]\n",
    "    }\n",
    "}\n",
    "\n",
    "fwrite(results_dt, paste0(RESULTS_DIR, \"/m1_m2_comparison.txt\"), sep = \"\\t\", quote = FALSE, row.names = FALSE)\n",
    "print(results_dt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "6f0a0e0f-44b6-494a-b117-57179538d139",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAy8AAAINCAYAAAA3PkC5AAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAco9JREFUeJzt3Xd4FOX+9/HPpoeEhBpDSUJNQgggTYpSRBQEUaTKj94EBFERKceKHAFRFM8B8QAhEURQBBE5gqCANEV6kSoGEEykk4SSQDLPHzzscQ0lC9lMhn2/rmuui52dmf1sdhP2u9+577EZhmEIAAAAAPI5D7MDAAAAAEBOULwAAAAAsASKFwAAAACWQPECAAAAwBIoXgAAAABYAsULAAAAAEugeAEAAABgCRQvAAAAACyB4gUAAACAJXiZHcAVzqZdMDsC8lDQn7vMjoA8lL5no9kRkIdO1e9udgTkodJFAk17bJ/qvVx27IytM1x2bMDd0HkBAAAAYAl3ZecFAADAGTYPT7MjAMgBihcAAOD2KF4Aa+C0MQAAAACWQOcFAAC4PTovgDXQeQEAAABgCXReAACA26PzAlgDnRcAAAAAlkDnBQAAuD2bJ50XwArovAAAAACwBDovAADA7Xkw5gWwBIoXAADg9hiwD1gDp40BAAAAsAQ6LwAAwO3ReQGsgc4LAAAAAEug8wIAANyezYPvcwEr4DcVAAAAgCXQeQEAAG6PMS+ANdB5AQAAAGAJpnZeUlJSFBgYKI+/nWeamZmp8+fPKygoyKRkAADAndB5AazBtM7Ll19+qVq1aunSpUvZ7rt06ZJq166tr7/+2oRkAADA3dg8PF22AMg9phUvU6ZM0bBhw1SgQIFs9wUEBGj48OGaNGmSCckAAAAA5EemFS+7du1S48aNb3h/w4YNtXPnzrwLBAAA3JbN09NlC4DcY1rxcubMGV25cuWG91++fFlnzpzJw0QAAAAA8jPTBuyXKVNGmzZtUnR09HXv37RpkyIiIvI4FQAAcEeMTQGswbTOS5s2bfTyyy/rzz//zHZfcnKyXnnlFbVt29aEZAAAAADyI9M6LyNGjNBXX32lihUrqkuXLoqKipIk7d27V7Nnz1ZYWJhGjBhhVjwAyFd6fzhfcc/whQ7gKnReAGswrXgpWLCg1q1bp5EjR+qzzz6zj28pVKiQunTporfeeksFCxY0Kx4AmG5r4h9Kv3xFdSPD7et+2n9Eft7eurdsCROTAQBgDlMvUhkcHKwPP/xQkydP1smTJ2UYhooXLy6bzSZJOn36tIoUKWJmRAAwTYlCBfXuojX6fudBXUjP0FvzV+rs+Yt6sVUDs6MBdx0POi+AJZg25uWvbDabihcvrpCQENlsNi1btkwdOnRQqVKlzI4GAKYJLVxQ73ZvoYJ+Ptpz9IQK+vvqnW4tFFqYrjSQ27hIJWAN+aJ4kaTDhw/r9ddfV5kyZdS+fXt5eHho5syZZscCANP8eS5Nw2ctUerFDFUqXVypF9I1fNYS/XkuzexoAACYwtTTxjIyMrRgwQJNnz5d69atU9OmTXX06FFt3bpVVapUydEx0tPTlZ6e7rjucqZ8fX1dERkA8swfp1P0ZJ3KqhsZrt4fztfL7R7UT/uP6I/TKbonONDseMBdhQ4JYA2mdV6effZZlSxZUh988IGefPJJHT16VF9//bVsNps8nbga7dixYxUcHOywvD/hXRcmB4C8Ub1sSYfB+pJUNzJc1cuWNCkRAADmshmGYZjxwF5eXho+fLhGjBjhMKuYt7e3tm/frpiYmBwd53qdl4t0XtxK0J+7zI6APJS+Z6PZEZCHTtXvbnYE5KHSRczrKJbqNNVlxz4252mXHRtwN6Z1XmbNmqWff/5ZJUqUUMeOHbV48WJlZmY6fRxfX18FBQU5LBQuAAAAwN3HtOKlU6dOWr58uXbu3Kno6GgNHDhQoaGhysrK0u7du82KBQAA3BCzjQHWYPpsY2XLltWoUaN06NAhffLJJ2rbtq26dOmi0qVLa/DgwWbHAwAAAJBPmF68XGOz2dSsWTN9/vnn+uOPP/TSSy/pww8/NDsWAABwA3ReAGswdark60lNTdUXX3yh2bNny6S5BAAAgJuxOTHTKQDz5JvOy+rVq9W9e3eVKFFC7777rpo0aaKffvrJ7FgAAAAA8glTOy/JyclKSEhQXFycUlJS1KFDB6Wnp2vhwoU5nioZAADgTnF6F2ANpnVeWrVqpaioKO3YsUMTJ07UH3/8oX//+99mxQEAAACQz5nWeVmyZIkGDx6sAQMGqGLFimbFAAAAoPMCWIRpnZe1a9cqNTVVNWvWVJ06dTRp0iSdPHnSrDgAAAAA8jnTipe6detq2rRpSkpKUr9+/TR37lyVLFlSWVlZWr58uVJTU82KBgAA3AxTJQPWYPpsYwEBAerVq5fWrl2rnTt36sUXX9S4ceMUEhKixx9/3Ox4AAAAAPIJ04uXv4qKitL48eN19OhRzZkzx+w4AADATXh42Fy2AMg9+e4ilZLk6emp1q1bq3Xr1mZHAQAAbsBGkQFYQr7qvAAAAADAjeTLzgsAAEBestnovABWQOcFAAAAgCXQeQEAAG6PgfWANdB5AQAAAGAJdF4AAIDbY7YxwBrovAAAAACwBDovAADA7dF5AayB4gUAALg9D6ZKBiyB08YAAAAAWAKdFwAA4PY4bQywBjovAAAAACyBzgsAAHB7dF4Aa6DzAgAAAMAS6LwAAAC350HnBbAEOi8AAAAALIHOCwAAcHs2vs4FLOGuLF6+2nfK7AjIQ6cuFDU7AvJScHOzEyAPbVi02+wIyEOf9bjPtMe2cZFKwBL4ngEAAACAJdyVnRcAAABnMGAfsAY6LwAAAAAsgc4LAABwe1ykErAGOi8AAAAALIHOCwAAcHt0XgBroPMCAAAAwBLovAAAALfnwXVeAEug8wIAAADAEui8AAAAt8eYF8AaKF4AAIDbo3gBrIHTxgAAAABYAp0XAADg9jzovACWQOcFAAAAgCXQeQEAAG7PxlTJgCXQeQEAAABgCXReAACA27PxdS5gCfyqAgAAALAEOi8AAMDtMdsYYA0ULwAAwO1xkUrAGjhtDAAAAIAl5Nvi5dKlS3r33XfNjgEAANyAzWZz2QIg95havJw4cUKLFy/WsmXLlJmZKUm6fPmyPvjgA5UpU0bjxo0zMx4AAACAfMS0MS9r167VY489ppSUFNlsNtWqVUvx8fFq3bq1vLy89MYbb6h79+5mxQMAAG6EAfuANZjWeXnllVfUokUL7dixQ0OGDNHGjRv15JNPasyYMdq9e7f69+8vf39/s+IBAAAAyGdMK1527typV155RbGxsXrzzTdls9k0fvx4tWvXzqxIAADATdk8bC5bAOQe04qXM2fOqFixYpIkf39/FShQQLGxsWbFuSt8MnqI2RGQhxaNH252BOQhXm/38lrzaLMjAEC+ZOp1Xnbv3q3k5GRJkmEY2rdvn86fP++wTdWqVc2IZhm/79ulKxnpKlulpn1d4s7N8vb1U+nIyiYmgyskH/hFVy5nqHRMdfu6o7u3ysvHV6EVYkxMBlfg9XYvUSGB8vH00M6kFPu6KiWClH4lS/tPpJmYzD140iEBLMHU4uWhhx6SYRj224899pikq9MVGoYhm81mn4UM1xdcLETfzfpI+zauVcbFi1o64wNdSD2npl36mx0NLhBYNEQ/fjZNiVvW6/Kli1rzyWRdSjuneh36mh0NLsDr7V5OpmWo233hui+isPy9PNW7boQK+npp1sbfzY7mFiheAGswrXhJTEw066HvKkFFQ9Tm+de0cu50JR86oHJVa6l5r+fMjgUXCSxSXA8P+Ic2zI/XySO/qnRsDTXoMtDsWHARXm/3cupCht5f9as61SitcsUCtP2Pc4r76bDZsQAgXzGteImIiDDroe8qqadP6rtPPpJfQKBCy1TUxfOp+vJf/1TTLv1VsEgxs+Mhl50/c1I/fj5dvgUCVSy8gjLOp+m7/4xTvQ59FFCY1/tuw+vtXgoX8Fa32uFKS7+i306eV4CPp55rVF4zNx7RmQuXzY5316PzAliDaQP2u3XrptTUVPvt7du36/Jl5/84p6enKyUlxWG5nJGem1HztbMnknXvg4/q0d7Py8ffX4/2fl73Pviozp5INjsaXCD15J+KbtBMDboOkrefvxp0HaToBs2UevJPs6PBBXi93UtIoK9W7D+huJ8O6+KVTMX9dFgr9p9QSKCv2dEAIN8wrXiZPXu2Ll68aL/doEED/f678+f1jh07VsHBwQ7L4vjJuRk1XwuLinUYrC9JZavUVFgUM7fdjUIrVnYYvC1JpWOqK7QikzPcjXi93cu+42kOg/UlaWdSivYdZ7B+XvD0sLlsAZB7bMZfR8znIQ8PDyUnJyskJESSVLBgQW3fvl3lypVz6jjp6elKT3fstHz+y3F5+/BNlbs4dSHD7AgAXGTDwVNmR0Ae+qzHfaY9duvpP7ns2Av71HXZsQF3Y+psY7nB19dXvr6OhYq3zzmT0gAAACuiQwJYQ766zsvevXuVlubYHuc6LwAAAACkfHqdl2u4zgsAAMgLXnReAEvI19d5+etsZAAAAK7CaWOANeS767ykpqZqzpw5iouL06ZNm+i8AAAAAJCUjwbsr169WnFxcZo/f75KliypNm3aaNKkSWbHAgAAboDOC2ANphYvycnJSkhIUFxcnFJSUtShQwelp6dr4cKFiomJMTMaAAAAgHzGtItUtmrVSlFRUdqxY4cmTpyoP/74Q//+97/NigMAANyYp4eHyxYAuce0zsuSJUs0ePBgDRgwQBUrVjQrBgAAAACLMO3rgLVr1yo1NVU1a9ZUnTp1NGnSJJ08edKsOAAAwI15ethctgDIPaYVL3Xr1tW0adOUlJSkfv36ae7cuSpZsqSysrK0fPlypkkGAAAA4MD0EzEDAgLUq1cvrV27Vjt37tSLL76ocePGKSQkRI8//rjZ8QAAgBug8wJYg+nFy19FRUVp/PjxOnr0qObMmWN2HAAA4CYoXgBryFfFyzWenp5q3bq1Fi1aZHYUAAAAAPlEvrlIJQAAgFk8bXRIACvIl50XAAAAAPg7Oi8AAMDtMTYFsAY6LwAAAAAsgc4LAABwe3ReAGug8wIAAADAEui8AAAAt+dF5wWwBIoXAADg9jhtDLAGThsDAAAAYAl0XgAAgNuj8wJYA50XAAAAAJZA5wUAALg9Oi+ANdB5AQAAAGAJdF4AAIDbo/MCWAOdFwAAAACWQOcFAAC4PTovgDVQvAAAALdH8QJYw11ZvHz07X6zIyAP+RXwNjsC8tCXfWqZHQF5aP66w2ZHAADkI3dl8QIAAOAMOi+ANTBgHwAAAIAl0HkBAABuj84LYA10XgAAAABYAp0XAADg9ui8ANZA5wUAAACAJdB5AQAAbo/OC2ANFC8AAMDtedooXgAr4LQxAAAAAJZA5wUAALg9DzovgCU43XmJj4/XhQsXXJEFAAAAAG7I6eJlxIgRCg0NVe/evbV+/XpXZAIAAMhTnjbXLQByj9PFy7Fjx/Txxx/r5MmTaty4saKjo/X2228rOTnZFfkAAAAAQNJtFC9eXl568skn9dVXX+n3339X3759NXv2bIWHh+vxxx/XV199paysLFdkBQAAcAkPD5vLFgC5545mG7vnnnv0wAMPqF69evLw8NDOnTvVvXt3lS9fXqtWrcqliAAAAABwm8XLn3/+qXfffVeVK1dW48aNlZKSosWLFysxMVHHjh1Thw4d1L1799zOCgAA4BKeNpvLFgC5x+nipVWrVgoLC1NCQoL69u2rY8eOac6cOWratKkkKSAgQC+++KJ+//33XA8LAAAAwH05fZ2XkJAQ/fDDD6pXr94NtylevLgSExPvKBgAAEBe4TovgDU4VbxcvnxZhw4dUrFixW66nc1mU0RExB0FAwAAyCtMaQxYg1OnjXl7e2vHjh2uygIAAAAAN+T0mJcuXbooLi7OFVkAAABMwVTJgDU4PeblypUrmjFjhr777jvVrFlTAQEBDve/9957uRYOAAAAAK5xunjZtWuXatSoIUnav39/rgcCAADIawzYB6zB6dPGVq5cedMFAAAA+deqVatks9l09uzZHO9TpkwZTZw40WWZbqRx48Z6/vnnc/WY13v+CxcuVIUKFeTp6annn39eCQkJKlSoUK4+7u2y2WxauHCh2THyDaeLl169eik1NTXb+vPnz6tXr165EgoAACAvedpctzijR48estls6t+/f7b7Bg4cKJvNph49euTOk85F06ZNU4MGDVS4cGEVLlxYTZs21c8//3zTfTIzMzVu3DhFR0fL399fRYoUUZ06dTR9+nT7NgsWLNDo0aNzNWv9+vWVlJSk4OBg+7p+/fqpXbt2+v333zV69Gh17Ngxz88weuONN3Tvvffm+nETEhJks9nsS2BgoGrWrKkFCxY4fZzrFXR5Xdg6Xbx8/PHHunjxYrb1Fy9e1MyZM3MlFAAAgLsKCwvT3LlzHT5vXbp0SZ9++qnCw8NNTHZjq1atUqdOnbRy5Ur9+OOPCgsL0yOPPKJjx47dcJ9Ro0bp/fff1+jRo7V7926tXLlSTz/9tENHpEiRIipYsGCuZvXx8VFoaKhs//9UwbS0NB0/flzNmjVTyZIlVbBgQfn7+yskJCRXH9dMQUFBSkpKUlJSkrZu3apmzZqpQ4cO2rdvn9nR7DIyMnK0XY6Ll5SUFJ07d06GYSg1NVUpKSn25cyZM/rmm2/uqhfZiiZ3rmF2BOSh99tVNTsCABf59/9VNzuC2/Gw2Vy2OKtGjRoKCwtz+GZ8wYIFCg8PV/Xqju+N9PR0DR48WCEhIfLz89MDDzygjRs3OmzzzTffKDIyUv7+/nrwwQd16NChbI+5du1aNWjQQP7+/goLC9PgwYN1/vz5HGeePXu2nnnmGd17772Kjo7W9OnTlZWVpe+///6G+yxatEjPPPOM2rdvr7Jly6patWrq3bu3hg4dat/m76eNJSUlqWXLlvL391fZsmX16aefZvvm32azafr06XryySdVoEABVaxYUYsWLbLf/9fTxlatWmUvjpo0aSKbzaZVq1Zdt8vw9ddfq3bt2vLz81OxYsX05JNP2u+bNWuWatWqpYIFCyo0NFT/93//p+PHj2d7zO+//161atVSgQIFVL9+fXvxkJCQoFGjRmn79u32DklCQkK2n1mTJk00aNAgh3UnTpyQj4/PTX/WNptNoaGhCg0NVcWKFfXPf/5THh4eDpdASU9P19ChQ1WqVCkFBASoTp06WrVqlT1/z549de7cOXu+N954Q40bN9bhw4f1wgsv2Ndfc6v3VJkyZTR69Gh169ZNQUFBevrpp2+Y/69yXLwUKlRIRYoUkc1mU2RkpL0tWLhwYRUrVky9evXSwIEDc3o45JKqpYNVu0xhh3W1yxRWlVLBN9gDVhZbIkg1wgo5rKsRVkiVSwSZEwhArqlSKli1Ihz/nteKKKzYUvx+5wVPD5vLlvT0dIcvfVNSUpSenn7TPL169VJ8fLz99owZM9SzZ89s2w0bNkzz58/Xxx9/rC1btqhChQpq1qyZTp8+LUn6/fff1aZNG7Vq1Urbtm1Tnz59NGLECIdjHDx4UM2bN1fbtm21Y8cOffbZZ1q7dm22D8nOuHDhgi5fvqwiRYrccJvQ0FCtWLFCJ06cyPFxu3Xrpj/++EOrVq3S/PnzNXXqVIci4ZpRo0apQ4cO2rFjh1q0aKHOnTvbfyZ/9dcCYv78+UpKSlL9+vWzbfff//5XTz75pFq0aKGtW7fq+++/13333We///Llyxo9erS2b9+uhQsX6tChQ9c9ve/ll1/WhAkTtGnTJnl5edmHXHTs2FEvvviiKleubO+QdOzYMdv+ffr00aeffurw/vnkk09UqlQpNWnS5NY/QF09Xe/jjz+WJPskXJI0aNAg/fjjj5o7d6527Nih9u3bq3nz5jpw4IDq16+viRMnOnRwhg4dqgULFqh06dJ688037eulnL+n3n33XVWrVk1bt27Vq6++mqP8OZ5tbOXKlTIMQ02aNNH8+fMd3ow+Pj6KiIhQyZIlc3q4HF3s0svLS6GhoTd947u75HOX9FzTimoUFaICPp4a2ixKhQp461/fHTA7Glzgz9R0PdOwnBpWKKYC3p56/sEKCvb31oerfzM7GoA79GfKJT37UEU1iiquAj6eevGRSAX7e2vSil/NjoY7NHbsWI0aNcph3euvv6433njjhvt06dJFI0eO1OHDhyVJ69at09y5c+3fhEtXxxtPmTJFCQkJevTRRyVdHXuyfPlyxcXF6aWXXtKUKVNUvnx5TZgwQZIUFRWlnTt36u2333bI17lzZ3uHo2LFivrXv/6lRo0aacqUKfLz83P6OQ8fPlwlS5ZU06ZNb7jNe++9p3bt2ik0NFSVK1dW/fr19cQTT9ify9/t3btX3333nTZu3KhatWpJkqZPn66KFStm27ZHjx7q1KmTJGnMmDH617/+pZ9//lnNmzd32M7Hx8d+5lCRIkUUGhp63cd+66239NRTTzm8jtWqVbP/+6/jvsuVK6d//etfql27ttLS0hQYGOhwnEaNGkmSRowYoZYtW+rSpUvy9/dXYGCg/bPvjbRp00aDBg3SV199pQ4dOki62rW5NlbqRs6dO2fPcfHiRXl7e2vq1KkqX768JOnIkSOKj4/XkSNH7J/nhw4dqqVLlyo+Pl5jxoxRcHCwvYPzV56envaO0zU5fU81adJEL7744g1zX0+Oi5drP+jExESFh4ff9AeUE/fee69sNpsMw7jpdjabTdWqVdPMmTMVGxt7R495Nzqemq6Xv9ylAY3LK7pEkDb8dlrvfpt/zl9E7jqRlq5R3+xRn/vLKPKegtp4+IwmruSDDXA3OJ6arlcX7lL/RuUUHRqkn387rQnLuCRBXnHlVMkjR47UkCFDHNb5+vredJ/ixYurZcuWSkhIkGEYatmypYoVK+awzcGDB3X58mXdf//99nXe3t667777tGfPHknSnj17VKdOHYf96tWr53B7+/bt2rFjh2bPnm1fZxiGsrKylJiYqEqVKuX8yUoaN26cvdC6WeETExOjXbt2afPmzVq3bp1Wr16tVq1aqUePHg6D9q/Zt2+fvLy8HLoFFSpUUOHChbNtW7Xq/06tDggIUFBQ0HU7NDm1bds29e3b94b3b968WW+88Ya2b9+uM2fOKCsrS9LVoiAmJua6uUqUKCFJOn78eI7HMvn5+alr166aMWOGOnTooC1btmjXrl0Op8VdT8GCBbVlyxZJV7ti3333nfr376+iRYuqVatW2rlzpzIzMxUZGemwX3p6uooWLZqjbH+V0/fUtSLUGU5f52XPnj36/fff9cADD0iSJk+erGnTpikmJkaTJ0++7hvoehITE2+5TUpKis6fP6933nlHAwYM0Jo1a5yNe9crHuirZ5tWVOqly9qblKKC/l56s3Ws/v3dAZ1Iu3lLGtZTLMBHAxqWU2r6Fe3/M1WBfl569dFoTVn9m06ez9lANwD5U7FAHz3b5P//PU9OUUE/L416vLL+veKATqbx+21lvr6+tyxWrqdXr17202wmT56c27Hs0tLS1K9fPw0ePDjbfc5OEPDuu+9q3Lhx+u677xw+qN+Ih4eHateurdq1a+v555/XJ598oq5du+rll19W2bJlnXrsv/L29na4bbPZ7AXF7fD397/hfefPn1ezZs3UrFkzzZ49W8WLF9eRI0fUrFmzbIPQ/5rrWiPA2Vx9+vTRvffeq6NHjyo+Pl5NmjRRRETETffx8PBQhQoV7LerVq2qZcuW6e2331arVq2UlpYmT09Pbd68WZ6eng77/rVzlFM5fU/9/WL3OeF08fLSSy/ZW407d+7UkCFD9OKLL2rlypUaMmSIw/mZN3OjH3JqaqrmzJmjuLg4bdq0SZmZmXr77bcdWnN/lZ6enu280awrGfLw8nHiWVlXiUJ++nrbMW08dEaTO9fQO0v3qXaZwipRyI/i5S4UGuSn/+5K1pbfz+r9dlU1ccWvqhFWSKFBfhQvgMWVCPbX19v/0KbDZ/Tv/6uud5ftV62IwioR7E/xkgecndI4LzRv3lwZGRmy2Wxq1qxZtvvLly8vHx8frVu3zv656vLly9q4caP9dJ1KlSpl+1b+p59+crhdo0YN7d692+HD7e0YP3683nrrLX377be39Y26JHuX4nqTBURFRenKlSvaunWratasKUn69ddfdebMmdsPnUNVq1bV999/f91xR3v37tWpU6c0btw4hYWFSZI2bdrk9GP4+PgoMzPzlttVqVJFtWrV0rRp0/Tpp59q0qRJTj+WdPV0r2sz2lWvXl2ZmZk6fvy4GjRo4FS+663PrffU9Tg9VXJiYqL9jTV//ny1atVKY8aM0eTJk7VkyZLbDrJ69Wp1795dJUqU0LvvvqsHH3zQ/stVtmxZrV+//rr7jR07VsHBwQ7LsR/m3HYOq9lx9Jw2HnL8pd146Ix2HD1nUiK40q6kFG35/azDui2/n9WupBRzAgHINTuPndOmw45/zzcdPqOdx/h77q48PT21Z88e7d69O9u34dLVb60HDBigl156SUuXLtXu3bvVt29fXbhwQb1795Yk9e/fXwcOHNBLL72kffv26dNPP802i9Xw4cO1fv16DRo0SNu2bdOBAwf01VdfOTVg/+2339arr76qGTNmqEyZMkpOTlZycrLS0tJuuE+7du30/vvva8OGDTp8+LBWrVqlgQMHKjIyUtHR0dm2j46OVtOmTfX000/r559/1tatW/X000/L39//jocz3Mrrr7+uOXPm6PXXX9eePXscxg2Fh4fLx8dH//73v/Xbb79p0aJFt3VtmjJlyigxMVHbtm3TyZMnbzqpQ58+fTRu3DgZhuEw69mNGIZhf00SExM1depUffvtt3riiSckSZGRkercubO6deumBQsWKDExUT///LPGjh2r//73v/Z8aWlp+v7773Xy5ElduHDBvn716tU6duyYTp48KSl33lM34nTx4uPjYw/73Xff6ZFHHpF0dZBTSopzH6CSk5M1btw4VaxYUe3bt1dQUJDS09O1cOFCjRs3TrVr15Z09Zf3Rp2XkSNH6ty5cw5LqUadnH1ad4WBs7eYHQF56IUvbj3pBQBrevbTrWZHcDv5aarkvwoKClJQ0I1nnBs3bpzatm2rrl27qkaNGvr111/17bff2k/jDw8P1/z587Vw4UJVq1ZNH330kcaMGeNwjKpVq+qHH37Q/v371aBBA1WvXl2vvfaaUxMxTZkyRRkZGWrXrp1KlChhX959990b7tOsWTN9/fXXatWqlSIjI9W9e3dFR0dr2bJl8vK6/slBM2fO1D333KOGDRvqySefVN++fVWwYMHbmlTAGY0bN9a8efO0aNEi3XvvvWrSpIn9IpzFixdXQkKC5s2bp5iYGI0bN+6mz/tG2rZtq+bNm+vBBx9U8eLFNWfOjb+M79Spk7y8vNSpU6ccPfeUlBT7a1KpUiVNmDBBb775pl5++WX7NvHx8erWrZtefPFFRUVFqXXr1tq4caP9NK/69eurf//+6tixo4oXL67x48dLkt58800dOnRI5cuXV/HixSXlznvqRmzGrUbM/83jjz+ujIwM3X///Ro9erQSExNVqlQpLVu2TIMGDcrx1UhbtWql1atXq2XLlurcubOaN28uT09PeXt7a/v27Q6Dm5xVb8yN57nG3cevgPetN8Jd48s+t3cqAqyp5aSfbr0R7hrrRuRsqldX+Hbf7Q/mvpVmUVwHz1WOHj2qsLAwfffdd3rooYfMjpNnrhULGzdudJjAwB04PeZl0qRJeuaZZ/TFF19oypQpKlWqlCRpyZIl2aafu5klS5Zo8ODBGjBgwHWnuAMAAMgrnh75cNALslmxYoXS0tJUpUoVJSUladiwYSpTpowaNmxodrQ8cfnyZZ06dUqvvPKK6tat63aFi3QbxUt4eLgWL16cbf3777/v1HHWrl2ruLg41axZU5UqVVLXrl311FNPORsHAADgjlG7WMPly5f1j3/8Q7/99psKFiyo+vXra/bs2dlmF7tbrVu3Tg8++KAiIyP1xRdfmB3HFE6PeZGuziv+yiuvqFOnTvY5s5csWaJffvklx8eoW7eupk2bpqSkJPXr109z585VyZIllZWVpeXLlys1NfV2ogEAAOAu1axZM+3atUsXLlzQn3/+qS+//PKW0wTfTRo3bizDMLRv3z5VqVLF7DimcLp4+eGHH1SlShVt2LBBCxYssM8isX37dr3++utOBwgICFCvXr20du1a7dy5Uy+++KLGjRunkJAQPf74404fDwAAwFmeNpvLFgC5x+niZcSIEfrnP/+p5cuXy8fnf9dSadKkSbZ5w50VFRWl8ePH6+jRozedYQEAAACA+3F6zMvOnTv16aefZlsfEhJin9v5Tnl6eqp169Zq3bp1rhwPAADgZu50SmMAecPpzkuhQoWUlJSUbf3WrVvtM48BAAAAQG5zunh56qmnNHz4cCUnJ8tmsykrK0vr1q3T0KFD1a1bN1dkBAAAcClPD9ctAHKP079SY8aMUXR0tMLCwpSWlqaYmBg1bNhQ9evX1yuvvOKKjAAAAADg/JgXHx8fTZs2Ta+99pp27typtLQ0Va9enQtNAgAAy2LMC2ANThcv14SFhSksLCw3swAAAJiCKY0Ba3D6tLG2bdvq7bffzrZ+/Pjxat++fa6EAgAAAIC/c7p4Wb16tVq0aJFt/aOPPqrVq1fnSigAAIC85GGzuWwBkHucLl7S0tIcLk55jbe3t1JSUnIlFAAAAAD8ndPFS5UqVfTZZ59lWz937lzFxMTkSigAAIC8xFTJgDU4PWD/1VdfVZs2bXTw4EE1adJEkvT9999rzpw5mjdvXq4HBAAAAADpNoqXVq1aaeHChRozZoy++OIL+fv7q2rVqvruu+/UqFEjV2QEAABwKcamANZwW1Mlt2zZUi1btsztLAAAAABwQ7d9nZeMjAwdP35cWVlZDuvDw8PvOBQAAEBeovECWIPTxcuBAwfUq1cvrV+/3mG9YRiy2WzKzMzMtXAAAAB5wUNUL4AVOF289OjRQ15eXlq8eLFKlCghG19VAAAAAMgDThcv27Zt0+bNmxUdHe2KPAAAAHmO72IBa3B69vGYmBidPHnSFVkAAAAA4IacLl7efvttDRs2TKtWrdKpU6eUkpLisAAAAFiNh811C4Dc4/RpY02bNpUkPfTQQw7rGbAPAAAAwJWcLl5WrlzpihwAAACmYcwLYA1OFy+NGjW64X27du26ozAAAAAAcCNOj3n5u9TUVE2dOlX33XefqlWrlhuZAAAA8pSHbC5bAOSe2y5eVq9ere7du6tEiRJ699131aRJE/3000+5mQ0AACBP2GyuWwDkHqdOG0tOTlZCQoLi4uKUkpKiDh06KD09XQsXLlRMTIyrMgIAAABAzouXVq1aafXq1WrZsqUmTpyo5s2by9PTUx999JEr892W8JIFzY6APDTqUS6YCtytErfuMzsC8lQT0x6ZKY0Ba8hx8bJkyRINHjxYAwYMUMWKFV2ZCQAAAACyyfGYl7Vr1yo1NVU1a9ZUnTp1NGnSJJ08edKV2QAAAPKEzYULgNyT4+Klbt26mjZtmpKSktSvXz/NnTtXJUuWVFZWlpYvX67U1FRX5gQAAADg5pyebSwgIEC9evXS2rVrtXPnTr344osaN26cQkJC9Pjjj7siIwAAgEt52GwuWwDknju6zktUVJTGjx+vo0ePas6cObmVCQAAAACycWqq5Bvx9PRU69at1bp169w4HAAAQJ6iQQJYQ64ULwAAAFZ2R6eiAMgz/K4CAAAAsAQ6LwAAwO3ZOG8MsAQ6LwAAAAAsIUedl0WLFuX4gEyXDAAArMaDxgtgCTkqXnI6i5jNZlNmZuad5AEAAACA68pR8ZKVleXqHAAAAKZhyAtgDXc05uXSpUu5lQMAAAAAbsrp4iUzM1OjR49WqVKlFBgYqN9++02S9OqrryouLi7XAwIAALiahwsXALnH6d+pt956SwkJCRo/frx8fHzs62NjYzV9+vRcDQcAAJAXbDabyxYAucfp4mXmzJmaOnWqOnfuLE9PT/v6atWqae/evbkaDgAAAACucfoilceOHVOFChWyrc/KytLly5dzJRQAAEBeYqpkwBqc7rzExMRozZo12dZ/8cUXql69eq6EAgAAAIC/c7rz8tprr6l79+46duyYsrKytGDBAu3bt08zZ87U4sWLcy3YpUuXNGnSJA0dOjTXjgkAAHA9NF4Aa3C68/LEE0/o66+/1nfffaeAgAC99tpr2rNnj77++ms9/PDDTh3rxIkTWrx4sZYtW2a/uOXly5f1wQcfqEyZMho3bpyz8QAAAADcpZzuvEhSgwYNtHz58jt64LVr1+qxxx5TSkqKbDabatWqpfj4eLVu3VpeXl5644031L179zt6DAAAgJxgzAtgDU53Xvr06aNVq1bd8QO/8soratGihXbs2KEhQ4Zo48aNevLJJzVmzBjt3r1b/fv3l7+//x0/DgAAAIC7g9PFy4kTJ9S8eXOFhYXppZde0rZt227rgXfu3KlXXnlFsbGxevPNN2Wz2TR+/Hi1a9futo4HAABwu7jOC2ANThcvX331lZKSkvTqq69q48aNqlmzpipXrqwxY8bo0KFDOT7OmTNnVKxYMUmSv7+/ChQooNjYWGfjAAAAAHATtzXmpXDhwnr66af19NNP6+jRo5ozZ45mzJih1157TVeuXMnxcXbv3q3k5GRJkmEY2rdvn86fP++wTdWqVW8nIgAAQI4x5gWwhtsqXq65fPmyNm3apA0bNujQoUO65557nNr/oYcekmEY9tuPPfaYpKutW8MwZLPZ7LOQ4dZeax6tN5fuNTsGAOAOffHa42r35iKzY7gVahfAGm6reFm5cqU+/fRTzZ8/X1lZWWrTpo0WL16sJk2a5PgYiYmJt/PQ+JuokED5eHpoZ1KKfV2VEkFKv5Kl/SfSTEwGAHBG7ahQ+fl4as3OY/Z1DaqU0sX0K9q0/08TkwFA/uF08VKqVCmdPn1azZs319SpU9WqVSv5+vo6/cARERFO74PsTqZlqNt94bovorD8vTzVu26ECvp6adbG382OBgBwwrGTqXq92/169L5yCvT30ZjeDVSkoJ/enLXe7GhuwYOB9YAlOF28vPHGG2rfvr0KFSp0Rw985MiRHG0XHh5+R49ztzt1IUPvr/pVnWqUVrliAdr+xznF/XTY7FgAACf9ceq8+r2/TCM71VHVcsW1avvv+kfcGrNjAUC+4nTx0rdvX0nSr7/+qoMHD6phw4by9/e3j1HJqbJly9r/fW3cy1/3Z8xLzhQu4K1utcOVln5Fv508rwAfTz3XqLxmbjyiMxcumx0PAJBDoYUD9Hq3ejqblqEdv51QcICvpjzXVKNm/qjkM+dvfQDcERovgDU4XbycOnVKHTp00MqVK2Wz2XTgwAGVK1dOvXv3VuHChTVhwoQcHcdms6l06dLq0aOHWrVqJS+v25s7ID09Xenp6Q7rMi9nyNPb57aOZzUhgb5asf+Edial6LXm0Yr76bCqlAhSSKAvxQsAWEhYSEF9umKP1uw8pi9ee1wj41arQZVSCgspSPECAP+f09d5eeGFF+Tt7a0jR46oQIEC9vUdO3bU0qVLc3yco0ePasCAAZo7d65atmypWbNmycfHR9WqVXNYbmXs2LEKDg52WPb892Nnn5Zl7Tue5jBYX5J2JqVo33EG6wOAlWzcl+wwWF+S1uw8po37kk1K5F5shuGyBUDusRmGc79VoaGh+vbbb1WtWjUVLFhQ27dvV7ly5fTbb7+patWqSktz/kPz2rVrFR8fr3nz5ikmJka9e/dW79695eFx69rqep2XXp/tcJvOC6RRj0abHQF5KDTgjmZ4h8XE9HafL6Mg/fHZANMe+9LFiy47tp+/v8uODbgbpzsv58+fd+i4XHP69OnbmnVMkh544AHFxcXpwIEDKlCggPr376+zZ8/maF9fX18FBQU5LBQuAADAKUaW6xYAucbp4qVBgwaaOXOm/bbNZlNWVpbGjx+vBx988LZCrF+/Xn369FFkZKTS0tI0efLkO57NDAAAIKdsRpbLFgC5x+nzL8aPH6+HHnpImzZtUkZGhoYNG6ZffvlFp0+f1rp163J8nKSkJM2cOVPx8fE6c+aMOnfurHXr1ik2NtbZSAAAAADcgNPFS2xsrPbv369JkyapYMGCSktLU5s2bTRw4ECVKFEix8cJDw9XqVKl1L17dz3++OPy9vZWVlaWduzY4bBd1apVnY0IAADgHDokgCU4PWA/t/x1MP6167v8PcrtXuelY8LPdxYOlsKAfffCgH33woB992LmgP30tHMuO7ZvYLDLjg24mxx9Cvh7N+RmctopSUxMvOU2qampOX5cAACA28aUxoAl5Kh4uffee2Wz2bJ1Rv7OmU5JRETEddenpqZqzpw5iouL06ZNm26r8wIAAADg7pOj4iUnXZI7tXr1asXFxWn+/PkqWbKk2rRpo0mTJrn8cQEAABjzAlhDjoqXG3VJ7lRycrISEhIUFxenlJQUdejQQenp6Vq4cKFiYmJc8pgAAAAArMnp67z8VVBQkH777bfb2rdVq1aKiorSjh07NHHiRP3xxx/697//fSdxAAAAbgvXeQGs4Y6m7bmTicqWLFmiwYMHa8CAAapYseKdxAAAALgzFBmAJdxR5+VOrF27VqmpqapZs6bq1KmjSZMm6eTJk2bFAQAAAJDP3VHx0qVLFwUFBd3WvnXr1tW0adOUlJSkfv36ae7cuSpZsqSysrK0fPlypkkGAAB5x8hy3QIg1zhdvMycOVPp6emSpClTpqhYsWKSpIyMDM2cOdPpAAEBAerVq5fWrl2rnTt36sUXX9S4ceMUEhKixx9/3OnjAQAAALg7OV289OzZU+fOZb8KbWpqqnr27HlHYaKiojR+/HgdPXpUc+bMuaNjAQAA5BidF8ASnC5eDMOQzWbLtv7o0aMKDg7OlVCenp5q3bq1Fi1alCvHAwAAAGB9OZ5trHr16rLZbLLZbHrooYfk5fW/XTMzM5WYmKjmzZu7JCQAAIBLZdEhAawgx8VL69atJUnbtm1Ts2bNFBgYaL/Px8dHZcqUUdu2bXM9IAAAAABIThQvr7/+uiSpTJky6tixo/z8/FwWCgAAIC9xMUnAGpy+SGX37t0lXZ1d7Pjx48r6W5s1PDw8d5IBAADkFYoXwBKcLl4OHDigXr16af369Q7rrw3kz8zMzLVwAAAAAHCN08VLjx495OXlpcWLF6tEiRLXnXkMAADAUgzD7AQAcsDp4mXbtm3avHmzoqOjXZEHAAAAAK7L6eIlJiZGJ0+edEUWAAAAczDmBbAEpy9S+fbbb2vYsGFatWqVTp06pZSUFIcFAAAAAFzB6c5L06ZNJUkPPfSQw3oG7AMAAKtiqmTAGpwuXlauXOmKHAAAAABwU04XL40aNXJFDgAAAPPQeQEsweni5ZoLFy7oyJEjysjIcFhftWrVOw4FAACQpyheAEtwung5ceKEevbsqSVLllz3fsa8AAAAAHAFp2cbe/7553X27Flt2LBB/v7+Wrp0qT7++GNVrFhRixYtckVGAAAA1zKyXLcAyDVOd15WrFihr776SrVq1ZKHh4ciIiL08MMPKygoSGPHjlXLli1dkdMpM9tGmh0BeeijHVx3yJ08Mv15syMgD7Xp8E+zIwAA8hGnOy/nz59XSEiIJKlw4cI6ceKEJKlKlSrasmVL7qYDAADIAzYjy2ULgNzjdPESFRWlffv2SZKqVaum//znPzp27Jg++ugjlShRItcDAgAAAIB0G6eNPffcc0pKSpIkvf7662revLlmz54tHx8fJSQk5HY+AAAA18uiQwJYgdPFS5cuXez/rlmzpg4fPqy9e/cqPDxcxYoVy9VwAAAAAHDNbV/nRZIMw5C/v79q1KiRW3kAAADynmGYnQBADjg95kWS4uLiFBsbKz8/P/n5+Sk2NlbTp0/P7WwAAAB5g6mSAUtwuvPy2muv6b333tOzzz6revXqSZJ+/PFHvfDCCzpy5IjefPPNXA8JAAAAAE4XL1OmTNG0adPUqVMn+7rHH39cVatW1bPPPkvxAgAALIcpjQFrcPq0scuXL6tWrVrZ1tesWVNXrlzJlVAAAAAA8HdOFy9du3bVlClTsq2fOnWqOnfunCuhAAAA8hRjXgBLuK3ZxuLi4rRs2TLVrVtXkrRhwwYdOXJE3bp105AhQ+zbvffee7mTEgAAAIDbc7p42bVrl31q5IMHD0qSihUrpmLFimnXrl327Ww2Wy5FBAAAcDE6JIAlOF28rFy50hU5AAAAAOCm7ugilQAAAHeFrEyzEwDIgRwVL23atFFCQoKCgoLUpk2bm267YMGCXAkGAACQV4wsThsDrCBHxUtwcLB9DEtwcLBLAwEAAADA9eSoeImPj7/uvwEAAO4KnDYGWILT13lJTEzUgQMHsq0/cOCADh06lBuZAAAAACAbp4uXHj16aP369dnWb9iwQT169MiNTAAAAHkrK9N1C4Bc43TxsnXrVt1///3Z1tetW1fbtm3LjUwAAAAAkI3TUyXbbDalpqZmW3/u3DllZvLtAgAAsB6DzzCAJTjdeWnYsKHGjh3rUKhkZmZq7NixeuCBB3Il1OHDh7V7925lMW0hAAAAgP/P6c7L22+/rYYNGyoqKkoNGjSQJK1Zs0YpKSlasWKFU8eaMWOGzp49qyFDhtjXPf3004qLi5MkRUVF6dtvv1VYWJizMQEAAHKOL0wBS3C68xITE6MdO3aoQ4cOOn78uFJTU9WtWzft3btXsbGxTh1r6tSpKly4sP320qVLFR8fr5kzZ2rjxo0qVKiQRo0a5WxEAAAA5zBgH7AEpzsvklSyZEmNGTPmjh/8wIEDqlWrlv32V199pSeeeEKdO3eWJI0ZM0Y9e/a848cBAAAAYH05Kl527Nih2NhYeXh4aMeOHTfdtmrVqjl+8IsXLyooKMh+e/369erdu7f9drly5ZScnJzj4wEAANwOgw4JYAk5Kl7uvfdeJScnKyQkRPfee69sNpsMw8i2nc1mc2rGsYiICG3evFkRERE6efKkfvnlF4dpmJOTkxUcHJzj4wHuZOHbw9R6+HizYyCPhA9/S0feftnsGMgjzzUspw9W/2Z2DADId3JUvCQmJqp48eL2f+eW7t27a+DAgfrll1+0YsUKRUdHq2bNmvb7169f7/Q4GuBulnTgF13JyFBY5er2db//slVevr4qUSHGxGRwBf8KlWTz8dGF3dvt6wrEVJORka6Lv+41MRlcoVzRAvL29NC+42n2dVEhgcq4kqXE0xdMTOYmGLAPWEKOipeIiAj7vw8fPqz69evLy8tx1ytXrmj9+vUO297KsGHDdOHCBS1YsEChoaGaN2+ew/3r1q3TU089lePjAXe7wKIhWj93qn7bsk6XL13UD7Mm6VJqiuo/1dfsaHCBy6dP6J6neiuoZj15+PkrtGt/eQYG6c/PZpgdDS5w+sJltalaQtVKBsnXy0Md7i2pQF8vLdiRZHY0AMg3bMb1zv+6CU9PTyUlJSkkJMRh/alTpxQSEpLrF6rMzMyUp6enU/ukp57N1QzI3z7acdrsCHnuxy/itfWbz1WjZUfVbdvD7Dh56pHpz5sdIc8Vb9dNxVq208nF83Ri/iyz4+SpyY/90+wIee7xyqF6JDpE3+49rq9/ca9xn5Pa5nzcbG7L+HG+y47tU6+ty44NuBunp0o2DEM2my3b+lOnTikgICBXQknS/v37NXz4cJUuXTrXjglYXdqZk1r20VhlXEhT8YgKSj+fqmUfjVXamZNmR4MLeBUqopL9h8rTP0AXD/0qzwKBKtl/qLwKFTE7Glwg2M9LPe8Ll7+Pp46cuaACPp7qeV+4gv1ua2JQ4IZWrVolm82ms2fP5nifMmXKaOLEiS7LdCM9evRQ69atc/WYhw4dks1m07Zt2+zr1q1bpypVqsjb21utW7e+rZ+Rq5j1s8+vcly8tGnTRm3atJHNZlOPHj3st9u0aaMnnnhCzZo1U/369e8ozIULFxQfH68GDRooJiZGP/zwg8MFLK8nPT1dKSkpDkt6evod5QDyq9STf6pSg+Zq1O1Zefv5q1G3Z1WpQXOlnvzT7GhwAe9i9+js6uVKnjVFWZcuKnnWFJ1dvVzexe4xOxpcoGiAj9YfOq3Pth5T+pUsfbb1mNYfOq2iAT5mR3MP+eQ6Lz169JDNZlP//v2z3Tdw4ED757D85pdfflHbtm1VpkwZ2Wy2HH/YnjZtmqpVq6bAwEAVKlRI1atX19ixY+33f/DBB0pISMjVrGFhYUpKSnIYVz1kyBDde++9SkxMVEJCgurXr6+kpKQ8nTgqISFBhQoVyvXjXivEri3+/v6qXLmypk6delvH+XtB17hxYz3//PO5F/gWcvx1zrUXzzAMFSxYUP7+/vb7fHx8VLduXfXte3vn3f/000+aPn265s2bp/DwcO3Zs0crV65UgwYNbrnv2LFjs13I8uURw/XqP0bcVhYgPytRsXK2dX8dvI+7y8Vf92Rb99fB+7i7/HYq+6D8vw7eh4vlowH7YWFhmjt3rt5//337561Lly7p008/VXh4uMnpru/ChQsqV66c2rdvrxdeeCFH+8yYMUPPP/+8/vWvf6lRo0ZKT0/Xjh07tGvXLvs2rigePD09FRoa6rDu4MGD6t+/v8MZP3/fxur27dunoKAgXbx4UV9//bUGDBig8uXL66GHHjI7miQpIyNDPj63/rImx52X+Ph4zZgxQ927d9f06dMVHx9vX/7zn/9o5MiRKlasmFMhJ0yYoMqVK6tdu3YqXLiwVq9erZ07d8pms6lo0aI5OsbIkSN17tw5h2XYizn7pQGsjGmS3QvTJLsXpkl2bzVq1FBYWJgWLFhgX7dgwQKFh4erenXHL6zS09M1ePBghYSEyM/PTw888IA2btzosM0333yjyMhI+fv768EHH9ShQ4eyPebatWvVoEED+fv7KywsTIMHD9b58+dznLl27dp655139NRTT8nX1zdH+yxatEgdOnRQ7969VaFCBVWuXFmdOnXSW2+9Zd/m76eNpaamqnPnzgoICFCJEiX0/vvvZ/vmv0yZMhozZox69eqlggULKjw83KHL8NfTxq79+9SpU+rVq5dsNpsSEhKu22VYt26dGjdurAIFCqhw4cJq1qyZzpw5I0launSpHnjgARUqVEhFixbVY489poMHD2Z7zAULFujBBx9UgQIFVK1aNf3444+SrnY1evbsqXPnztk7JG+88Ua2n1mvXr302GOPOay7fPmyQkJCFBcXd9Ofd0hIiEJDQ1W2bFkNHjxYZcuW1ZYtW+z3Z2VlaezYsSpbtqz8/f1VrVo1ffHFF/b8Dz74oCSpcOHC9g5gjx499MMPP+iDDz6w5772/tq1a5ceffRRBQYG6p577lHXrl118uT/TnNv3LixBg0apOeff17FihVTs2bNbpr/GqfGvBiGodmzZyspKXdmPhk+fLhat26tw4cP65133lG1atWcPoavr6+CgoIclpz+0gAAAEiSkZnpsuV29OrVS/Hx8fbbM2bMUM+ePbNtN2zYMM2fP18ff/yxtmzZogoVKqhZs2Y6ffrqZDa///672rRpo1atWmnbtm3q06ePRoxwPDvl4MGDat68udq2basdO3bos88+09q1azVo0KDbyp5ToaGh+umnn3T48OEc7zNkyBCtW7dOixYt0vLly7VmzRqHD+DXTJgwQbVq1dLWrVv1zDPPaMCAAdq3b1+27a6dQhYUFKSJEycqKSlJHTt2zLbdtm3b9NBDDykmJkY//vij1q5dq1atWtknqjp//ryGDBmiTZs26fvvv5eHh4eefPJJZf2to/fyyy9r6NCh2rZtmyIjI9WpUydduXJF9evX18SJExUUFKSkpCQlJSVp6NCh2XL06dNHS5cudfgsvnjxYl24cOG6ua/HMAwtXbpUR44cUZ06dezrx44dq5kzZ+qjjz7SL7/8ohdeeEFdunTRDz/8oLCwMM2ff3VSi3379ikpKUkffPCBPvjgA9WrV099+/a15w4LC9PZs2fVpEkTVa9eXZs2bdLSpUv1559/qkOHDg5ZPv74Y/n4+GjdunX66KOPcpTfqVGAHh4eqlixok6dOqWKFSs6s+t1jR49WvHx8Zo1a5Y6deqkrl27cl0XAABwV0lPT882HtfX1/emX7Z26dJFI0eOtH+wX7dunebOnatVq1bZtzl//rymTJmihIQEPfroo5KujiFZvny54uLi9NJLL2nKlCkqX768JkyYIEmKiorSzp079fbbb9uPM3bsWHXu3NnevahYsaL9VK4pU6bIz88vN34M2bz++utq06aNypQpo8jISNWrV08tWrRQu3bt5OGR/fv11NRUffzxx/r000/tpzrFx8erZMmS2bZt0aKFnnnmGUlXvyx///33tXLlSkVFRTlsd+0UMpvNpuDg4BueKjZ+/HjVqlVLH374oX1d5cr/O5W7bVvHGeVmzJih4sWLa/fu3Q6fbYcOHaqWLVtKkkaNGqXKlSvr119/VXR0tIKDg2Wz2W56ulr9+vUVFRWlWbNmadiwYfafQfv27RUYGHjD/STZT4lLT09XVlaW3nzzTTVs2NC+bsyYMfruu+9Ur149SVK5cuW0du1a/ec//1GjRo1UpMjVyWJCQkIcxub4+PioQIECDrknTZqk6tWra8yYMQ4/k7CwMO3fv1+RkZGSrr7Xxo937kwSp2cbGzdunF566SWH8xFv18iRI7V//37NmjVLycnJqlOnjqpVqybDMOxtOAAAAJdz4YD9sWPHKjg42GH566D06ylevLhatmyphIQExcfHq2XLltlOzz948KAuX76s+++/377O29tb9913n/bsuTpmbs+ePQ7frkuyfzi9Zvv27UpISFBgYKB9adasmbKysnL14uR/V6JECf3444/auXOnnnvuOV25ckXdu3dX8+bNs3UsJOm3337T5cuXdd9999nXBQcHZytIJKlq1f9Nu32tIDh+/PhtZ73WebmRAwcOqFOnTipXrpyCgoJUpkwZSdKRI0dumKtEiRKS5HSuPn362Ltyf/75p5YsWaJevXrdcr81a9Zo27Zt2rZtm6ZPn64xY8ZoypQpkqRff/1VFy5c0MMPP+zwPpg5c6bD6W85tX37dq1cudLhWNHR0ZLkcLy/Xpw+p5yef7Fbt266cOGCqlWrJh8fH4eB+5LsbUpnNGrUSI0aNdKkSZP06aefasaMGWrUqJHuu+8+tWvX7pYzjgEAAORXI0eOzPZZJienuPfq1ct+6tbkyZNdkk2S0tLS1K9fPw0ePDjbfXkxQUBsbKxiY2P1zDPPqH///mrQoIF++OEH+xiL2+Ht7e1w22azXbcgyqm/f979u1atWikiIkLTpk1TyZIllZWVpdjYWGVkZNww17VLjzibq1u3bhoxYoR+/PFHrV+/XmXLls3RJFdly5a1d0wqV66sDRs26K233tKAAQOUlnZ1cpD//ve/KlWqlMN+tzMcIy0tTa1atXLo8F1zrWiTdFuXWXG6eHHlPNMFCxZUv3791K9fP+3cuVNxcXEaN24cxQsAAHAtJ6c0dsatThG7kebNmysjI0M2m+26g5nLly9vHy8QEREh6erg7Y0bN9pPAatUqZIWLVrksN9PP/3kcLtGjRravXu3KlSo4HTG3BYTEyNJ150soFy5cvL29tbGjRvtRdW5c+e0f/9+++lPrlK1alV9//332Wa4la5e63Dfvn2aNm2avYhYu3at04/h4+OTo4u9Fy1aVK1bt1Z8fLx+/PHH646FyglPT09dvHhR0tWfu6+vr44cOaJGjRrdMJ+kbBmvl7tGjRqaP3++ypQpIy+v3L1WldNH6969e649eIsWLTRnzhz7NHjjxo1T//79VahQIVWpUkWvvvqqFi9enGuPBwAAYBWenp720788PT2z3R8QEKABAwbopZdeUpEiRRQeHq7x48frwoUL6t27tySpf//+mjBhgl566SX16dNHmzdvznbdlOHDh6tu3boaNGiQ+vTpo4CAAO3evVvLly/XpEmTcpQ1IyNDu3fvtv/72LFj2rZtmwIDA29YFA0YMEAlS5ZUkyZNVLp0aSUlJemf//ynihcvnu3UNunql9zdu3e3P9+QkBC9/vrr8vDwuO4F1HPTyJEjVaVKFXt3yMfHRytXrlT79u1VpEgRFS1aVFOnTlWJEiV05MiRbJMi5ESZMmWUlpam77//XtWqVVOBAgVUoECB627bp08fPfbYY8rMzMzxZ/Pjx4/r0qVLSk9P188//6xZs2apXbt2kq7+bIcOHaoXXnhBWVlZeuCBB3Tu3DmtW7dOQUFB6t69uyIiImSz2bR48WK1aNFC/v7+CgwMVJkyZbRhwwYdOnRIgYGBKlKkiAYOHKhp06apU6dOGjZsmIoUKaJff/1Vc+fO1fTp06/7fs4pp8e8/NWlS5eyXSDSGd9++63DALYxY8Y4nHZ25coVl55rCQAAIElGVpbLljtxbSbVGxk3bpzatm2rrl27qkaNGvr111/17bffqnDhwpKunvY1f/58LVy4UNWqVdNHH33kMIhautpV+OGHH7R//341aNBA1atX12uvvXbdgfA38scff6h69eqqXr26kpKS9O6776p69erq06fPDfdp2rSpfvrpJ7Vv316RkZFq27at/Pz89P3339/wkhnvvfee6tWrp8cee0xNmzbV/fffr0qVKrlsUoFrIiMjtWzZMm3fvl333Xef6tWrp6+++kpeXl7y8PDQ3LlztXnzZsXGxuqFF17QO++84/Rj1K9fX/3791fHjh1VvHjxmw5kb9q0qUqUKKFmzZrl+HWKiopSiRIlVKFCBQ0fPlz9+vXTv//9b/v9o0eP1quvvqqxY8eqUqVKat68uf773/+qbNmykqRSpUpp1KhRGjFihO655x77KY1Dhw6Vp6enYmJiVLx4cR05ckQlS5bUunXrlJmZqUceeURVqlTR888/r0KFCl13MgZn2AzDMJzZ4fz58xo+fLg+//xznTp1Ktv9OWl3XePh4aHk5GSFhIRIulr1bd++XeXKlZN0dRBSyZIlnTqmJKWnnnVqe1jbRzucH2cF63pk+vNmR0AemvzYP82OgDw0qW3VW2/kIpeW3fwaGXfC75HeLju2uzt//rxKlSqlCRMm2LtN7iAtLU2lSpVSfHy82rRpY3acPOV06TNs2DCtWLFCU6ZMka+vr6ZPn65Ro0apZMmSmjlzpisyAgAAuJYLZxtD7tm6davmzJmjgwcPasuWLercubMk6YknnjA5Wd7IysrS8ePHNXr0aBUqVEiPP/642ZHynNNjXr7++mvNnDlTjRs3Vs+ePdWgQQNVqFBBERERmj17tv1NlBPXrsT593UAAAB5iiLDMt59913t27dPPj4+qlmzptasWZNtGum71ZEjR1S2bFmVLl1aCQkJuT4Y3gqcfsanT5+2n9YVFBRkH6PywAMPaMCAAU4dyzAM9ejRwz4Dx6VLl9S/f3/7tGl/v6ATAAAA3Ff16tW1efNms2OYpkyZMnJyxMddx+nipVy5ckpMTFR4eLiio6P1+eef67777tPXX3/tcLXNnPj77AhdunTJtk23bt2cjQgAAOCUOx1YDyBvOF289OzZU9u3b1ejRo00YsQItWrVSpMmTdLly5f13nvvOXWsa1cHBQAAAIBbcbp4eeGFF+z/btq0qfbu3avNmzerQoUKqlrVvFlCAAAAbhtjXgBLyHHxkpWVpXfeeUeLFi1SRkaGHnroIb3++uuKiIiwX9UVAAAAAFwlx1Mlv/XWW/rHP/6hwMBAlSpVSh988IEGDhzoymwAAAB5g6mSAUvIcfEyc+ZMffjhh/r222+1cOFCff3115o9e7ayGOAGAAAAIA/k+LSxI0eOqEWLFvbbTZs2lc1m0x9//KHSpUu7JBwAAEBeMDLpkABWkOPi5cqVK/Lz83NY5+3trcuXL+d6KAAAgDzFmSSAJeS4ePn7BSWl7BeVlKQFCxbkbkIAAAAAkBPFy98vKCld/6KSAAAAlsPAesAScly8cEFJAAAAAGZy+iKVAAAAdxuDzgtgCTmeKhkAAAAAzETnBQAAuD2D2cYAS6DzAgAAAMAS6LwAAAC3Z2TSeQGsgOIFAAC4PYoXwBo4bQwAAACAJdB5AQAAbo8B+4A10HkBAAAAYAl3ZeelYMMhZkdAHirfuLXZEZCH+rV/1OwIyENzp3xjdgTkoUltq5r22Ix5AayBzgsAAAAAS7grOy8AAADOoPMCWAOdFwAAAACWQOcFAAC4vazMTLMjAMgBihcAAOD2mCoZsAZOGwMAAABgCXReAACA22PAPmANdF4AAAAAWAKdFwAA4PbovADWQOcFAAAAgCXQeQEAAG6P2cYAa6DzAgAAAMAS6LwAAAC3l8WYF8ASKF4AAIDbY8A+YA2cNgYAAADAEui8AAAAt0fnBbAGOi8AAAAALIHOCwAAcHtMlQxYA50XAAAAAJZA5wUAALg9xrwA1kDnBQAAAIAl5OviZcGCBapatarZMQAAwF3OyMxy2QIg95hevPznP/9Ru3bt9H//93/asGGDJGnFihWqXr26unbtqvvvv9/khAAA4G6XlZXlsgVA7jG1eBk3bpyeffZZHTp0SIsWLVKTJk00ZswYde7cWR07dtTRo0c1ZcoUMyMCAAAAyCdMHbAfHx+vadOmqXv37lqzZo0aNWqk9evX69dff1VAQICZ0QAAgBvh9C7AGkztvBw5ckRNmjSRJDVo0EDe3t4aNWoUhQsAAACAbEwtXtLT0+Xn52e/7ePjoyJFipiYyNqWTxtmdgTkoY8H1jc7AvJQ70nzzI6APPTVmP8zO4LbMTIzXbYAyD2mX+fl1VdfVYECBSRJGRkZ+uc//6ng4GCHbd577z0zollCvWoV5O/noxUbdtvXNakTowuX0vXT9oMmJoMrVC9bRH7eHvpx/0n7unqRxXQxI1PbDp0xMRlcYetvfyj98hXVjQq3r/tp3xH5+Xjp3rIlTUwGV7ivUin5+Xhp9fbD9nUNq0XoYvoVbdx7zMRkAJB/mFq8NGzYUPv27bPfrl+/vn777TeHbWw2W17HspTfk0/pnRefUusmNVSwgJ/+NbKLihUqqGHvzTU7Glwg6cwFDX8iVg9XKaEAXy+91raKCgX6aPxXv5gdDS5QonBBvfvVan2/81edT7+st75YobNpF/XiEw3NjgYXOHoiRaN7P6TH6kcp0N9H4wc8oiJB/notboXZ0dyCwaxggCWYWrysWrXKzIe/Kxz984w6DZuifz7bVjViymj5j7s0eOwnZseCiySfvaQXPt6kF1pWUuWwQlq797jenL/T7FhwkdDCBfVuj5b6YPFa7Tl6XPWjI/RyuyZmx4KL/HEyVb3fXqhXujVStQqhWrHlNw2bsszsWACQr5h+nRfcmZLFC+mTcf0UXLCAtuw+pEJBAfpkXD+VLF7I7GhwgZBgP03oWlNB/t765fezCi7gowldayok2O/WO8Ny/jybpmEzv1HqxXRVKh2ilAuXNGzmN/rzbJrZ0eACoUUCNe2lxxUc4KvtvyarUKCfpr30uEKLBJodzS1wkUrAGkztvLRp0+a664ODgxUZGak+ffqoePHiNz1Genq60tPTHdYZWZmyeXjmWs78LKJkMc34co1WbNit5dOG6dkxs9SkTowiShbTHyfOmh0PuaxUkQL6YsNh/bj/pD4eWF+jvtihepHFVKpIAR0/d8nseMhlf5xOUZs6saobFa7ek+bplfYP6ad9R/TH6RTdU4gPtHeb8HuCNWvZdq3eflhfjfk/vTRlmRpWi1D4PcFKPk3B6moUGYA1mFq8/H1g/jVnz57VtGnT9M4772j16tWKjY294THGjh2rUaNGOazzuOdeeZaonqtZ86sft/+abd1fB+/j7rI18XS2dX8dvI+7S/Vy2Qfl/3XwPu4uP+/JPij/r4P3AQCSzTAMw+wQ15OVlaW+ffvq+PHj+vrrr2+43fU6L0UbPOs2nRdI5Ru3NjsC8tDWh5l1yZ2UnnLO7AjIQye/HmHaY+97+vpng+SGqKkLXHZswN2YPlXyjXh4eGjw4MF69NFHb7qdr6+vfH19HdZRuAAAAAB3n3xbvEhSQECALly4YHYMAABwl2OqZMAa8vVsY8uXL1dkZKTZMQAAAADkA6Z2XhYtWnTd9efOndPmzZs1ffp0TZ8+PY9TAQAAd8NsY4A1mFq8tG7d+rrrCxYsqKioKE2fPl1PPfVU3oYCAAAAkC+ZWrxkcX4pAADIB4zMfDn5KoC/MXXMS4sWLXTu3P+mwRw3bpzOnj1rv33q1CnFxMSYkAwAALiTrMwsly0Aco+pxcvSpUsdrtEyZswYnT79v4vwXblyRfv27TMjGgAAAIB8Jl9NlZxPr5cJAADuckYWn0EAK8jXUyUDAAAAwDWmdl5sNptsNlu2dQAAAHkpiwH7gCWYWrwYhqEePXrI19dXknTp0iX1799fAQEBkuQwHgYAAACAezO1eOnevbvD7S5dumTbplu3bnkVBwAAuCkuUglYg6nFS3x8vJkPDwAAAMBC8tVsYwAAAGbgIpWANTDbGAAAAABLoPMCAADcHrONAdZA8QIAANweA/YBa+C0MQAAAACWQOcFAAC4vawsThsDrIDOCwAAAABLoPMCAADcHlMlA9ZA5wUAAACAJdB5AQAAbi+L2cYAS6DzAgAAAMAS6LwAAAC3x5gXwBooXgAAgNujeAGsgdPGAAAAAFgCnRcAAOD2GLAPWAOdFwAAAACWQOcFAAC4PSOLMS+AFdyVxcvMhNFmR0AeKuDtaXYE5KHfihQwOwLy0Ir/8IESAPA/d2XxAgAA4IwsZhsDLIExLwAAAAAsgc4LAABwewazjQGWQPECAADcHhepBKyB08YAAAAAWAKdFwAA4PYYsA9YA50XAAAAAJZA5wUAALg9I4sB+4AV0HkBAAAAYAl0XgAAgNtjzAtgDXReAAAAAFgCnRcAAOD2uM4LYA0ULwAAwO0ZmQzYB6yA08YAAAAAWAKdFwAA4PYYsA9YA50XAAAAAJZA5wUAALg9BuwD1kDnBQAAAIAl0HkBAABuL8ug8wJYAZ0XAAAAAJaQ7zovhw8f1vnz5xUdHS0PD2orAADgepl0XgBLMK06mDFjht577z2HdU8//bTKlSunKlWqKDY2Vr///rtJ6QAAgDvJNFy3AMg9phUvU6dOVeHChe23ly5dqvj4eM2cOVMbN25UoUKFNGrUKLPiAQAAAMhnTDtt7MCBA6pVq5b99ldffaUnnnhCnTt3liSNGTNGPXv2NCseAABwI5w2BliDaZ2XixcvKigoyH57/fr1atiwof12uXLllJycbEY0AAAAAPmQacVLRESENm/eLEk6efKkfvnlF91///32+5OTkxUcHGxWPAAA4EYY8wJYg2nFS/fu3TVw4ECNHj1a7du3V3R0tGrWrGm/f/369YqNjTUrniXFv/GC2RGQh6a88pzZEQAAAPKUaWNehg0bpgsXLmjBggUKDQ3VvHnzHO5ft26dnnrqKZPSWceRvbt0OSNd5av+r/A7uGOzvH19FR5F8Xe3SdyzU5cz0hVZ7X/jxfZv3yQfXz+Vieb1BoDbxZgXwBpshpF/f1szMzPl6enp9H5ztx9zQZr86dzJ41o680MFBBXS0QN7VLpiJV1IOadm3QYouFiI2fHyRAFv598jVnX25HF9PWOSAoIL6cj+PQqPrKTzKefUqudAFXKT17tCkQJmR0AeupKVb/+LggtULWne6eKzilVy2bG7ntzjsmMD7iZfXgVy//79Gj58uEqXLm12lHwvuFiIOg55Q74FApSUeEB+BQLVYcjrblO4uJtCxULUddib8isQqGO/7Zd/QKC6vjTKbQoXAHAVxrwA1pBvipcLFy4oPj5eDRo0UExMjH744QcNGTLE7Fj5XsrpE5o3cbQunU9TibIVdfF8quZNHK2U0yfMjgYXOHfqhD6ZMEoXz6eqVLlIXUhL1ScTRuncKV5vALgTmYbhsgVA7jFtzMs1P/30k6ZPn6558+YpPDxce/bs0cqVK9WgQYMc7Z+enq709HSHdZcz0uXt4+uKuPnO2eN/qkaTFipftabi33hBrfq+oIM7Nuvs8T8VVKS42fGQy04fT9Z9TVsqslotTXnlObXt/6L2b9+k08eTFVyU1xsAANzdTOu8TJgwQZUrV1a7du1UuHBhrV69Wjt37pTNZlPRokVzfJyxY8cqODjYYfkqbpILk+cv4dGxDoP1Jal81ZoKZ/D2XalspSoOg/UlKbJaLZWtVMWkRABwd+C0McAaTBuw7+XlpeHDh+vNN990GJTv7e2t7du3KyYmJkfHuV7n5at9J92m8wL3GrAPBuy7GwbsuxczB+xPLRztsmM/fWavy44NuBvTOi+jR4/WvHnzVLZsWQ0fPly7du26reP4+voqKCjIYaFwAQAAzqDzAliDacXLyJEjtX//fs2aNUvJycmqU6eOqlWrJsMwdObMGbNiAQAAAMinTJ9trFGjRvr444+VnJysZ555RjVr1lSjRo1Uv359vffee2bHAwAAboDZxgBrML14uaZgwYLq16+fNmzYoG3btqlOnToaN26c2bEAAAAA5BOmFS8rVqxQTEyMUlJSst0XFhamb7/9Vp9++qkJyQAAgLthzAtgDaYVLxMnTlTfvn0VFBSU7b7g4GD1799fkydPNiEZAABwN5w2BliDacXL9u3b1bx58xve/8gjj2jz5s15mAgAAABAfuZl1gP/+eef8vb2vuH9Xl5eOnHiRB4mAgAA7orTuwBrMK3zUqpUqZte22XHjh0qUaJEHiYCAAAAkJ+ZVry0aNFCr776qi5dupTtvosXL+r111/XY489ZkIyAADgbhjzAliDaaeNvfLKK1qwYIEiIyM1aNAgRUVFSZL27t2ryZMnKzMzUy+//LJZ8QAAAADkM6YVL/fcc4/Wr1+vAQMGaOTIkTL+/zcTNptNzZo10+TJk3XPPfeYFQ8AALgRxrwA1mBa8SJJERER+uabb3TmzBn9+uuvMgxDFStWVOHChc2MBQAAACAfMrV4uaZw4cKqXbu22TEAAICbYmwKYA35ongBAAAwU5bZAQDkiGmzjQEAAACAM+i8AAAAt8dpY4A10HkBAAAAYAl0XgAAgNtjqmTAGui8AAAAALAEOi8AAMDtMeYFsAY6LwAAAAAsgc4LAABwe4x5AayBzgsAAAAAS6DzAgAA3B5jXgBroHgBAABuj9PGAGvgtDEAAAAAlkDnBQAAuD1OGwOsgc4LAAAAAEug8wIAANweY14Aa6DzAgAAAMASbIbBSZ53g/T0dI0dO1YjR46Ur6+v2XHgYrze7oXX273wegPAjVG83CVSUlIUHBysc+fOKSgoyOw4cDFeb/fC6+1eeL0B4MY4bQwAAACAJVC8AAAAALAEihcAAAAAlkDxcpfw9fXV66+/zuBON8Hr7V54vd0LrzcA3BgD9gEAAABYAp0XAAAAAJZA8QIAAADAEiheAAAAAFgCxQsAAAAAS6B4yed+/PFHeXp6qmXLljfcZs6cOfL09NTAgQOve39KSopefvllRUdHy8/PT6GhoWratKkWLFgg5mvIP3r06CGbzZZtad68ucN2Y8eOlaenp955551sx8jMzNS4ceMUHR0tf39/FSlSRHXq1NH06dPz6mkgB271WttsNi1cuPC6+7Vu3dp+u3HjxvZ9/fz8FBkZqbFjx/J7nQ/d6G/5oUOHZLPZ5OnpqWPHjjncl5SUJC8vL9lsNh06dMhh+2tLkSJF1KhRI61ZsyavngoAmIriJZ+Li4vTs88+q9WrV+uPP/644TbDhg3TnDlzdOnSJYf7zp49q/r162vmzJkaOXKktmzZotWrV6tjx44aNmyYzp07lxdPAznUvHlzJSUlOSxz5sxx2GbGjBkaNmyYZsyYkW3/UaNG6f3339fo0aO1e/durVy5Uk8//bTOnj2bR88AOZWT1zon+vbtq6SkJO3bt08jR47Ua6+9po8++sgFiXEnbvW3vFSpUpo5c6bDuo8//lilSpW67vG+++47JSUlafXq1SpZsqQee+wx/fnnny7JDgD5CcVLPpaWlqbPPvtMAwYMUMuWLZWQkJBtm8TERK1fv14jRoxQZGSkFixY4HD/P/7xDx06dEgbNmxQ9+7dFRMTo8jISPXt21fbtm1TYGBgHj0b5ISvr69CQ0MdlsKFC9vv/+GHH3Tx4kW9+eabSklJ0fr16x32X7RokZ555hm1b99eZcuWVbVq1dS7d28NHTo0r58KbuFWr3VOFShQQKGhoYqIiFDPnj1VtWpVLV++3AWJcbty8re8e/fuio+Pd1gXHx+v7t27X/eYRYsWVWhoqGJjY/WPf/xDKSkp2rBhgyviA0C+QvGSj33++eeKjo5WVFSUunTpohkzZmQ7HSQ+Pl4tW7ZUcHCwunTpori4OPt9WVlZmjt3rjp37qySJUtmO35gYKC8vLxc/jyQe+Li4tSpUyd5e3urU6dODq+3JIWGhmrFihU6ceKESQlhFsMwtGbNGu3du1c+Pj5mx8Ff5ORv+eOPP64zZ85o7dq1kqS1a9fqzJkzatWq1U2PffHiRXvHhtcdgDugeMnH4uLi1KVLF0lXTzE5d+6cfvjhB/v9WVlZSkhIsG/z1FNPae3atUpMTJQknTx5UmfOnFF0dHTeh8dtWbx4sQIDAx2WMWPGSLo6dumLL76wv95dunTR559/rrS0NPv+7733nk6cOKHQ0FBVrVpV/fv315IlS0x5Lri5m73Wzvjwww8VGBgoX19fNWzYUFlZWRo8eLALEuN23epvuSR5e3vbCxvp6umhXbp0kbe393WPWb9+fQUGBiogIEDvvvuuatasqYceesi1TwQA8gGKl3xq3759+vnnn9WpUydJkpeXlzp27OjwTfvy5ct1/vx5tWjRQpJUrFgxPfzww/b//Bi0az0PPvigtm3b5rD0799f0tWJGcqXL69q1apJku69915FRETos88+s+8fExOjXbt26aefflKvXr10/PhxtWrVSn369DHl+eDGbvZaO6Nz587atm2b1q1bp0cffVQvv/yy6tev74LEuB05+Vt+Ta9evTRv3jwlJydr3rx56tWr1w2P+9lnn2nr1q2aP3++KlSooISEhBsWOgBwN+GcoXwqLi5OV65ccTjdyzAM+fr6atKkSQoODlZcXJxOnz4tf39/+zZZWVnasWOHRo0apeLFi6tQoULau3evGU8BtyEgIEAVKlS47n1xcXH65ZdfHE71y8rK0owZM9S7d2/7Og8PD9WuXVu1a9fW888/r08++URdu3bVyy+/rLJly7r8OSBnbvZaFyxY8LqTaZw9e1bBwcEO64KDg+3H+fzzz1WhQgXVrVtXTZs2zf3QcNqt/pb/VZUqVRQdHa1OnTqpUqVKio2N1bZt26573LCwMFWsWFEVK1bUlStX9OSTT2rXrl3y9fV15dMBANPRecmHrly5opkzZ2rChAkO38pu375dJUuW1Jw5c3Tq1Cl99dVXmjt3rsM2W7du1ZkzZ7Rs2TJ5eHjoqaee0uzZs687u01aWpquXLliwjOEs3bu3KlNmzZp1apVDq/3qlWr9OOPP960QI2JiZEknT9/Pq/i4g5FRUVp8+bNDusyMzO1fft2RUZG3nC/wMBAPffccxo6dCid13wgJ3/L/65Xr15atWrVTbsuf9euXTt5eXnpww8/zM34AJAv0XnJhxYvXqwzZ86od+/e2b5lbdu2reLi4nTp0iUVLVpUHTp0kM1mc9imRYsWiouLU/PmzfXWW29p1apVqlOnjt566y3VqlVL3t7eWrNmjcaOHauNGzeqUKFCefjscDPp6elKTk52WOfl5aW4uDjdd999atiwYbZ9ateurbi4OL3zzjtq166d7r//ftWvX1+hoaFKTEzUyJEjFRkZydinfOZGr3WxYsU0ZMgQ9e7dW9HR0Xr44Yd1/vx5/fvf/9aZM2dueQpgv379NHr0aM2fP1/t2rVz5VPALeTkb/nfr+PUt29ftW/f3qm/yzabTYMHD9Ybb7yhfv36qUCBArkRHwDyJwP5zmOPPWa0aNHiuvdt2LDBkGTYbDbjmWeeue42n332meHj42OcOHHCMAzDOHv2rDFixAijYsWKho+Pj3HPPfcYTZs2Nb788ksjKyvLZc8DzunevbshKdtStmxZo2jRosb48eOvu9/bb79thISEGBkZGcbUqVONBx980ChevLjh4+NjhIeHGz169DAOHTqUx88GN3Oj1zoqKsq+zezZs42aNWsaBQsWNO655x6jRYsWxvbt2x2O06hRI+O5557Ldvx+/foZlStXNjIzM139VHATOflbvn37dkOSsXXr1utut3XrVkOSkZiYaBiGYSQmJl53+/PnzxuFCxc23n777Vx8BgCQ/9gMg3MLAAAAAOR/jHkBAAAAYAkULwAAAAAsgeIFAAAAgCVQvAAAAACwBIoXAAAAAJZA8QIAAADAEiheAAAAAFgCxQuQz6xatUo2m01nz57N8T5lypTRxIkTXZYJAAAgP6B4AZzQo0cP2Ww29e/fP9t9AwcOlM1mU48ePfI+2C388ssvatu2rcqUKSObzZajQudaEXVt8ff3V+XKlTV16lTXB76Fv+YKCAhQxYoV1aNHD23evNnpYzVu3FjPP/98ruZr1qyZPD09tXHjxlw9bm5wxfMFACCvULwATgoLC9PcuXN18eJF+7pLly7p008/VXh4uInJbuzChQsqV66cxo0bp9DQUKf23bdvn5KSkrR7927169dPAwYM0Pfff++ipDkXHx+vpKQk/fLLL5o8ebLS0tJUp04dzZw509RcR44c0fr16zVo0CDNmDHD1Cy3yzAMXblyxewYAABkQ/ECOKlGjRoKCwvTggUL7OsWLFig8PBwVa9e3WHb9PR0DR48WCEhIfLz89MDDzyQ7dv4b775RpGRkfL399eDDz6oQ4cOZXvMtWvXqkGDBvL391dYWJgGDx6s8+fP5zhz7dq19c477+ipp56Sr6+vU883JCREoaGhKlu2rAYPHqyyZctqy5Yt9vuXLl2qBx54QIUKFVLRokX12GOP6eDBg/b7MzIyNGjQIJUoUUJ+fn6KiIjQ2LFj7fefPXtWffr0UfHixRUUFKQmTZpo+/btt8xVqFAhhYaGqkyZMnrkkUf0xRdfqHPnzho0aJDOnDkjSTp16pQ6deqkUqVKqUCBAqpSpYrmzJljP0aPHj30ww8/6IMPPrB3cg4dOqTMzEz17t1bZcuWlb+/v6KiovTBBx/k6OcVHx+vxx57TAMGDNCcOXMcilzpaudj8ODBGjZsmIoUKaLQ0FC98cYbDtucPXtW/fr10z333CM/Pz/FxsZq8eLF9vtv9X748MMPVbFiRfn5+emee+5Ru3btbvp8r3XZlixZopo1a8rX11effPKJPDw8tGnTJodsEydOVEREhLKysnL08wAAIDdRvAC3oVevXoqPj7ffnjFjhnr27Jltu2HDhmn+/Pn6+OOPtWXLFlWoUEHNmjXT6dOnJUm///672rRpo1atWmnbtm3q06ePRowY4XCMgwcPqnnz5mrbtq127Nihzz77TGvXrtWgQYNc+yT/xjAMLV26VEeOHFGdOnXs68+fP68hQ4Zo06ZN+v777+Xh4aEnn3zS/uH2X//6lxYtWqTPP/9c+/bt0+zZs1WmTBn7/u3bt9fx48e1ZMkSbd68WTVq1NBDDz1k/xk544UXXlBqaqqWL18u6WpHrGbNmvrvf/+rXbt26emnn1bXrl31888/S5I++OAD1atXT3379lVSUpKSkpIUFhamrKwslS5dWvPmzdPu3bv12muv6R//+Ic+//zzW/6M4uPj1aVLF0VHR6tChQr64osvsm338ccfKyAgQBs2bND48eP15ptv2jNnZWXp0Ucf1bp16/TJJ59o9+7dGjdunDw9PSXd+v2wadMmDR48WG+++ab27dunpUuXqmHDhjd9vteMGDFC48aN0549e/T444+radOmDu9z6Wpx1qNHD3l48N8HAMAEBoAc6969u/HEE08Yx48fN3x9fY1Dhw4Zhw4dMvz8/IwTJ04YTzzxhNG9e3fDMAwjLS3N8Pb2NmbPnm3fPyMjwyhZsqQxfvx4wzAMY+TIkUZMTIzDYwwfPtyQZJw5c8YwDMPo3bu38fTTTztss2bNGsPDw8O4ePGiYRiGERERYbz//vs5eg453XblypWGJCMgIMAICAgwvLy8DA8PD+Of//znTfc7ceKEIcnYuXOnYRiG8eyzzxpNmjQxsrKysm27Zs0aIygoyLh06ZLD+vLlyxv/+c9/bvgYkowvv/wy2/qLFy8akoy33377hvu2bNnSePHFF+23GzVqZDz33HM3fU6GYRgDBw402rZte9Ntli1bZhQvXty4fPmyYRiG8f777xuNGjVy2KZRo0bGAw884LCudu3axvDhww3DMIxvv/3W8PDwMPbt23fdx7jV+2H+/PlGUFCQkZKSct39r/d8r73WCxcudFj/2WefGYULF7a/Pps3bzZsNpuRmJh4w58BAACuxFdnwG0oXry4WrZsqYSEBMXHx6tly5YqVqyYwzYHDx7U5cuXdf/999vXeXt767777tOePXskSXv27HHoYkhSvXr1HG5v375dCQkJCgwMtC/NmjVTVlaWEhMTXfQM/2fNmjXatm2btm3bpunTp2vMmDGaMmWK/f4DBw6oU6dOKleunIKCguxdlSNHjki6eqrStm3bFBUVpcGDB2vZsmUOzy0tLU1FixZ1eH6JiYkOp57llGEYkq4O6JekzMxMjR49WlWqVFGRIkUUGBiob7/91p7tZiZPnqyaNWuqePHiCgwM1NSpU2+534wZM9SxY0d5eXlJkjp16qR169Zley5Vq1Z1uF2iRAkdP35ckrRt2zaVLl1akZGR132MW70fHn74YUVERKhcuXLq2rWrZs+erQsXLtzy+UpSrVq1HG63bt1anp6e+vLLLyVJCQkJevDBBx06ZwAA5CUvswMAVtWrVy/7qTqTJ0922eOkpaWpX79+Gjx4cLb78mKCgLJly6pQoUKSpMqVK2vDhg166623NGDAAElSq1atFBERoWnTpqlkyZLKyspSbGysMjIyJF0dI5SYmKglS5bou+++U4cOHdS0aVN98cUXSktLU4kSJbRq1apsj3vtMZ1xrSgsW7asJOmdd97RBx98oIkTJ6pKlSoKCAjQ888/b892I3PnztXQoUM1YcIE1atXTwULFtQ777yjDRs23HCf06dP68svv9Tly5cdirvMzEzNmDFDb731ln2dt7e3w742m81+mp2/v/9Ns93q/eDj46MtW7Zo1apVWrZsmV577TW98cYb2rhx4y1/pgEBAQ63fXx81K1bN8XHx6tNmzb69NNPczz2BwAAV6B4AW5T8+bNlZGRIZvNpmbNmmW7v3z58vLx8dG6desUEREhSbp8+bI2btxon6q2UqVKWrRokcN+P/30k8PtGjVqaPfu3apQoYJrnoiTPD097YPQT506pX379mnatGlq0KCBpKuDyf8uKChIHTt2VMeOHdWuXTs1b95cp0+fVo0aNZScnCwvL69c+TZ/4sSJCgoKUtOmTSVJ69at0xNPPKEuXbpIujqeZP/+/YqJibHv4+Pjo8zMTIfjrFu3TvXr19czzzxjX3erTtDs2bNVunRpLVy40GH9smXLNGHCBL355pv2cSs3U7VqVR09elT79++/bvclJ+8HLy8vNW3aVE2bNtXrr7+uQoUKacWKFWrTps11n+/N9OnTR7Gxsfrwww915coVtWnTJsf7AgCQ2yhegNvk6elp/6b/eh9KAwICNGDAAL300ksqUqSIwsPDNX78eF24cEG9e/eWJPXv318TJkzQSy+9pD59+mjz5s1KSEhwOM7w4cNVt25dDRo0SH369FFAQIB2796t5cuXa9KkSTnKmpGRod27d9v/fezYMW3btk2BgYG3LIqOHz+uS5cuKT09XT///LNmzZpln72qcOHCKlq0qKZOnaoSJUroyJEj2SYceO+991SiRAlVr15dHh4emjdvnkJDQ1WoUCE1bdpU9erVU+vWrTV+/HhFRkbqjz/+0H//+189+eST2U5j+quzZ88qOTlZ6enp2r9/v/7zn/9o4cKFmjlzpr3DULFiRX3xxRdav369ChcurPfee09//vmnQ/FSpkwZbdiwQYcOHVJgYKCKFCmiihUraubMmfr2229VtmxZzZo1Sxs3brR3dK4nLi5O7dq1U2xsrMP6sLAwjRw5UkuXLlXLli1v+rOWpEaNGqlhw4Zq27at3nvvPVWoUEF79+6VzWZT8+bNb/l+WLx4sX777Tc1bNhQhQsX1jfffKOsrCxFRUXd8PneTKVKlVS3bl0NHz5cvXr1umVnCAAAlzJ70A1gJdcG7N/IXwfsG8bVAeTPPvusUaxYMcPX19e4//77jZ9//tlhn6+//tqoUKGC4evrazRo0MCYMWOGw4B9wzCMn3/+2Xj44YeNwMBAIyAgwKhatarx1ltv2e+/1SD8xMREQ1K25e+Dyf/q2iDua4uXl5dRtmxZY+jQoUZaWpp9u+XLlxuVKlUyfH19japVqxqrVq1yGFA/depU49577zUCAgKMoKAg46GHHjK2bNli3z8lJcV49tlnjZIlSxre3t5GWFiY0blzZ+PIkSM3zPbXXH5+fkb58uWN7t27G5s3b3bY7tSpU8YTTzxhBAYGGiEhIcYrr7xidOvWzeE13Ldvn1G3bl3D39/fkGQkJiYaly5dMnr06GEEBwcbhQoVMgYMGGCMGDHCqFat2nXzbNq0yZCU7bW95tFHHzWefPJJwzCuP2D+7++bU6dOGT179jSKFi1q+Pn5GbGxscbixYvt99/s/bBmzRqjUaNGRuHChQ1/f3+jatWqxmeffXbT53vttf7re+6v4uLibvr8AADIKzbD+P8jXAEAuI7Ro0dr3rx52rFjh9lRAABujtnGAADXlZaWpl27dmnSpEl69tlnzY4DAADFCwDg+gYNGqSaNWuqcePG6tWrl9lxAAAQp40BAAAAsAQ6LwAAAAAsgeIFAAAAgCVQvAAAAACwBIoXAAAAAJZA8QIAAADAEiheAAAAAFgCxQsAAAAAS6B4AQAAAGAJFC8AAAAALOH/AS/1cUexReAgAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 800x600 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "compare_rocs(\n",
    "    f\"{RESULTS_DIR}/m1_m2_comparison.txt\", \n",
    "    f\"{RESULTS_DIR}/figure5.png\", \n",
    "    \"Model 1\", \n",
    "    \"Model 2\", \n",
    "    \"Model 1 Base Data Ancestry\", \n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b14a8b5-eeb0-4692-8db0-a9652f71135e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Density plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "390a4fab-7244-473d-962c-a70722eb23e2",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "model2_full <- fread(paste0(RESULTS_DIR, \"/model2_risk_results.txt\"), select = c(\"CASE\", \"probDisease\", \"Group\", \"Group.name\"))\n",
    "for (ancestry in ancestries) {\n",
    "    model1_to_plot <- fread(paste0(RESULTS_DIR, \"/\", ancestry, \"/model1_risk_results.txt\"), select = c(\"CASE\", \"probDisease\", \"Group\"))\n",
    "    model2_to_plot <- subset(model2_full, Group.name == paste0(\"Model 2 \", ancestry))\n",
    "    model2_to_plot[,Group.name:=NULL]\n",
    "    \n",
    "    to_plot <- rbind(model1_to_plot, model2_to_plot)    \n",
    "    \n",
    "    to_plot_cases <- subset(to_plot, CASE == 1)\n",
    "    density_rocs_plot <- ggplot(to_plot_cases, aes(probDisease, color=Group)) + geom_density(alpha = 0.5) + theme_bw()\n",
    "    ggsave(plot = density_rocs_plot, filename = paste0(RESULTS_DIR, \"/supp_fig5_\", ancestry, \".png\"), width = 8, height = 5, units = \"in\", dpi = 300)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c49ce7b-9732-418c-9edc-3a7a5999f182",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Forest plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "b9df7de1-66ca-4541-89e9-bdb5b377abbf",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "for (ancestry in ancestries) {\n",
    "    level_order <- c('Model 1 (EUR)', 'Model 1 (EAS)', 'Model 1 (AMR)', 'Model 1 (AAC)', 'Model 2')\n",
    "    model1 <- fread(paste0(RESULTS_DIR, \"/\", ancestry, \"/forest_risk_pcs_\", ancestry, \".txt\"))\n",
    "    model1[, COHORT := paste0(\"Model 1 (\", COHORT, \")\")]\n",
    "    model2 <- fread(paste0(RESULTS_DIR, \"/\", ancestry, \"/model2_\", ancestry, \"_best.prsice\"))\n",
    "    model2$COHORT <- 'Model 2'\n",
    "    model2$L95 <- model2$Coefficient - 1.96 * model2$Standard.Error\n",
    "    model2$U95 <- model2$Coefficient + 1.96 * model2$Standard.Error\n",
    "    cols_to_keep <- c(\"COHORT\", \"Coefficient\", \"Standard.Error\", \"P\", \"L95\", \"U95\")\n",
    "    model2 <- model2[, ..cols_to_keep]\n",
    "    colnames(model2) <- c(\"COHORT\", \"BETA\", \"SE\", \"P\", \"L95\", \"U95\")\n",
    "    dat <- rbind(model1, model2)\n",
    "    \n",
    "    plot <- ggplot(\n",
    "        data=dat, \n",
    "        aes(\n",
    "            x = factor(COHORT,levels = level_order),\n",
    "            y = BETA,\n",
    "            ymin = U95,\n",
    "            ymax = L95,\n",
    "            color = COHORT,\n",
    "        ),\n",
    "    ) +\n",
    "    geom_pointrange(\n",
    "        aes(ymin = L95, ymax = U95),\n",
    "        cex = 0.7,\n",
    "    ) +\n",
    "    geom_hline(\n",
    "        yintercept = 0,\n",
    "        linetype = 2\n",
    "    ) +\n",
    "    scale_color_manual(\n",
    "        values = c(\n",
    "            \"Model 2\" = \"#FF0000\", \n",
    "            \"Model 1 (EUR)\" = \"#0000FF\", \n",
    "            \"Model 1 (EAS)\" = \"#0000FF\", \n",
    "            \"Model 1 (AMR)\" = \"#0000FF\", \n",
    "            \"Model 1 (AAC)\" = \"#0000FF\"\n",
    "        ),\n",
    "        guide = \"none\",\n",
    "    ) +\n",
    "    theme(\n",
    "        plot.title = element_text(size = 20, face = \"bold\"),\n",
    "        axis.text.y = element_text(size = 8, face = 'bold'),\n",
    "        axis.ticks.y = element_blank(),\n",
    "        axis.text.x = element_text(face=\"bold\"),\n",
    "        axis.title = element_text(size = 16, face=\"bold\"),\n",
    "        legend.position = \"none\",\n",
    "    ) +\n",
    "    xlab(' ') +\n",
    "    ylab(\"Beta coefficient\") +\n",
    "    coord_flip() +\n",
    "    theme_minimal() +\n",
    "    ggtitle(paste0(ancestry, \" - Individual level data - GP2 release 9\")) +\n",
    "    theme(plot.title = element_text(hjust=0.5))\n",
    "\n",
    "    ggsave(\n",
    "        paste0(RESULTS_DIR, \"/figure3_\", ancestry, \".jpg\"), \n",
    "        plot = plot,\n",
    "        width = 8, \n",
    "        height = 5, \n",
    "        units = \"in\",\n",
    "        dpi = 500,\n",
    "    )\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f507706-3b8d-442c-bd69-6289e03e158d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Estimate SNPs contributing most to PRS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a62122a4-49e6-4ece-9c0e-4883bcae748a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AAC/bestsnps_AAC.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AAC/model1_plink/merged_AAC\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AAC/bestsnps_AAC\n",
      "  --score /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:36:34 2025\n",
      "12984 MiB RAM detected, ~3323 available; reserving 3259 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "1196 samples (698 females, 498 males; 1196 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AAC/model1_plink/merged_AAC.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AAC/model1_plink/merged_AAC.bim.\n",
      "1 binary phenotype loaded (325 cases, 819 controls).\n",
      "--extract: 84 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "84 variants remaining after main filters.\n",
      "--export-allele: 84 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324252526262727282829293030313132323333343435353636373738383939404041414242434344444545464647474848495050515152525353545455555656575758585959606061616262636364646565666667676868696970707171727273737475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AAC/bestsnps_AAC.raw\n",
      "written.\n",
      "--score: 84 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AAC/bestsnps_AAC.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AAC/bestsnps_AAC.sscore .\n",
      "End time: Wed Feb 12 14:36:35 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AFR/bestsnps_AAC.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AFR/model1_plink/merged_AFR\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AFR/bestsnps_AAC\n",
      "  --score /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:36:35 2025\n",
      "12984 MiB RAM detected, ~3327 available; reserving 3263 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "2648 samples (1190 females, 1458 males; 2648 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AFR/model1_plink/merged_AFR.fam.\n",
      "88 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AFR/model1_plink/merged_AFR.bim.\n",
      "1 binary phenotype loaded (954 cases, 1652 controls).\n",
      "--extract: 84 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "84 variants remaining after main filters.\n",
      "--export-allele: 84 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324252526262727282829293030313132323333343435353636373738383939404041414242434344444545464647474848495050515152525353545455555656575758585959606061616262636364646565666667676868696970707171727273737475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AFR/bestsnps_AAC.raw\n",
      "written.\n",
      "--score: 84 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AFR/bestsnps_AAC.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AFR/bestsnps_AAC.sscore .\n",
      "End time: Wed Feb 12 14:36:36 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AJ/bestsnps_AAC.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AJ/model1_plink/merged_AJ\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AJ/bestsnps_AAC\n",
      "  --score /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:36:36 2025\n",
      "12984 MiB RAM detected, ~3324 available; reserving 3260 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "3126 samples (1168 females, 1958 males; 3126 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AJ/model1_plink/merged_AJ.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AJ/model1_plink/merged_AJ.bim.\n",
      "1 binary phenotype loaded (1396 cases, 423 controls).\n",
      "--extract: 84 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "84 variants remaining after main filters.\n",
      "--export-allele: 84 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 10101111121213131414151516161717181819192020212122222323242425252626272728282929303031313232333334343535363637373838393940404141424243434444454546464747484849505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AJ/bestsnps_AAC.raw\n",
      "written.\n",
      "--score: 84 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AJ/bestsnps_AAC.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AJ/bestsnps_AAC.sscore .\n",
      "End time: Wed Feb 12 14:36:37 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AMR/bestsnps_AAC.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AMR/model1_plink/merged_AMR\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AMR/bestsnps_AAC\n",
      "  --score /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:36:37 2025\n",
      "12984 MiB RAM detected, ~3305 available; reserving 3241 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "3496 samples (1835 females, 1661 males; 3496 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AMR/model1_plink/merged_AMR.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AMR/model1_plink/merged_AMR.bim.\n",
      "1 binary phenotype loaded (1928 cases, 1430 controls).\n",
      "--extract: 84 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "84 variants remaining after main filters.\n",
      "--export-allele: 84 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324252526262727282829293030313132323333343435353636373738383939404041414242434344444545464647474848495050515152525353545455555656575758585959606061616262636364646565666667676868696970707171727273737475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AMR/bestsnps_AAC.raw\n",
      "written.\n",
      "--score: 84 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AMR/bestsnps_AAC.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AMR/bestsnps_AAC.sscore .\n",
      "End time: Wed Feb 12 14:36:38 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS/bestsnps_AAC.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/EAS/model1_plink/merged_EAS\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS/bestsnps_AAC\n",
      "  --score /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:36:38 2025\n",
      "12984 MiB RAM detected, ~3307 available; reserving 3243 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EAS/model1_plink/merged_EAS.fam.\n",
      "87 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EAS/model1_plink/merged_EAS.bim.\n",
      "1 binary phenotype loaded (1819 cases, 2349 controls).\n",
      "--extract: 83 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "83 variants remaining after main filters.\n",
      "--export-allele: 83 variants updated, 1 variant ID not present.\n",
      "--export A pass 1/1: loading... 0writing... 10101112121313141415161617171818192020212122222324242525262627282829293030313232333334343536363737383839404041414242434444454546464748484949505051525253535454555656575758585960606161626263646465656666676868696970707172727373747475767677777878798080818182828384848585868687888889899090919292939394949596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/EAS/bestsnps_AAC.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 1 entry in\n",
      "/home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt was skipped\n",
      "due to a missing variant ID.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 83 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/EAS/bestsnps_AAC.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/EAS/bestsnps_AAC.sscore .\n",
      "End time: Wed Feb 12 14:36:39 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EUR/bestsnps_AAC.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/EUR/model1_plink/merged_EUR\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/EUR/bestsnps_AAC\n",
      "  --score /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:36:39 2025\n",
      "12984 MiB RAM detected, ~3308 available; reserving 3244 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "48909 samples (21569 females, 27340 males; 48909 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EUR/model1_plink/merged_EUR.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EUR/model1_plink/merged_EUR.bim.\n",
      "1 binary phenotype loaded (22093 cases, 9706 controls).\n",
      "--extract: 84 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "84 variants remaining after main filters.\n",
      "--export-allele: 84 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324242525262627272828292930303131323233333434353536363737383839394040414142424343444445454646474748484949505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/EUR/bestsnps_AAC.raw\n",
      "written.\n",
      "--score: 84 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/EUR/bestsnps_AAC.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/EUR/bestsnps_AAC.sscore .\n",
      "End time: Wed Feb 12 14:36:41 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/CAS/bestsnps_AAC.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/CAS/model1_plink/merged_CAS\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/CAS/bestsnps_AAC\n",
      "  --score /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:36:41 2025\n",
      "12984 MiB RAM detected, ~3281 available; reserving 3217 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "1003 samples (545 females, 458 males; 1003 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/CAS/model1_plink/merged_CAS.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/CAS/model1_plink/merged_CAS.bim.\n",
      "1 binary phenotype loaded (582 cases, 323 controls).\n",
      "--extract: 84 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "84 variants remaining after main filters.\n",
      "--export-allele: 84 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324242525262627272828292930303131323233333434353536363737383839394040414142424343444445454646474748484949505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/CAS/bestsnps_AAC.raw\n",
      "written.\n",
      "--score: 84 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/CAS/bestsnps_AAC.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/CAS/bestsnps_AAC.sscore .\n",
      "End time: Wed Feb 12 14:36:42 2025\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "for ancestry in {\"AAC\",\"AFR\",\"AJ\",\"AMR\",\"EAS\",\"EUR\",\"CAS\"}; do\n",
    "    /home/jupyter/plink2 \\\n",
    "    --bfile /home/jupyter/workspace/ws_files/results/${ancestry}/model1_plink/merged_${ancestry} \\\n",
    "    --score /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt list-variants \\\n",
    "    --extract /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt \\\n",
    "    --export A --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt \\\n",
    "    --out /home/jupyter/workspace/ws_files/results/${ancestry}/bestsnps_AAC\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "736fc71d-e16a-45c1-b4bd-a983d9f7f34d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AAC/bestsnps_EAS.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AAC/model1_plink/merged_AAC\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AAC/bestsnps_EAS\n",
      "  --score /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:37:02 2025\n",
      "12984 MiB RAM detected, ~3269 available; reserving 3205 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "1196 samples (698 females, 498 males; 1196 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AAC/model1_plink/merged_AAC.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AAC/model1_plink/merged_AAC.bim.\n",
      "1 binary phenotype loaded (325 cases, 819 controls).\n",
      "--extract: 65 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "65 variants remaining after main filters.\n",
      "--export-allele: 65 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324252526262727282829293030313132323333343435353636373738383939404041414242434344444545464647474848495050515152525353545455555656575758585959606061616262636364646565666667676868696970707171727273737475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AAC/bestsnps_EAS.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 1 entry in\n",
      "/home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "was skipped due to a mismatching allele code.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 64 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AAC/bestsnps_EAS.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AAC/bestsnps_EAS.sscore .\n",
      "End time: Wed Feb 12 14:37:02 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AFR/bestsnps_EAS.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AFR/model1_plink/merged_AFR\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AFR/bestsnps_EAS\n",
      "  --score /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:37:03 2025\n",
      "12984 MiB RAM detected, ~3258 available; reserving 3194 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "2648 samples (1190 females, 1458 males; 2648 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AFR/model1_plink/merged_AFR.fam.\n",
      "88 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AFR/model1_plink/merged_AFR.bim.\n",
      "1 binary phenotype loaded (954 cases, 1652 controls).\n",
      "--extract: 65 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "65 variants remaining after main filters.\n",
      "--export-allele: 65 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324252526262727282829293030313132323333343435353636373738383939404041414242434344444545464647474848495050515152525353545455555656575758585959606061616262636364646565666667676868696970707171727273737475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AFR/bestsnps_EAS.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 1 entry in\n",
      "/home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "was skipped due to a mismatching allele code.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 64 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AFR/bestsnps_EAS.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AFR/bestsnps_EAS.sscore .\n",
      "End time: Wed Feb 12 14:37:03 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AJ/bestsnps_EAS.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AJ/model1_plink/merged_AJ\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AJ/bestsnps_EAS\n",
      "  --score /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:37:03 2025\n",
      "12984 MiB RAM detected, ~3281 available; reserving 3217 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "3126 samples (1168 females, 1958 males; 3126 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AJ/model1_plink/merged_AJ.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AJ/model1_plink/merged_AJ.bim.\n",
      "1 binary phenotype loaded (1396 cases, 423 controls).\n",
      "--extract: 65 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "65 variants remaining after main filters.\n",
      "--export-allele: 65 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 10101111121213131414151516161717181819192020212122222323242425252626272728282929303031313232333334343535363637373838393940404141424243434444454546464747484849505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AJ/bestsnps_EAS.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 1 entry in\n",
      "/home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "was skipped due to a mismatching allele code.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 64 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AJ/bestsnps_EAS.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AJ/bestsnps_EAS.sscore .\n",
      "End time: Wed Feb 12 14:37:04 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AMR/bestsnps_EAS.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AMR/model1_plink/merged_AMR\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AMR/bestsnps_EAS\n",
      "  --score /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:37:04 2025\n",
      "12984 MiB RAM detected, ~3285 available; reserving 3221 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "3496 samples (1835 females, 1661 males; 3496 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AMR/model1_plink/merged_AMR.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AMR/model1_plink/merged_AMR.bim.\n",
      "1 binary phenotype loaded (1928 cases, 1430 controls).\n",
      "--extract: 65 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "65 variants remaining after main filters.\n",
      "--export-allele: 65 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324252526262727282829293030313132323333343435353636373738383939404041414242434344444545464647474848495050515152525353545455555656575758585959606061616262636364646565666667676868696970707171727273737475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AMR/bestsnps_EAS.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 1 entry in\n",
      "/home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "was skipped due to a mismatching allele code.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 64 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AMR/bestsnps_EAS.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AMR/bestsnps_EAS.sscore .\n",
      "End time: Wed Feb 12 14:37:05 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS/bestsnps_EAS.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/EAS/model1_plink/merged_EAS\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS/bestsnps_EAS\n",
      "  --score /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:37:05 2025\n",
      "12984 MiB RAM detected, ~3302 available; reserving 3238 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EAS/model1_plink/merged_EAS.fam.\n",
      "87 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EAS/model1_plink/merged_EAS.bim.\n",
      "1 binary phenotype loaded (1819 cases, 2349 controls).\n",
      "--extract: 64 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "64 variants remaining after main filters.\n",
      "--export-allele: 64 variants updated, 1 variant ID not present.\n",
      "--export A pass 1/1: loading... 0writing... 10101112121313141415161617171818192020212122222324242525262627282829293030313232333334343536363737383839404041414242434444454546464748484949505051525253535454555656575758585960606161626263646465656666676868696970707172727373747475767677777878798080818182828384848585868687888889899090919292939394949596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/EAS/bestsnps_EAS.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 1 entry in\n",
      "/home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "was skipped due to a missing variant ID.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 64 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/EAS/bestsnps_EAS.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/EAS/bestsnps_EAS.sscore .\n",
      "End time: Wed Feb 12 14:37:06 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EUR/bestsnps_EAS.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/EUR/model1_plink/merged_EUR\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/EUR/bestsnps_EAS\n",
      "  --score /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:37:06 2025\n",
      "12984 MiB RAM detected, ~3307 available; reserving 3243 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "48909 samples (21569 females, 27340 males; 48909 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EUR/model1_plink/merged_EUR.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EUR/model1_plink/merged_EUR.bim.\n",
      "1 binary phenotype loaded (22093 cases, 9706 controls).\n",
      "--extract: 65 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "65 variants remaining after main filters.\n",
      "--export-allele: 65 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324242525262627272828292930303131323233333434353536363737383839394040414142424343444445454646474748484949505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/EUR/bestsnps_EAS.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 1 entry in\n",
      "/home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "was skipped due to a mismatching allele code.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 64 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/EUR/bestsnps_EAS.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/EUR/bestsnps_EAS.sscore .\n",
      "End time: Wed Feb 12 14:37:07 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/CAS/bestsnps_EAS.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/CAS/model1_plink/merged_CAS\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/CAS/bestsnps_EAS\n",
      "  --score /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:37:07 2025\n",
      "12984 MiB RAM detected, ~3303 available; reserving 3239 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "1003 samples (545 females, 458 males; 1003 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/CAS/model1_plink/merged_CAS.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/CAS/model1_plink/merged_CAS.bim.\n",
      "1 binary phenotype loaded (582 cases, 323 controls).\n",
      "--extract: 65 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "65 variants remaining after main filters.\n",
      "--export-allele: 65 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324242525262627272828292930303131323233333434353536363737383839394040414142424343444445454646474748484949505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/CAS/bestsnps_EAS.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 1 entry in\n",
      "/home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\n",
      "was skipped due to a mismatching allele code.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 64 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/CAS/bestsnps_EAS.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/CAS/bestsnps_EAS.sscore .\n",
      "End time: Wed Feb 12 14:37:07 2025\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "for ancestry in {\"AAC\",\"AFR\",\"AJ\",\"AMR\",\"EAS\",\"EUR\",\"CAS\"}; do\n",
    "    /home/jupyter/plink2 \\\n",
    "    --bfile /home/jupyter/workspace/ws_files/results/${ancestry}/model1_plink/merged_${ancestry} \\\n",
    "    --score /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt list-variants \\\n",
    "    --extract /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt \\\n",
    "    --export A --export-allele /home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt \\\n",
    "    --out /home/jupyter/workspace/ws_files/results/${ancestry}/bestsnps_EAS\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "553a403b-2c25-4993-ab30-11383daddd01",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AAC/bestsnps_AMR.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AAC/model1_plink/merged_AAC\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AAC/bestsnps_AMR\n",
      "  --score /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:37:47 2025\n",
      "12984 MiB RAM detected, ~3350 available; reserving 3286 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "1196 samples (698 females, 498 males; 1196 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AAC/model1_plink/merged_AAC.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AAC/model1_plink/merged_AAC.bim.\n",
      "1 binary phenotype loaded (325 cases, 819 controls).\n",
      "--extract: 69 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "69 variants remaining after main filters.\n",
      "--export-allele: 69 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324252526262727282829293030313132323333343435353636373738383939404041414242434344444545464647474848495050515152525353545455555656575758585959606061616262636364646565666667676868696970707171727273737475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AAC/bestsnps_AMR.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 1 entry in\n",
      "/home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "was skipped due to a mismatching allele code.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 68 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AAC/bestsnps_AMR.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AAC/bestsnps_AMR.sscore .\n",
      "End time: Wed Feb 12 14:37:48 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AFR/bestsnps_AMR.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AFR/model1_plink/merged_AFR\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AFR/bestsnps_AMR\n",
      "  --score /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:37:48 2025\n",
      "12984 MiB RAM detected, ~3297 available; reserving 3233 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "2648 samples (1190 females, 1458 males; 2648 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AFR/model1_plink/merged_AFR.fam.\n",
      "88 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AFR/model1_plink/merged_AFR.bim.\n",
      "1 binary phenotype loaded (954 cases, 1652 controls).\n",
      "--extract: 69 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "69 variants remaining after main filters.\n",
      "--export-allele: 69 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324252526262727282829293030313132323333343435353636373738383939404041414242434344444545464647474848495050515152525353545455555656575758585959606061616262636364646565666667676868696970707171727273737475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AFR/bestsnps_AMR.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 1 entry in\n",
      "/home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "was skipped due to a mismatching allele code.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 68 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AFR/bestsnps_AMR.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AFR/bestsnps_AMR.sscore .\n",
      "End time: Wed Feb 12 14:37:49 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AJ/bestsnps_AMR.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AJ/model1_plink/merged_AJ\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AJ/bestsnps_AMR\n",
      "  --score /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:37:49 2025\n",
      "12984 MiB RAM detected, ~3305 available; reserving 3241 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "3126 samples (1168 females, 1958 males; 3126 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AJ/model1_plink/merged_AJ.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AJ/model1_plink/merged_AJ.bim.\n",
      "1 binary phenotype loaded (1396 cases, 423 controls).\n",
      "--extract: 69 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "69 variants remaining after main filters.\n",
      "--export-allele: 69 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 10101111121213131414151516161717181819192020212122222323242425252626272728282929303031313232333334343535363637373838393940404141424243434444454546464747484849505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AJ/bestsnps_AMR.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 1 entry in\n",
      "/home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "was skipped due to a mismatching allele code.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 68 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AJ/bestsnps_AMR.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AJ/bestsnps_AMR.sscore .\n",
      "End time: Wed Feb 12 14:37:49 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AMR/bestsnps_AMR.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AMR/model1_plink/merged_AMR\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AMR/bestsnps_AMR\n",
      "  --score /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:37:50 2025\n",
      "12984 MiB RAM detected, ~3303 available; reserving 3239 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "3496 samples (1835 females, 1661 males; 3496 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AMR/model1_plink/merged_AMR.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AMR/model1_plink/merged_AMR.bim.\n",
      "1 binary phenotype loaded (1928 cases, 1430 controls).\n",
      "--extract: 69 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "69 variants remaining after main filters.\n",
      "--export-allele: 69 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324252526262727282829293030313132323333343435353636373738383939404041414242434344444545464647474848495050515152525353545455555656575758585959606061616262636364646565666667676868696970707171727273737475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AMR/bestsnps_AMR.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 1 entry in\n",
      "/home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "was skipped due to a mismatching allele code.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 68 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AMR/bestsnps_AMR.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AMR/bestsnps_AMR.sscore .\n",
      "End time: Wed Feb 12 14:37:50 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS/bestsnps_AMR.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/EAS/model1_plink/merged_EAS\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS/bestsnps_AMR\n",
      "  --score /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:37:50 2025\n",
      "12984 MiB RAM detected, ~3305 available; reserving 3241 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EAS/model1_plink/merged_EAS.fam.\n",
      "87 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EAS/model1_plink/merged_EAS.bim.\n",
      "1 binary phenotype loaded (1819 cases, 2349 controls).\n",
      "--extract: 69 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "69 variants remaining after main filters.\n",
      "--export-allele: 69 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 10101112121313141415161617171818192020212122222324242525262627282829293030313232333334343536363737383839404041414242434444454546464748484949505051525253535454555656575758585960606161626263646465656666676868696970707172727373747475767677777878798080818182828384848585868687888889899090919292939394949596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/EAS/bestsnps_AMR.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 1 entry in\n",
      "/home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "was skipped due to a mismatching allele code.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 68 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/EAS/bestsnps_AMR.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/EAS/bestsnps_AMR.sscore .\n",
      "End time: Wed Feb 12 14:37:51 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EUR/bestsnps_AMR.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/EUR/model1_plink/merged_EUR\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/EUR/bestsnps_AMR\n",
      "  --score /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:37:51 2025\n",
      "12984 MiB RAM detected, ~3307 available; reserving 3243 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "48909 samples (21569 females, 27340 males; 48909 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EUR/model1_plink/merged_EUR.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EUR/model1_plink/merged_EUR.bim.\n",
      "1 binary phenotype loaded (22093 cases, 9706 controls).\n",
      "--extract: 69 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "69 variants remaining after main filters.\n",
      "--export-allele: 69 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324242525262627272828292930303131323233333434353536363737383839394040414142424343444445454646474748484949505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/EUR/bestsnps_AMR.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 1 entry in\n",
      "/home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "was skipped due to a mismatching allele code.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 68 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/EUR/bestsnps_AMR.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/EUR/bestsnps_AMR.sscore .\n",
      "End time: Wed Feb 12 14:37:52 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/CAS/bestsnps_AMR.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/CAS/model1_plink/merged_CAS\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/CAS/bestsnps_AMR\n",
      "  --score /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:37:52 2025\n",
      "12984 MiB RAM detected, ~3307 available; reserving 3243 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "1003 samples (545 females, 458 males; 1003 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/CAS/model1_plink/merged_CAS.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/CAS/model1_plink/merged_CAS.bim.\n",
      "1 binary phenotype loaded (582 cases, 323 controls).\n",
      "--extract: 69 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "69 variants remaining after main filters.\n",
      "--export-allele: 69 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324242525262627272828292930303131323233333434353536363737383839394040414142424343444445454646474748484949505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/CAS/bestsnps_AMR.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 1 entry in\n",
      "/home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\n",
      "was skipped due to a mismatching allele code.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 68 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/CAS/bestsnps_AMR.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/CAS/bestsnps_AMR.sscore .\n",
      "End time: Wed Feb 12 14:37:53 2025\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "for ancestry in {\"AAC\",\"AFR\",\"AJ\",\"AMR\",\"EAS\",\"EUR\",\"CAS\"}; do\n",
    "    /home/jupyter/plink2 \\\n",
    "    --bfile /home/jupyter/workspace/ws_files/results/${ancestry}/model1_plink/merged_${ancestry} \\\n",
    "    --score /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt list-variants \\\n",
    "    --extract /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt \\\n",
    "    --export A --export-allele /home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt \\\n",
    "    --out /home/jupyter/workspace/ws_files/results/${ancestry}/bestsnps_AMR\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c346e2f3-b696-4331-8077-1ce8c688550b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AAC/bestsnps_EUR.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AAC/model1_plink/merged_AAC\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AAC/bestsnps_EUR\n",
      "  --score /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:38:00 2025\n",
      "12984 MiB RAM detected, ~3306 available; reserving 3242 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "1196 samples (698 females, 498 males; 1196 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AAC/model1_plink/merged_AAC.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AAC/model1_plink/merged_AAC.bim.\n",
      "1 binary phenotype loaded (325 cases, 819 controls).\n",
      "--extract: 90 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "90 variants remaining after main filters.\n",
      "--export-allele: 90 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324252526262727282829293030313132323333343435353636373738383939404041414242434344444545464647474848495050515152525353545455555656575758585959606061616262636364646565666667676868696970707171727273737475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AAC/bestsnps_EUR.raw\n",
      "written.\n",
      "--score: 90 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AAC/bestsnps_EUR.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AAC/bestsnps_EUR.sscore .\n",
      "End time: Wed Feb 12 14:38:00 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AFR/bestsnps_EUR.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AFR/model1_plink/merged_AFR\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AFR/bestsnps_EUR\n",
      "  --score /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:38:00 2025\n",
      "12984 MiB RAM detected, ~3307 available; reserving 3243 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "2648 samples (1190 females, 1458 males; 2648 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AFR/model1_plink/merged_AFR.fam.\n",
      "88 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AFR/model1_plink/merged_AFR.bim.\n",
      "1 binary phenotype loaded (954 cases, 1652 controls).\n",
      "--extract: 88 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "88 variants remaining after main filters.\n",
      "--export-allele: 88 variants updated, 2 variant IDs not present.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324252526262727282829293030313132323333343435353636373738383939404041414242434344444545464647474848495050515152525353545455555656575758585959606061616262636364646565666667676868696970707171727273737475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AFR/bestsnps_EUR.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 2 entries in\n",
      "/home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt were skipped\n",
      "due to missing variant IDs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 88 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AFR/bestsnps_EUR.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AFR/bestsnps_EUR.sscore .\n",
      "End time: Wed Feb 12 14:38:01 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AJ/bestsnps_EUR.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AJ/model1_plink/merged_AJ\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AJ/bestsnps_EUR\n",
      "  --score /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:38:01 2025\n",
      "12984 MiB RAM detected, ~3277 available; reserving 3213 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "3126 samples (1168 females, 1958 males; 3126 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AJ/model1_plink/merged_AJ.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AJ/model1_plink/merged_AJ.bim.\n",
      "1 binary phenotype loaded (1396 cases, 423 controls).\n",
      "--extract: 90 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "90 variants remaining after main filters.\n",
      "--export-allele: 90 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 10101111121213131414151516161717181819192020212122222323242425252626272728282929303031313232333334343535363637373838393940404141424243434444454546464747484849505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AJ/bestsnps_EUR.raw\n",
      "written.\n",
      "--score: 90 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AJ/bestsnps_EUR.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AJ/bestsnps_EUR.sscore .\n",
      "End time: Wed Feb 12 14:38:02 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/AMR/bestsnps_EUR.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/AMR/model1_plink/merged_AMR\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/AMR/bestsnps_EUR\n",
      "  --score /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:38:02 2025\n",
      "12984 MiB RAM detected, ~3291 available; reserving 3227 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "3496 samples (1835 females, 1661 males; 3496 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AMR/model1_plink/merged_AMR.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/AMR/model1_plink/merged_AMR.bim.\n",
      "1 binary phenotype loaded (1928 cases, 1430 controls).\n",
      "--extract: 90 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "90 variants remaining after main filters.\n",
      "--export-allele: 90 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324252526262727282829293030313132323333343435353636373738383939404041414242434344444545464647474848495050515152525353545455555656575758585959606061616262636364646565666667676868696970707171727273737475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/AMR/bestsnps_EUR.raw\n",
      "written.\n",
      "--score: 90 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/AMR/bestsnps_EUR.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/AMR/bestsnps_EUR.sscore .\n",
      "End time: Wed Feb 12 14:38:02 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS/bestsnps_EUR.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/EAS/model1_plink/merged_EAS\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS/bestsnps_EUR\n",
      "  --score /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:38:03 2025\n",
      "12984 MiB RAM detected, ~3299 available; reserving 3235 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EAS/model1_plink/merged_EAS.fam.\n",
      "87 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EAS/model1_plink/merged_EAS.bim.\n",
      "1 binary phenotype loaded (1819 cases, 2349 controls).\n",
      "--extract: 87 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "87 variants remaining after main filters.\n",
      "--export-allele: 87 variants updated, 3 variant IDs not present.\n",
      "--export A pass 1/1: loading... 0writing... 10101112121313141415161617171818192020212122222324242525262627282829293030313232333334343536363737383839404041414242434444454546464748484949505051525253535454555656575758585960606161626263646465656666676868696970707172727373747475767677777878798080818182828384848585868687888889899090919292939394949596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/EAS/bestsnps_EUR.raw\n",
      "written.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: --score: 3 entries in\n",
      "/home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt were skipped\n",
      "due to missing variant IDs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--score: 87 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/EAS/bestsnps_EUR.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/EAS/bestsnps_EUR.sscore .\n",
      "End time: Wed Feb 12 14:38:03 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EUR/bestsnps_EUR.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/EUR/model1_plink/merged_EUR\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/EUR/bestsnps_EUR\n",
      "  --score /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:38:03 2025\n",
      "12984 MiB RAM detected, ~3281 available; reserving 3217 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "48909 samples (21569 females, 27340 males; 48909 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EUR/model1_plink/merged_EUR.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EUR/model1_plink/merged_EUR.bim.\n",
      "1 binary phenotype loaded (22093 cases, 9706 controls).\n",
      "--extract: 90 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "90 variants remaining after main filters.\n",
      "--export-allele: 90 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324242525262627272828292930303131323233333434353536363737383839394040414142424343444445454646474748484949505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/EUR/bestsnps_EUR.raw\n",
      "written.\n",
      "--score: 90 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/EUR/bestsnps_EUR.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/EUR/bestsnps_EUR.sscore .\n",
      "End time: Wed Feb 12 14:38:04 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/CAS/bestsnps_EUR.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/CAS/model1_plink/merged_CAS\n",
      "  --export A\n",
      "  --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\n",
      "  --extract /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/CAS/bestsnps_EUR\n",
      "  --score /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt list-variants\n",
      "\n",
      "Start time: Wed Feb 12 14:38:04 2025\n",
      "12984 MiB RAM detected, ~3305 available; reserving 3241 MiB for main workspace.\n",
      "Using up to 2 compute threads.\n",
      "1003 samples (545 females, 458 males; 1003 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/CAS/model1_plink/merged_CAS.fam.\n",
      "90 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/CAS/model1_plink/merged_CAS.bim.\n",
      "1 binary phenotype loaded (582 cases, 323 controls).\n",
      "--extract: 90 variants remaining.\n",
      "Calculating allele frequencies... done.\n",
      "90 variants remaining after main filters.\n",
      "--export-allele: 90 variants updated.\n",
      "--export A pass 1/1: loading... 0writing... 1010111112121313141415151616171718181919202021212222232324242525262627272828292930303131323233333434353536363737383839394040414142424343444445454646474748484949505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: /home/jupyter/workspace/ws_files/results/CAS/bestsnps_EUR.raw\n",
      "written.\n",
      "--score: 90 variants processed.\n",
      "--score: Variant list written to /home/jupyter/workspace/ws_files/results/CAS/bestsnps_EUR.sscore.vars .\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/CAS/bestsnps_EUR.sscore .\n",
      "End time: Wed Feb 12 14:38:05 2025\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "for ancestry in {\"AAC\",\"AFR\",\"AJ\",\"AMR\",\"EAS\",\"EUR\",\"CAS\"}; do\n",
    "    /home/jupyter/plink2 \\\n",
    "    --bfile /home/jupyter/workspace/ws_files/results/${ancestry}/model1_plink/merged_${ancestry} \\\n",
    "    --score /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt list-variants \\\n",
    "    --extract /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt \\\n",
    "    --export A --export-allele /home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt \\\n",
    "    --out /home/jupyter/workspace/ws_files/results/${ancestry}/bestsnps_EUR\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "6294c369-718e-4611-b0f4-a328232f3f73",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAC\n",
      "                       SNP A1    BETA   CONTROL        PD  \\\n",
      "1:155235843_T  1:155235843  T -0.7467  1.998779  1.997347   \n",
      "1:155162560_A  1:155162560  A  0.6068  0.015892  0.015915   \n",
      "\n",
      "               PD_allele_count_and_beta  CONTROL_allele_count_and_beta  \\\n",
      "1:155235843_T                 -1.491419                      -1.492488   \n",
      "1:155162560_A                  0.009657                       0.009644   \n",
      "\n",
      "               DIFFERENCE  DIFFERENCE_absolute  \n",
      "1:155235843_T    0.001069             0.001069  \n",
      "1:155162560_A    0.000014             0.000014  \n",
      "AFR\n",
      "                       SNP A1    BETA   CONTROL   PD  \\\n",
      "1:155162560_A  1:155162560  A  0.6068  0.001211  0.0   \n",
      "1:155235843_T  1:155235843  T -0.7467       NaN  NaN   \n",
      "\n",
      "               PD_allele_count_and_beta  CONTROL_allele_count_and_beta  \\\n",
      "1:155162560_A                       0.0                       0.000735   \n",
      "1:155235843_T                       NaN                            NaN   \n",
      "\n",
      "               DIFFERENCE  DIFFERENCE_absolute  \n",
      "1:155162560_A   -0.000735             0.000735  \n",
      "1:155235843_T         NaN                  NaN  \n",
      "AJ\n",
      "                       SNP A1    BETA   CONTROL        PD  \\\n",
      "1:155235843_T  1:155235843  T -0.7467  1.945626  1.806236   \n",
      "1:155162560_A  1:155162560  A  0.6068  0.004728  0.015908   \n",
      "\n",
      "               PD_allele_count_and_beta  CONTROL_allele_count_and_beta  \\\n",
      "1:155235843_T                 -1.348716                      -1.452799   \n",
      "1:155162560_A                  0.009653                       0.002869   \n",
      "\n",
      "               DIFFERENCE  DIFFERENCE_absolute  \n",
      "1:155235843_T    0.104083             0.104083  \n",
      "1:155162560_A    0.006784             0.006784  \n",
      "AMR\n",
      "                       SNP A1    BETA   CONTROL        PD  \\\n",
      "1:155162560_A  1:155162560  A  0.6068  0.003497  0.012585   \n",
      "1:155235843_T  1:155235843  T -0.7467  1.998601  1.994192   \n",
      "\n",
      "               PD_allele_count_and_beta  CONTROL_allele_count_and_beta  \\\n",
      "1:155162560_A                  0.007636                       0.002122   \n",
      "1:155235843_T                 -1.489063                      -1.492356   \n",
      "\n",
      "               DIFFERENCE  DIFFERENCE_absolute  \n",
      "1:155162560_A    0.005515             0.005515  \n",
      "1:155235843_T    0.003293             0.003293  \n",
      "EAS\n",
      "                       SNP A1    BETA  CONTROL  PD  PD_allele_count_and_beta  \\\n",
      "1:155162560_A  1:155162560  A  0.6068      NaN NaN                       NaN   \n",
      "1:155235843_T  1:155235843  T -0.7467      NaN NaN                       NaN   \n",
      "\n",
      "               CONTROL_allele_count_and_beta  DIFFERENCE  DIFFERENCE_absolute  \n",
      "1:155162560_A                            NaN         NaN                  NaN  \n",
      "1:155235843_T                            NaN         NaN                  NaN  \n",
      "EUR\n",
      "                       SNP A1    BETA   CONTROL        PD  \\\n",
      "1:155162560_A  1:155162560  A  0.6068  0.025863  0.037714   \n",
      "1:155235843_T  1:155235843  T -0.7467  1.995776  1.990256   \n",
      "\n",
      "               PD_allele_count_and_beta  CONTROL_allele_count_and_beta  \\\n",
      "1:155162560_A                  0.022885                       0.015694   \n",
      "1:155235843_T                 -1.486124                      -1.490246   \n",
      "\n",
      "               DIFFERENCE  DIFFERENCE_absolute  \n",
      "1:155162560_A    0.007191             0.007191  \n",
      "1:155235843_T    0.004122             0.004122  \n",
      "CAS\n",
      "                       SNP A1    BETA   CONTROL        PD  \\\n",
      "1:155162560_A  1:155162560  A  0.6068  0.003096  0.014706   \n",
      "1:155235843_T  1:155235843  T -0.7467  1.996904  1.995588   \n",
      "\n",
      "               PD_allele_count_and_beta  CONTROL_allele_count_and_beta  \\\n",
      "1:155162560_A                  0.008924                       0.001879   \n",
      "1:155235843_T                 -1.490106                      -1.491088   \n",
      "\n",
      "               DIFFERENCE  DIFFERENCE_absolute  \n",
      "1:155162560_A    0.007045             0.007045  \n",
      "1:155235843_T    0.000982             0.000982  \n"
     ]
    }
   ],
   "source": [
    "filenames = [\n",
    "    \"/home/jupyter/workspace/ws_files/data/90LOCI_AFRICAN_CHR_POS.txt\",\n",
    "    \"/home/jupyter/workspace/ws_files/results/meta/META_FOO_23ANDME_90LOCI_EASTASIAN_CHR_POS.txt\",\n",
    "    \"/home/jupyter/workspace/ws_files/data/90LOCI_EUROPEANS_CHR_POS.txt\",\n",
    "    \"/home/jupyter/workspace/ws_files/results/meta/META_LOESCH_23ANDME_90LOCI_LATINO_CHR_POS.txt\",\n",
    "]\n",
    "for summstat_ancestry, base_data_file in zip(summstat_ancestries, filenames):\n",
    "    df_bestperforming = []\n",
    "    for ancestry in ancestries:\n",
    "        ## Read in PRS\n",
    "        prs_df = pd.read_csv(f\"{RESULTS_DIR}/{ancestry}/bestsnps_{summstat_ancestry}.sscore\", sep=\"\\s+\", usecols=['IID', 'SCORE1_AVG'])\n",
    "        prs_df.columns=['IID','PRS']\n",
    "\n",
    "        ## Read in the covariate file and the transposed PLINK .raw file with dosages\n",
    "        temp = pd.read_csv(f\"{RESULTS_DIR}/covariates_pheno.txt\", sep=\"\\s+\")\n",
    "        cov_df = temp.rename(columns={'GP2ID': 'IID'})\n",
    "        transpose_df = pd.read_csv(f\"{RESULTS_DIR}/{ancestry}/bestsnps_{summstat_ancestry}.raw\", sep='\\t').drop(columns=['FID', 'PAT', 'MAT', 'SEX'])\n",
    "\n",
    "        ## Merge dataframes\n",
    "        merged_df1 = pd.merge(prs_df, cov_df, on=\"IID\", how=\"left\")\n",
    "        merged_df2 = pd.merge(merged_df1, transpose_df, on=\"IID\", how=\"left\")\n",
    "        merged_df2.head()\n",
    "\n",
    "        ## Recode status\n",
    "        merged_df2['STATUS'] = np.where(merged_df2['PHENO']==1, \"CONTROL\", \"PD\")\n",
    "\n",
    "        ## Aggregate the information\n",
    "        pivot_df = merged_df2.loc[\n",
    "            merged_df2.STATUS.isin(['CONTROL', 'PD']),\n",
    "            ['STATUS']+list(merged_df2.columns[merged_df2.columns.str.contains(':')])\n",
    "        ].pivot_table(\n",
    "            index=['STATUS'],\n",
    "            aggfunc=['mean']\n",
    "        ).T.reset_index().drop(columns='level_0').set_index('level_1')\n",
    "\n",
    "        ## Read in the full Nalls et al., scoring file\n",
    "        scores = pd.read_csv(base_data_file, sep='\\t', header=None, names=['SNP', 'A1', 'BETA'])\n",
    "        scores.index = scores['SNP'] + '_' + scores['A1']\n",
    "\n",
    "        ## Merge pivot table and scores file to calculate mean effect\n",
    "        snp_beta_df = pd.concat([scores,pivot_df],axis=1)\n",
    "        snp_beta_df['PD_allele_count_and_beta'] = snp_beta_df['PD']*snp_beta_df['BETA']\n",
    "        snp_beta_df['CONTROL_allele_count_and_beta'] = snp_beta_df['CONTROL']*snp_beta_df['BETA']\n",
    "        snp_beta_df['DIFFERENCE'] = snp_beta_df['PD_allele_count_and_beta'] - snp_beta_df['CONTROL_allele_count_and_beta']\n",
    "        snp_beta_df['DIFFERENCE_absolute'] = snp_beta_df['DIFFERENCE'].abs()\n",
    "        snp_beta_df.sort_values(by=\"DIFFERENCE_absolute\", ascending=False, inplace=True)\n",
    "        \n",
    "        if summstat_ancestry == \"EUR\":\n",
    "            print(ancestry)\n",
    "            print(snp_beta_df[snp_beta_df[\"SNP\"].isin([\"1:155162560\", \"1:155235843\"])])\n",
    "            \n",
    "        snp_beta_df = snp_beta_df.iloc[:5,:]\n",
    "        snp_beta_df.insert(0, \"Cohort\", ancestry)\n",
    "        snp_beta_df = snp_beta_df[[\"Cohort\",\"SNP\",\"DIFFERENCE\"]]\n",
    "        snp_beta_df.rename(columns={\"SNP\":\"SNP (hg38)\", \"DIFFERENCE\":\"Mean Effect\"}, inplace=True)\n",
    "\n",
    "        df_bestperforming.append(snp_beta_df)\n",
    "    df_bestperforming = pd.concat(df_bestperforming)\n",
    "    df_bestperforming.to_csv(f\"{RESULTS_DIR}/supp_tab5_{summstat_ancestry}.txt\", sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c270a52-2d9e-48ff-a571-dc71a73649c9",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Foo et al. EAS analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67f1369a-c125-4f02-aeb2-b0dcf8673b98",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Extract 11 SNPs from Foo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "c7a81bf8-d33f-45c2-9c79-506c44a5031a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr1.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr1\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr1_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:31 2025\n",
      "12984 MiB RAM detected, ~11296 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr1_EAS_release9_vwb.psam.\n",
      "3879127 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr1_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 3879125 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n",
      "2 variants remaining after main filters.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr1.fam ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr1.bim ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr1.bed ... 50done.\n",
      "End time: Fri Feb 14 00:12:32 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr2.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr2\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr2_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:33 2025\n",
      "12984 MiB RAM detected, ~11265 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr2_EAS_release9_vwb.psam.\n",
      "4160904 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr2_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 4160904 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: No variants remaining after main filters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "End time: Fri Feb 14 00:12:33 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr3.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr3\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr3_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:33 2025\n",
      "12984 MiB RAM detected, ~11265 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr3_EAS_release9_vwb.psam.\n",
      "3466328 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr3_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 3466327 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n",
      "1 variant remaining after main filters.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr3.fam ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr3.bim ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr3.bed ... done.\n",
      "End time: Fri Feb 14 00:12:34 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr4.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr4\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr4_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:35 2025\n",
      "12984 MiB RAM detected, ~11265 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr4_EAS_release9_vwb.psam.\n",
      "3341933 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr4_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 3341931 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n",
      "2 variants remaining after main filters.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr4.fam ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr4.bim ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr4.bed ... 50done.\n",
      "End time: Fri Feb 14 00:12:36 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr5.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr5\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr5_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:36 2025\n",
      "12984 MiB RAM detected, ~11267 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr5_EAS_release9_vwb.psam.\n",
      "3109654 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr5_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 3109653 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n",
      "1 variant remaining after main filters.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr5.fam ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr5.bim ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr5.bed ... done.\n",
      "End time: Fri Feb 14 00:12:37 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr6.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr6\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr6_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:37 2025\n",
      "12984 MiB RAM detected, ~11266 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr6_EAS_release9_vwb.psam.\n",
      "3002660 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr6_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 3002659 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n",
      "1 variant remaining after main filters.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr6.fam ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr6.bim ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr6.bed ... done.\n",
      "End time: Fri Feb 14 00:12:38 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr7.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr7\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr7_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:38 2025\n",
      "12984 MiB RAM detected, ~11267 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr7_EAS_release9_vwb.psam.\n",
      "2780134 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr7_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 2780133 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n",
      "1 variant remaining after main filters.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr7.fam ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr7.bim ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr7.bed ... done.\n",
      "End time: Fri Feb 14 00:12:39 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr8.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr8\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr8_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:39 2025\n",
      "12984 MiB RAM detected, ~11267 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr8_EAS_release9_vwb.psam.\n",
      "2672830 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr8_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 2672830 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: No variants remaining after main filters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "End time: Fri Feb 14 00:12:39 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr9.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr9\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr9_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:39 2025\n",
      "12984 MiB RAM detected, ~11267 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr9_EAS_release9_vwb.psam.\n",
      "2133556 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr9_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 2133556 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: No variants remaining after main filters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "End time: Fri Feb 14 00:12:40 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr10.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr10\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr10_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:40 2025\n",
      "12984 MiB RAM detected, ~11269 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr10_EAS_release9_vwb.psam.\n",
      "2389291 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr10_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 2389291 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: No variants remaining after main filters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "End time: Fri Feb 14 00:12:41 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr11.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr11\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr11_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:41 2025\n",
      "12984 MiB RAM detected, ~11266 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr11_EAS_release9_vwb.psam.\n",
      "2324503 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr11_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 2324503 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: No variants remaining after main filters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "End time: Fri Feb 14 00:12:41 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr12.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr12\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr12_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:41 2025\n",
      "12984 MiB RAM detected, ~11268 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr12_EAS_release9_vwb.psam.\n",
      "2301328 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr12_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 2301327 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n",
      "1 variant remaining after main filters.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr12.fam ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr12.bim ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr12.bed ... done.\n",
      "End time: Fri Feb 14 00:12:42 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr13.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr13\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr13_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:42 2025\n",
      "12984 MiB RAM detected, ~11266 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr13_EAS_release9_vwb.psam.\n",
      "1745120 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr13_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 1745120 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: No variants remaining after main filters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "End time: Fri Feb 14 00:12:43 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr14.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr14\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr14_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:43 2025\n",
      "12984 MiB RAM detected, ~11266 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr14_EAS_release9_vwb.psam.\n",
      "1557805 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr14_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 1557805 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: No variants remaining after main filters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "End time: Fri Feb 14 00:12:43 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr15.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr15\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr15_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:43 2025\n",
      "12984 MiB RAM detected, ~11268 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr15_EAS_release9_vwb.psam.\n",
      "1457164 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr15_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 1457164 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: No variants remaining after main filters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "End time: Fri Feb 14 00:12:44 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr16.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr16\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr16_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:44 2025\n",
      "12984 MiB RAM detected, ~11267 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr16_EAS_release9_vwb.psam.\n",
      "1573363 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr16_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 1573363 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: No variants remaining after main filters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "End time: Fri Feb 14 00:12:44 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr17.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr17\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr17_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:44 2025\n",
      "12984 MiB RAM detected, ~11267 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr17_EAS_release9_vwb.psam.\n",
      "1404704 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr17_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 1404703 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n",
      "1 variant remaining after main filters.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr17.fam ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr17.bim ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr17.bed ... done.\n",
      "End time: Fri Feb 14 00:12:45 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr18.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr18\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr18_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:45 2025\n",
      "12984 MiB RAM detected, ~11267 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr18_EAS_release9_vwb.psam.\n",
      "1381992 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr18_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 1381992 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: No variants remaining after main filters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "End time: Fri Feb 14 00:12:46 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr19.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr19\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr19_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:46 2025\n",
      "12984 MiB RAM detected, ~11267 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr19_EAS_release9_vwb.psam.\n",
      "1098413 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr19_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 1098412 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n",
      "1 variant remaining after main filters.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr19.fam ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr19.bim ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/chr19.bed ... done.\n",
      "End time: Fri Feb 14 00:12:46 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr20.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr20\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr20_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:47 2025\n",
      "12984 MiB RAM detected, ~11266 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr20_EAS_release9_vwb.psam.\n",
      "1114827 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr20_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 1114827 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: No variants remaining after main filters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "End time: Fri Feb 14 00:12:47 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr21.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr21\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr21_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:47 2025\n",
      "12984 MiB RAM detected, ~11268 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr21_EAS_release9_vwb.psam.\n",
      "656069 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr21_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 656069 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: No variants remaining after main filters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "End time: Fri Feb 14 00:12:47 2025\n",
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/chr22.log.\n",
      "Options in effect:\n",
      "  --extract range /home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/chr22\n",
      "  --pfile /home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr22_EAS_release9_vwb\n",
      "  --remove /home/jupyter/workspace/ws_files/results/EAS/related.txt\n",
      "  --update-sex /home/jupyter/workspace/ws_files/results/sex.txt\n",
      "\n",
      "Start time: Fri Feb 14 00:12:47 2025\n",
      "12984 MiB RAM detected, ~11267 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr22_EAS_release9_vwb.psam.\n",
      "670038 variants loaded from\n",
      "/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr22_EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract bed1: 670038 variants excluded.\n",
      "--update-sex: 5902 samples updated, 75673 IDs not present.\n",
      "--remove: 5625 samples remaining.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) remaining after main\n",
      "filters.\n",
      "1819 cases and 2349 controls remaining after main filters.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: No variants remaining after main filters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "End time: Fri Feb 14 00:12:48 2025\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'b'\\nfor CHRNUM in {1..22}; do\\n    /home/jupyter/plink2 \\\\\\n    --pfile \"/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr${CHRNUM}_EAS_release9_vwb\" \\\\\\n    --remove \"/home/jupyter/workspace/ws_files/results/EAS/related.txt\" \\\\\\n    --update-sex \"/home/jupyter/workspace/ws_files/results/sex.txt\" \\\\\\n    --extract range \"/home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\" \\\\\\n    --make-bed \\\\\\n    --out \"/home/jupyter/workspace/ws_files/results/EAS_Foo/chr${CHRNUM}\"\\ndone\\n'' returned non-zero exit status 13.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[55], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mget_ipython\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_cell_magic\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mbash\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mfor CHRNUM in \u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;124;43m1..22}; do\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    /home/jupyter/plink2 \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    --pfile \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr$\u001b[39;49m\u001b[38;5;132;43;01m{CHRNUM}\u001b[39;49;00m\u001b[38;5;124;43m_EAS_release9_vwb\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    --remove \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/home/jupyter/workspace/ws_files/results/EAS/related.txt\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    --update-sex \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/home/jupyter/workspace/ws_files/results/sex.txt\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    --extract range \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    --make-bed \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    --out \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/home/jupyter/workspace/ws_files/results/EAS_Foo/chr$\u001b[39;49m\u001b[38;5;132;43;01m{CHRNUM}\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mdone\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.10/site-packages/IPython/core/interactiveshell.py:2517\u001b[0m, in \u001b[0;36mInteractiveShell.run_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2515\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[1;32m   2516\u001b[0m     args \u001b[38;5;241m=\u001b[39m (magic_arg_s, cell)\n\u001b[0;32m-> 2517\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2519\u001b[0m \u001b[38;5;66;03m# The code below prevents the output from being displayed\u001b[39;00m\n\u001b[1;32m   2520\u001b[0m \u001b[38;5;66;03m# when using magics with decorator @output_can_be_silenced\u001b[39;00m\n\u001b[1;32m   2521\u001b[0m \u001b[38;5;66;03m# when the last Python token in the expression is a ';'.\u001b[39;00m\n\u001b[1;32m   2522\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(fn, magic\u001b[38;5;241m.\u001b[39mMAGIC_OUTPUT_CAN_BE_SILENCED, \u001b[38;5;28;01mFalse\u001b[39;00m):\n",
      "File \u001b[0;32m/opt/conda/lib/python3.10/site-packages/IPython/core/magics/script.py:154\u001b[0m, in \u001b[0;36mScriptMagics._make_script_magic.<locals>.named_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    152\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    153\u001b[0m     line \u001b[38;5;241m=\u001b[39m script\n\u001b[0;32m--> 154\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshebang\u001b[49m\u001b[43m(\u001b[49m\u001b[43mline\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcell\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.10/site-packages/IPython/core/magics/script.py:314\u001b[0m, in \u001b[0;36mScriptMagics.shebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    309\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m args\u001b[38;5;241m.\u001b[39mraise_error \u001b[38;5;129;01mand\u001b[39;00m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    310\u001b[0m     \u001b[38;5;66;03m# If we get here and p.returncode is still None, we must have\u001b[39;00m\n\u001b[1;32m    311\u001b[0m     \u001b[38;5;66;03m# killed it but not yet seen its return code. We don't wait for it,\u001b[39;00m\n\u001b[1;32m    312\u001b[0m     \u001b[38;5;66;03m# in case it's stuck in uninterruptible sleep. -9 = SIGKILL\u001b[39;00m\n\u001b[1;32m    313\u001b[0m     rc \u001b[38;5;241m=\u001b[39m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m9\u001b[39m\n\u001b[0;32m--> 314\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(rc, cell)\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'b'\\nfor CHRNUM in {1..22}; do\\n    /home/jupyter/plink2 \\\\\\n    --pfile \"/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr${CHRNUM}_EAS_release9_vwb\" \\\\\\n    --remove \"/home/jupyter/workspace/ws_files/results/EAS/related.txt\" \\\\\\n    --update-sex \"/home/jupyter/workspace/ws_files/results/sex.txt\" \\\\\\n    --extract range \"/home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\" \\\\\\n    --make-bed \\\\\\n    --out \"/home/jupyter/workspace/ws_files/results/EAS_Foo/chr${CHRNUM}\"\\ndone\\n'' returned non-zero exit status 13."
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "for CHRNUM in {1..22}; do\n",
    "    /home/jupyter/plink2 \\\n",
    "    --pfile \"/home/jupyter/workspace/gp2_tier2_eu_release9_18122024/imputed_genotypes/EAS/chr${CHRNUM}_EAS_release9_vwb\" \\\n",
    "    --remove \"/home/jupyter/workspace/ws_files/results/EAS/related.txt\" \\\n",
    "    --update-sex \"/home/jupyter/workspace/ws_files/results/sex.txt\" \\\n",
    "    --extract range \"/home/jupyter/workspace/ws_files/data/chr_pos_foo.txt\" \\\n",
    "    --make-bed \\\n",
    "    --out \"/home/jupyter/workspace/ws_files/results/EAS_Foo/chr${CHRNUM}\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdd3ff8a-fa27-417b-8b4e-ab1ada6cef8a",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Merge files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "c2b61803-0cd1-420b-9797-4ecfd87537da",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final list of files for EAS_Foo:\n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/chr1\n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/chr3\n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/chr4\n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/chr5\n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/chr6\n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/chr7\n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/chr12\n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/chr17\n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/chr19\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "> /home/jupyter/workspace/ws_files/results/EAS_Foo/files.txt\n",
    "for CHRNUM in {1..22}; do\n",
    "    BED_FILE=\"/home/jupyter/workspace/ws_files/results/EAS_Foo/chr${CHRNUM}.bed\"\n",
    "    if [[ -f \"${BED_FILE}\" ]]; then\n",
    "        echo \"/home/jupyter/workspace/ws_files/results/EAS_Foo/chr${CHRNUM}\" >> /home/jupyter/workspace/ws_files/results/EAS_Foo/files.txt\n",
    "    fi\n",
    "done\n",
    "\n",
    "echo \"Final list of files for EAS_Foo:\"\n",
    "cat /home/jupyter/workspace/ws_files/results/EAS_Foo/files.txt \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "9a6cc965-3a3f-4b97-a676-5d2759627ad8",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/merged_orig.log.\n",
      "Options in effect:\n",
      "  --make-bed\n",
      "  --merge-list /home/jupyter/workspace/ws_files/results/EAS_Foo/files.txt\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/merged_orig\n",
      "\n",
      "12984 MB RAM detected; reserving 6492 MB for main workspace.\n",
      "Performing single-pass merge (5625 people, 11 variants).\n",
      "Merged fileset written to                     \n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/merged_orig-merge.bed +\n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/merged_orig-merge.bim +\n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/merged_orig-merge.fam .\n",
      "11 variants loaded from .bim file.\n",
      "5625 people (3473 males, 2152 females) loaded from .fam.\n",
      "4168 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 5625 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.945115.\n",
      "11 variants and 5625 people pass filters and QC.\n",
      "Among remaining phenotypes, 1819 are cases and 2349 are controls.  (1457\n",
      "phenotypes are missing.)\n",
      "--make-bed to /home/jupyter/workspace/ws_files/results/EAS_Foo/merged_orig.bed\n",
      "+ /home/jupyter/workspace/ws_files/results/EAS_Foo/merged_orig.bim +\n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/merged_orig.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "/home/jupyter/plink \\\n",
    "--merge-list \"/home/jupyter/workspace/ws_files/results/EAS_Foo/files.txt\" \\\n",
    "--make-bed \\\n",
    "--out \"/home/jupyter/workspace/ws_files/results/EAS_Foo/merged_orig\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "63772edf-bf67-41a2-a98f-4ef8eb72b76e",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.0.0-a.6.7LM 64-bit Intel (20 Jan 2025)    cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/merged.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/EAS_Foo/merged_orig\n",
      "  --make-bed\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/merged\n",
      "  --rm-dup force-first\n",
      "  --set-all-var-ids @:#\n",
      "\n",
      "Start time: Fri Feb 14 00:14:54 2025\n",
      "12984 MiB RAM detected, ~11263 available; reserving 6492 MiB for main\n",
      "workspace.\n",
      "Using up to 2 compute threads.\n",
      "5625 samples (2152 females, 3473 males; 5625 founders) loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/merged_orig.fam.\n",
      "11 variants loaded from\n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/merged_orig.bim.\n",
      "1 binary phenotype loaded (1819 cases, 2349 controls).\n",
      "Note: Skipping --rm-dup since no duplicate IDs are present.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/merged.fam ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/merged.bim ... done.\n",
      "Writing /home/jupyter/workspace/ws_files/results/EAS_Foo/merged.bed ... done.\n",
      "End time: Fri Feb 14 00:14:54 2025\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "/home/jupyter/plink2 \\\n",
    "--bfile \"/home/jupyter/workspace/ws_files/results/EAS_Foo/merged_orig\" \\\n",
    "--rm-dup force-first \\\n",
    "--set-all-var-ids @:# \\\n",
    "--make-bed \\\n",
    "--out \"/home/jupyter/workspace/ws_files/results/EAS_Foo/merged\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb07775c-9a8b-47a1-a4d0-2ec6eff3531d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Generate scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "4194efe2-9678-433d-9821-b50348142ff9",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/workspace/ws_files/results/EAS_Foo/merged.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/workspace/ws_files/results/EAS_Foo/merged\n",
      "  --out /home/jupyter/workspace/ws_files/results/EAS_Foo/merged\n",
      "  --score /home/jupyter/workspace/ws_files/results_final/meta/foo.txt\n",
      "\n",
      "12984 MB RAM detected; reserving 6492 MB for main workspace.\n",
      "11 variants loaded from .bim file.\n",
      "5625 people (3473 males, 2152 females) loaded from .fam.\n",
      "4168 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 5625 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.945115.\n",
      "11 variants and 5625 people pass filters and QC.\n",
      "Among remaining phenotypes, 1819 are cases and 2349 are controls.  (1457\n",
      "phenotypes are missing.)\n",
      "--score: 11 valid predictors loaded.\n",
      "--score: Results written to\n",
      "/home/jupyter/workspace/ws_files/results/EAS_Foo/merged.profile .\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "/home/jupyter/plink \\\n",
    "--bfile \"/home/jupyter/workspace/ws_files/results/EAS_Foo/merged\" \\\n",
    "--score \"/home/jupyter/workspace/ws_files/results_final/meta/foo.txt\" \\\n",
    "--out \"/home/jupyter/workspace/ws_files/results/EAS_Foo/merged\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da2024fa-5034-4e73-a9ff-2446d4709382",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Estimate PRS vs PD Risk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "b303bdc8-a4d0-4657-98f1-a7b5fc33c05e",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Call:\n",
      "glm(formula = CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + \n",
      "    PC5 + PC6 + PC7 + PC8 + PC9 + PC10, family = \"binomial\", \n",
      "    data = dat)\n",
      "\n",
      "Coefficients:\n",
      "              Estimate Std. Error z value Pr(>|z|)    \n",
      "(Intercept)   3.129102   0.305639  10.238  < 2e-16 ***\n",
      "zSCORE        0.300234   0.039044   7.690 1.48e-14 ***\n",
      "SEX          -0.212803   0.085182  -2.498 0.012482 *  \n",
      "AGE          -0.050642   0.004492 -11.275  < 2e-16 ***\n",
      "PC1          84.278636   3.358306  25.096  < 2e-16 ***\n",
      "PC2         -19.183015   3.009553  -6.374 1.84e-10 ***\n",
      "PC3          81.092429   4.070465  19.922  < 2e-16 ***\n",
      "PC4          17.545156   4.106848   4.272 1.94e-05 ***\n",
      "PC5          18.220521   9.710231   1.876 0.060597 .  \n",
      "PC6          12.450490   9.348828   1.332 0.182936    \n",
      "PC7          44.830436   5.322699   8.423  < 2e-16 ***\n",
      "PC8           5.090496   4.477832   1.137 0.255613    \n",
      "PC9         -10.908316   6.535556  -1.669 0.095103 .  \n",
      "PC10         18.858205   4.883558   3.862 0.000113 ***\n",
      "---\n",
      "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
      "\n",
      "(Dispersion parameter for binomial family taken to be 1)\n",
      "\n",
      "    Null deviance: 5710.5  on 4167  degrees of freedom\n",
      "Residual deviance: 3781.7  on 4154  degrees of freedom\n",
      "AIC: 3809.7\n",
      "\n",
      "Number of Fisher Scoring iterations: 5\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Waiting for profiling to be done...\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "temp_data <- read.table(paste0(RESULTS_DIR, \"/EAS_Foo/merged.profile\"), header = T) \n",
    "temp_covs <- read.table(paste0(RESULTS_DIR, \"/covariates.txt\"), header = T, sep=\" \")\n",
    "colnames(temp_covs) <- c(\"FID\", \"IID\", \"SEX\", \"AGE\", \"ANCESTRY\", \"PC1\", \"PC2\", \"PC3\", \"PC4\", \"PC5\", \"PC6\", \"PC7\", \"PC8\", \"PC9\", \"PC10\")\n",
    "data <- merge(temp_data, temp_covs, by = \"IID\")\n",
    "data$CASE <- data$PHENO - 1\n",
    "\n",
    "dat <- subset(data, CASE != -10) \n",
    "meanPop <- mean(subset(dat, CASE == 0)$SCORE)\n",
    "sdPop <- sd(subset(dat, CASE == 0)$SCORE)\n",
    "dat$zSCORE <- (dat$SCORE - meanPop)/sdPop\n",
    "\n",
    "grsTests <- glm(CASE ~ zSCORE + SEX + AGE + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10, data = dat, family = \"binomial\")\n",
    "print(summary(grsTests))\n",
    "grs_summary <- summary(grsTests)$coefficients\n",
    "conf_intervals <- confint(grsTests, level = 0.95)\n",
    "zscore_summary <- grs_summary[\"zSCORE\", , drop = FALSE]\n",
    "zscore_conf <- conf_intervals[\"zSCORE\", , drop = FALSE]\n",
    "zscore <- as.data.table(cbind(\"EAS_Foo\", zscore_summary, zscore_conf))\n",
    "colnames(zscore) <- c(\"COHORT\", \"BETA\", \"SE\", \"T\", \"P\", \"L95\", \"U95\")\n",
    "zscore <- zscore[, !\"T\", with = FALSE]\n",
    "\n",
    "fwrite(zscore, paste0(RESULTS_DIR, \"/EAS_Foo/forest_risk.txt\"), sep = \"\\t\", col.names = TRUE, quote = FALSE)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "114d0615-e936-4076-a3b4-00c5fe964fdf",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Generate plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "ddb0cabd-4dab-4cb6-9526-a243bbbde65c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "data <- read.table(paste0(RESULTS_DIR, \"/EAS_Foo/merged.profile\"), header = T) \n",
    "data$CASE <- data$PHENO - 1\n",
    "dat <- subset(data, CASE != -10)\n",
    "Model <- glm(CASE ~ SCORE, data = dat, family = 'binomial')\n",
    "dat$probDisease <- predict(Model, dat, type = c(\"response\"))\n",
    "dat$predicted <- ifelse(dat$probDisease > 0.5, \"DISEASE\", \"CONTROL\")\n",
    "dat$reported <- ifelse(dat$CASE == 1, \"DISEASE\",\"CONTROL\")\n",
    "dat$Group <- \"EAS_Foo\"\n",
    "fwrite(dat, paste0(RESULTS_DIR, \"/EAS_Foo/traditional_input.txt\"), sep = \"\\t\", col.names = TRUE, quote = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "306ec3b7-8585-4b04-8f18-9cd21b3ac146",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "dat_cases <- subset(dat, CASE == 1)\n",
    "density_rocs_plot <- ggplot(dat_cases, aes(probDisease, color=Group)) + geom_density(alpha = 0.5) + theme_bw()\n",
    "ggsave(plot = density_rocs_plot, filename = paste0(RESULTS_DIR, \"/EAS_Foo/density_joint.png\"), width = 8, height = 5, units = \"in\", dpi = 300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "c561539f-880a-4aee-a52f-57ded9795907",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8, 5))\n",
    "\n",
    "dat = pd.read_csv(f\"{RESULTS_DIR}/EAS_Foo/traditional_input.txt\", sep=\"\\t\")\n",
    "fpr, tpr, _ = roc_curve(dat[\"CASE\"], dat[\"probDisease\"])\n",
    "roc_auc = auc(fpr, tpr)\n",
    "\n",
    "plt.plot(fpr, tpr, label=f\"EAS_Foo (AUC = {roc_auc:.2f})\")\n",
    "plt.plot([0, 1], [0, 1], \"k--\", lw=2)\n",
    "\n",
    "plt.xlabel(\"False Positive Rate\")\n",
    "plt.ylabel(\"True Positive Rate\")\n",
    "plt.title(f\"ROC Curve for EAS_Foo\")\n",
    "plt.legend(loc=\"lower right\")\n",
    "output_path = f\"{RESULTS_DIR}/EAS_Foo/traditional_roc.png\"\n",
    "plt.savefig(output_path, dpi=300)\n",
    "plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "67f06a5f-7556-4b14-930e-354512a9b971",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "level_order <- c('EAS_Foo')\n",
    "dat <- fread(paste0(RESULTS_DIR, \"/EAS_Foo/forest_risk.txt\"))\n",
    "plot <- ggplot(data=dat, \n",
    "               aes(x = factor(COHORT,levels = level_order),\n",
    "                   y = BETA,\n",
    "                   ymin = U95,\n",
    "                   ymax = L95)\n",
    "              ) +\n",
    "  geom_pointrange(aes(ymin = L95, ymax = U95), cex = 0.7) +\n",
    "  geom_hline(yintercept = 0, linetype = 2) +\n",
    "  theme(plot.title = element_text(size = 20, face = \"bold\"),\n",
    "        axis.text.y = element_text(size = 8, face = 'bold'),\n",
    "        axis.ticks.y = element_blank(),\n",
    "        axis.text.x = element_text(face=\"bold\"),\n",
    "        axis.title = element_text(size = 16, face=\"bold\"),\n",
    "        legend.position = \"none\"\n",
    "  ) +\n",
    "  xlab(' ') +\n",
    "  ylab(\"Beta coefficient\") +\n",
    "  coord_flip() +\n",
    "  theme_minimal() +\n",
    "  ggtitle(\"EAS_Foo - Individual level data - GP2 release 9\") +\n",
    "  theme(plot.title = element_text(hjust=0.5))\n",
    "\n",
    "\n",
    "ggsave(\n",
    "    paste0(RESULTS_DIR, \"/EAS_Foo/forest_risk.jpg\"), \n",
    "    plot = plot,\n",
    "    width = 8, \n",
    "    height = 5, \n",
    "    units = \"in\",\n",
    "    dpi = 500,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "33ed50eb-752f-4454-a6e7-4ca7d0bd7a29",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confusion Matrix and Statistics\n",
      "\n",
      "          Reference\n",
      "Prediction CONTROL DISEASE\n",
      "   CONTROL    1395     816\n",
      "   DISEASE     954    1003\n",
      "                                          \n",
      "               Accuracy : 0.5753          \n",
      "                 95% CI : (0.5602, 0.5904)\n",
      "    No Information Rate : 0.5636          \n",
      "    P-Value [Acc > NIR] : 0.064801        \n",
      "                                          \n",
      "                  Kappa : 0.144           \n",
      "                                          \n",
      " Mcnemar's Test P-Value : 0.001128        \n",
      "                                          \n",
      "            Sensitivity : 0.5514          \n",
      "            Specificity : 0.5939          \n",
      "         Pos Pred Value : 0.5125          \n",
      "         Neg Pred Value : 0.6309          \n",
      "             Prevalence : 0.4364          \n",
      "         Detection Rate : 0.2406          \n",
      "   Detection Prevalence : 0.4695          \n",
      "      Balanced Accuracy : 0.5726          \n",
      "                                          \n",
      "       'Positive' Class : DISEASE         \n",
      "                                          \n",
      "\n",
      "Call:\n",
      "roc.default(response = dat$reported, predictor = dat$probDisease)\n",
      "\n",
      "Data: dat$probDisease in 2349 controls (dat$reported CONTROL) < 1819 cases (dat$reported DISEASE).\n",
      "Area under the curve: 0.6004\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Setting levels: control = CONTROL, case = DISEASE\n",
       "Setting direction: controls < cases\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "dat <- read.table(paste0(RESULTS_DIR, \"/EAS_Foo/merged.profile\"), header = T)\n",
    "dat$CASE <- dat$PHENO - 1\n",
    "dat <- subset(dat, CASE != -10)\n",
    "meanControls <- mean(dat$SCORE[dat$CASE == 0])\n",
    "sdControls <- sd(dat$SCORE[dat$CASE == 0])\n",
    "dat$zSCORE <- (dat$SCORE - meanControls)/sdControls\n",
    "grsTests <- glm(CASE ~ zSCORE, family=\"binomial\", data = dat)\n",
    "dat$probDisease <- predict(grsTests, dat, type = c(\"response\"))\n",
    "dat$reported <- ifelse(dat$CASE == 1, \"DISEASE\", \"CONTROL\")\n",
    "roc <- roc(response = dat$reported, predictor = dat$probDisease)\n",
    "\n",
    "result.coords <- coords(\n",
    "    roc, \n",
    "    \"best\", \n",
    "    best.method = \"closest.topleft\", \n",
    "    ret = c(\"threshold\", \"accuracy\", \"specificity\", \"sensitivity\", \"youden\"),\n",
    ")\n",
    "\n",
    "dat$predicted <- ifelse(dat$probDisease > result.coords$threshold, \"DISEASE\", \"CONTROL\")\n",
    "confMat <- confusionMatrix(data = as.factor(dat$predicted), reference = as.factor(dat$reported), positive = \"DISEASE\")\n",
    "print(confMat)\n",
    "print(roc)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4461366-469c-4389-bcc1-158346dbd92d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "python3",
   "name": "r-cpu.m127",
   "type": "gcloud",
   "uri": "us-docker.pkg.dev/deeplearning-platform-release/gcr.io/r-cpu:m127"
  },
  "kernelspec": {
   "display_name": "Python 3 (Local)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
